---
title: "Dreamlet analysis NEMP25"
author: "Reed McMullen"
date: "2025-02-11"
output: html_document
---

# Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r installations, message=FALSE, warning=FALSE}
#Install required packages.
install.packages(c('dplyr',
                   "ComplexUpset",
                   'ggplot2',
                   'tidytable',
                   "viridis",
                   "RColorBrewer",
                   "tidyverse",
                   'tidyr',
                   "ggrepel",
                   "umap",
                   "tibble",
                   'biomaRt',
                   'stringr',
                   "openxlsx"))

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(c("clusterProfiler",
                       "dreamlet",
                       "SingleCellExperiment",
                       "zenith",
                       "org.Hs.eg.db",
                       "ComplexHeatmap",
                       "PCAtools",
                       "dorothea",
                       "biomaRt",
                       "AnnotationDbi"))

install_gitlab("medbio/disgenet2r")
```

Load required libraries
```{r load libraries, message=FALSE, warning=FALSE}
library(ComplexUpset)
library(ggplot2)
library(msigdbr)
library(tidytable)
library(tidyr)
library(viridis)
library(RColorBrewer)
library(tidyverse)
library(ggrepel)
library(ggtext)
library(umap)
library(tibble)
library(dplyr)
library(igraph)
library(ggraph)
library(stringr)
library(circlize)
library(RColorBrewer)
library(purrr)

library(clusterProfiler)
library(dreamlet)
library(SingleCellExperiment)
library(zenith)
library(ComplexHeatmap)
library(PCAtools)
library(dorothea)
library(biomaRt)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(devtools)
library(GO.db)
library(AnnotationDbi)

library(disgenet2r)
library(forcats)
library(scales)
library(openxlsx)

```

Set paths
```{r set paths}
dir_path <- "/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/dreamlet_v2/"
fig_path <- paste0(dir_path, 'figures_v2/')
```

Define named vectors for consistent plotting aesthetics
```{r aesthetic mappings}
pert_colors = c("Untreated" = 'black',
                "TreMan" = "#848884", 
                 "DMSO" = "#C0C0C0",
                 
                 "BMP7" = '#40E0D0',
                 
                 "LY411575" = '#7F00FF',

                 "FGF2" = '#4F7942',
                 "FGF8b" = '#4CBB17',
                 "NRG1" = 'olivedrab',
                 "EGF" = 'olivedrab2',
                 
                 "ATRA" = '#89CFF0',
                 "TRULI" = '#4169E1',
                 "CHIR99021" = '#000080',
                 
                 "WNTC59"= '#C04000',
                 "BGJ398" = '#800000',
                 "SANT1" = '#C41E3A',
                 
                 "IGF1LR3" = '#FF69B4',
                 "JAG1" = '#F8C8DC',
                 "DLL1" = '#D8BFD8',

                 'SHH' = '#FAA0A0',
                 "HX531" = '#FFA500',
                 "AZD8931"= '#DAA520',
                 "LDN193189" = '#FFD700',
                 "Rapamycin" = '#E1C16E'
)

species_colors = c('Human' = '#ee7600',
                   'Chimp' = '#1874cd',
                   'Orangutan' = '#4CBB17')

time_point_colors = c('6hr' = '#006400',
                      '54hr' = '#003f7f',
                      '7day' = '#800080')

significant_colors <- c('FALSE' = 'lightgrey',
                        'TRUE' = 'red1')

time_point_shapes <- c('6hr' = 21,
                       '54hr' = 22,
                       '7day' = 24)

signaling_category_colors <- c("SHH signaling" = "salmon1",
                              "WNT signaling" = 'royalblue2',
                              "WNT & MAPK ERK signaling" = 'royalblue4',
                              "BMP signaling" = "turquoise2",
                              "BMP & MAPK ERK signaling" = "turquoise4",
                              "WNT & BMP signaling" = 'steelblue3',
                              "NOTCH signaling" = 'hotpink1',
                              "FGF signaling" = 'forestgreen',
                              "FGF & MAPK ERK signaling" = 'darkgreen',
                              "ERBB signaling" = 'olivedrab3',
                              "ERBB & MAPK ERK signaling" = "olivedrab4",
                              "FGF & ERBB signaling" = "seagreen3",
                              "FGF & ERBB & MAPK ERK signaling" = "chartreuse1",
                              "HIPPO signaling" = 'purple1',
                              "RA signaling" = 'firebrick2',
                              "mTOR signaling" = 'gold1',
                              #"IGF signaling" = 'gold1',
                              "MAPK ERK signaling" = 'springgreen2',
                              #"RTK signaling" = 'turquoise3', 
                              "Multiple pathways" = 'orange2'
                              )

rgene_colors <- c("Non-rGene" = 'black',
                  "Other rGene" = 'grey',
                  "Conserved rGene" = "royalblue2",
                  "Divergent rGene" = "firebrick2"
                  )

div_rgene_colors <- c("Baseline Only" = 'magenta',
                  "Response Only" = 'green',
                  "Reversing" = "navy",
                  "Enhancing" = "cyan",
                  "Opposing"= "red",
                  'Other' = 'grey'
                  )
```

# Dreamlet results

Load dreamlet results_df
```{r load results_df}
results_df <- readRDS(paste0(dir_path, "dreamlet_results_v16_df.rds"))
#results_df <- results_df %>% mutate(significant = ifelse(adj.P.Val < 0.10, TRUE, FALSE)) %>% # Add significance flag.
head(results_df)
```

# Response df

```{r generate response_df}
# Specify contrasts for response differences
contrasts <- unique(results_df$coef)
responsediff_contrasts <- contrasts[grepl("ResponseDiff", contrasts) & !grepl("SpeciesDiffResponseDiff", contrasts) & !grepl("AvgResponseDiff", contrasts)]

# Filter to response difference data and separate out metadata
response_df <- results_df %>% filter(coef %in% responsediff_contrasts) %>%
  separate(coef, into = c("perturbation", "time_point", "species"), sep = "_", remove = FALSE)

# Set factor levels
response_df$time_point <- factor(response_df$time_point, levels = c('6hr', '54hr', '7day'))
response_df$species <- factor(response_df$species, levels = c('Human', 'Chimp', 'Orangutan'))
head(response_df)
```

Define categories of signaling pathway response genes from GO:BP terms
```{r generate signaling_genes_df}
shh_terms <- c('GO:0007224', # GOBP_SMOOTHENED_SIGNALING_PATHWAY
               'GO:0008589', # GOBP_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY
               'GO:0045879', # GOBP_NEGATIVE_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY
               'GO:0045880' # GOBP_POSITIVE_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY
               )

wnt_terms <- c('GO:0016055', # GOBP_WNT_SIGNALING_PATHWAY
               'GO:0030111', # GOBP_REGULATION_OF_WNT_SIGNALING_PATHWAY
               'GO:0030178', # GOBP_NEGATIVE_REGULATION_OF_WNT_SIGNALING_PATHWAY
               'GO:0030177', # GOBP_POSITIVE_REGULATION_OF_WNT_SIGNALING_PATHWAY
               'GO:0060070', # GOBP_CANONICAL_WNT_SIGNALING_PATHWAY
               'GO:0060828', # GOBP_REGULATION_OF_CANONICAL_WNT_SIGNALING_PATHWAY
               'GO:0090090', # GOBP_NEGATIVE_REGULATION_OF_CANONICAL_WNT_SIGNALING_PATHWAY
               'GO:0090263' # GOBP_POSITIVE_REGULATION_OF_CANONICAL_WNT_SIGNALING_PATHWAY
               )

notch_terms <- c('GO:0007219', # GOBP_NOTCH_SIGNALING_PATHWAY
                 'GO:0008593', # GOBP_REGULATION_OF_NOTCH_SIGNALING_PATHWAY
                 'GO:0045746', # GOBP_NEGATIVE_REGULATION_OF_NOTCH_SIGNALING_PATHWAY
                 'GO:0045747', # GOBP_POSITIVE_REGULATION_OF_NOTCH_SIGNALING_PATHWAY
                 'GO:0007221' # GOBP_POSITIVE_REGULATION_OF_TRANSCRIPTION_OF_NOTCH_RECEPTOR_TARGET
                 )

fgf_terms <- c('GO:0008543', # GOBP_FIBROBLAST_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY
               'GO:0040036', # GOBP_REGULATION_OF_FIBROBLAST_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY
               'GO:0040037', # GOBP_NEGATIVE_REGULATION_OF_FIBROBLAST_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY
               'GO:0040038', # GOBP_POSITIVE_REGULATION_OF_FIBROBLAST_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY
               'GO:0071774', # GOBP_RESPONSE_TO_FIBROBLAST_GROWTH_FACTOR
               'GO:0044344' # GOBP_CELLULAR_RESPONSE_TO_FIBROBLAST_GROWTH_FACTOR_STIMULUS
               )

hippo_terms <- c('GO:0035329', # GOBP_HIPPO_SIGNALING
                 'GO:0035330', # GOBP_REGULATION_OF_HIPPO_SIGNALING
                 'GO:0035331', # GOBP_NEGATIVE_REGULATION_OF_HIPPO_SIGNALING
                 'GO:0035332' # GOBP_POSITIVE_REGULATION_OF_HIPPO_SIGNALING
                 )

erbb_terms <- c('GO:0007173', # epidermal growth factor receptor signaling pathway
               'GO:0042058', # regulation of epidermal growth factor receptor signaling pathway
               'GO:0042059', # negative regulation of epidermal growth factor receptor signaling pathway
               'GO:0045742', # positive regulation of epidermal growth factor receptor signaling pathway
               'GO:0038137', # ERBB4-EGFR signaling pathway
               'GO:0038134', # ERBB2-EGFR signaling pathway
               'GO:0070849', # response to epidermal growth factor
               'GO:0071364', # cellular response to epidermal growth factor stimulus
               
                'GO:0038127', # ERBB signaling pathway
                'GO:1901184', # regulation of ERBB signaling pathway
                'GO:1901185', # negative regulation of ERBB signaling pathway
                'GO:1901186', # positive regulation of ERBB signaling pathway
                'GO:0038129', # ERBB3 signaling pathway
                'GO:1905578', # regulation of ERBB3 signaling pathway
                'GO:1905579', # negative regulation of ERBB3 signaling pathway
                'GO:1905580', # positive regulation of ERBB3 signaling pathway
                'GO:0038128', # ERBB2 signaling pathway
                'GO:0038130', # ERBB4 signaling pathway
                'GO:0120154', # negative regulation of ERBB4 signaling pathway
                'GO:0038138', # ERBB4-ERBB4 signaling pathway
                'GO:0038133', # ERBB2-ERBB3 signaling pathway
                'GO:0038136', # ERBB3-ERBB4 signaling pathway
                'GO:0038135' # ERBB2-ERBB4 signaling pathway
                )

bmp_terms <- c('GO:0030509', # BMP signaling pathway
              'GO:0030510', # regulation of BMP signaling pathway
              'GO:0030514', # negative regulation of BMP signaling pathway
              'GO:0030513', # positive regulation of BMP signaling pathway
              'GO:0071772', # response to BMP
              'GO:0071773', # cellular response to BMP stimulus
               
              'GO:0060395', # SMAD protein signal transduction
              'GO:0060390', # regulation of SMAD protein signal transduction
              'GO:0060391', # positive regulation of SMAD protein signal transduction
              'GO:0060392' # negative regulation of SMAD protein signal transduction
               
                    #'GO:0007179', # transforming growth factor beta receptor signaling pathway
                    #'GO:0017015', # regulation of transforming growth factor beta receptor signaling pathway
                    #'GO:0030512', # negative regulation of transforming growth factor beta receptor signaling pathway
                    #'GO:0030511', # positive regulation of transforming growth factor beta receptor signaling pathway
                    #'GO:0141091', # transforming growth factor beta receptor superfamily signaling pathway
                    #'GO:0071559', # response to transforming growth factor beta
                    #'GO:1903845', # negative regulation of cellular response to transforming growth factor beta stimulus
                    #'GO:1903846', # positive regulation of cellular response to transforming growth factor beta stimulus
                    #'GO:0071560', # cellular response to transforming growth factor beta stimulus
                    #'GO:1903844', # regulation of cellular response to transforming growth factor beta stimulus
                    #'GO:1903845', # negative regulation of cellular response to transforming growth factor beta stimulus
                    #'GO:1903846' # positive regulation of cellular response to transforming growth factor beta stimulus
                    )

ra_terms <- c('GO:0048384', # retinoic acid receptor signaling pathway
              'GO:0048385', # regulation of retinoic acid receptor signaling pathway
              'GO:0048387', # negative regulation of retinoic acid receptor signaling pathway
              'GO:0048386', # positive regulation of retinoic acid receptor signaling pathway
              'GO:0032526', # response to retinoic acid
              'GO:0071300', # cellular response to retinoic acid
              'GO:0002138' # retinoic acid biosynthetic process
              )

mtor_terms <- c('GO:0031929', # TOR signaling
                'GO:0032006', # regulation of TOR signaling
                'GO:0032007', # negative regulation of TOR signaling
                'GO:0032008', # positive regulation of TOR signaling
                'GO:0038202', # TORC1 signaling
                'GO:0038203', # TORC2 signaling
                'GO:1901355', # response to rapamycin
                'GO:0072752' # cellular response to rapamycin
                )

#igf_terms <- c('GO:0048009', # insulin-like growth factor receptor signaling pathway
#               'GO:0043567', # regulation of insulin-like growth factor receptor signaling pathway
#               'GO:0043569', # negative regulation of insulin-like growth factor receptor signaling pathway
#               'GO:0043568', # positive regulation of insulin-like growth factor receptor signaling pathway
#               'GO:1990418', # response to insulin-like growth factor stimulus
#               'GO:1990314' # cellular response to insulin-like growth factor stimulus
#               )

mapk_erk_terms <- c('GO:0000165', # MAPK cascade
                    'GO:0043408', # regulation of MAPK cascade
                    'GO:0043409', # negative regulation of MAPK cascade
                    'GO:0043410', # positive regulation of MAPK cascade
                    'GO:0043405', # regulation of MAP kinase activity
                    'GO:0043407', # negative regulation of MAP kinase activity
                    'GO:0043406', # positive regulation of MAP kinase activity

                    'GO:0070371', # ERK1 and ERK2 cascade
                    'GO:0070372', # regulation of ERK1 and ERK2 cascade
                    'GO:0070373', # negative regulation of ERK1 and ERK2 cascade
                    'GO:0070374' # positive regulation of ERK1 and ERK2 cascade
                    )

# Connect to Ensembl
mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

go_term_sets <- list('SHH signaling' = shh_terms,
                  'WNT signaling' = wnt_terms,
                  'NOTCH signaling' = notch_terms,
                  'FGF signaling' = fgf_terms,
                  'HIPPO signaling' = hippo_terms,
                  'ERBB signaling' = erbb_terms,
                  'BMP signaling' = bmp_terms,
                  'RA signaling' = ra_terms,
                  'mTOR signaling' = mtor_terms,
                  #'IGF signaling' = igf_terms,
                  'MAPK ERK signaling' = mapk_erk_terms
                  )

# Initialize an empty list to store data frames
signaling_gene_list <- list()

# Iterate through GO term sets
for (key in names(go_term_sets)) {
  # Retrieve genes for GO terms
  genes <- getBM(attributes = c("hgnc_symbol", "go_id"), 
                 filters = "go", 
                 values = go_term_sets[[key]], 
                 mart = mart)
  
  # Add signaling category column
  genes$signaling_category <- key
  
  # Append to the list
  signaling_gene_list[[key]] <- genes
}

# Combine all data frames into one
signaling_genes_df <- bind_rows(signaling_gene_list)
head(signaling_genes_df)
```

Clean up signaling genes data frame
```{r clean up signaling_genes_df}
signaling_genes_df <- signaling_genes_df %>%
  dplyr::select(-go_id) %>%
  dplyr::rename(ID = hgnc_symbol) %>%
  distinct(ID, signaling_category, .keep_all = TRUE) %>%
  group_by(ID) %>%
  summarise(signaling_category = paste(unique(signaling_category), collapse = ", "), .groups = "drop") %>%
  dplyr::slice(-1)

head(signaling_genes_df)
```

Look for frequently repeated combinations
```{r frequent signaling combos}
normalized_counts <- signaling_genes_df %>%
  dplyr::count(signaling_category) %>%
  mutate(pct = n / sum(n)*100) %>%
  arrange(desc(pct))

head(normalized_counts, 30)
```

Customize signaling genes data frame
```{r rename frequent combos}
signaling_genes_df <- signaling_genes_df %>%
  mutate(
    signaling_category = ifelse(signaling_category == "ERBB signaling, MAPK ERK signaling", "ERBB & MAPK ERK signaling", signaling_category),
    signaling_category = ifelse(signaling_category == "FGF signaling, MAPK ERK signaling", "FGF & MAPK ERK signaling", signaling_category),
    signaling_category = ifelse(signaling_category == "FGF signaling, ERBB signaling", "FGF & ERBB signaling", signaling_category),
    signaling_category = ifelse(signaling_category == "FGF signaling, ERBB signaling, MAPK ERK signaling", "FGF & ERBB & MAPK ERK signaling", signaling_category),
    signaling_category = ifelse(signaling_category == "WNT signaling, BMP signaling", "WNT & BMP signaling", signaling_category),
    signaling_category = ifelse(signaling_category == "WNT signaling, MAPK ERK signaling", "WNT & MAPK ERK signaling", signaling_category),
    signaling_category = ifelse(signaling_category == "BMP signaling, MAPK ERK signaling", "BMP & MAPK ERK signaling", signaling_category)
  )
head(signaling_genes_df)
```

Customize signaling gene categories
```{r customize signaling categories}
new_row <- setNames(as.list(rep(NA, ncol(signaling_genes_df))), names(signaling_genes_df))

new_row[c("ID", "signaling_category")] <- list("ID2", "BMP signaling")
signaling_genes_df <- rbind(signaling_genes_df, new_row)

new_row[c("ID", "signaling_category")] <- list("ID3", "BMP signaling")
signaling_genes_df <- rbind(signaling_genes_df, new_row)
```

Save signaling genes data frame
```{r save signaling_genes_df}
saveRDS(signaling_genes_df, file = paste0(dir_path, "signaling_genes_df.rds"))
```

Add signaling pathway response gene categories and category colors to the response data frame
```{r add signaling gene metadata to response_df}
# Add signaling categories
response_df <- response_df %>%
  dplyr::left_join(signaling_genes_df %>% dplyr::select(ID, signaling_category), by = "ID")

response_df <- response_df %>%
  dplyr::mutate(simple_signaling_category = case_when(
    is.na(signaling_category) ~ NA_character_,
    grepl(",", signaling_category) ~ "Multiple pathways",
    TRUE ~ signaling_category
  ))

# Add signaling category colors
response_df <- response_df %>%
  dplyr::mutate(signaling_category_color = dplyr::recode(simple_signaling_category, !!!signaling_category_colors)) %>%
  replace_na(list(signaling_category_color = "grey25"))

head(response_df)
```

Save response data frame
```{r save response_df rds}
saveRDS(response_df, paste0(dir_path, "response_df.rds"))
```

Write response data frame to CSV file
```{r save csv response_df}
write.csv(response_df, file = paste0(dir_path, "response_df.csv"), row.names = FALSE)
```

```{r}
write.xlsx(response_df, file = paste0(dir_path, "Table_S4.xlsx"), rowNames = FALSE)
```


Load response data frame
```{r load response_df}
response_df <- readRDS(paste0(dir_path, "response_df.rds"))
head(response_df)
```

# rGenes data frame

Generate a summary data frame storing the number, identity, direction, and conservation of rGenes for each sample group
```{r}
# Add group metadata term
response_df <- response_df %>%
  mutate(group = paste(perturbation, time_point, sep = "_"))

# Helper to collapse vectors to "/"-separated strings with a stable type
collapse_or_na <- function(x) {
  x <- unique(na.omit(x))
  if (length(x) > 0) paste(x, collapse = "/") else NA_character_
}

rgenes_list <- list()

for (pert in unique(response_df$perturbation)) {
  for (time in unique(response_df$time_point)) {

    filt_df <- response_df %>% filter(perturbation == pert, time_point == time)

    # Handle empty combination (optional safety)
    if (nrow(filt_df) == 0) next

    grp <- unique(filt_df$group)

    human_up_rgenes   <- filt_df %>% filter(species == "Human", significant, logFCsign ==  1) %>% pull(ID)
    human_down_rgenes <- filt_df %>% filter(species == "Human", significant, logFCsign == -1) %>% pull(ID)
    all_human_rgenes  <- c(human_up_rgenes, human_down_rgenes)

    chimp_up_rgenes   <- filt_df %>% filter(species == "Chimp", significant, logFCsign ==  1) %>% pull(ID)
    chimp_down_rgenes <- filt_df %>% filter(species == "Chimp", significant, logFCsign == -1) %>% pull(ID)
    all_chimp_rgenes  <- c(chimp_up_rgenes, chimp_down_rgenes)

    all_rgenes            <- unique(c(all_human_rgenes, all_chimp_rgenes))
    conserved_up_rgenes   <- intersect(human_up_rgenes, chimp_up_rgenes)
    conserved_down_rgenes <- intersect(human_down_rgenes, chimp_down_rgenes)
    all_conserved_rgenes  <- c(conserved_down_rgenes, conserved_up_rgenes)

    rgenes_list[[paste(pert, time, sep = "_")]] <- tibble(
      perturbation = pert,
      time_point   = time,
      group        = grp,

      N_all_rGenes           = length(unique(na.omit(all_rgenes))),
      N_all_human_rGenes     = length(unique(na.omit(all_human_rgenes))),
      N_human_up_rGenes      = length(unique(na.omit(human_up_rgenes))),
      N_human_down_rGenes    = length(unique(na.omit(human_down_rgenes))),
      N_all_chimp_rGenes     = length(unique(na.omit(all_chimp_rgenes))),
      N_chimp_up_rGenes      = length(unique(na.omit(chimp_up_rgenes))),
      N_chimp_down_rGenes    = length(unique(na.omit(chimp_down_rgenes))),
      N_conserved_up_rGenes  = length(unique(na.omit(conserved_up_rgenes))),
      N_conserved_down_rGenes= length(unique(na.omit(conserved_down_rgenes))),
      N_all_conserved_rGenes = length(unique(na.omit(all_conserved_rgenes))),

      all_rGenes          = collapse_or_na(all_rgenes),
      all_human_rGenes    = collapse_or_na(all_human_rgenes),
      human_up_rGenes     = collapse_or_na(human_up_rgenes),
      human_down_rGenes   = collapse_or_na(human_down_rgenes),
      all_chimp_rGenes    = collapse_or_na(all_chimp_rgenes),
      chimp_up_rGenes     = collapse_or_na(chimp_up_rgenes),
      chimp_down_rGenes   = collapse_or_na(chimp_down_rgenes),
      all_conserved_rGenes= collapse_or_na(all_conserved_rgenes),
      conserved_up_rGenes = collapse_or_na(conserved_up_rgenes),
      conserved_down_rGenes=collapse_or_na(conserved_down_rgenes)
    )
  }
}

rgenes_df <- bind_rows(rgenes_list)
head(rgenes_df)
```

Generate a data frame storing the number, identity, direction, and conservation of rGenes for each perturbation across time points.
```{r generate simple_rgenes_df}
simple_rgenes_list <- list() # Initialize an empty list

# Iterate through each perturbation
for (pert in unique(response_df$perturbation)) {
  # Filter data to current cell type, perturbation, and time point.
  filt_df <- response_df %>% filter(perturbation == pert)
      
  # Pull rGenes
  human_up_rgenes <- filt_df %>% filter(species == 'Human' & significant == TRUE & logFCsign == 1 & abs(logFC) > 0.5) %>% pull(ID)
  human_down_rgenes <- filt_df %>% filter(species == 'Human' & significant == TRUE & logFCsign == -1 & abs(logFC) > 0.5) %>% pull(ID)
  all_human_rgenes <- c(human_up_rgenes, human_down_rgenes)
  chimp_up_rgenes <- filt_df %>% filter(species == 'Chimp' & significant == TRUE & logFCsign == 1 & abs(logFC) > 0.5) %>% pull(ID)
  chimp_down_rgenes <- filt_df %>% filter(species == 'Chimp' & significant == TRUE & logFCsign == -1 & abs(logFC) > 0.5) %>% pull(ID)
  all_chimp_rgenes <- c(chimp_up_rgenes, chimp_down_rgenes)

  # Define rGene sets
  all_rgenes <- unique(c(all_human_rgenes, all_chimp_rgenes))
  conserved_up_rgenes <- intersect(human_up_rgenes, chimp_up_rgenes)
  conserved_down_rgenes <- intersect(human_down_rgenes, chimp_down_rgenes)
  all_conserved_rgenes <- c(conserved_down_rgenes, conserved_up_rgenes)
      
  # Create a new list entry for the results for the current perturbation
  simple_rgenes_list[[pert]] <- tibble(
    # Metadata
    perturbation = pert,

    # Number of rGenes
    N_all_rGenes = length(all_rgenes),
        
    N_all_human_rGenes = length(all_human_rgenes),
    N_human_up_rGenes = length(human_up_rgenes),
    N_human_down_rGenes = length(human_down_rgenes),
        
    N_all_chimp_rGenes = length(all_chimp_rgenes),
    N_chimp_up_rGenes = length(chimp_up_rgenes),
    N_chimp_down_rGenes = length(chimp_down_rgenes),
        
    N_conserved_up_rGenes = length(conserved_up_rgenes),
    N_conserved_down_rGenes = length(conserved_down_rgenes),
    N_all_conserved_rGenes = length(all_conserved_rgenes),
        
    # Lists of rGenes
    all_rGenes = ifelse(length(all_rgenes) > 0, paste(all_rgenes, collapse = "/"), NA),
        
    all_human_rGenes = ifelse(length(all_human_rgenes) > 0, paste(all_human_rgenes, collapse = "/"), NA),
    human_up_rGenes = ifelse(length(human_up_rgenes) > 0, paste(human_up_rgenes, collapse = "/"), NA),
    human_down_rGenes = ifelse(length(human_down_rgenes) > 0, paste(human_down_rgenes, collapse = "/"), NA),

    all_chimp_rGenes = ifelse(length(all_chimp_rgenes) > 0, paste(all_chimp_rgenes, collapse = "/"), NA),
    chimp_up_rGenes = ifelse(length(chimp_up_rgenes) > 0, paste(chimp_up_rgenes, collapse = "/"), NA),
    chimp_down_rGenes = ifelse(length(chimp_down_rgenes) > 0, paste(chimp_down_rgenes, collapse = "/"), NA),
        
    all_conserved_rGenes = ifelse(length(all_conserved_rgenes) > 0, paste(all_conserved_rgenes, collapse = "/"), NA),
    conserved_up_rGenes = ifelse(length(conserved_up_rgenes) > 0, paste(conserved_up_rgenes, collapse = "/"), NA),
    conserved_down_rGenes = ifelse(length(conserved_down_rgenes) > 0, paste(conserved_down_rgenes, collapse = "/"), NA)
  )
}

# Combine all results into a single data frame
simple_rgenes_df <- bind_rows(simple_rgenes_list)
head(simple_rgenes_df)
```

Save rGenes data frames
```{r save rds rgenes_df}
saveRDS(rgenes_df, file = paste0(dir_path, "rGenes_df.rds"))
saveRDS(simple_rgenes_df, file = paste0(dir_path, "simple_rGenes_df.rds"))
```

Write rGenes data frames to CSV files
```{r save csv rgenes_df}
write.csv(rgenes_df, file = paste0(dir_path, "rGenes_df.csv"), row.names = FALSE)
write.csv(simple_rgenes_df, file = paste0(dir_path, "simple_rGenes_df.csv"), row.names = FALSE)
```

Load rGenes data frame
```{r load rgenes_df}
rgenes_df <- readRDS(paste0(dir_path, "rGenes_df.rds"))
head(rgenes_df)
```

Load simple rGenes data frame
```{r load simple_rgenes_df}
simple_rgenes_df <- readRDS(paste0(dir_path, "simple_rGenes_df.rds"))
head(simple_rgenes_df)
```

# Average response data frame
```{r generate avg_response_df}
# Specify contrasts for average response data
contrasts <- unique(results_df$coef)
avg_response_contrasts <- contrasts[grepl("AvgResponseDiff", contrasts)]

# Filter to response data and separate out metadata.
avg_response_df <- results_df %>% filter(coef %in% avg_response_contrasts) %>%
  separate(coef, into = c("perturbation"), sep = "_", remove = FALSE)

head(avg_response_df)
```

```{r load signaling_genes_df}
signaling_genes_df <- readRDS(paste0(dir_path, "signaling_genes_df.rds"))
head(signaling_genes_df)
```

Add signaling pathway categories and category colors to the avg response data frame
```{r add signaling gene metadata to avg_response_df}
# Add signaling categories.
avg_response_df <- avg_response_df %>%
  left_join(signaling_genes_df %>% dplyr::select(ID, signaling_category), by = "ID")

# Ensure signaling category is a character
avg_response_df$signaling_category <- as.character(avg_response_df$signaling_category)

# Simplify signaling categories in cases when a gene has multiple pathways.
avg_response_df <- avg_response_df %>%
  mutate(simple_signaling_category = case_when(
    is.na(signaling_category) ~ NA_character_,
    grepl(",", signaling_category) ~ "Multiple pathways",
    TRUE ~ signaling_category
  ))

# Add signaling category colors
avg_response_df <- avg_response_df %>%
  mutate(signaling_category_color = recode(simple_signaling_category, !!!signaling_category_colors)) %>%
  replace_na(list(signaling_category_color = "grey25"))

head(avg_response_df)
```

Save avg response data frame
```{r save avg_response_df}
saveRDS(avg_response_df, paste0(dir_path, "avg_response_df.rds"))
```

Write average response data frame to CSV file
```{r save csv avg_response_df}
write.csv(avg_response_df, file = paste0(dir_path, "avg_response_df.csv"), row.names = FALSE)
```

Load avg response data frame
```{r load avg_response_df}
avg_response_df <- readRDS(paste0(dir_path, "avg_response_df.rds"))
head(avg_response_df)
```

# Species difference data frame

```{r generate speciesdiff_df}
# Specify contrasts for species differences
contrasts <- unique(results_df$coef)
speciesdiff_contrasts <- contrasts[grepl("SpeciesDiff", contrasts) & !grepl("SpeciesDiffResponseDiff", contrasts)]

# Filter to species difference data and separate out metadata
speciesdiff_df <- results_df %>% filter(coef %in% speciesdiff_contrasts) %>%
  separate(coef, into = c("perturbation", "time_point"), sep = "_", remove = FALSE)

# Set factor levels
speciesdiff_df$time_point <- factor(speciesdiff_df$time_point, levels = c('6hr', '54hr', '7day'))
head(speciesdiff_df)
```

Save species differences data frame
```{r save speciesdiff_df rds}
saveRDS(speciesdiff_df, paste0(dir_path, "speciesdiff_df.rds"))
```

Write species differences data frame to CSV file
```{r save csv avg_response_df}
write.csv(speciesdiff_df, file = paste0(dir_path, "speciesdiff_df.csv"), row.names = FALSE)
```

```{r}
write.xlsx(speciesdiff_df, file = paste0(dir_path, "speciesdiff_df.xlsx"), rowNames = FALSE)
```

Load species diff data frame
```{r load speciesdiff_df}
speciesdiff_df <- readRDS(paste0(dir_path, "speciesdiff_df.rds"))
head(speciesdiff_df)
```

# Divergent response data frame

```{r generate divergent_response_df}
# Specify contrasts for divergent responses
contrasts <- unique(results_df$coef)
speciesdiff_responsediff_contrasts <- contrasts[grepl("SpeciesDiffResponseDiff", contrasts)]

# Filter to divergent response data and separate out metadata
divergent_response_df <- results_df %>% filter(coef %in% speciesdiff_responsediff_contrasts) %>%
  separate(coef, into = c("perturbation", "time_point"), sep = "_", remove = FALSE) %>%
  mutate(significant = ifelse(adj.P.Val < 0.10, TRUE, FALSE))

# Set factor levels.
divergent_response_df$time_point <- factor(divergent_response_df$time_point, levels = c('6hr', '54hr', '7day'))
head(divergent_response_df)
```

```{r load signaling_genes_df}
signaling_genes_df <- readRDS(paste0(dir_path, "signaling_genes_df.rds"))
head(signaling_genes_df)
```

Add signaling pathway response gene categories and category colors to the divergent response data frame
```{r add signaling gene metadata to divergent_response_df}
# Add signaling categories
divergent_response_df <- divergent_response_df %>%
  left_join(signaling_genes_df %>% dplyr::select(ID, signaling_category), by = "ID")

# Simplify signaling categories in cases when a gene has multiple pathways
divergent_response_df$signaling_category <- as.character(divergent_response_df$signaling_category)

divergent_response_df <- divergent_response_df %>%
  mutate(simple_signaling_category = case_when(
    is.na(signaling_category) ~ NA_character_,
    grepl(",", signaling_category) ~ "Multiple pathways",
    TRUE ~ signaling_category
  ))

# Add signaling category colors
divergent_response_df <- divergent_response_df %>%
  mutate(signaling_category_color = recode(simple_signaling_category, !!!signaling_category_colors)) %>%
  replace_na(list(signaling_category_color = "grey25"))

head(divergent_response_df)
```

Load in NDD genes as a dataframe
```{r generate ndd_genes_df}
# Define file paths and names in a list
ndd_files <- c(
  SCZ = "disgenet_SCZ_genes.csv",
  ASD = "disgenet_ASD_genes.csv",
  BPD = "disgenet_BPD_genes.csv",
  EPI = "disgenet_EPI_genes.csv",
  ID = "disgenet_ID_genes.csv",
  MACRO = "disgenet_macrocephaly_genes.csv", 
  MICRO = "disgenet_microcephaly_genes.csv"
)

# Read each CSV and build a data frame
ndd_genes_list <- lapply(names(ndd_files), function(key) {
  df <- read_csv(ndd_files[[key]])
  tibble(ID = df$Gene, NDD = key)
})

# Combine them into one data frame
ndd_genes_df <- bind_rows(ndd_genes_list)

# Set relaxed significance threshold


# Inspect result
head(ndd_genes_df)
```

Clean up NDD genes data frame
```{r clean up signaling_genes_df}
ndd_genes_df <- ndd_genes_df %>%
  distinct(ID, NDD, .keep_all = TRUE) %>%
  group_by(ID) %>%
  summarise(NDD = paste(unique(NDD), collapse = ", "), .groups = "drop")

head(ndd_genes_df)
```

Add NDD gene categories to the divergent response data frame
```{r add NDD gene metadata to divergent_response_df}
# Add NDD categories
divergent_response_df <- divergent_response_df %>%
  left_join(ndd_genes_df %>% dplyr::select(ID, NDD), by = "ID")

# Simplify NDD categories in cases when a gene has multiple pathways
divergent_response_df$NDD<- as.character(divergent_response_df$NDD)

divergent_response_df <- divergent_response_df %>%
  mutate(simple_NDD = case_when(
    is.na(NDD) ~ NA_character_,
    grepl(",", NDD) ~ "Multiple NDDs",
    TRUE ~ NDD
  ))

head(divergent_response_df)
```

Save divergent response data frame
```{r save divergent_response_df}
saveRDS(divergent_response_df, paste0(dir_path, "divergent_response_df.rds"))
```

Write divergent response data frame to CSV file
```{r save csv avg_response_df}
write.csv(divergent_response_df, file = paste0(dir_path, "divergent_response_df.csv"), row.names = FALSE)
```

```{r}
write.xlsx(divergent_response_df, file = paste0(dir_path, "divergent_response_df.xlsx"), rowNames = FALSE)
```

Load divergent response data frame
```{r load divergent_response_df}
divergent_response_df <- readRDS(paste0(dir_path, "divergent_response_df.rds"))
head(divergent_response_df)
```

# Divergent rGenes data frame

Generate a summary data frame storing divergent rGenes for each combination of perturbation x time point.
```{r generate divergent_rgenes_df}
# Initialize an empty list to store the results
divergent_rgenes_list <- list()

# Iterate through each time point
for (pert in unique(divergent_response_df$perturbation)){
  for (time in unique(divergent_response_df$time_point)){
    
    # Filter results to current cell type and time point
    filt_df <- divergent_response_df %>% filter(perturbation == pert & time_point == time)

    # Pull divergent rGenes
    response_all <- filt_df %>% filter(significant == TRUE) %>% pull(ID)
    response_up_human <- filt_df %>% filter(significant == TRUE & logFCsign == 1) %>% pull(ID)
    response_up_chimp <- filt_df %>% filter(significant == TRUE & logFCsign == -1) %>% pull(ID)
      
    # Create a new list entry for the results for the current time point and perturbation
    divergent_rgenes_list[[paste(pert, time, sep = "_")]] <- tibble(
      perturbation = pert,
      time_point = time,
        
      N_div_rGenes = length(response_all),
      N_up_human_div_rGenes = length(response_up_human),
      N_up_chimp_div_rGenes = length(response_up_chimp),
        
      div_rGenes = ifelse(length(response_all) > 0, paste(response_all, collapse = "/"), NA),
      up_human_div_rGenes = ifelse(length(response_up_human) > 0, paste(response_up_human, collapse = "/"), NA),
      up_chimp_div_rGenes = ifelse(length(response_up_chimp) > 0, paste(response_up_chimp, collapse = "/"), NA)
      )
  }
}

# Combine all results into a single data frame
divergent_rgenes_df <- bind_rows(divergent_rgenes_list)
head(divergent_rgenes_df)
```

Save the divergent rGenes data frame
```{r save divergent_rgenes_df}
saveRDS(divergent_rgenes_df, file = paste0(dir_path, "divergent_rGenes_df.rds"))
```

Generate a data frame storing the number, identity, and direction of divergent rGenes for each perturbation
```{r generate simple_divergent_rgenes_df}
# Initialize an empty list to store the results
simple_degs_list <- list()

# Iterate through each perturbation
for (pert in unique(divergent_rgenes_df$perturbation)){
  # Filter results to the current perturbation
  filt_df <- divergent_rgenes_df %>% filter(perturbation == pert)
    
  simple_df <- filt_df %>% group_by(perturbation) %>%
    summarise(div_rGenes = div_rGenes %>%
                str_split("/") %>% unlist() %>% na.omit() %>% unique() %>% paste(collapse = "/"),
              up_human_div_rGenes = up_human_div_rGenes %>%
                str_split("/") %>% unlist() %>% na.omit() %>%unique() %>% paste(collapse = "/"),
              up_chimp_div_rGenes = up_chimp_div_rGenes %>%
                str_split("/") %>% unlist() %>% na.omit() %>%unique() %>% paste(collapse = "/")
              )
  simple_degs_list[[pert]] <- simple_df
}
simple_divergent_rgenes_df <- bind_rows(simple_degs_list)

# Add DEG counts for each category
simple_divergent_rgenes_df <- simple_divergent_rgenes_df %>%
  mutate(
    N_all = str_split(div_rGenes, "/") %>% sapply(length),
    N_all_up_human = str_split(up_human_div_rGenes, "/") %>% sapply(length),
    N_all_up_chimp = str_split(up_chimp_div_rGenes, "/") %>% sapply(length)
  )
head(simple_divergent_rgenes_df)
```

Save the simple divergent rGenes data frame
```{r save simple_divergent_rgenes_df}
saveRDS(simple_divergent_rgenes_df, file = paste0(dir_path, "simple_divergent_rGenes_df.rds"))
```

Write divergent rGene data frames to CSV files
```{r save csv rgenes_df}
write.csv(divergent_rgenes_df, file = paste0(dir_path, "divergent_rGenes_df.csv"), row.names = FALSE)
write.csv(simple_divergent_rgenes_df, file = paste0(dir_path, "simple_divergent_rGenes_df.csv"), row.names = FALSE)
```

Load the divergent rGenes data frame
```{r load divergent_rgenes_df}
divergent_rgenes_df <- readRDS(paste0(dir_path, "divergent_rGenes_df.rds"))
head(divergent_rgenes_df)
```

Load the simple divergent rGenes data frame
```{r load simple_divergent_rgenes_df}
simple_divergent_rgenes_df <- readRDS(paste0(dir_path, "simple_divergent_rGenes_df.rds"))
head(simple_divergent_rgenes_df)
```

# Conserved rGenes data frame

Generate conserved rGenes data frame by exclusion of divergent rGenes from initial conserved rGenes
```{r generate conserved_response_genes_df}
# Prepare an empty list to collect updates
conserved_rgenes_list <- list()

for (pert in unique(rgenes_df$perturbation)) {
  for (tp in unique(rgenes_df$time_point)) {

    # Filter to current perturbation and time point
    filt_r_df  <- rgenes_df %>% dplyr::filter(perturbation == pert, time_point == tp)
    filt_dr_df <- divergent_rgenes_df %>% dplyr::filter(perturbation == pert, time_point == tp)

    # Helper to split slash-delimited strings safely (handles 0-row filters)
    split_slash <- function(x) {
      x <- x[!is.na(x)]
      if (length(x) == 0) return(character(0))
      out <- unlist(stringr::str_split(paste(x, collapse = "/"), "/"))
      out <- out[!is.na(out) & out != ""]
      unique(as.character(out))
    }

    # Pull rGenes, conserved rGenes, and divergent rGenes
    rgenes          <- split_slash(filt_r_df  %>% dplyr::pull(.data$all_rGenes))
    con_rgenes      <- split_slash(filt_r_df  %>% dplyr::pull(.data$all_conserved_rGenes))
    up_con_rgenes   <- split_slash(filt_r_df  %>% dplyr::pull(.data$conserved_up_rGenes))
    down_con_rgenes <- split_slash(filt_r_df  %>% dplyr::pull(.data$conserved_down_rGenes))
    div_rgenes      <- split_slash(filt_dr_df %>% dplyr::pull(.data$div_rGenes))  # if your column is actually named `all`, use .data$`all`

    # Update final conserved rGenes by excluding divergent rGenes
    con_rgenes_final      <- setdiff(con_rgenes, div_rgenes)
    up_con_rgenes_final   <- intersect(con_rgenes_final, up_con_rgenes)
    down_con_rgenes_final <- intersect(con_rgenes_final, down_con_rgenes)

    # Save the updates as a small tibble
    conserved_rgenes_list[[paste("conserved", pert, tp, sep = "_")]] <- tibble::tibble(
      perturbation            = pert,
      time_point              = tp,
      N_rGenes                = length(rgenes),
      N_conserved_rGenes      = length(con_rgenes_final),
      N_up_conserved_rGenes   = length(up_con_rgenes_final),
      N_down_conserved_rGenes = length(down_con_rgenes_final),
      N_div_rGenes            = length(div_rgenes),
      conserved_rGenes        = paste(con_rgenes_final, collapse = "/"),
      up_conserved_rGenes     = paste(up_con_rgenes_final, collapse = "/"),
      down_conserved_rGenes   = paste(down_con_rgenes_final, collapse = "/")
    )
  }
}

# Combine all updates into a single dataframe
conserved_rgenes_df <- dplyr::bind_rows(conserved_rgenes_list)
```

```{r generate simple_conserved_response_degs_df}
# Initialize an empty list to store the results.
simple_conserved_rgenes_list <- list()

# Iterate through each perturbation.
for (pert in unique(conserved_rgenes_df$perturbation)){

  # Filter results to the current perturbation
  filt_df <- conserved_rgenes_df %>% filter(perturbation == pert)
    
  simple_df <- filt_df %>% group_by(perturbation) %>%
    summarise(conserved_rGenes = conserved_rGenes %>%
                str_split("/") %>% unlist() %>% na.omit() %>% unique() %>% paste(collapse = "/"),
              up_conserved_rGenes = up_conserved_rGenes%>%
                str_split("/") %>% unlist() %>% na.omit() %>%unique() %>% paste(collapse = "/"),
              down_conserved_rGenes = down_conserved_rGenes %>%
                str_split("/") %>% unlist() %>% na.omit() %>%unique() %>% paste(collapse = "/")
              )
  simple_conserved_rgenes_list[[pert]] <- simple_df
}
simple_conserved_rgenes_df <- bind_rows(simple_conserved_rgenes_list)

# Add DEG counts for each category
simple_conserved_rgenes_df <- simple_conserved_rgenes_df %>%
  mutate(
    N_conserved_rGenes = str_split(conserved_rGenes, "/") %>% sapply(length),
    N_up_conserved_rGenes = str_split(up_conserved_rGenes, "/") %>% sapply(length),
    N_down_conserved_rGenes = str_split(down_conserved_rGenes, "/") %>% sapply(length)
    ) %>%
  dplyr::select(perturbation,  N_conserved_rGenes, N_up_conserved_rGenes, N_down_conserved_rGenes,
                conserved_rGenes, up_conserved_rGenes, down_conserved_rGenes)

head(simple_conserved_rgenes_df)
```

Save conserved rGenes data frames
```{r save conserved_rgenes_df}
saveRDS(conserved_rgenes_df, paste0(dir_path, "conserved_rGenes_df.rds"))
saveRDS(simple_conserved_rgenes_df, paste0(dir_path, "simple_conserved_rGenes_df.rds"))
```

Write conserved rGenes data frames to CSV files
```{r save csv conserved_rgenes_df}
write.csv(conserved_rgenes_df, file = paste0(dir_path, "conserved_rGenes_df.csv"), row.names = FALSE)
write.csv(simple_conserved_rgenes_df, file = paste0(dir_path, "simple_conserved_rGenes_df.csv"), row.names = FALSE)
```

Load the conserved rGenes data frame
```{r load conserved_rgenes_df}
conserved_rgenes_df <- readRDS(paste0(dir_path, "conserved_rGenes_df.rds"))
head(conserved_rgenes_df)
```

Load the simple conserved rGenes data frame
```{r load simple_conserved_rgenes_df}
simple_conserved_rgenes_df <- readRDS(paste0(dir_path, "simple_conserved_rGenes_df.rds"))
head(simple_conserved_rgenes_df)
```

# Response LogFC data frame

Generate a wide format data frame:
  rows= each combo of sample group and gene, 
  cols= gene ID, perturbation, time point, Human logFC, Chimp LogFC, and Significant.
```{r generate response_logfc_df}
# Add grouping terms
response_df$group <- paste(response_df$perturbation, response_df$time_point, sep = "_")
response_df$ID_group <- paste(response_df$ID, response_df$group, sep = "_")

# Pivot data frame to wide format based on species, removing rows for which there is no value for either Human or Chimp.
response_logfc_df <- response_df %>%
  group_by(ID_group) %>%
  summarize(
    Chimp_logFC = if ("Chimp" %in% species) logFC[species == "Chimp"][1] else NA_real_,
    Human_logFC = if ("Human" %in% species) logFC[species == "Human"][1] else NA_real_,
    Chimp_rGene_sig = if ("Chimp" %in% species) significant[species == "Chimp"][1] else NA,
    Human_rGene_sig = if ("Human" %in% species) significant[species == "Human"][1] else NA,
    perturbation = perturbation[1],
    time_point = time_point[1],
    ID = ID[1],
    .groups = "drop"
  ) %>%
  filter(!is.na(Chimp_logFC) & !is.na(Human_logFC))

head(response_logfc_df)
```

Add conserved and divergent rGene significance flags to the response logFC data frame
```{r add rGene significance flags}
# Extract conserved rGene significance flags
conserved_rgenes_df$group <- paste(conserved_rgenes_df$perturbation, conserved_rgenes_df$time_point, sep="_")
conserved_flags_df <- conserved_rgenes_df %>%
  filter(!is.na(conserved_rGenes) & conserved_rGenes != "") %>%
  tidyr::separate_rows(conserved_rGenes, sep = "/") %>%
  filter(conserved_rGenes != "") %>%
  dplyr::rename(ID = conserved_rGenes) %>%
  mutate(
    cons_rGene_sig = TRUE,
    ID_group = paste(ID, group, sep = "_")
  ) %>%
  dplyr::select(ID, group, ID_group, cons_rGene_sig)

# Add conserved rGene significance flags to response LogFC data frame
response_logfc_df <- left_join(response_logfc_df, conserved_flags_df, by = 'ID_group')
response_logfc_df <- response_logfc_df %>%
  mutate(cons_rGene_sig = if_else(is.na(cons_rGene_sig), FALSE, cons_rGene_sig))

# Extract divergent rGene significance flags
divergent_response_df <- divergent_response_df %>% dplyr::mutate(div_rGene_sig = significant)
divergent_response_df$group <- paste(divergent_response_df$perturbation, divergent_response_df$time_point, sep="_")
divergent_response_df$ID_group <- paste(divergent_response_df$ID, divergent_response_df$group, sep="_")
divergent_flags_df <- divergent_response_df %>% dplyr::select(ID, group, ID_group, div_rGene_sig)

# Add divergent rGene significance flags to response LogFC data frame
response_logfc_df <- left_join(response_logfc_df, divergent_flags_df, by = 'ID_group')

# Reorder columns
response_logfc_df <- response_logfc_df %>% dplyr::select(ID, perturbation, time_point, Chimp_logFC, Human_logFC,
                                                         Chimp_rGene_sig, Human_rGene_sig, cons_rGene_sig, div_rGene_sig)

head(response_logfc_df)
```

Add rGene category metadata
```{r add rGene category metadata}
response_logfc_df <- response_logfc_df %>%
  mutate(rGene_category = case_when(
    cons_rGene_sig ~ "Conserved rGene",
    div_rGene_sig ~ "Divergent rGene",
    Human_rGene_sig & !cons_rGene_sig & !div_rGene_sig ~ "Other rGene",
    Chimp_rGene_sig & !cons_rGene_sig & !div_rGene_sig ~ "Other rGene",
    !Human_rGene_sig & !Chimp_rGene_sig & !cons_rGene_sig & !div_rGene_sig ~ "Non-rGene",
    TRUE ~ NA_character_
  ))
```

```{r load signaling_genes_df}
signaling_genes_df <- readRDS(paste0(dir_path, "signaling_genes_df.rds"))
head(signaling_genes_df)
```

Add signaling pathway categories and category colors to the response logfc df
```{r add signaling gene metadata to response_logfc_df}
# Add signaling categories.
response_logfc_df <- response_logfc_df %>%
  left_join(signaling_genes_df %>% dplyr::select(ID, signaling_category), by = "ID")

# Ensure signaling category is a character
response_logfc_df$signaling_category <- as.character(response_logfc_df$signaling_category)

# Simplify signaling categories in cases when a gene has multiple pathways.
response_logfc_df <- response_logfc_df %>%
  mutate(simple_signaling_category = case_when(
    is.na(signaling_category) ~ NA_character_,
    grepl(",", signaling_category) ~ "Multiple pathways",
    TRUE ~ signaling_category
  ))

# Add signaling category colors
response_logfc_df <- response_logfc_df %>%
  mutate(signaling_category_color = recode(simple_signaling_category, !!!signaling_category_colors)) %>%
  replace_na(list(signaling_category_color = "grey25"))

head(response_logfc_df)
```

Save response LogFC data frame
```{r save response_logfc_df}
saveRDS(response_logfc_df, file = paste0(dir_path, "response_LogFC_df.rds"))
```

Write response LogFC  data frame to CSV file
```{r save csv response_logfc_df}
write.csv(response_logfc_df, file = paste0(dir_path, "response_LogFC_df.csv"), row.names = FALSE)
```

```{r}
write.xlsx(response_logfc_df, file = paste0(dir_path, "response_LogFC_df.xlsx"), rowNames = FALSE)
```

Load response LogFC data frame
```{r load response_logfc_df}
response_logfc_df <- readRDS(paste0(dir_path, "response_LogFC_df.rds"))
head(response_logfc_df)
```

# Species Difference LogFC data frame

Generate baseline vs stimulus speciesdiff LogFC data frame
```{r generate speciesdiff_logfc_df}
# Define perturbation vectors
perts <- c('TRULI', 'ATRA', 'HX531', 'WNTC59', 'AZD8931', 'LDN193189', 'CHIR99021', 'Rapamycin', 'LY411575',
           'BGJ398', 'SANT1', 'BMP7', 'IGF1LR3', 'JAG1', 'DLL1', 'FGF2', 'FGF8b', 'NRG1', 'SHH', 'EGF')

# Initialize an empty list to store the results.
speciesdiff_logfc_list <- list()

# Iterate through time point
for (time in unique(speciesdiff_df$time_point)){
  # Filter dataset to current cell type, time point, and Baseline.
  b_logfc_df <- speciesdiff_df %>% filter(time_point == time & perturbation == 'Baseline') %>%
          
  # Add logFC, significance, and logFC sign columns for baseline
  mutate(Baseline_logFC = logFC,
         Baseline_sig = significant,
         Baseline_logFCsign = logFCsign) %>%
        
  # Select relevant columns only
  dplyr::select(time_point, ID, Baseline_logFC, Baseline_logFCsign, Baseline_sig)
    
  # Iterate through perturbations
  for (pert in perts){
    # Filter dataset to current cell type, time point, and perturbation
    r_logfc_df <- speciesdiff_df %>% filter(time_point == time & perturbation == pert) %>%
          
    # Add Response_logFC column
    mutate(Response_logFC = logFC,
           Response_sig = significant,
           Response_logFCsign = logFCsign) %>%
        
    # Select relevant columns only
    dplyr::select(time_point, ID, Response_logFC, Response_logFCsign, Response_sig) 
      
    # Join the data frames
    logfc_df <- inner_join(b_logfc_df, r_logfc_df, by = c('time_point', 'ID')) %>%
    mutate(perturbation = pert)
      
    # Add the results to the list
    speciesdiff_logfc_list[[paste(pert, time, sep='_')]] <- logfc_df
  }
}

# Join list into a data frame
speciesdiff_logfc_df <- bind_rows(speciesdiff_logfc_list)

# Calculate residuals
speciesdiff_logfc_df$Baseline_logFC <- as.numeric(speciesdiff_logfc_df$Baseline_logFC)
speciesdiff_logfc_df$Response_logFC <- as.numeric(speciesdiff_logfc_df$Response_logFC)

speciesdiff_logfc_df <- speciesdiff_logfc_df %>% mutate(residual = Response_logFC - Baseline_logFC) %>%
    arrange(desc(residual))

# Reorder columns
speciesdiff_logfc_df <- speciesdiff_logfc_df %>% dplyr::select(ID, perturbation, time_point, Baseline_logFC, Response_logFC, Baseline_sig, Response_sig, Baseline_logFCsign, Response_logFCsign, residual)

head(speciesdiff_logfc_df)
```

Add conserved and divergent rGene significance flags
```{r add rGene significance flags}
# Extract conserved rGene significance flags
conserved_flags_df <- conserved_rgenes_df %>%
  filter(!is.na(conserved_rGenes) & conserved_rGenes != "") %>%
  tidyr::separate_rows(conserved_rGenes, sep = "/") %>%
  filter(conserved_rGenes != "") %>%
  dplyr::rename(ID = conserved_rGenes) %>%
  mutate(cons_rGene_sig = TRUE) %>%
  dplyr::select(ID, perturbation, time_point, cons_rGene_sig)

# Add conserved rGene significance flags to speciesdiff LogFC data frame
speciesdiff_logfc_df <- left_join(speciesdiff_logfc_df, conserved_flags_df, by = c('ID', 'perturbation', 'time_point'))
speciesdiff_logfc_df <- speciesdiff_logfc_df %>%
  mutate(cons_rGene_sig = if_else(is.na(cons_rGene_sig), FALSE, cons_rGene_sig))

# Extract divergent rGene significance flags
divergent_response_df <- divergent_response_df %>% dplyr::mutate(div_rGene_sig = significant)
divergent_flags_df <- divergent_response_df %>%
  group_by(ID, perturbation, time_point) %>%
  summarize(div_rGene_sig = any(significant), .groups = "drop") %>%
  dplyr::select(ID, perturbation, time_point, div_rGene_sig)

# Add divergent rGene significance flags to speciesdiff LogFC data frame
speciesdiff_logfc_df <- left_join(speciesdiff_logfc_df, divergent_flags_df, by = c('ID', 'perturbation', 'time_point'))

head(speciesdiff_logfc_df)
```

Add rGene category metadata
```{r add rGene category metadata}
speciesdiff_logfc_df <- speciesdiff_logfc_df %>%
  mutate(rGene_category = case_when(
    cons_rGene_sig ~ "Conserved rGene",
    div_rGene_sig ~ "Divergent rGene",
    TRUE ~ NA_character_
  ))
```

Categorize divergent rGenes
```{r categorize divergent rGenes}
speciesdiff_logfc_df <- speciesdiff_logfc_df %>%
  dplyr::mutate(
    div_rGene_category = dplyr::case_when(
      # Only classify divergent rGenes
      # Baseline only: response not significant and baseline significant
      rGene_category == "Divergent rGene" &
        div_rGene_sig & !Response_sig & Baseline_sig ~ "Baseline Only",
      
     # Response only: response significant and baseline not significant
      rGene_category == "Divergent rGene" &
        div_rGene_sig & Response_sig & !Baseline_sig ~ "Response Only",
     
    # Reversing: response significant and baseline significant and logFC sign is different
      rGene_category == "Divergent rGene" &
        div_rGene_sig & Baseline_sig & Response_sig &
        (Baseline_logFCsign * Response_logFCsign == -1) ~ "Reversing",
    
    # Enhancing: response significant and baseline significant and logFC sign the same and response logFC > baseline logFC absolute value
      rGene_category == "Divergent rGene" &
        div_rGene_sig & Baseline_sig & Response_sig &
        (Baseline_logFCsign * Response_logFCsign == 1) &
        (abs(Response_logFC) > abs(Baseline_logFC)) ~ "Enhancing",
    # Opposing: response significant and baseline significant and logFC sign the same and response logFC < baseline logFC absolute value
      rGene_category == "Divergent rGene" &
        div_rGene_sig & Baseline_sig & Response_sig &
        (Baseline_logFCsign * Response_logFCsign == 1) &
        (abs(Response_logFC) < abs(Baseline_logFC)) ~ "Opposing",

      # Divergent rGenes that don't meet any of the above criteria
      rGene_category == "Divergent rGene" ~ "Other",

      # Non-divergent rGenes stay NA
      TRUE ~ NA_character_
    )
  )
```

Save speciesdiff LogFC data frame
```{r save speciesdiff_logfc_df}
saveRDS(speciesdiff_logfc_df, file = paste0(dir_path, "speciesdiff_LogFC_df.rds"))
```

Write speciesdiff LogFC  data frame to CSV file
```{r save csv response_logfc_df}
write.csv(speciesdiff_logfc_df, file = paste0(dir_path, "speciesdiff_LogFC_df.csv"), row.names = FALSE)
```

```{r}
write.xlsx(speciesdiff_logfc_df, file = paste0(dir_path, "speciesdiff_LogFC_df.xlsx"), rowNames = FALSE)
```

Load speciesdiff LogFC data frame
```{r load speciesdiff_logfc_df}
speciesdiff_logfc_df <- readRDS(paste0(dir_path, "speciesdiff_LogFC_df.rds"))
head(speciesdiff_logfc_df)
```

# E-distance to vehicle

## Load E-distance data

```{r load pert_etest df}
# Load and format E-dist and E-test data
pert_etest_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/pertpy/edistances_etest/pert_etest_v2_results_X_pca_harmony_alpha0.0005_nperms20000.csv")
pert_etest_df <- pert_etest_df %>% 
  dplyr::rename(perturbation = X) %>%
  filter(!(perturbation %in% c('Untreated', 'DMSO', 'TreMan'))) %>%
  arrange(desc(distance))

head(pert_etest_df)
```

```{r save pert_etest_df}
saveRDS(pert_etest_df, paste0(dir_path, "perturbation_EDistanceToControl_ETest_df.rds"))
```

```{r load sample group etest df}
# Load and format E-dist and E-test data
group_etest_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/pertpy/edistances_etest/group_etest_v2_results_X_pca_harmony_alpha5e-05_nperms20000.csv")
group_etest_df <- group_etest_df %>% 
  dplyr::rename(sample_group = X) %>%
  separate(sample_group, into = c("species", "perturbation", "time_point"), sep = "_", remove = FALSE) %>%
  filter(!(perturbation %in% c('Untreated', 'DMSO', 'TreMan'))) %>%
  mutate(time_point = factor(time_point, levels = c("6hr", "54hr", "7day"))) %>%
  arrange(desc(distance))

head(group_etest_df)
```

```{r save pert_etest_df}
saveRDS(group_etest_df, paste0(dir_path, "sample_group_EDistanceToControl_ETest_df.rds"))
```

## E-distance bar plot by perturbation

```{r bar plot E-distance by perturbation}
pert_etest_df <- pert_etest_df %>% arrange(desc(distance)) %>%
    mutate(perturbation = factor(perturbation, levels = perturbation))

p <- ggplot(pert_etest_df, aes(x = perturbation, y = distance, fill = perturbation)) +
  geom_col(color = 'black', linewidth = 0.25) +
  labs(
    title = 'Stimuli Ranked by E-Distance',
    x = "",
    y = "E-Distance from Vehicle") +
  scale_fill_manual(name = "Stimulus", values = pert_colors) +
  theme_minimal() +
  theme(
    panel.grid = element_blank(),
    axis.line = element_line(color = "black"),
    legend.title = element_text(),
    legend.position = "none",
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size=12),
    axis.title.y = element_text(size=14),
    plot.title = element_text(size=16, hjust=0.5)
  )
ggsave(filename = paste0(fig_path, "barplot_EdistanceByPerturbation.png"),
       plot = p, height = 4, width = 8, bg = 'white', dpi=600, units='in')

print(p)
```

## E-distance vs E-test significance scatter plot

```{r scatter plot E-dist vs E-test sig}
# Plot scatter plot
p <- ggplot(pert_etest_df, aes(x = distance, y = -log10(q_val), fill = perturbation)) +
  geom_point(size = 4, color = 'black', stroke = 0.5, shape = 23) +
  labs(
    x = "E-distance",
    y = expression("-" ~ Log[10] ~ "(adj. p-value)"),
    color = "Stimulus"
  ) +
  scale_fill_manual(name = "Perturbation", values = pert_colors,
    guide = guide_legend(override.aes = list(shape = 23, color = "black", stroke = 0.5))
  ) + 
  theme_minimal() + 
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),  # Center and size the title
    legend.position = "right",  # Place legend on the right
    axis.line = element_line(color = "black"),  # Add axis lines
    panel.grid.major = element_blank(),  # Remove major gridlines
    panel.grid.minor = element_blank()   # Remove minor gridlines
  ) + 
  geom_abline(slope = 0, intercept = -log10(alpha), linetype = "dashed", color = "black") # Add dashed threshold line

# Print and save
print(p)
ggsave(filename = paste0(fig_path, "scatterplot_Perturbation_Edistance_vs_ETest.png"),
                           plot = p, height = 7, width = 8, bg = 'white')
```

## Pie chart E-test significance

```{r pie chart significant}
pie_df <- group_etest_df %>%
  dplyr::count(sig_q, name = "n", sort = TRUE) %>%
  mutate(
    frac       = n / sum(n),
    ymax       = cumsum(frac),
    ymin       = dplyr::lag(ymax, default = 0),
    ymid       = (ymin + ymax) / 2,
    pct_label  = scales::percent(frac, accuracy = 0.1),
    Significant = ifelse(sig_q, "Significant", "Not significant")
  )

# Plot pie chart with percentage labels
p <- ggplot(pie_df, aes(ymax = ymax, ymin = ymin, xmax = 1, xmin = 0, fill = Significant)) +
  geom_rect(color = "white") +
  coord_polar(theta = "y") +
  scale_fill_manual(values = c("Significant" = "steelblue3",
                               "Not significant" = "firebrick")) +
  xlim(c(0, 1.6)) +  # add room for labels
  theme_void() +
  geom_text_repel(
    data = pie_df,
    aes(x = 1.3, y = ymid, label = pct_label),
    inherit.aes   = FALSE,
    nudge_x       = 0.05,
    direction     = "y",
    vjust         = 0.5,
    size          = 3,
    segment.size  = 0.3,
    segment.color = "grey50",
    show.legend   = FALSE
  ) +
  ggtitle("E-Test of Sample Group \nE-distances") +
  theme(
    plot.title = element_text(hjust = 0.5)   # center title
  )
ggsave(filename = paste0(fig_path, "PieChart_SampleGroupETestSignificant_ColorBySignificant.png"), plot = p, height = 3, width = 4, units = "in", bg = "white")

print(p)
```

## E-distance vs N rGenes scatter plot

```{r pert E-distance vs N rGenes scatter plot}
# Load perturbation E-distance to control data, adding perturbation and Log2 E-distance columns
pert_etest_df <- readRDS(paste0(dir_path, "perturbation_EDistanceToControl_ETest_df.rds"))

# Select relevant columns from data frames
pert_edist_df <- pert_etest_df %>%
  dplyr::select(perturbation, distance)

pert_rgenes_df <- simple_rgenes_df %>%
  dplyr::select(perturbation, N_all_rGenes) 

# Inner join E-distance and N rGenes data frames
pert_df <- left_join(pert_edist_df, pert_rgenes_df, by='perturbation')

# Calculate Spearman's correlation coefficient
spearman_res <- cor.test(pert_df$distance, pert_df$N_all_rGenes, method = "spearman", exact = FALSE)
spearman_rho <- unname(spearman_res$estimate)
spearman_p   <- spearman_res$p.value

# Plot scatter plot of E-distance vs N rGenes for each perturbation
p <- ggplot(pert_df, aes(x = distance, y = N_all_rGenes, fill = perturbation)) +
  geom_point(shape = 23, size = 4, color = "black", stroke = 0.5, show.legend = TRUE) +
  labs(
    x = "E-Distance",
    y = "N rGenes",
    fill = "Stimulus"
  ) +
  scale_fill_manual(
    name = "Stimulus",
    values = pert_colors,
    guide = guide_legend(ncol = 2, override.aes = list(shape = 21, color = "black", stroke = 0.5))
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    legend.position = "none",
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title = element_text(size=14),
    axis.text = element_text(size=12),
    legend.title = element_text(size=14),
    legend.text = element_text(size=12),
  ) +
  annotate(
    "text",
    x = min(pert_df$distance, na.rm = TRUE),
    y = max(pert_df$N_all_rGenes, na.rm = TRUE),
    hjust = 0, vjust = 1,
    label = sprintf("rho = %.2f\np = %.2g", spearman_rho, spearman_p),
    size = 4
  )

ggsave(filename = paste0(fig_path, "ScatterPlot_Edist_vs_NrGenes_ByPerturbation.png"),
       plot = p, height = 4, width = 4, bg = "white", dpi = 600, units='in')

print(p)
```

```{r E-distance vs N rGenes scatter plot by sample group}
# Load perturbation E-distance to control data, adding perturbation and Log2 E-distance columns
group_etest_df <- readRDS(paste0(dir_path, "sample_group_EDistanceToControl_ETest_df.rds"))

# Select relevant columns from data frames
group_edist_df <- group_etest_df %>%
  dplyr::select(species, perturbation, time_point, distance)

group_rgenes_df <- response_df %>% filter(significant==TRUE & !(perturbation %in% c('Untreated', 'DMSO', 'TreMan'))) %>%
  group_by(species, perturbation, time_point) %>%
  summarise(N_rGenes = n_distinct(ID), .groups = "drop")

# Inner join E-distance and N rGenes data frames
group_df <- left_join(group_edist_df, group_rgenes_df, by=c('species', 'time_point', 'perturbation'))

# Calculate Spearman's correlation coefficient
spearman_res <- cor.test(group_df$distance, group_df$N_rGenes, method = "spearman", exact = FALSE)
spearman_rho <- unname(spearman_res$estimate)
spearman_p   <- spearman_res$p.value

# Plot scatter plot of E-distance vs N rGenes for each perturbation
p <- ggplot(group_df, aes(x = distance, y = N_rGenes, fill = perturbation, shape = time_point)) +
  geom_point(size = 4, color = "black", stroke = 0.5, show.legend = TRUE) +
  labs(
    x = "E-Distance",
    y = "N rGenes",
    fill = "Stimulus"
  ) +
  scale_fill_manual(
    name = "Stimulus",
    values = pert_colors,
    guide = guide_legend(ncol = 2, override.aes = list(shape = 21, color = "black", stroke = 0.5, size=4))
  ) +
  scale_shape_manual(name = "Time Point", values = time_point_shapes,
    guide = guide_legend(override.aes = list(fill = "black", size=4))
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    legend.position = "none",
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title = element_text(size=14),
    axis.text = element_text(size=12),
    legend.title = element_text(size=14),
    legend.text = element_text(size=12),
  ) +
  annotate(
    "text",
    x = min(group_df$distance, na.rm = TRUE),
    y = max(group_df$N_rGenes, na.rm = TRUE),
    hjust = 0, vjust = 1,
    label = sprintf("rho = %.2f\np = %.2g", spearman_rho, spearman_p),
    size = 4
  )

ggsave(filename = paste0(fig_path, "ScatterPlot_Edist_vs_NrGenes_BySampleGroup.png"),
       plot = p, height = 4, width = 4, bg = "white", dpi = 600, units='in')

print(p)
```

## N cells vs E-distance scatter plots

```{r scatter plot N cells vs E-distance by perturbation}
# Load in counts data for perturbations
pert_counts_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/scanpy/telencephalon_final/NEMP25_Tel_v5_perturbation_AllCells_counts.csv")
pert_counts_df <- pert_counts_df %>% filter(!(perturbation %in% c('TreMan', 'Untreated', 'DMSO'))) %>%
    dplyr::rename(N_cells = n)
  
# Join counts and E-distance data frames
pert_edist_cell_counts_df <- dplyr::left_join(pert_counts_df, pert_edist_df, by = c('perturbation'))

# Calculate Spearman's correlation coefficient
spearman_res <- cor.test(
  pert_edist_cell_counts_df$N_cells,
  pert_edist_cell_counts_df$distance,
  method = "spearman",
  exact = FALSE
)
spearman_rho <- unname(spearman_res$estimate)
spearman_p <- spearman_res$p.value

# Add padding
x_pad <- diff(range(pert_edist_cell_counts_df$N_cells, na.rm = TRUE)) * 0.025
y_pad <- diff(range(pert_edist_cell_counts_df$distance, na.rm = TRUE)) * 0.1

# Plot a scatter plot of N rGenes/pert vs. N TelNECs/pert
p <- ggplot(pert_edist_cell_counts_df, aes(x = N_cells, y = distance, fill = perturbation)) +
  geom_point(shape = 23, size = 4, color = "black", fill='black', stroke = 0.5, show.legend = TRUE) +
  labs(
    x = "N Cells",
    y = "E-Distance",
    fill = "Stimulus"
  ) +
  #scale_fill_manual(
    #name = "Stimulus",
    #values = pert_colors,
    #guide = guide_legend(ncol=2, title.position='top', override.aes = list(shape = 21, color = "black", stroke = 0.5))
  #) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    legend.position = "none",
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.title.align = 0.5,
    axis.title = element_text(size = 14),
    axis.text  = element_text(size = 12),
    legend.title = element_text(size=14),
    legend.text = element_text(size=12)
  ) +
  annotate(
    "text",
    x = min(pert_edist_cell_counts_df$N_cells, na.rm = TRUE) + x_pad,
    y = max(pert_edist_cell_counts_df$distance, na.rm = TRUE) - y_pad,
    hjust = 0, vjust = 1,
    label = sprintf("rho = %.2f\np = %.2g", spearman_rho, spearman_p),
    size = 5
  )

print(p)
ggsave(filename = paste0(fig_path, "ScatterPlot_NCells_vs_EDistance_ByPerturbation.png"), plot = p, units='in', height = 4, width = 4, bg = "white", dpi=600)
```

```{r scatter plot N rGenes vs N TelNECs by sample group}
# Load in counts data for sample groups
sg_counts_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/scanpy/telencephalon_final/NEMP25_Tel_v5_sample_group_AllCells_counts.csv")
sg_counts_df <- sg_counts_df %>% filter(!(perturbation %in% c('TreMan', 'Untreated', 'DMSO'))) %>%
    dplyr::rename(N_cells = n)

# Join rGene and TelNEC counts data frames
sg_edist_cell_counts_df <- dplyr::left_join(group_edist_df, sg_counts_df, by = c('species', 'perturbation', 'time_point')) %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Calculate Spearman's correlation coefficient
spearman_res <- cor.test(sg_edist_cell_counts_df$N_cells, sg_edist_cell_counts_df$distance, method = "spearman", exact = FALSE)
spearman_rho <- unname(spearman_res$estimate)
spearman_p <- spearman_res$p.value

# Add padding
x_pad <- diff(range(sg_edist_cell_counts_df$N_cells, na.rm = TRUE)) * 0.01
y_pad <- diff(range(sg_edist_cell_counts_df$distance, na.rm = TRUE)) * 0.02

# Plot a scatter plot of N rGenes/sample group vs. N TelNECs/sample group
p <- ggplot(sg_edist_cell_counts_df, aes(x = N_cells, y = distance, fill = perturbation, shape = time_point)) +
  geom_point(size = 4, color = "black", fill='black', stroke = 0.5) +
  labs(x = "N Cells",
       y = "E-Distance") +
  #scale_fill_manual(values = pert_colors, guide = "none") +
  scale_shape_manual(
    name = "Time Point",
    values = time_point_shapes,
    guide = guide_legend(override.aes = list(fill = "black"))
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title = element_text(size = 14),
    axis.text  = element_text(size = 12),
    legend.position = "right"
  ) +
  annotate(
    "text",
    x = min(sg_edist_cell_counts_df$N_cells, na.rm = TRUE) + x_pad,
    y = max(sg_edist_cell_counts_df$distance, na.rm = TRUE) - y_pad,
    hjust = 0, vjust = 1,
    label = sprintf("rho = %.2f\np = %.2g", spearman_rho, spearman_p),
    size = 5
  )

print(p)
ggsave(filename = paste0(fig_path, "ScatterPlot_NCells_vs_EDistance_BySampleGroup.png"), plot = p, units='in', height = 4, width = 5, bg = "white", dpi=600)
```

## E-distance across time points box plot

```{r box plot E-distance across time points}
group_etest_df <- group_etest_df %>% mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))
                                  
p <- ggplot(group_etest_df, aes(x = time_point, y = distance, fill=time_point)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.15, alpha = 0.6, size = 1) +
  scale_fill_manual(values = time_point_colors, name = "Time point") +
  labs(
    x = "",
    y = "E-Distance"
  ) +
  theme_classic() +
  theme(
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    legend.position = 'none'
  )
ggsave(filename = paste0(fig_path, "boxplot_EDistance_ByTimePoint.png"), plot = p, height = 4, width = 2, bg = 'white', dpi=600)
print(p)
```

## E-distance Human vs Chimp scatter plot

```{r etest_df pivot wide}
# Reshape to wide format
etest_wide <- group_etest_df %>%
  filter(species %in% c("Human", "Chimp") & !(perturbation %in% c('DMSO', 'TreMan', 'Untreated'))) %>%
  pivot_wider(id_cols = c(perturbation, time_point), names_from = species, values_from = distance) %>%
  dplyr::rename(
    Human_Edistance = Human,
    Chimp_Edistance = Chimp
  ) %>%
  mutate(
    Human_log2Edistance = log2(Human_Edistance + 0.0001),
    Chimp_log2Edistance = log2(Chimp_Edistance + 0.0001)
      )

head(etest_wide)
```

```{r E-dist human vs chimp scatter plot}
# Calculate Pearson's correlation coefficient
pearson_res <- cor.test(etest_wide$Human_Edistance, etest_wide$Chimp_Edistance, method = "pearson")
pearson_r <- unname(pearson_res$estimate)
pearson_p <- pearson_res$p.value

# Make square limits: use the same max (and min) for both axes
xy_min <- min(c(etest_wide$Human_Edistance, etest_wide$Chimp_Edistance), na.rm = TRUE)
xy_max <- max(c(etest_wide$Human_Edistance, etest_wide$Chimp_Edistance), na.rm = TRUE)
pad    <- (xy_max - xy_min) * 0.05
if (!is.finite(pad) || pad == 0) pad <- 0

lims <- c(xy_min - pad, xy_max + pad)

# Plot scatter plot
p <- ggplot(etest_wide, aes(x = Human_Edistance, y = Chimp_Edistance, shape = time_point, fill = perturbation)) +
  geom_point(size = 4, color = "black", stroke = 0.5, show.legend = TRUE) +
  labs(
    x = "Human E-Distance",
    y = "Chimp E-Distance",
    fill = "Stimulus",
    shape = "Time Point"
  ) +
  scale_fill_manual(
    name = "Stimulus",
    values = pert_colors,
    guide = guide_legend(ncol = 2, override.aes = list(shape = 21, color = "black", stroke = 0.5))
  ) +
  scale_shape_manual(
    name = "Time Point",
    values = time_point_shapes,
    guide  = guide_legend(override.aes = list(fill = "black"))
  ) +
  annotate(
    "text",
    x = lims[1] + pad,
    y = lims[2] - pad,
    hjust = 0, vjust = 1,
    label = sprintf("r = %.2f\np = %.2g", pearson_r, pearson_p),
    size = 5
  ) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
  coord_fixed(ratio = 1, xlim = lims, ylim = lims, expand = FALSE) +
  theme_minimal() +
  theme(
    legend.position = "none",
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14)
  )

print(p)

ggsave(
  filename = paste0(fig_path, "ScatterPlot_EDistance_HumanVsChimp.png"),
  plot = p, height = 4.75, width = 4.75, bg = "white", dpi=600, units='in'
)
```

# Pairwise E-distances

## Load data and generate metadata data frames

```{r load pert_edist_df}
# Load pairwise E-distance data for perturbations
pert_edist_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/pertpy/edistances_etest/pert_edist_v2_df.csv", row.names = 1)

# Remove columns/rows for controls
pert_edist_df <- pert_edist_df[!grepl("Untreated", rownames(pert_edist_df)), ]
pert_edist_df <- pert_edist_df[, !grepl("Untreated", colnames(pert_edist_df))]
pert_edist_df <- pert_edist_df[!grepl("DMSO", rownames(pert_edist_df)), ]
pert_edist_df <- pert_edist_df[, !grepl("DMSO", colnames(pert_edist_df))]
pert_edist_df <- pert_edist_df[!grepl("TreMan", rownames(pert_edist_df)), ]
pert_edist_df <- pert_edist_df[, !grepl("TreMan", colnames(pert_edist_df))]

# Convert to a matrix.
pert_edist_mat <- as.matrix(pert_edist_df)
head(pert_edist_df)
```

```{r generate pert_metadata_df}
# Create metadata data frame for perturbations
pert_metadata_df <- data.frame(perturbation = rownames(pert_edist_df), stringsAsFactors = FALSE)

# Set factor levels
pert_metadata_df$perturbation <- factor(pert_metadata_df$perturbation, levels = names(pert_colors[-c(1, 2)]))

head(pert_metadata_df)
```

```{r load sample_metagroup_df}
# Load pairwise E-distance data for sample metagroups
metagroup_edist_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/pertpy/edistances_etest/sample_metagroup_edist_v2_df.csv", row.names = 1)

# Remove columns/rows for controls
metagroup_edist_df <- metagroup_edist_df[!grepl("Untreated", rownames(metagroup_edist_df)), ]
metagroup_edist_df <- metagroup_edist_df[, !grepl("Untreated", colnames(metagroup_edist_df))]
metagroup_edist_df <- metagroup_edist_df[!grepl("DMSO", rownames(metagroup_edist_df)), ]
metagroup_edist_df <- metagroup_edist_df[, !grepl("DMSO", colnames(metagroup_edist_df))]
metagroup_edist_df <- metagroup_edist_df[!grepl("TreMan", rownames(metagroup_edist_df)), ]
metagroup_edist_df <- metagroup_edist_df[, !grepl("TreMan", colnames(metagroup_edist_df))]

# Convert to a matrix
metagroup_edist_mat <- as.matrix(metagroup_edist_df)
```

```{r generate metagroup metadata data frame}
# Create metadata data frame for annotating pairwise E-distance for sample groups heatmap
metagroup_metadata_df <- data.frame(metagroup = rownames(metagroup_edist_df), stringsAsFactors = FALSE)

# Add metadata columns
metagroup_metadata_df <- metagroup_metadata_df %>%
  separate(metagroup, into = c("species", "perturbation"), sep = "_", remove = TRUE) %>%
  filter(!(perturbation %in% c('Untreated', 'DMSO', 'Treman')))

# Redo sample name ordering
metagroup_metadata_df$metagroup <- paste(metagroup_metadata_df$species, metagroup_metadata_df$perturbation, sep = "_")
# Set sample as rownames
metagroup_metadata_df <- metagroup_metadata_df %>%
  column_to_rownames(var = "metagroup")
# Reorder columns
metagroup_metadata_df <- metagroup_metadata_df %>%
  dplyr::select(species, perturbation)

# Set factor levels
metagroup_metadata_df$perturbation <- factor(metagroup_metadata_df$perturbation, levels = names(pert_colors[-c(1, 2)]))
metagroup_metadata_df$species <- factor(metagroup_metadata_df$species, levels = c("Human", "Chimp"))

head(metagroup_metadata_df)
```

Load pairwise energy distance results for sample groups and convert to a matrix
```{r load group_edist_df}
group_edist_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/pertpy/edistances_etest/sample_group_edist_v2_df.csv", row.names = 1)

# Remove columns/rows for controls
group_edist_df <- group_edist_df[!grepl("Untreated", rownames(group_edist_df)), ]
group_edist_df <- group_edist_df[, !grepl("Untreated", colnames(group_edist_df))]
group_edist_df <- group_edist_df[!grepl("DMSO", rownames(group_edist_df)), ]
group_edist_df <- group_edist_df[, !grepl("DMSO", colnames(group_edist_df))]
group_edist_df <- group_edist_df[!grepl("TreMan", rownames(group_edist_df)), ]
group_edist_df <- group_edist_df[, !grepl("TreMan", colnames(group_edist_df))]

# Convert to a matrix.
group_edist_mat <- as.matrix(group_edist_df)
head(group_edist_df)
```

```{r generate group_metadata_df}
# Create metadata data frame for annotating pairwise E-distance for sample groups heatmap
group_metadata_df <- data.frame(sample_group = rownames(group_edist_df), stringsAsFactors = FALSE)

# Add metadata columns
group_metadata_df <- group_metadata_df %>%
  separate(sample_group, into = c("species", "perturbation", "time_point"), sep = "_", remove = FALSE) %>%
  column_to_rownames(var = "sample_group") %>% # Set sample group as rownames
  dplyr::select(species, perturbation, time_point)

# Set factor levels
group_metadata_df$time_point <- factor(group_metadata_df$time_point, levels = c('6hr', '54hr', '7day'))
group_metadata_df$perturbation <- factor(group_metadata_df$perturbation, levels = names(pert_colors[-c(1, 2)]))
group_metadata_df$species <- factor(group_metadata_df$species, levels = c("Human", "Chimp"))

head(group_metadata_df)
```

## Pairwise E-distance complex heatmaps

```{r pairwise E-distance complex heatmap by perturbation}
# Set metadata colors
colors_list <- list(
  perturbation = pert_colors[-c(1,3)]
)

# Set E-distance color map
min_val <- min(pert_edist_mat, na.rm = TRUE)
max_val <- max(pert_edist_mat, na.rm = TRUE)

color_map <- circlize::colorRamp2(
  seq(min_val, max_val, length.out = 11),
  viridisLite::magma(11)
)

# Perform hierarchical clustering of pairwise energy distance matrix
hc <- hclust(dist(pert_edist_mat), method = 'centroid')

# Set up annotations.
column_ha <- HeatmapAnnotation(df = pert_metadata_df,
                               col = colors_list)
row_ha <- rowAnnotation(df = pert_metadata_df,
                        col = colors_list,
                        show_annotation_name = FALSE,
                        show_legend = FALSE)

# Plot annotated complexheatmap
p <- Heatmap(pert_edist_mat,
          name = "E-distance",
          cluster_rows = hc,  # Cluster rows
          cluster_columns = hc,  # Cluster columns
          show_column_names = FALSE,  # Display column names
          show_row_names = FALSE,  # Display row names
          col = color_map,
          top_annotation = column_ha,
          left_annotation = row_ha
          )
pdf(paste0(fig_path, "complexheatmap_PairwiseEdistance_ByPertrubation.pdf"), width = 8, height = 6, useDingbats = FALSE)
draw(p)
dev.off()
draw(p)
```

```{r pairwise E-distance complex heatmap by sample metagroup}
# Set metadata colors
colors_list <- list(
  perturbation = pert_colors[-c(1,3)],
  species = species_colors[-c(3)],
  time_point = time_point_colors
)

# Set E-distance color map
min_val <- min(metagroup_edist_mat, na.rm = TRUE)
max_val <- max(metagroup_edist_mat, na.rm = TRUE)

color_map <- circlize::colorRamp2(
  seq(min_val, max_val, length.out = 11),
  viridisLite::magma(11)
)

# Perform hierarchical clustering of pairwise energy distance matrix.
hc <- hclust(dist(metagroup_edist_mat), method = 'average')

# Set up annotations.
column_ha <- HeatmapAnnotation(df = metagroup_metadata_df,
                               col = colors_list,
                               show_annotation_name = TRUE,
                               show_legend = TRUE
                               )

row_ha <- rowAnnotation(df = metagroup_metadata_df,
                        col = colors_list,
                        show_annotation_name = FALSE,   # hide labels next to annotation tracks
                        show_legend = FALSE
                        )

# Plot annotated complexheatmap.
p <- Heatmap(metagroup_edist_mat,
          name = "E-distance",
          cluster_rows = hc,  # Cluster rows
          cluster_columns = hc,  # Cluster columns
          show_column_names = FALSE,  # Display column names
          show_row_names = FALSE,  # Display row names
          col = color_map,
          top_annotation = column_ha,
          left_annotation = row_ha
          )
pdf(paste0(fig_path, "complexheatmap_PairwiseEdistance_BySampleMetagroup.pdf"), width = 8, height = 6, useDingbats = FALSE)
draw(p)
dev.off()
draw(p)
```

```{r pairwise E-distance complex heatmap by sample group}
# Set metadata colors
colors_list <- list(
  perturbation = pert_colors[-seq_len(3)],
  species = species_colors,
  time_point = time_point_colors
)

# Set E-distance color map
min_val <- min(pert_edist_mat, na.rm = TRUE)
max_val <- max(pert_edist_mat, na.rm = TRUE)

color_map <- circlize::colorRamp2(
  seq(min_val, max_val, length.out = 11),
  viridisLite::magma(11)
)

# Perform hierarchical clustering of pairwise energy distance matrix.
hc <- hclust(dist(group_edist_mat), method = 'average')

# Set up annotations.
column_ha <- HeatmapAnnotation(df = group_metadata_df,
                               col = colors_list,
                               show_annotation_name = FALSE,   # hide labels next to annotation tracks
                               show_legend = TRUE              # keep legends
                               )

row_ha <- rowAnnotation(df = group_metadata_df,
                        col = colors_list,
                        show_annotation_name = FALSE,   # hide labels next to annotation tracks
                        show_legend = FALSE
                        )

# Plot annotated complexheatmap.
p <- Heatmap(group_edist_mat,
          name = "E-distance",
          cluster_rows = hc,  # Cluster rows
          cluster_columns = hc,  # Cluster columns
          show_column_names = FALSE,  # Display column names
          show_row_names = FALSE,  # Display row names
          col = color_map,
          top_annotation = column_ha,
          left_annotation = row_ha
          )
pdf(paste0(fig_path, "complexheatmap_PairwiseEdistance_BySampleGroup.pdf"), width = 8, height = 6, useDingbats = FALSE)
draw(p)
dev.off()
draw(p)
```

## Mean parwise E-distance bar plot by perturbation

```{r pert mean pairwise E-distance}
m <- pert_edist_mat

# Exclude self-distances from pairwise pert E-distances
diag(m) <- NA

pert_mp_edist_df <- data.frame(
  perturbation = rownames(m),
  mp_edist = rowMeans(m, na.rm = TRUE),
  row.names = NULL,
  stringsAsFactors = FALSE
)

pert_mp_edist_df <- pert_mp_edist_df %>%
  arrange(desc(mp_edist), perturbation)

head(pert_mp_edist_df)
```

```{r bar plot mean pairwise E-distance by perturbation}
pert_mp_edist_df <- pert_mp_edist_df %>% arrange(mp_edist) %>%
    mutate(perturbation = factor(perturbation, levels = perturbation))

p <- ggplot(pert_mp_edist_df, aes(x = mp_edist, y = perturbation, fill = perturbation)) +
  geom_col(color = 'black', linewidth = 0.25) +
  labs(
    title = 'Stimuli Ranked by \nMean Pairwise E-Distance',
    y = "",
    x = "Mean Pairwise E-Distance\nfrom Other Stimuli") +
  scale_fill_manual(name = "Stimulus", values = pert_colors) +
  theme_minimal() +
  theme(
    panel.grid = element_blank(),
    axis.line = element_line(color = "black"),
    plot.title = element_text(size=16, hjust = 0.5),
    legend.title = element_text(),
    legend.position = "none",
    axis.text = element_text(size=12),
    axis.title = element_text(size=14)
  )
ggsave(filename = paste0(fig_path, "barplot_MeanPairwiseEdistance_ByPerturbation.png"),
       plot = p, height = 9, width = 4, bg = 'white', dpi=600, units='in')

print(p)
```

## Mean pairwise E-distance box plot across time points by sample group

```{r sample group mean pairwise E-distance}
m <- group_edist_mat

# Exclude self-distances
diag(m) <- NA

sg_mean_pairwise_edist_df <- data.frame(
  sample_group = rownames(m),
  mean_pairwise_edist = rowMeans(m, na.rm = TRUE),
  row.names = NULL,
  stringsAsFactors = FALSE
)

sg_mean_pairwise_edist_df <- sg_mean_pairwise_edist_df %>%
  separate(sample_group, into = c("species", "perturbation", "time_point"), remove = FALSE) %>%
  arrange(desc(mean_pairwise_edist), sample_group)

head(sg_mean_pairwise_edist_df)
```

```{r box plot pairwise E-distance across time points}
sg_mean_pairwise_edist_df <- sg_mean_pairwise_edist_df %>% mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))
                                  
p <- ggplot(sg_mean_pairwise_edist_df, aes(x = time_point, y = mean_pairwise_edist, fill=time_point)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.15, alpha = 0.6, size = 1) +
  scale_fill_manual(values = time_point_colors, name = "Time point") +
  labs(
    x = "",
    y = "Mean Pairwise E-Distance"
  ) +
  theme_classic() +
  theme(
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    legend.position = 'None'
  )
ggsave(filename = paste0(fig_path, "boxplot_MeanPairwiseEDistance_ByTimePoint.png"), plot = p, height = 4, width = 2, bg = 'white', dpi=600)
print(p)
```

# Response correlation analysis

## Calculate correlation

Define genes of interest (i.e. rGenes)
```{r define genes of interest}
# Define species and cell types of interest.
specs <- c('Human', 'Chimp')

print(paste0("n genes = ", length(unique(results_df$ID))))

# Pull rGenes
rgenes <- response_df %>% filter(species %in% specs & !(perturbation %in% c('TreMan', 'DMSO')) & significant == TRUE) %>%   
  distinct(ID) %>%
  pull(ID)

print(paste0("n rGenes = ", length(rgenes)))
```

Generate a wide LogFC data frame for genes of interest
```{r generate logFC_df}
# Filter response data frame to the cell types, species, and genes of interest
filt_df <- response_df %>% filter(species %in% specs & ID %in% rgenes) %>%
  # Select only relevant columns
  dplyr::select(ID, logFC, coef, significant)

# Initialize an empty list to store data frames for each cell type
logfc_list <- list()

# Select only relevant columns and Pivot data frame to a wide format
logfc_df <- filt_df %>% 
  dplyr::select(ID, logFC, coef) %>%
  pivot_wider(names_from = coef, values_from = logFC) %>%
  column_to_rownames('ID') %>%
  mutate(across(everything(), as.numeric))  

head(logfc_df)
```

Compute pairwise Pearson's correlation and hierarchical clustering
```{r pearson's r and hclust}
# Compute pairwise Pearson's correlation
cor_matrix <- cor(logfc_df, use = "pairwise.complete.obs", method = "pearson")

# Remove contrast name from row and column names
rownames(cor_matrix) <- sub("_ResponseDiff$", "", rownames(cor_matrix))
colnames(cor_matrix) <- sub("_ResponseDiff$", "", colnames(cor_matrix))

rownames(cor_matrix) <- sub("_ResponseDiff$", "", rownames(cor_matrix))
colnames(cor_matrix) <- sub("_ResponseDiff$", "", colnames(cor_matrix))

# Perform hierarchical clustering.
hc <- hclust(dist(cor_matrix), method = 'complete')
```

```{r}
write.xlsx(cor_matrix, file = paste0(dir_path, "rGene_logFC_correlation_matrix.xlsx"), rowNames = TRUE)
```

## Pairwise correlation heatmap

```{r}
cor_group_metadata_df <- as.data.frame(cor_matrix) %>% rownames_to_column('sample_group') %>%
  dplyr::select(sample_group) %>%
  separate(sample_group, into = c('perturbation', 'time_point', 'species'), remove = FALSE) %>%
  column_to_rownames('sample_group')
  
group_metadata_sub_df <- cor_group_metadata_df[rownames(cor_group_metadata_df) %in% rownames(cor_matrix),]

# Set factor levels
group_metadata_sub_df$time_point <-   factor(group_metadata_sub_df$time_point, levels = c('6hr', '54hr', '7day'))
group_metadata_sub_df$perturbation <- factor(group_metadata_sub_df$perturbation, levels = names(pert_colors))
group_metadata_sub_df$species <-      factor(group_metadata_sub_df$species, levels = c("Human", "Chimp"))

head(group_metadata_sub_df)
```

Plot a hierarchically clustered Pearson's correlation matrix of sample group annotated by metadata variables.
```{r complexheatmap correlation matrix}
# Define a list for colors used for annotating the heatmap
colors_list <- list(
  perturbation = pert_colors,
  species = species_colors[-c(3)],
  time_point = time_point_colors
)

# Set color map as bwr.
color_map = colorRamp2(c(-1, 0, 1), c("darkblue", "white", "darkred"))

#Set column and row annotations
column_ha <- HeatmapAnnotation(df = group_metadata_sub_df, col = colors_list, show_annotation_name = FALSE, show_legend = FALSE)
row_ha <- rowAnnotation(df = group_metadata_sub_df, col = colors_list, show_annotation_name = FALSE)

p <- Heatmap(cor_matrix,
          name = "Pearson\nCorrelation",
          cluster_rows = hc,  # Cluster rows
          cluster_columns = hc,  # Cluster columns
          show_column_names = FALSE,  # Display column names
          show_row_names = FALSE,  # Display row names
          col = color_map,
          top_annotation = column_ha,
          left_annotation = row_ha
          )
png(paste0(fig_path, "ComplexHeatmap_CorrelationCoefficient_rGene_Log2FCs.png"), width = 8, height = 6, units = "in", res = 600)
draw(p)
dev.off()
draw(p)
```

## Correlation density plots

```{r}
# Reconfigure group metadata df
group_metadata_df <- group_metadata_df %>%
  mutate(sample_group = paste(perturbation, time_point, species, sep = "_"))

head(group_metadata_df)
```

```{r}
# Filter to perturbations with significant effects
sig_perts <- simple_rgenes_df %>%
  dplyr::filter(N_all_rGenes > 50, !(perturbation %in% c("TreMan", "DMSO"))) %>%
  dplyr::pull(perturbation) %>%
  unique()

sig_perts
```

### Species/stimuli sample group pairs

```{r calculate correlation between species for the same sample group}
# Find HumanChimp species sample pairs within each perturbation x time_point
species_pairs_df <- group_metadata_df %>%
  dplyr::filter(perturbation %in% sig_perts) %>%
  dplyr::select(sample_group, perturbation, time_point, species) %>%
  tidyr::drop_na(sample_group, perturbation, time_point, species) %>%
  dplyr::group_by(perturbation, time_point, species, sample_group) %>% 
  dplyr::summarise(.groups = "drop") %>%  # ensure unique rows before joining
  dplyr::filter(species %in% c("Human", "Chimp")) %>%
  tidyr::pivot_wider(
    names_from  = species,
    values_from = sample_group
  ) %>%
  dplyr::filter(!is.na(Human), !is.na(Chimp)) %>%
  dplyr::transmute(human = Human, chimp = Chimp) %>%
  dplyr::distinct()

# Get correlations for pairs (drop pairs that arent present in the matrix, if any)
species_pairs_df <- species_pairs_df %>%
  filter(human %in% rownames(cor_matrix),
         chimp %in% colnames(cor_matrix)) %>%
  mutate(cor = mapply(function(h, c) cor_matrix[h, c], human, chimp)) %>%
  filter(!is.na(cor)) %>%
  arrange(desc(cor))

head(species_pairs_df)
```

```{r calculate correlation between perturbation and control for the same species and time point}
# Specify vehicle control for perturbations
dmso_perts   <- c('TRULI','ATRA','HX531','WNTC59','AZD8931', 'LDN193189','CHIR99021','Rapamycin','LY411575','BGJ398','SANT1')
treman_perts <- c('BMP7','IGF1LR3','JAG1','DLL1','FGF2','FGF8b','NRG1','SHH','EGF')

# For each treatment, determine which vehicle it should use
treat_meta <- group_metadata_sub_df %>%
  dplyr::filter(perturbation %in% sig_perts) %>%
  rownames_to_column('sample_group') %>%
  mutate(vehicle_needed = case_when(
    perturbation %in% dmso_perts   ~ "DMSO",
    perturbation %in% treman_perts ~ "TreMan",
    TRUE                           ~ NA_character_
  )) %>%
  filter(!is.na(vehicle_needed)) %>%
    dplyr::select(sample_group, species, time_point, perturbation, vehicle_needed)

# Join to find the matching vehicle within the same species and time_point
veh_meta <- group_metadata_sub_df %>%
  rownames_to_column('sample_group') %>%
  filter(perturbation %in% c("DMSO", "TreMan")) %>%
  dplyr::select(vehicle_sample_group = sample_group, species, time_point, vehicle_perturbation = perturbation)

pert_pairs_df <- treat_meta %>%
  left_join(veh_meta, by = c("species", "time_point"), relationship = "many-to-many") %>%
  filter(vehicle_perturbation == vehicle_needed) %>%
  transmute(
    treatment_group = sample_group,
    vehicle_group = vehicle_sample_group,
    species,
    time_point,
    treatment_perturbation = perturbation,
    vehicle_type = vehicle_needed
  ) %>%
  distinct() %>%
  filter(!is.na(vehicle_group))

# Pull correlations for those treatmentvehicle pairs (drop any missing in the matrix)
pert_pairs_df <- pert_pairs_df %>%
  filter(treatment_group %in% rownames(cor_matrix),
         vehicle_group   %in% colnames(cor_matrix)) %>%
  mutate(cor = mapply(function(tg, vg) cor_matrix[tg, vg],
                      treatment_group, vehicle_group)) %>%
  filter(!is.na(cor)) %>%
  arrange(desc(cor))

head(pert_pairs_df)
```

```{r density plot human/chimp and stimulus/vehicle pair response correlations}
# Bind data with desired group labels
both_df <- bind_rows(species_pairs_df %>% transmute(cor, group = "Human/Chimp Pairs"),
                     pert_pairs_df %>% transmute(cor, group = "Stimulus/Vehicle Pairs")) %>%
  filter(!is.na(cor))

# Colors
col_vals <- c(
  "Human/Chimp Pairs" = "gold2",
  "Stimulus/Vehicle Pairs" = "magenta2"
)

p <- ggplot(both_df, aes(x = cor, color = group, fill = group)) +
  geom_density(alpha = 0.50) +
  scale_color_manual(values = col_vals) +
  scale_fill_manual(values = col_vals) +
  labs(title = "Distributions of Response Correlations\nBetween Sample Group Pairs", x = "Correlation", y = "Density") +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size=14),
    axis.text = element_text(size=12),
    axis.title = element_text(size=12),
    panel.grid = element_blank(),
    axis.line = element_line(),
    axis.ticks = element_line(),
    legend.title = element_blank(),
    legend.text = element_text(size=12),
    legend.position = 'bottom'
    ) +
  guides(
    color = guide_legend(ncol = 1, direction = "vertical"),
    fill  = guide_legend(ncol = 1, direction = "vertical")
  )

ggsave(filename = paste0(fig_path, "DensityPlot_rGeneResponseLogFCs_SpeciesStimulus_PairCorrelations.png"),
       plot = p, height = 4, width = 3.75, bg = 'white')
print(p)
```

### Activators/Inhibitors

```{r calculate correlation between perturbation and control for the same species and time point}
# Specify activator pairs
activator_pairs <- list(
  FGF   = c("FGF2", "FGF8b"),
  NOTCH = c("DLL1", "JAG1"),
  ERBB  = c("NRG1", "EGF")
)

# Subset metadata to perturbations in activator pairs
meta_act <- group_metadata_df %>%
  filter(perturbation %in% unlist(activator_pairs)) %>%
  dplyr::select(sample_group, species, time_point, perturbation) %>%
  distinct()

# Expand pairs into table
activator_pairs_df <- tibble(
  pair = names(activator_pairs),
  act1 = vapply(activator_pairs, `[`, character(1), 1),
  act2 = vapply(activator_pairs, `[`, character(1), 2)
)

# Find matching sample groups
activator_match_df <- activator_pairs_df %>%
  inner_join(meta_act, by = c("act1" = "perturbation")) %>% 
  dplyr::rename(group1 = sample_group) %>%
  inner_join(meta_act, by = c("act2" = "perturbation", "species", "time_point")) %>%
  dplyr::rename(group2 = sample_group) %>%
  transmute(pair, species, time_point, act1, act2, group1, group2) %>%
  distinct()

# Pull the correlation from the matrix for each pair
activator_cor_df <- activator_match_df %>%
  filter(group1 %in% rownames(cor_matrix),
         group2 %in% colnames(cor_matrix)) %>%
  mutate(cor = mapply(function(g1, g2) cor_matrix[g1, g2], group1, group2)) %>%
  filter(!is.na(cor)) %>%
  arrange(desc(cor))

head(activator_cor_df)
```

```{r}
# Specify activator/inhibitor pairs
activator_inhibitor_pairs <- list(FGF_1 = c('FGF2', 'BGJ398'),
                               FGF_2 = c('FGF8b', 'BGJ398'),
                               NOTCH_1 = c('DLL1', 'LY411575'),
                               NOTCH_2 = c('JAG1', 'LY411575'),
                               ERBB_1 = c('NRG1', 'AZD8931'),
                               ERBB_2 = c('EGF', 'AZD8931'),
                               #RA = c('ATRA', 'HX531'),
                               WNT = c('CHIR99021', 'WNTC59'),
                               #BMP = c('BMP7', 'LDN193189'),
                               #mTOR = c('IGF1LR3', 'Rapamycin'),
                               SHH = c('SHH', 'SANT1')
)

# Subset metadata to perturbations in activato/inhibitor pairs
meta_act <- group_metadata_df %>%
  filter(perturbation %in% unlist(activator_inhibitor_pairs)) %>%
  dplyr::select(sample_group, species, time_point, perturbation) %>%
  distinct()

# Expand pairs into table
activator_inhibitor_pairs_df <- tibble(
  pair = names(activator_inhibitor_pairs),
  act = vapply(activator_inhibitor_pairs, `[`, character(1), 1),
  inhib = vapply(activator_inhibitor_pairs, `[`, character(1), 2)
)

# Find matching sample group
activator_inhibitor_match_df <- activator_inhibitor_pairs_df %>%
  inner_join(meta_act, by = c("act" = "perturbation")) %>%
  dplyr::rename(group1 = sample_group) %>%
  inner_join(meta_act, by = c("inhib" = "perturbation", "species", "time_point")) %>%
  dplyr::rename(group2 = sample_group) %>%
  transmute(pair, species, time_point, act, inhib, group1, group2) %>%
  distinct()

# Pull the correlation from the matrix for each pair
activator_inhibitor_cor_df <- activator_inhibitor_match_df %>%
  filter(group1 %in% rownames(cor_matrix),
         group2 %in% colnames(cor_matrix)) %>%
  mutate(cor = mapply(function(g1, g2) cor_matrix[g1, g2], group1, group2)) %>%
  filter(!is.na(cor)) %>%
  arrange(cor)

head(activator_inhibitor_cor_df)
```

```{r density plot activator/activator and activator/inhibitor pair response correlations}
# Bind data with desired group labels
both_df <- bind_rows(activator_cor_df %>% transmute(cor, group = "Same Pathway Activator/Activator Pairs"),
                     activator_inhibitor_cor_df %>% transmute(cor, group = "Same Pathway Activator/Inhibitor Pairs")) %>%
  filter(!is.na(cor))

# Colors
col_vals <- c(
  "Same Pathway Activator/Activator Pairs" = "darkgreen",
  "Same Pathway Activator/Inhibitor Pairs" = "darkorange"
)

p <- ggplot(both_df, aes(x = cor, color = group, fill = group)) +
  geom_density(alpha = 0.50) +
  scale_color_manual(values = col_vals) +
  scale_fill_manual(values = col_vals) +
  labs(title = "Distributions of Response Correlations\nBetween Sample Group Pairs", x = "Correlation", y = "Density") +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size=14),
    axis.text = element_text(size=12),
    axis.title = element_text(size=12),
    panel.grid = element_blank(),
    axis.line = element_line(),
    axis.ticks = element_line(),
    legend.title = element_blank(),
    legend.text = element_text(size=12),
    legend.position = 'bottom'
    ) +
  guides(
    color = guide_legend(ncol = 1, direction = "vertical"),
    fill  = guide_legend(ncol = 1, direction = "vertical")
  )

ggsave(filename = paste0(fig_path, "DensityPlot_rGeneResponseLogFCs_ActInhib_PairCorrelations.png"),
       plot = p, height = 4, width = 3.75, bg = 'white')
print(p)
```

# Average rGenes analysis

## Reciprocal rGenes

```{r}
intersect_2activators_1inhibitor <- function(activator1, activator2, inhibitor, pathways){
  # helper to stringify gene vectors
  genes_to_string <- function(x) if (length(x) == 0) "" else paste(sort(unique(x)), collapse = "/")
  pattern <- paste(pathways, collapse = "|")

  # --- Activator 1: up ---
  activator1_up <- avg_response_df %>%
    dplyr::filter(
      perturbation == activator1,
      significant == TRUE,
      logFCsign == 1
    )

  filt_activator1_up <- activator1_up %>%
    dplyr::filter(!is.na(signaling_category) & grepl(pattern, signaling_category))

  # --- Activator 2: up ---
  activator2_up <- avg_response_df %>%
    dplyr::filter(
      perturbation == activator2,
      significant == TRUE,
      logFCsign == 1
    )

  filt_activator2_up <- activator2_up %>%
    dplyr::filter(!is.na(signaling_category) & grepl(pattern, signaling_category))

  # --- Inhibitor: down ---
  inhibitor_down <- avg_response_df %>%
    dplyr::filter(
      perturbation == inhibitor,
      significant == TRUE,
      logFCsign == -1
    )

  filt_inhibitor_down <- inhibitor_down %>%
    dplyr::filter(!is.na(signaling_category) & grepl(pattern, signaling_category))

  # --- Sets ---
  activators_up_genes        <- intersect(unique(activator1_up$ID), unique(activator2_up$ID))
  filt_activators_up_genes   <- intersect(unique(filt_activator1_up$ID), unique(filt_activator2_up$ID))
  inhibitor_down_genes       <- unique(inhibitor_down$ID)
  filt_inhibitor_down_genes  <- unique(filt_inhibitor_down$ID)
  reciprocal_genes           <- intersect(activators_up_genes, inhibitor_down_genes)
  filt_reciprocal_genes      <- intersect(filt_activators_up_genes, filt_inhibitor_down_genes)

  # --- Counts & percent ---
  n_activators_up                    <- length(activators_up_genes)
  n_filt_activators_up               <- length(filt_activators_up_genes)
  n_inhibitor_down                   <- length(inhibitor_down_genes)
  n_filt_inhibitor_down              <- length(filt_inhibitor_down_genes)
  n_reciprocal                       <- length(reciprocal_genes)
  n_filt_reciprocal                  <- length(filt_reciprocal_genes)
  pct_filt_reciprocal_of_reciprocal <- if (n_reciprocal == 0) NA_real_ else (n_filt_reciprocal / n_reciprocal) * 100

  # --- Summary table with genes column ---
  summary_tbl <- tibble::tibble(
    metric = c(
      "N activators up genes",
      "N signaling-associated activators up genes",
      "N inhibitor down genes",
      "N signaling-associated inhibitor down genes",
      "N reciprocal genes",
      "N signaling-associated reciprocal genes",
      "Signaling-associated reciprocal / reciprocal genes (%)"
    ),
    value = c(
      n_activators_up,
      n_filt_activators_up,
      n_inhibitor_down,
      n_filt_inhibitor_down,
      n_reciprocal,
      n_filt_reciprocal,
      pct_filt_reciprocal_of_reciprocal
    ),
    genes = c(
      genes_to_string(activators_up_genes),
      genes_to_string(filt_activators_up_genes),
      genes_to_string(inhibitor_down_genes),
      genes_to_string(filt_inhibitor_down_genes),
      genes_to_string(reciprocal_genes),
      genes_to_string(filt_reciprocal_genes),
      ""  # percentage row has no gene list
    )
  )

  return(summary_tbl)
}
```

```{r}
activator1 <- 'FGF2'
activator2 <- 'FGF8b'
inhibitor <- 'BGJ398'
pathways <- c('FGF signaling', 'FGF & MAPK ERK signaling', 'FGF & ERBB & MAPK ERK signaling', 'FGF')
summary_tbl <- intersect_2activators_1inhibitor(activator1, activator2, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator1 <- 'NRG1'
activator2 <- 'EGF'
inhibitor <- 'AZD8931'
pathways <- c('ERBB signaling', 'ERBB & MAPK ERK signaling', 'FGF & ERBB & MAPK ERK signaling', 'ERBB', 'MAPK ERK signaling')
summary_tbl <- intersect_2activators_1inhibitor(activator1, activator2, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator1 <- 'DLL1'
activator2 <- 'JAG1'
inhibitor <- 'LY411575'
pathways <- c('NOTCH signaling', 'NOTCH')
summary_tbl <- intersect_2activators_1inhibitor(activator1, activator2, inhibitor, pathways)
View(summary_tbl)
```

```{r}
intersect_1activator_1inhibitor <- function(activator, inhibitor, pathways){
  # helper to stringify gene vectors
  genes_to_string <- function(x) if (length(x) == 0) "" else paste(sort(unique(x)), collapse = "/")
  pattern <- paste(pathways, collapse = "|")

  # --- Activator 1: up ---
  activator_up <- avg_response_df %>%
    dplyr::filter(
      perturbation == activator,
      significant == TRUE,
      logFCsign == 1
    )

  filt_activator_up <- activator_up %>%
    dplyr::filter(!is.na(signaling_category) & grepl(pattern, signaling_category))

  # --- Inhibitor: down ---
  inhibitor_down <- avg_response_df %>%
    dplyr::filter(
      perturbation == inhibitor,
      significant == TRUE,
      logFCsign == -1
    )

  filt_inhibitor_down <- inhibitor_down %>%
    dplyr::filter(!is.na(signaling_category) & grepl(pattern, signaling_category))

  # --- Sets ---
  activator_up_genes        <- activator_up$ID
  filt_activator_up_genes   <- filt_activator_up$ID
  inhibitor_down_genes       <- unique(inhibitor_down$ID)
  filt_inhibitor_down_genes  <- unique(filt_inhibitor_down$ID)
  reciprocal_genes           <- intersect(activator_up_genes, inhibitor_down_genes)
  filt_reciprocal_genes      <- intersect(filt_activator_up_genes, filt_inhibitor_down_genes)

  # --- Counts & percent ---
  n_activator_up                    <- length(activator_up_genes)
  n_filt_activator_up               <- length(filt_activator_up_genes)
  n_inhibitor_down                   <- length(inhibitor_down_genes)
  n_filt_inhibitor_down              <- length(filt_inhibitor_down_genes)
  n_reciprocal                       <- length(reciprocal_genes)
  n_filt_reciprocal                  <- length(filt_reciprocal_genes)
  pct_filt_reciprocal_of_reciprocal <- if (n_reciprocal == 0) NA_real_ else (n_filt_reciprocal / n_reciprocal) * 100

  # --- Summary table with genes column ---
  summary_tbl <- tibble::tibble(
    metric = c(
      "N activator up genes",
      "N signaling-associated activator up genes",
      "N inhibitor down genes",
      "N signaling-associated inhibitor down genes",
      "N reciprocal genes",
      "N signaling-associated reciprocal genes",
      "Signaling-associated reciprocal / reciprocal genes (%)"
    ),
    value = c(
      n_activator_up,
      n_filt_activator_up,
      n_inhibitor_down,
      n_filt_inhibitor_down,
      n_reciprocal,
      n_filt_reciprocal,
      pct_filt_reciprocal_of_reciprocal
    ),
    genes = c(
      genes_to_string(activator_up_genes),
      genes_to_string(filt_activator_up_genes),
      genes_to_string(inhibitor_down_genes),
      genes_to_string(filt_inhibitor_down_genes),
      genes_to_string(reciprocal_genes),
      genes_to_string(filt_reciprocal_genes),
      ""  # percentage row has no gene list
    )
  )

  return(summary_tbl)
}
```

```{r}
activator <- 'CHIR99021'
inhibitor <- 'WNTC59'
pathways <- c('WNT signaling', 'WNT')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'BMP7'
inhibitor <- 'LDN193189'
pathways <- c('BMP signaling', 'BMP')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'ATRA'
inhibitor <- 'HX531'
pathways <- c('RA signaling', 'RA')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'SHH'
inhibitor <- 'SANT1'
pathways <- c('SHH signaling', 'SHH')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'IGF1LR3'
inhibitor <- 'Rapamycin'
pathways <- c('mTOR signaling', 'mTOR', 'IGF signaling', 'IGF')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'TRULI'
inhibitor <- 'Verteporfin'
pathways <- c('HIPPO signaling', 'HIPPO')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r inspect specific avg responses}
filt_df <- avg_response_df %>%
    filter(perturbation == 'JAG1'
           & significant == TRUE
           #& logFCsign == 1
           #& !(is.na(signaling_category))
          )

View(filt_df)
```

```{r define genes to label}
genes_to_label = list('BMP7' = c('ID3', 'ID3', 'BMP7', 'BMPR1A'),
                      'LDN193189' = c('ID3'),
                      'FGF2' = c('SPRY1', 'SPRY2', 'SPRY4', 'TNC', 'IER2'),
                      'FGF8b' = c('SPRY1', 'SPRY2', 'SPRY4', 'TNC', 'IER2'),
                      'BGJ398' = c('SPRY1', 'SPRY2', 'SPRY4', 'TNC', 'IER2'),
                      'NRG1' = c('GAREM1', 'SPRY1', 'SPRY2',  'DUSP4', 'DUSP6'),
                      'EGF' = c('GAREM1', 'SPRY1', 'SPRY2',  'DUSP4', 'DUSP6'),
                      'AZD8931' = c(),
                      'SHH' = c('PTCH1', 'PTCH2', 'GLI1'),
                      'SANT1' = c(),
                      'ATRA' = c(),
                      'HX531' = c('SREBF1', 'ABCA1'),
                      'CHIR99021' = c('AXIN2', 'NKD1', 'DKK1', 'ZNRF3', 'SULF1'),
                      'WNTC59' = c('AXIN2', 'LEF1', 'LGR5', 'ZNRF3', 'NKD1'),
                      'DLL1' = c('HEY2'),
                      'JAG1' = c('HEY2'),
                      'LY411575' = c('HEY1', 'HES1', 'NOTCH1', 'NOTCH3', 'LFNG'),
                      'TRULI' = c('WWC2', 'LATS2', 'WTIP', 'NUAK2'),
                      'Rapamycin' = c("EIF4EBP1",  "LAMTOR5", "GSK3B"),
                      'IGF1LR3' = c('GATA3')
                      )
```

```{r signaling category value counts}
filt_df <- avg_response_df %>% filter(significant == TRUE)
value_counts <- as.data.frame(table(avg_response_df$signaling_category))
colnames(value_counts) <- c("category", "count")
value_counts <- value_counts  %>%
  arrange(desc(count))
value_counts
```

## Average response scatter plot

```{r}
# Set perturbation order
pert_order <- c("BMP7","LDN193189","FGF2","FGF8b","BGJ398","NRG1","EGF","AZD8931","ATRA","HX531",
                "CHIR99021","WNTC59","SHH","SANT1","DLL1","JAG1","LY411575","IGF1LR3","Rapamycin","TRULI")

pert_pathway_list <- list(
  'BMP7'      = c('BMP'),
  'LDN193189' = c('BMP'),
  'FGF2'      = c('FGF','MAPK ERK'),
  'FGF8b'     = c('FGF','MAPK ERK'),
  'BGJ398'    = c('FGF','MAPK ERK'),
  'NRG1'      = c('ERBB','MAPK ERK'),
  'EGF'       = c('ERBB','MAPK ERK'),
  'AZD8931'   = c('ERBB','MAPK ERK'),
  'ATRA'      = c('RA'),
  'HX531'     = c('RA'),
  'CHIR99021' = c('WNT'),
  'WNTC59'    = c('WNT'),
  'SHH'       = c('SHH'),
  'SANT1'     = c('SHH'),
  'DLL1'      = c('NOTCH'),
  'JAG1'      = c('NOTCH'),
  'LY411575'  = c('NOTCH'),
  'IGF1LR3'   = c('mTOR'),
  'Rapamycin' = c('mTOR'),
  'TRULI'     = c('HIPPO')
)

# Jitter (uses n_distinct for robustness)
set.seed(1234)
jit_width <- 2 * dplyr::n_distinct(avg_response_df$perturbation) * 0.8

filt_df <- avg_response_df %>%
  filter(
    !(perturbation %in% c("Untreated","DMSO","TreMan","Verteporfin"))
  ) %>%
  mutate(
    perturbation = factor(perturbation, levels = pert_order)
  ) %>%
  rowwise() %>%
  mutate(
    # list-column of expected pathways for this perturbation
    expected_pathway = list(pert_pathway_list[[as.character(perturbation)]]),

    # TRUE if signaling_category contains ANY expected pathway (substring, case-insensitive)
    match_expected = {
      sc <- as.character(signaling_category)
      ep <- expected_pathway
      if (is.na(sc) || is.null(ep) || length(ep) == 0) {
        FALSE
      } else {
        any(map_lgl(ep, ~ str_detect(sc, regex(.x, ignore_case = TRUE))))
      }
    },

    legend = dplyr::case_when(
      !isTRUE(significant)                  ~ "Not significant",
      isTRUE(significant) & match_expected  ~ "Significant expected pathway",
      isTRUE(significant) & !match_expected ~ "Significant uncategorized"
    )
  ) %>%
  ungroup() %>%
  group_by(perturbation) %>%
  mutate(
    x_jit = runif(dplyr::n(), min = -jit_width/2, max = jit_width/2)
  ) %>%
  ungroup()

# Split by legend (keeps x_jit)
ns_df    <- dplyr::filter(filt_df, legend == "Not significant")
uncat_df <- dplyr::filter(filt_df, legend == "Significant uncategorized")
exp_df   <- dplyr::filter(filt_df, legend == "Significant expected pathway")

# y-range and helpers
y_rng <- range(filt_df$logFC, na.rm = TRUE)
y_min <- y_rng[1]
y_max <- y_rng[2]
dy    <- diff(y_rng)

# facet colors
levs <- levels(droplevels(filt_df$perturbation))  # drop unused levels if any

bg_df <- data.frame(
  perturbation = factor(levs, levels = levs),
  xmin = -Inf, xmax =  Inf, ymin = -Inf, ymax =  Inf,
  bg   = unname(pert_colors[levs])
)

# NA-safe labeller (falls back to black if color missing)
col_labeller <- function(labels) {
  sapply(labels, function(lb) {
    col <- tryCatch(pert_colors[[lb]], error = function(e) NA)
    if (is.null(col) || is.na(col) || length(col) == 0) col <- "black"
    sprintf("<span style='color:%s'><b>%s</b></span>", col, lb)
  }, USE.NAMES = FALSE)
}

# labels (guard against NULL list entries)
has_label <- function(pert, id) {
  labs <- genes_to_label[[pert]]
  !is.null(labs) && id %in% labs
}

label_df <- filt_df %>%
  mutate(perturbation_chr = as.character(perturbation)) %>%
  rowwise() %>%
  filter(has_label(perturbation_chr, ID)) %>%
  ungroup()

label_df_up   <- dplyr::filter(label_df, logFC > 0)
label_df_down <- dplyr::filter(label_df, logFC < 0)

library(ggrepel)
library(ggtext)

p <- ggplot(filt_df, aes(y = logFC, size = abs(logFC))) +
  # colored panel background
  geom_rect(
    data = bg_df,
    aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax, fill = bg),
    inherit.aes = FALSE, alpha = 0.3
  ) +
  scale_fill_identity(guide = "none") +

  # jittered points using precomputed x_jit
  geom_point(data = ns_df,    aes(x = x_jit, color = legend), alpha = 0.6, na.rm = TRUE) +
  geom_point(data = uncat_df, aes(x = x_jit, color = legend), alpha = 0.8, na.rm = TRUE) +
  geom_point(data = exp_df,   aes(x = x_jit, color = legend), alpha = 1.0, na.rm = TRUE) +

  # labels: anchor segments to jittered points with x = x_jit
  ggrepel::geom_text_repel(
    data = label_df_up,
    aes(x = x_jit, label = ID),
    size          = 2.5,
    box.padding   = 0.20,
    point.padding = 0.20,
    max.overlaps  = Inf,
    force         = 7,
    force_pull    = 3,
    direction     = "y",
    ylim          = c(y_max - 0.35*dy, y_max + 0.05*dy),
    xlim          = c(0.15, 0.98),
    hjust         = 0,
    min.segment.length = 0,
    segment.size  = 0.3,
    segment.color = "black",
    color         = "black",
    seed          = 1
  ) +
  ggrepel::geom_text_repel(
    data = label_df_down,
    aes(x = x_jit, label = ID),
    size          = 2.5,
    box.padding   = 0.20,
    point.padding = 0.20,
    max.overlaps  = Inf,
    force         = 7,
    force_pull    = 3,
    direction     = "y",
    ylim          = c(y_min - 0.05*dy, y_min + 0.15*dy),
    xlim          = c(-0.98, -0.15),
    hjust         = 1,
    min.segment.length = 0,
    segment.size  = 0.3,
    segment.color = "black",
    color         = "black",
    seed          = 2
  ) +

  facet_wrap(
    ~ perturbation, nrow = 1, scales = "fixed",
    labeller = as_labeller(col_labeller)
  ) +

  # keep swarm centered around 0; allow labels to spill slightly
  scale_x_continuous(limits = c(-1.2, 1.2), breaks = NULL, expand = ggplot2::expansion(mult = c(0.05, 0.05))) +
  scale_y_continuous(limits = c(y_rng[1]-0.20*dy, y_rng[2]+0.10*dy)) +
  coord_cartesian(clip = "off") +

  scale_color_manual(
    values = c(
      "Not significant"              = "grey70",
      "Significant uncategorized"    = "grey35",
      "Significant expected pathway" = "firebrick3"
    ),
    breaks = c("Not significant","Significant uncategorized","Significant expected pathway"),
    name = NULL
  ) +
  scale_size_continuous(name = expression("|Log"[2]*"FC|"), range = c(0.3, 3)) +
  labs(x = NULL, y = expression(Log[2]~"(Stimulus/Vehicle)")) +
  theme_bw() +
  theme(
    panel.grid       = element_blank(),
    panel.border     = element_blank(),
    panel.background = element_blank(),
    axis.line.y      = element_line(color = "black"),
    axis.line.x      = element_blank(),
    axis.text.x      = element_blank(),
    axis.ticks.x     = element_blank(),
    strip.background = element_blank(),
    strip.text       = ggtext::element_markdown(size = 18),
    legend.position   = "bottom",
    legend.direction  = "horizontal",
    legend.justification = "center"
  )

# Save
out_w <- 2 * length(unique(filt_df$perturbation))
ggsave(
  filename = file.path(fig_path, "scatterplot_AvgResponseLogFC_Jittered_all_v2.png"),
  plot = p, height = 6, width = out_w, bg = "white", dpi = 300
)
print(p)
```


```{r average response LogFC jittered scatter plot}
pert_order <- c("BMP7","LDN193189","FGF2","FGF8b","BGJ398","NRG1","EGF","AZD8931","ATRA","HX531",
                "CHIR99021","WNTC59","SHH","SANT1","DLL1","JAG1","LY411575","IGF1LR3","Rapamycin","TRULI")

pert_pathway_list <- c(
  'BMP7'      = 'BMP',
  'LDN193189' = 'BMP',
  'FGF2'      = 'FGF',
  'FGF8b'     = 'FGF',
  'BGJ398'    = 'FGF',
  'NRG1'      = 'ERBB',
  'EGF'       = 'ERBB',
  'AZD8931'   = 'ERBB',
  'ATRA'      = 'RA',
  'HX531'     = 'RA',
  'CHIR99021' = 'WNT',
  'WNTC59'    = 'WNT',
  'SHH'       = 'SHH',
  'SANT1'     = 'SHH',
  'DLL1'      = 'NOTCH',
  'JAG1'      = 'NOTCH',
  'LY411575'  = 'NOTCH',
  'IGF1LR3'   = 'mTOR',
  'Rapamycin' = 'mTOR',
  'TRULI'     = 'HIPPO'
)

# -----------------------------
# Params for jitter & reproducibility
# -----------------------------
set.seed(1234)         # reproducible jitter across runs
jit_width <- 2.5      # total horizontal spread allowed in each facet (x  [-0.475, 0.475])

# -----------------------------
# Prep data
# -----------------------------
filt_df <- avg_response_df %>%
  filter(
    !(perturbation %in% c("Untreated", "DMSO", "TreMan", "Verteporfin"))
  ) %>%
  mutate(
    perturbation     = factor(perturbation, levels = pert_order),
    expected_pathway = unname(pert_pathway_list[as.character(perturbation)]),
    legend = dplyr::case_when(
      !significant ~ "Not significant",
      significant & !is.na(expected_pathway) &
        !is.na(signaling_category) &
        stringr::str_detect(signaling_category, stringr::fixed(expected_pathway)) ~ "Significant expected pathway",
      significant ~ "Significant uncategorized"
    )
  ) %>%
  group_by(perturbation) %>%
  mutate(
    # Reproducible per-point horizontal jitter within each facet
    x_jit = runif(n(), min = -jit_width/2, max = jit_width/2)
  ) %>%
  ungroup()

# Split by legend (keeps x_jit)
ns_df    <- dplyr::filter(filt_df, legend == "Not significant")
uncat_df <- dplyr::filter(filt_df, legend == "Significant uncategorized")
exp_df   <- dplyr::filter(filt_df, legend == "Significant expected pathway")

# y-range and helpers
y_rng <- range(filt_df$logFC, na.rm = TRUE)
y_min <- y_rng[1]
y_max <- y_rng[2]
dy    <- diff(y_rng)

# facet colors
levs <- levels(droplevels(filt_df$perturbation))  # drop unused levels if any

bg_df <- data.frame(
  perturbation = factor(levs, levels = levs),
  xmin = -Inf, xmax =  Inf, ymin = -Inf, ymax =  Inf,
  bg   = unname(pert_colors[levs])
)

# NA-safe labeller (falls back to black if color missing)
col_labeller <- function(labels) {
  sapply(labels, function(lb) {
    col <- tryCatch(pert_colors[[lb]], error = function(e) NA)
    if (is.null(col) || is.na(col) || length(col) == 0) col <- "black"
    sprintf("<span style='color:%s'><b>%s</b></span>", col, lb)
  }, USE.NAMES = FALSE)
}

# labels (guard against NULL list entries)
has_label <- function(pert, id) {
  labs <- genes_to_label[[pert]]
  !is.null(labs) && id %in% labs
}

label_df <- filt_df %>%
  mutate(perturbation_chr = as.character(perturbation)) %>%
  rowwise() %>%
  filter(has_label(perturbation_chr, ID)) %>%
  ungroup()

label_df_up   <- dplyr::filter(label_df, logFC > 0)
label_df_down <- dplyr::filter(label_df, logFC < 0)


# -----------------------------
# Plot
# -----------------------------
p <- ggplot(filt_df, aes(y = logFC, size = abs(logFC))) +
  # colored panel background
  geom_rect(
    data = bg_df,
    aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax, fill = bg),
    inherit.aes = FALSE, alpha = 0.3
  ) +
  scale_fill_identity(guide = "none") +

  # jittered points using precomputed x_jit
  geom_point(data = ns_df,    aes(x = x_jit, color = legend), alpha = 0.6) +
  geom_point(data = uncat_df, aes(x = x_jit, color = legend), alpha = 0.8) +
  geom_point(data = exp_df,   aes(x = x_jit, color = legend), alpha = 1.0) +

  # labels: anchor segments to jittered points with x = x_jit
  ggrepel::geom_text_repel(
    data = label_df_up,
    aes(x = x_jit, label = ID),
    size          = 2.5,
    box.padding   = 0.50,
    point.padding = 0.50,
    max.overlaps  = Inf,
    force         = 10,
    force_pull    = 3,
    direction     = "y",
    # push towards top/right by restricting allowable region
    ylim          = c(y_max - 0.35*dy, y_max + 0.05*dy),
    xlim          = c(0.15, 0.98),
    hjust         = 0,
    min.segment.length = 0,
    segment.size  = 0.3,
    segment.color = "black",
    color         = "black",
    seed          = 1
  ) +
  ggrepel::geom_text_repel(
    data = label_df_down,
    aes(x = x_jit, label = ID),
    size          = 2.5,
    box.padding   = 0.50,
    point.padding = 0.50,
    max.overlaps  = Inf,
    force         = 10,
    force_pull    = 3,
    direction     = "y",
    # push towards bottom/left by restricting allowable region
    ylim          = c(y_min - 0.05*dy, y_min + 0.15*dy),
    xlim          = c(-0.98, -0.15),
    hjust         = 1,
    min.segment.length = 0,
    segment.size  = 0.3,
    segment.color = "black",
    color         = "black",
    seed          = 2
  ) +

  facet_wrap(
    ~ perturbation, nrow = 1, scales = "fixed",
    labeller = as_labeller(col_labeller)
  ) +

  # keep swarm centered around 0; allow labels to spill slightly
  scale_x_continuous(limits = c(-1.2, 1.2), breaks = NULL, expand = expansion(mult = c(0.05, 0.05))) +
  scale_y_continuous(limits = c(y_rng[1]-0.20*dy, y_rng[2]+0.10*dy)) +
  coord_cartesian(clip = "off") +

  scale_color_manual(
    values = c(
      "Not significant"              = "grey70",
      "Significant uncategorized"    = "grey35",
      "Significant expected pathway" = "firebrick3"
    ),
    breaks = c("Not significant","Significant uncategorized","Significant expected pathway"),
    name = NULL
  ) +
  scale_size_continuous(name = expression("|log"[2]*"FC|"), range = c(0.3, 3)) +
  labs(x = NULL, y = expression(Log[2]~"(Stimulus/Vehicle)")) +
  theme_bw() +
  theme(
    panel.grid       = element_blank(),
    panel.border     = element_blank(),
    panel.background = element_blank(),
    axis.line.y      = element_line(color = "black"),
    axis.line.x      = element_blank(),
    axis.text.x      = element_blank(),
    axis.ticks.x     = element_blank(),
    strip.background = element_blank(),
    strip.text       = element_markdown(),
    plot.title       = element_text(hjust = 0.5)
  )

# -----------------------------
# Save
# -----------------------------
out_w <- 1.5 * length(unique(filt_df$perturbation))
ggsave(
  filename = file.path(fig_path, "scatterplot_AvgResponseLogFC_Jittered_all_v1.png"),
  plot = p, height = 5, width = out_w, bg = "white", dpi = 300
)
print(p)
```

## Average rGenes GSEA GO:BP

Define a function for performing GSEA with a dreamlet results data frame as input and a GSEA results data frame as output.
```{r GSEA function}
gsea_df <- function(average_response_df) {
  
  # Initialize an empty list to store results
  gsea_list <- list()
  
  # Iterate through species x perturbation x time point
  for (pert in unique(average_response_df$perturbation)) {
    # Pull genes and their logFCs, rank by logFC, and create a named numeric vector for input
    gene_list <- average_response_df %>% filter(perturbation == pert) %>%
      ungroup() %>%
      dplyr::select(ID, logFC) %>%
      dplyr::arrange(desc(logFC)) %>%
      deframe()
        
    # Skip iteration if gene_list is empty
    if (length(gene_list) == 0) next  
    
    # Run GSEA using GO:BP terms
    gsea_go <- gseGO(
      geneList = gene_list,
      OrgDb = org.Hs.eg.db,
      keyType = "SYMBOL",  
      ont = "BP", 
      pvalueCutoff = 1,
      minGSSize = 5,
      eps = 0,
      verbose = FALSE,
      by = "fgsea"
      )
          
    # Check if GSEA returned results
    if (!is.null(gsea_go) && nrow(gsea_go@result) > 0) {
              
    # Format GSEA results data frame.
    gsea_results <- gsea_go@result %>%
      mutate(perturbation = pert)
        
    # Add results to list
    gsea_list[[pert]] <- gsea_results
    }
  }
  
  # Combine list into a data frame and return
  gsea_df <- bind_rows(gsea_list)
  return(gsea_df)
}
```

Perform GO:BP term GSEA of rGenes
```{r rGenes gsea_df}
avg_response_gsea_df <- gsea_df(avg_response_df)
head(avg_response_gsea_df)
```


# Sample size analysis

## N cells across species box plots

```{r box plot N telNECs per sample group by species}
# Load in counts data for sample groups
sg_counts_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/scanpy/telencephalon_final/NEMP25_Tel_v5_sample_group_TelNEC_counts.csv")
sg_counts_df <- sg_counts_df %>%
    dplyr::rename(N_cells = n) %>%
    mutate(species = factor(species, levels = c('Human', 'Chimp', 'Orangutan')))

# Plot a box plot of unique rGenes within perturbation per time point                                  
p <- ggplot(sg_counts_df, aes(x = species, y = N_cells, fill=species)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.15, alpha = 0.6, size = 1) +
  scale_fill_manual(values = species_colors, name = "Species") +
  labs(
    title = 'TelNEC\nSample Group Sizes',
    x = "",
    y = "N Cells"
  ) +
  theme_classic() +
  theme(
    plot.title = element_text(hjust=0.5, size=14),
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    legend.position = "none",
  )
ggsave(filename = paste0(fig_path, "boxplot_NtelNECs_BySpecies.pdf"),
       plot = p, height = 4, width = 2.75, bg = 'white', dpi=600)
print(p)
```

```{r box plot N all telencephalon cells per sample group by species}
# Load in counts data for sample groups
sg_counts_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/scanpy/telencephalon_final/NEMP25_Tel_v5_sample_group_AllCells_counts.csv")
sg_counts_df <- sg_counts_df %>%
    dplyr::rename(N_cells = n) %>%
    mutate(species = factor(species, levels = c('Human', 'Chimp', 'Orangutan')))

# Plot a box plot                               
p <- ggplot(sg_counts_df, aes(x = species, y = N_cells, fill=species)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.15, alpha = 0.6, size = 1) +
  scale_fill_manual(values = species_colors, name = "Species") +
  labs(
    title = 'All Telencephalon\nSample Group Sizes',
    x = "",
    y = "N Cells"
  ) +
  theme_classic() +
  theme(
    plot.title = element_text(hjust=0.5, size=14),
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    legend.position = "none",
  )
ggsave(filename = paste0(fig_path, "boxplot_NTelAllCells_BySpecies.pdf"),
       plot = p, height = 4, width = 2.75, bg = 'white', dpi=600)
print(p)
```

```{r}
# Load in counts data for perturbations
pert_counts_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/scanpy/telencephalon_final/NEMP25_Tel_v5_perturbation_AllCells_counts.csv")
pert_counts_df <- pert_counts_df %>%
  dplyr::rename(N_cells = n) %>%
  arrange(desc(N_cells)) %>%
  mutate(perturbation = factor(perturbation, levels=perturbation))

# Plot bar plot
p <- ggplot(pert_counts_df, aes(x = perturbation, y=N_cells, fill = perturbation)) +
  geom_col(color = 'black', linewidth = 0.25) +
    labs(
      title = 'Telencephalic Cells Stimulus Group Sizes',
      x = "",
      y = "N Cells") +
  scale_fill_manual(name = "Stimuli", values = pert_colors) +
  theme_minimal() +
  theme(
    panel.grid = element_blank(),
    axis.line = element_line(color = "black"),
    plot.title = element_text(hjust = 0.5),
    legend.title = element_text(),
    legend.position = "none",
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
  )
  

ggsave(filename = paste0(fig_path, "barplot_NTelAllCells_ByPerturbation.pdf"), plot = p, height = 4, width = 5, bg = "white", dpi = 600)
print(p)
```


## N TelNECs across time points box plot

```{r box plot N telNECs across time points by sample group}
# Load in counts data for sample groups
sg_counts_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/scanpy/telencephalon_final/NEMP25_Tel_v5_sample_group_TelNEC_counts.csv")
sg_counts_df <- sg_counts_df %>% filter(!(perturbation %in% c('TreMan', 'Untreated', 'DMSO')) & species != 'Orangutan') %>%
    dplyr::rename(N_cells = n) %>%
    mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Plot a box plot of unique rGenes within perturbation per time point                                  
p <- ggplot(sg_counts_df, aes(x = time_point, y = N_cells, fill=time_point)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.15, alpha = 0.6, size = 1) +
  scale_fill_manual(values = time_point_colors, name = "Time point") +
  labs(
    x = "",
    y = "N TelNECs"
  ) +
  theme_classic() +
  theme(
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    legend.position = "none",
  )
ggsave(filename = paste0(fig_path, "boxplot_NtelNECs_ByTimePoint.png"),
       plot = p, height = 4, width = 2.25, bg = 'white', dpi=600)
print(p)
```

```{r box plot N all cells across time points by sample group}
# Load in counts data for sample groups
sg_counts_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/scanpy/telencephalon_final/NEMP25_Tel_v5_sample_group_AllCells_counts.csv")
sg_counts_df <- sg_counts_df %>% filter(!(perturbation %in% c('TreMan', 'Untreated', 'DMSO')) & species != 'Orangutan') %>%
    dplyr::rename(N_cells = n) %>%
    mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Plot a box plot of unique rGenes within perturbation per time point                                  
p <- ggplot(sg_counts_df, aes(x = time_point, y = N_cells, fill=time_point)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.15, alpha = 0.6, size = 1) +
  scale_fill_manual(values = time_point_colors, name = "Time point") +
  labs(
    x = "",
    y = "N Telencephalic Cells"
  ) +
  theme_classic() +
  theme(
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    legend.position = "none",
  )
ggsave(filename = paste0(fig_path, "boxplot_NTelencephalicCells_ByTimePoint.png"),
       plot = p, height = 4, width = 2.25, bg = 'white', dpi=600)
print(p)
```

# rGene analysis

## N rGenes vs N TelNECs scatter plots

```{r scatter plot N rGenes vs N TelNECs by perturbation}
# Load in counts data for perturbations
pert_counts_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/scanpy/telencephalon_final/NEMP25_Tel_v5_perturbation_TelNEC_counts.csv")
pert_counts_df <- pert_counts_df %>% filter(!(perturbation %in% c('TreMan', 'Untreated', 'DMSO'))) %>%
    dplyr::rename(N_cells = n)

# Format rGene data frame
pert_rgenes_df <- simple_rgenes_df %>% dplyr::select(perturbation, N_all_rGenes) %>%
  dplyr::rename(N_rGenes = N_all_rGenes)
  
# Join rGene and TelNEC counts data frames
pert_gene_cell_counts_df <- dplyr::left_join(pert_counts_df, pert_rgenes_df, by = c('perturbation'))

# Calculate Spearman's correlation coefficient
spearman_res <- cor.test(
  pert_gene_cell_counts_df$N_cells,
  pert_gene_cell_counts_df$N_rGenes,
  method = "spearman",
  exact = FALSE
)
spearman_rho <- unname(spearman_res$estimate)
spearman_p <- spearman_res$p.value

# Add padding
x_pad <- diff(range(pert_gene_cell_counts_df$N_cells, na.rm = TRUE)) * 0.025
y_pad <- diff(range(pert_gene_cell_counts_df$N_rGenes, na.rm = TRUE)) * 0.1

# Plot a scatter plot of N rGenes/pert vs. N TelNECs/pert
p <- ggplot(pert_gene_cell_counts_df, aes(x = N_cells, y = N_rGenes)) +
  geom_point(shape = 23, size = 4, color = "black", fill='black', stroke = 0.5, show.legend = TRUE) +
  labs(
    x = "N Cells",
    y = "N rGenes",
    fill = "Stimulus"
  ) +
  #scale_fill_manual(
    #name = "Stimulus",
    #values = pert_colors,
    #guide = guide_legend(ncol=2, title.position='top', override.aes = list(shape = 21, color = "black", stroke = 0.5))
  #) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    legend.position = "right",
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.title.align = 0.5,
    axis.title = element_text(size = 14),
    axis.text  = element_text(size = 12),
    legend.title = element_text(size=14),
    legend.text = element_text(size=12)
  ) +
  annotate(
    "text",
    x = min(pert_gene_cell_counts_df$N_cells, na.rm = TRUE) + x_pad,
    y = max(pert_gene_cell_counts_df$N_rGenes, na.rm = TRUE) - y_pad,
    hjust = 0, vjust = 1,
    label = sprintf("rho = %.2f\np = %.2g", spearman_rho, spearman_p),
    size = 5
  )

print(p)
ggsave(filename = paste0(fig_path, "ScatterPlot_NCells_vs_NrGenes_ByPerturbation.png"), plot = p, units='in', height = 4, width = 4, bg = "white", dpi=600)
```

```{r scatter plot N rGenes vs N TelNECs by sample group}
# Load in counts data for sample groups
sg_counts_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/scanpy/telencephalon_final/NEMP25_Tel_v5_sample_group_TelNEC_counts.csv")
sg_counts_df <- sg_counts_df %>% filter(!(perturbation %in% c('TreMan', 'Untreated', 'DMSO')) & species != 'Orangutan') %>%
    dplyr::rename(N_cells = n)

# Generate rGene data frame for sample groups
sg_rgenes_df <- response_df %>% filter(significant == TRUE & !(perturbation %in% c('TreMan', 'Untreated', 'DMSO')) & species != 'Orangutan') %>% 
  dplyr::group_by(species, perturbation, time_point) %>%
  dplyr::summarise(N_rGenes = sum(!is.na(ID) & ID != ""), .groups = "drop") %>%
  complete(species, perturbation, time_point, fill = list(N_rGenes = 0))
  
# Join rGene and TelNEC counts data frames
sg_gene_cell_counts_df <- dplyr::left_join(sg_counts_df, sg_rgenes_df, by = c('species', 'perturbation', 'time_point')) %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Calculate Spearman's correlation coefficient
spearman_res <- cor.test(
  sg_gene_cell_counts_df$N_cells,
  sg_gene_cell_counts_df$N_rGenes,
  method = "spearman",
  exact = FALSE
)
spearman_rho <- unname(spearman_res$estimate)
spearman_p <- spearman_res$p.value

# Add padding
x_pad <- diff(range(sg_gene_cell_counts_df$N_cells, na.rm = TRUE)) * 0.01
y_pad <- diff(range(sg_gene_cell_counts_df$N_rGenes, na.rm = TRUE)) * 0.01

# Plot a scatter plot of N rGenes/sample group vs. N TelNECs/sample group
p <- ggplot(sg_gene_cell_counts_df, aes(x = N_cells, y = N_rGenes, shape = time_point)) +
  geom_point(size = 4, color = "black", fill='black', stroke = 0.5) +
  labs(x = "N Cells", y = "N rGenes") +
  #scale_fill_manual(
    #name = 'Stimulus',
    #values = pert_colors,
    #guide = guide_legend(ncol=2, title.position='top', override.aes = list(shape = 21, color = "black", stroke = 0.5))) +
  scale_shape_manual(
    name = "Time Point",
    values = time_point_shapes,
    guide = guide_legend(override.aes = list(fill = "black"))
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title = element_text(size = 14),
    axis.text  = element_text(size = 12),
    legend.position = "right"
  ) +
  annotate(
    "text",
    x = min(sg_gene_cell_counts_df$N_cells, na.rm = TRUE) + x_pad,
    y = max(sg_gene_cell_counts_df$N_rGenes, na.rm = TRUE) - y_pad,
    hjust = 0, vjust = 1,
    label = sprintf("rho = %.2f\np = %.2g", spearman_rho, spearman_p),
    size = 5
  )

print(p)
ggsave(filename = paste0(fig_path, "ScatterPlot_NCells_vs_NrGenes_BySampleGroup.png"), plot = p, units='in', height = 4, width = 5, bg = "white", dpi=600)
```

## N rGenes per perturbation bar plot

```{r bar plot N rGenes by perturbation}
sum_rgenes_df <- rgenes_df %>%
  filter(!(perturbation %in% c('Untreated', 'DMSO', 'TreMan'))) %>%
  group_by(perturbation) %>%
  summarise(N_rGenes = sum(N_all_rGenes, na.rm = TRUE), .groups = "drop") %>%
  arrange(desc(N_rGenes)) %>%
  mutate(perturbation = factor(perturbation, levels = perturbation))

p <- ggplot(sum_rgenes_df, aes(x = perturbation, y = N_rGenes, fill = perturbation)) +
  geom_col(color = 'black', linewidth = 0.25) +
  labs(
    title = 'Stimuli Ranked by N rGenes',
    x = "",
    y = "Sum of rGenes Across Time Points") +
  scale_fill_manual(name = "Stimuli", values = pert_colors) +
  theme_minimal() +
  theme(
    panel.grid = element_blank(),
    axis.line = element_line(color = "black"),
    plot.title = element_text(hjust = 0.5),
    legend.title = element_text(),
    legend.position = "none",
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
  )
ggsave(filename = paste0(fig_path, "barplot_NTotalrGenes.png"), plot = p, height = 4, width = 8, bg = 'white', dpi=600)

print(p)
```

## N rGenes across time points box plot

```{r box plot N rGenes across time points}
rgenes_df <- rgenes_df %>% mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))
                                  
p <- ggplot(rgenes_df, aes(x = time_point, y = N_all_rGenes, fill=time_point)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.15, alpha = 0.6, size = 1) +
  scale_fill_manual(values = time_point_colors, name = "Time point") +
  labs(
    x = "",
    y = "N rGenes"
  ) +
  theme_classic() +
  theme(
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
  )
ggsave(filename = paste0(fig_path, "boxplot_NrGenes_ByTimePoint.png"), plot = p, height = 4, width = 3, bg = 'white', dpi=600)
print(p)
```

## N unique rGenes across time points box plot

```{r box plot N unique rGenes across time points}
# Find number of rGenes within a perturbation that are unique to a time point
unique_rgenes_df <- response_df %>% filter(significant==TRUE) %>%
  dplyr::select(perturbation, time_point, ID) %>%
  group_by(perturbation, ID) %>%
  mutate(n_timepoints_in_pert = n_distinct(time_point)) %>%
  ungroup() %>%
  filter(n_timepoints_in_pert == 1) %>%                  # unique within perturbation across time points
  dplyr::count(perturbation, time_point, name = "N_unique_rGenes") %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Plot a box plot of unique rGenes within perturbation per time point                                  
p <- ggplot(unique_rgenes_df, aes(x = time_point, y = N_unique_rGenes, fill=time_point)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.15, alpha = 0.6, size = 1) +
  scale_fill_manual(values = time_point_colors, name = "Time point") +
  labs(
    x = "",
    y = "N Unique rGenes (Within Stimulus)"
  ) +
  theme_classic() +
  theme(
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    legend.position = "none",
  )
ggsave(filename = paste0(fig_path, "boxplot_NUniquerGenes_AcrossTimePointsWithinPerturbation_ByTimePoint.png"),
       plot = p, height = 4, width = 2.25, bg = 'white', dpi=600)
print(p)
```

## N rGenes over time by perturbation connected scatter plot

```{r scatter plot log2 N rGenes over time}
# Set factor levels and filter to data of interest
filt_df <- rgenes_df %>%
  mutate(
    log2NrGenes = log2(N_all_rGenes + 1),
    time_point = factor(time_point, levels = c("6hr", "54hr", "7day"))
  )

# Plot scatter plot across time points
p <- ggplot(filt_df, aes(x = time_point, y = N_all_rGenes, color = perturbation, group = perturbation)) +
  geom_point(size = 3) + # Scatter plot.
  geom_line(size = 1) +  # Connecting lines.
  labs(
    x = "Time Point",
    y = "N rGenes",
    color = "Perturbation"
    ) +
    scale_color_manual(values = pert_colors) +  # Add custom color map.
    scale_x_discrete(expand = c(0.02, 0.02)) +  # Edit padding around the x-axis
    theme_minimal() +                     # Clean theme
    theme(plot.title = element_text(hjust=0.5, size=16), # Edit plot title.
        plot.background = element_blank(),  # Remove the background around the plot.
        panel.background = element_blank(),  # Remove the background inside the plot.
        panel.grid.major = element_blank(),  # Remove major gridlines
        panel.grid.minor = element_blank(),   # Remove minor gridlines
        axis.text  = element_text(size = 12),
        axis.title = element_text(size = 14),
        legend.position = "right", # Edit legend location.
        axis.line = element_line(color = "black", size = 0.5),  # Add axis lines
        axis.ticks = element_line(color = "black", size = 0.5)  # Add axis ticks
    )
ggsave(filename = paste0(fig_path, "connectedscatterplot_NrGenes_ByPerturbation_ByTimePoint.png"),
         plot = p, height = 5, width = 7, bg = 'white')
print(p)
```

## N unique rGenes across perturbations by perturbation bar plot

```{r bar plot mean pairwise E-distance by perturbation}
# Find number of rGenes that are unique to each perturbation
pert_unique_rgenes_df <- response_df %>% filter(significant == TRUE & !(perturbation %in% c('Untreated', 'DMSO', 'TreMan'))) %>%
  dplyr::select(perturbation, ID) %>%
  distinct() %>% 
  add_count(ID, name = "n_perts") %>%      # in how many perturbations does this gene appear?
  filter(n_perts == 1) %>%                 # keep genes unique to a single perturbation
  dplyr::count(perturbation, name = "N_unique_rGenes") %>%
  arrange(N_unique_rGenes) %>%
  mutate(perturbation = factor(perturbation, levels = perturbation))

p <- ggplot(pert_unique_rgenes_df, aes(x = N_unique_rGenes, y = perturbation, fill = perturbation)) +
  geom_col(color = 'black', linewidth = 0.25) +
  labs(
    title = 'Stimuli Ranked by \nN Unique rGenes',
    y = "",
    x = "N Unique rGenes\nAcross All Stimuli") +
  scale_fill_manual(name = "Stimulus", values = pert_colors) +
  theme_minimal() +
  theme(
    panel.grid = element_blank(),
    axis.line = element_line(color = "black"),
    plot.title = element_text(hjust = 0.5),
    legend.title = element_text(),
    legend.position = "none",
    axis.text = element_text(size=12),
    axis.title = element_text(size=14),
    title.text = element_text(size=16)
  )
ggsave(filename = paste0(fig_path, "barplot_NUniquerGenes_ByPerturbation.png"),
       plot = p, height = 9, width = 4, bg = 'white', dpi=600, units='in')

print(p)
```

## N rGenes Human vs Chimp

```{r N rGenes human vs chimp scatter plot by sample group}
# Filter data
filt_df <- rgenes_df %>%
  mutate(time_point = factor(time_point, levels = c("6hr", "54hr", "7day")))

# Calculate Pearson's correlation coefficient
pearson_res <- cor.test(filt_df$N_all_human_rGenes, filt_df$N_all_chimp_rGenes, method = "pearson")
pearson_r <- unname(pearson_res$estimate)
pearson_p <- pearson_res$p.value

# Make square limits: use the same min/max for both axes
xy_min <- min(c(filt_df$N_all_human_rGenes, filt_df$N_all_chimp_rGenes), na.rm = TRUE)
xy_max <- max(c(filt_df$N_all_human_rGenes, filt_df$N_all_chimp_rGenes), na.rm = TRUE)
pad    <- (xy_max - xy_min) * 0.05
if (!is.finite(pad) || pad == 0) pad <- 1  # safe pad for count data

lims <- c(xy_min - pad, xy_max + pad)

# Plot scatter plot
p <- ggplot(filt_df, aes(x = N_all_human_rGenes, y = N_all_chimp_rGenes, shape = time_point, fill = perturbation)) +
  geom_point(size = 4, color = "black", stroke = 0.5, show.legend = TRUE) +
  labs(
    x = "Human N rGenes",
    y = "Chimp N rGenes",
    fill = "Stimulus",
    shape = "Time Point"
  ) +
  scale_fill_manual(
    name = "Stimulus", values = pert_colors,
    guide = guide_legend(ncol = 2, override.aes = list(shape = 21, color = "black", stroke = 0.5))
  ) +
  scale_shape_manual(
    name = "Time Point", values = time_point_shapes,
    guide = guide_legend(override.aes = list(fill = "black"))
  ) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
  annotate(
    "text",
    x = lims[1] + pad,
    y = lims[2] - pad,
    hjust = 0, vjust = 1,
    label = sprintf("r = %.2f\np = %.2g", pearson_r, pearson_p),
    size = 5
  ) +
  coord_fixed(ratio = 1, xlim = lims, ylim = lims, expand = FALSE) +
  theme_minimal() +
  theme(
    legend.position = "right",
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14)
  )

ggsave(
  filename = paste0(fig_path, "ScatterPlot_NrGenes_HumanVsChimp.png"),
  plot = p, height = 4.75, width = 6, bg = "white", units = "in", dpi = 600
)

print(p)
```

## rGenes upset plot

Plot an upset plot of overlapping rGenes 
```{r upset plot simple rGenes}
binary_df <- simple_rgenes_df %>%
  dplyr::filter(!perturbation %in% c("TreMan", "DMSO")) %>%
  tidyr::separate_rows(all_rGenes, sep = "/") %>%
  dplyr::mutate(all_rGenes = trimws(all_rGenes)) %>%
  dplyr::filter(!is.na(all_rGenes), all_rGenes != "") %>%
  dplyr::distinct(gene = all_rGenes, perturbation) %>%
  dplyr::mutate(present = 1L) %>%
  tidyr::pivot_wider(
    names_from  = perturbation,
    values_from = present,
    values_fill = 0L
  )

set_cols <- setdiff(names(binary_df), "gene")

# --- NEW: compute intersections with > 25 genes ---
mat <- as.matrix(binary_df[, set_cols, drop = FALSE])

# Build a key per gene that identifies the exact intersection (e.g., "BMP7&FGF2")
intersection_key <- apply(mat, 1, function(v) {
  on <- set_cols[as.logical(v)]
  if (length(on) == 0) return(NA_character_)  # ignore genes in no sets
  paste(on, collapse = "&")
})

intersection_counts <- sort(table(intersection_key, useNA = "no"), decreasing = TRUE)

# Keep only intersection patterns with > 25 genes
keep_keys <- names(intersection_counts)[intersection_counts > 25]

# Convert to UpSetR format: list of character vectors
intersections_keep <- lapply(keep_keys, function(k) strsplit(k, "&", fixed = TRUE)[[1]])

# Optional: if you only want TRUE intersections (>=2 sets), uncomment:
# intersections_keep <- intersections_keep[vapply(intersections_keep, length, integer(1)) >= 2]

png(
  paste0(fig_path, "upsetplot_simple_rGenes.png"),
  width = 24, height = 8, units = "in", res = 600
)

if (length(intersections_keep) == 0) {
  warning("No intersections with > 25 genes found ensured; plotting default UpSetR output instead.")
  UpSetR::upset(
    as.data.frame(binary_df),
    sets = set_cols,
    keep.order = FALSE,
    sets.bar.color = "gray20",
    order.by = "freq",
    mb.ratio = c(0.35, 0.65),
    sets.x.label = "Set Size",
    mainbar.y.label = "Intersection Size",
    decreasing = c(TRUE, NA)
  )
} else {
  UpSetR::upset(
    as.data.frame(binary_df),
    sets = set_cols,
    intersections = intersections_keep,  # <-- restrict plotted intersections here
    keep.order = FALSE,
    sets.bar.color = "gray20",
    order.by = "freq",
    mb.ratio = c(0.35, 0.65),
    sets.x.label = "Set Size",
    mainbar.y.label = "Intersection Size",
    decreasing = c(TRUE, NA),
    text.scale = c(2.5, #intersection_size_title,   
                   2.0, #intersection_size_ticks,
                   2.5, #set_size_title,
                   2.0, #set_size_ticks,
                   2.0  #set_names
    ),
    show.numbers = FALSE
  )
}
dev.off()
```






## N unique rGenes vs. mean pairwise E-distance scatter plot by perturbation

```{r}
# Join data frames
pert_distinctiveness_df <- left_join(pert_mp_edist_df, pert_unique_rgenes_df, by='perturbation')

# Calculate Spearman's correlation coefficient
spearman_res <- cor.test(pert_distinctiveness_df$mp_edist, pert_distinctiveness_df$N_unique_rGenes, method = "spearman", exact = FALSE)
spearman_rho <- unname(spearman_res$estimate)
spearman_p   <- spearman_res$p.value

# Plot scatter plot of E-distance vs N rGenes for each perturbation
p <- ggplot(pert_distinctiveness_df, aes(x = mp_edist, y = N_unique_rGenes, fill = perturbation)) +
  geom_point(shape = 23, size = 4, color = "black", stroke = 0.5, show.legend = TRUE) +
  labs(
    x = "Mean Pairwise E-Distance",
    y = "N Unique rGenes",
    fill = "Stimulus"
  ) +
  scale_fill_manual(
    name = "Stimulus",
    values = pert_colors,
    guide = guide_legend(ncol = 2, override.aes = list(shape = 21, color = "black", stroke = 0.5))
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    legend.position = "none",
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title = element_text(size=14),
    axis.text = element_text(size=12),
    legend.title = element_text(size=14),
    legend.text = element_text(size=12),
  ) +
  annotate(
    "text",
    x = min(pert_distinctiveness_df$mp_edist, na.rm = TRUE),
    y = max(pert_distinctiveness_df$N_unique_rGenes, na.rm = TRUE),
    hjust = 0, vjust = 1,
    label = sprintf("rho = %.2f\np = %.2g", spearman_rho, spearman_p),
    size = 4
  )

ggsave(filename = paste0(fig_path, "ScatterPlot_MeanPairwiseEdist_vs_NUniquerGenes_ByPerturbation.png"),
       plot = p, height = 4, width = 4, bg = "white", dpi = 600, units='in')

print(p)

```

## Reciprocal rGenes

```{r}
intersect_2activators_1inhibitor <- function(activator1, activator2, inhibitor, pathways){
  # helper to stringify gene vectors
  genes_to_string <- function(x) if (length(x) == 0) "" else paste(sort(unique(x)), collapse = "/")
  pattern <- paste(pathways, collapse = "|")

  # --- Activator 1: up ---
  activator1_up <- response_df %>%
    dplyr::filter(
      perturbation == activator1,
      significant == TRUE,
      logFCsign == 1
    )

  filt_activator1_up <- activator1_up %>%
    dplyr::filter(!is.na(signaling_category) & grepl(pattern, signaling_category))

  # --- Activator 2: up ---
  activator2_up <- response_df %>%
    dplyr::filter(
      perturbation == activator2,
      significant == TRUE,
      logFCsign == 1
    )

  filt_activator2_up <- activator2_up %>%
    dplyr::filter(!is.na(signaling_category) & grepl(pattern, signaling_category))

  # --- Inhibitor: down ---
  inhibitor_down <- response_df %>%
    dplyr::filter(
      perturbation == inhibitor,
      significant == TRUE,
      logFCsign == -1
    )

  filt_inhibitor_down <- inhibitor_down %>%
    dplyr::filter(!is.na(signaling_category) & grepl(pattern, signaling_category))

  # --- Sets ---
  activators_up_genes        <- intersect(unique(activator1_up$ID), unique(activator2_up$ID))
  filt_activators_up_genes   <- intersect(unique(filt_activator1_up$ID), unique(filt_activator2_up$ID))
  inhibitor_down_genes       <- unique(inhibitor_down$ID)
  filt_inhibitor_down_genes  <- unique(filt_inhibitor_down$ID)
  reciprocal_genes           <- intersect(activators_up_genes, inhibitor_down_genes)
  filt_reciprocal_genes      <- intersect(filt_activators_up_genes, filt_inhibitor_down_genes)

  # --- Counts & percent ---
  n_activators_up                    <- length(activators_up_genes)
  n_filt_activators_up               <- length(filt_activators_up_genes)
  n_inhibitor_down                   <- length(inhibitor_down_genes)
  n_filt_inhibitor_down              <- length(filt_inhibitor_down_genes)
  n_reciprocal                       <- length(reciprocal_genes)
  n_filt_reciprocal                  <- length(filt_reciprocal_genes)
  pct_filt_reciprocal_of_reciprocal <- if (n_reciprocal == 0) NA_real_ else (n_filt_reciprocal / n_reciprocal) * 100

  # --- Summary table with genes column ---
  summary_tbl <- tibble::tibble(
    metric = c(
      "N activators up genes",
      "N signaling-associated activators up genes",
      "N inhibitor down genes",
      "N signaling-associated inhibitor down genes",
      "N reciprocal genes",
      "N signaling-associated reciprocal genes",
      "Signaling-associated reciprocal / reciprocal genes (%)"
    ),
    value = c(
      n_activators_up,
      n_filt_activators_up,
      n_inhibitor_down,
      n_filt_inhibitor_down,
      n_reciprocal,
      n_filt_reciprocal,
      pct_filt_reciprocal_of_reciprocal
    ),
    genes = c(
      genes_to_string(activators_up_genes),
      genes_to_string(filt_activators_up_genes),
      genes_to_string(inhibitor_down_genes),
      genes_to_string(filt_inhibitor_down_genes),
      genes_to_string(reciprocal_genes),
      genes_to_string(filt_reciprocal_genes),
      ""  # percentage row has no gene list
    )
  )

  return(summary_tbl)
}
```

```{r}
activator1 <- 'FGF2'
activator2 <- 'FGF8b'
inhibitor <- 'BGJ398'
pathways <- c('FGF signaling', 'FGF & MAPK ERK signaling', 'FGF & ERBB & MAPK ERK signaling', 'FGF')
summary_tbl <- intersect_2activators_1inhibitor(activator1, activator2, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator1 <- 'NRG1'
activator2 <- 'EGF'
inhibitor <- 'AZD8931'
pathways <- c('ERBB signaling', 'ERBB & MAPK ERK signaling', 'FGF & ERBB & MAPK ERK signaling', 'ERBB', 'MAPK ERK signaling')
summary_tbl <- intersect_2activators_1inhibitor(activator1, activator2, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator1 <- 'DLL1'
activator2 <- 'JAG1'
inhibitor <- 'LY411575'
pathways <- c('NOTCH signaling', 'NOTCH')
summary_tbl <- intersect_2activators_1inhibitor(activator1, activator2, inhibitor, pathways)
View(summary_tbl)
```

```{r}
intersect_1activator_1inhibitor <- function(activator, inhibitor, pathways){
  # helper to stringify gene vectors
  genes_to_string <- function(x) if (length(x) == 0) "" else paste(sort(unique(x)), collapse = "/")
  pattern <- paste(pathways, collapse = "|")

  # --- Activator 1: up ---
  activator_up <- response_df %>%
    dplyr::filter(
      perturbation == activator,
      significant == TRUE,
      logFCsign == 1
    )

  filt_activator_up <- activator_up %>%
    dplyr::filter(!is.na(signaling_category) & grepl(pattern, signaling_category))

  # --- Inhibitor: down ---
  inhibitor_down <- response_df %>%
    dplyr::filter(
      perturbation == inhibitor,
      significant == TRUE,
      logFCsign == -1
    )

  filt_inhibitor_down <- inhibitor_down %>%
    dplyr::filter(!is.na(signaling_category) & grepl(pattern, signaling_category))

  # --- Sets ---
  activator_up_genes        <- activator_up$ID
  filt_activator_up_genes   <- filt_activator_up$ID
  inhibitor_down_genes       <- unique(inhibitor_down$ID)
  filt_inhibitor_down_genes  <- unique(filt_inhibitor_down$ID)
  reciprocal_genes           <- intersect(activator_up_genes, inhibitor_down_genes)
  filt_reciprocal_genes      <- intersect(filt_activator_up_genes, filt_inhibitor_down_genes)

  # --- Counts & percent ---
  n_activator_up                    <- length(activator_up_genes)
  n_filt_activator_up               <- length(filt_activator_up_genes)
  n_inhibitor_down                   <- length(inhibitor_down_genes)
  n_filt_inhibitor_down              <- length(filt_inhibitor_down_genes)
  n_reciprocal                       <- length(reciprocal_genes)
  n_filt_reciprocal                  <- length(filt_reciprocal_genes)
  pct_filt_reciprocal_of_reciprocal <- if (n_reciprocal == 0) NA_real_ else (n_filt_reciprocal / n_reciprocal) * 100

  # --- Summary table with genes column ---
  summary_tbl <- tibble::tibble(
    metric = c(
      "N activator up genes",
      "N signaling-associated activator up genes",
      "N inhibitor down genes",
      "N signaling-associated inhibitor down genes",
      "N reciprocal genes",
      "N signaling-associated reciprocal genes",
      "Signaling-associated reciprocal / reciprocal genes (%)"
    ),
    value = c(
      n_activator_up,
      n_filt_activator_up,
      n_inhibitor_down,
      n_filt_inhibitor_down,
      n_reciprocal,
      n_filt_reciprocal,
      pct_filt_reciprocal_of_reciprocal
    ),
    genes = c(
      genes_to_string(activator_up_genes),
      genes_to_string(filt_activator_up_genes),
      genes_to_string(inhibitor_down_genes),
      genes_to_string(filt_inhibitor_down_genes),
      genes_to_string(reciprocal_genes),
      genes_to_string(filt_reciprocal_genes),
      ""  # percentage row has no gene list
    )
  )

  return(summary_tbl)
}
```

```{r}
activator <- 'EGF'
inhibitor <- 'AZD8931'
pathways <- c('ERBB signaling', 'ERBB & MAPK ERK signaling', 'FGF & ERBB & MAPK ERK signaling', 'ERBB', 'MAPK ERK signaling')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'DLL1'
inhibitor <- 'LY411575'
pathways <- c('NOTCH signaling', 'NOTCH')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'JAG1'
inhibitor <- 'LY411575'
pathways <- c('NOTCH signaling', 'NOTCH')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'NRG1'
inhibitor <- 'AZD8931'
pathways <- c('ERBB signaling', 'ERBB & MAPK ERK signaling', 'FGF & ERBB & MAPK ERK signaling', 'ERBB', 'MAPK ERK signaling')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'CHIR99021'
inhibitor <- 'WNTC59'
pathways <- c('WNT signaling', 'WNT')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'BMP7'
inhibitor <- 'LDN193189'
pathways <- c('BMP signaling', 'BMP')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'ATRA'
inhibitor <- 'HX531'
pathways <- c('RA signaling', 'RA')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'SHH'
inhibitor <- 'SANT1'
pathways <- c('SHH signaling', 'SHH')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'IGF1LR3'
inhibitor <- 'Rapamycin'
pathways <- c('mTOR signaling', 'mTOR', 'IGF signaling', 'IGF')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

```{r}
activator <- 'TRULI'
inhibitor <- 'Verteporfin'
pathways <- c('HIPPO signaling', 'HIPPO')
summary_tbl <- intersect_1activator_1inhibitor(activator, inhibitor, pathways)
View(summary_tbl)
```

## Total rGenes scatter plot across perturbations

```{r define genes to label}
genes_to_label = list('BMP7' = c('ID3'),
                      'LDN193189' = c('ID3'),
                      'FGF2' = c('SPRY1'),
                      'FGF8b' = c('SPRY1'),
                      'BGJ398' = c('SPRY1'),
                      'NRG1' = c('GAREM1'),
                      'EGF' = c('GAREM1'),
                      'AZD8931' = c(),
                      'SHH' = c('PTCH1'),
                      'SANT1' = c('PTCH1'),
                      'ATRA' = c('RARA'),
                      'HX531' = c('SREBF1'),
                      'CHIR99021' = c('LEF1'),
                      'WNTC59' = c('LEF1'),
                      'DLL1' = c(),
                      'JAG1' = c(),
                      'LY411575' = c('HES1'),
                      'TRULI' = c('LATS2'),
                      'Rapamycin' = c("LAMTOR4"),
                      'IGF1LR3' = c()
                      )

# Specify pathway for each perturbation
pert_pathway_list <- list(
  'BMP7'      = c('BMP'),
  'LDN193189' = c('BMP'),
  'FGF2'      = c('FGF','MAPK ERK'),
  'FGF8b'     = c('FGF','MAPK ERK'),
  'BGJ398'    = c('FGF','MAPK ERK'),
  'NRG1'      = c('ERBB','MAPK ERK'),
  'EGF'       = c('ERBB','MAPK ERK'),
  'AZD8931'   = c('ERBB','MAPK ERK'),
  'ATRA'      = c('RA'),
  'HX531'     = c('RA'),
  'CHIR99021' = c('WNT'),
  'WNTC59'    = c('WNT'),
  'SHH'       = c('SHH'),
  'SANT1'     = c('SHH'),
  'DLL1'      = c('NOTCH'),
  'JAG1'      = c('NOTCH'),
  'LY411575'  = c('NOTCH'),
  'IGF1LR3'   = c('mTOR', 'IGF'),
  'Rapamycin' = c('mTOR'),
  'TRULI'     = c('HIPPO')
)

# Set perturbation order
pert_order <- c("BMP7", "LDN193189",
                "FGF2", "FGF8b", "BGJ398",
                "NRG1", "EGF", "AZD8931",
                "ATRA", "HX531",
                "CHIR99021", "WNTC59",
                "SHH", "SANT1",
                "DLL1", "JAG1", "LY411575",
                "IGF1LR3", "Rapamycin",
                "TRULI")
```

```{r}
# Add a column with a flag for whether the gene is an expected response gene for each perturbation
filt_df <- response_df %>%
  dplyr::filter(!(perturbation %in% c("Untreated", "DMSO", "TreMan"))) %>%
  dplyr::mutate(perturbation = factor(perturbation, levels = pert_order)) %>%
  dplyr::rowwise() %>%
  dplyr::mutate(
    expected_pathway = list(pert_pathway_list[[as.character(perturbation)]]),
    match_expected = {
      sc <- as.character(signaling_category)
      ep <- expected_pathway[[1]]   # <-- FIX

      if (length(sc) == 0 || is.na(sc) || is.null(ep) || length(ep) == 0) {
        FALSE
      } else {
        any(purrr::map_lgl(
          ep,
          ~ stringr::str_detect(sc, stringr::regex(.x, ignore_case = TRUE))
        ))
      }
    },
    legend = dplyr::case_when(
      !isTRUE(significant)                  ~ "NS",
      isTRUE(significant) & match_expected  ~ "S, Expected Pathway",
      isTRUE(significant) & !match_expected ~ "S, Uncategorized",
      TRUE                                  ~ "NS"  # optional, but safer
    )
  ) %>%
  dplyr::ungroup()

head(filt_df)
```

```{r define genes to label}
# Add a column with a flag for whether the gene is a gene to label for each perturbation
filt_df <- filt_df %>%
  dplyr::rowwise() %>%
  dplyr::mutate(
    genes_label_vec = list({
      p <- as.character(perturbation)
      if (is.na(p) || !p %in% names(genes_to_label)) {
        character(0)
      } else {
        unique(genes_to_label[[p]])
      }
    }),
    label_gene = {
      g <- as.character(ID)
      if (is.na(g) || length(genes_label_vec[[1]]) == 0) {
        FALSE
      } else {
        g %in% genes_label_vec[[1]]
      }
    }
  ) %>%
  dplyr::ungroup()

head(filt_df)
```

```{r}
# Find y range, min, max, and difference
y_rng <- range(filt_df$logFC, na.rm = TRUE)
y_min <- y_rng[1]
y_max <- y_rng[2]
dy <- diff(y_rng)

# Make x axis wider per perturbation AND widen jitter accordingly
x_half <- 2
jit_width <- 0.85 * (2 * x_half)

set.seed(11)
filt_df <- filt_df %>%
  group_by(perturbation) %>%
  mutate(x_jit = runif(n(), min = -jit_width/2, max = jit_width/2)) %>%
  ungroup()

# Order facets by pert_order
filt_df <- filt_df %>%
  mutate(perturbation = factor(as.character(perturbation), levels = pert_order))

# Split data by categories to color by in legend
ns_df    <- filt_df %>% filter(legend == "NS")
uncat_df <- filt_df %>% filter(legend == "S, Uncategorized")
exp_df   <- filt_df %>% filter(legend == "S, Expected Pathway")

# Split data by label 
label_df <- exp_df %>% filter(isTRUE(label_gene) | label_gene == TRUE)
label_df_up   <- label_df %>% filter(logFC > 0)
label_df_down <- label_df %>% filter(logFC < 0)

top_band    <- c(y_max - 0.20*dy, y_max + 0.10*dy)   # narrower + higher
bottom_band <- c(y_min - 0.15*dy, y_min + 0.05*dy)   # narrower + lower

# Plot jittered scatter plot faceted by perturbation
p <- ggplot(filt_df, aes(y = logFC, size = abs(logFC))) +
  geom_point(data = ns_df,    aes(x = x_jit, color = legend), alpha = 0.8, na.rm = TRUE) +
  geom_point(data = uncat_df, aes(x = x_jit, color = legend), alpha = 0.9, na.rm = TRUE) +
  geom_point(data = exp_df,   aes(x = x_jit, color = legend), alpha = 1.0, na.rm = TRUE) +

  ggrepel::geom_text_repel(
    data = label_df_up,
    aes(x = x_jit, label = ID),
    size               = 5,
    box.padding        = 0.15,
    point.padding      = 0.15,
    max.overlaps       = Inf,
    direction          = "y",
    ylim               = top_band,
    nudge_y            = 0.18 * dy,      # strong upward bias
    force              = 30,             # stronger repel
    force_pull         = 0.5,            # weaker pull back to point
    nudge_x            = 0.30 * x_half,
    hjust              = 0,
    min.segment.length = 0,
    segment.size       = 0.3,
    segment.color      = "black",
    color              = "black",
    seed               = 1
  ) +
  
  ggrepel::geom_text_repel(
    data = label_df_down,
    aes(x = x_jit, label = ID),
    size               = 5,
    box.padding        = 0.15,
    point.padding      = 0.15,
    max.overlaps       = Inf,
    direction          = "y",
    ylim               = bottom_band,
    nudge_y            = -0.18 * dy,     # strong downward bias
    force              = 35,
    force_pull         = 0.5,
    nudge_x            = -0.30 * x_half,
    hjust              = 1,
    min.segment.length = 0,
    segment.size       = 0.3,
    segment.color      = "black",
    color              = "black",
    seed               = 2
  ) +

  # Facet order comes from factor levels of perturbation_md
  facet_wrap(~ perturbation, nrow = 1, scales = "fixed", drop = TRUE) +

  scale_x_continuous(
    limits = c(-x_half, x_half),
    breaks = NULL,
    expand = ggplot2::expansion(mult = c(0.02, 0.02))
  ) +
  scale_y_continuous(limits = c(y_min - 0.15*dy, y_max + 0.15*dy)) +
  coord_cartesian(clip = "off") +
  scale_color_manual(
    values = c(
      "NS"              = "grey80",
      "S, Uncategorized"    = "grey30",
      "S, Expected Pathway" = "firebrick3"
    ),
    breaks = c("NS","S, Uncategorized","S, Expected Pathway"),
    name = NULL,
    guide = guide_legend(override.aes = list(size = 3))
  ) +
  scale_size_continuous(name = expression("|Log"[2]*"FC|"), range = c(0.3, 3)) +
  labs(x = NULL, y = expression(Log[2]~"(Stimulus/Vehicle)")) +
  theme_bw() +
  theme(
    plot.title            = element_blank(), 
    panel.grid            = element_blank(),
    panel.border          = element_blank(),
    panel.background      = element_blank(),
    axis.line.y           = element_line(color = "black"),
    axis.line.x           = element_blank(),
    axis.text.x           = element_blank(),
    axis.text.y           = element_text(size=26),
    axis.title.y           = element_text(size=20),
    axis.ticks.x          = element_blank(),
    strip.background      = element_blank(),
    legend.position       = "bottom",
    legend.direction      = "horizontal",
    legend.justification  = "center",
    legend.title  = element_text(size=20),
    legend.text = element_text(size=26)
  )

# Save and print
ggsave(file.path(fig_path, "scatterplot_ResponseLogFC_Jittered_all.png"), plot = p, height = 8, width = 32, bg = "white", dpi = 600)

print(p)
```

## Simple rGenes ORA GO:BP

```{r define ORA function}
simple_ora_df <- function(rgenes_df, results_df, rgenes_df_colname, species_name, category, subcategory) {
  
  # Initialize an empty list for storing the ORA results
  ora_results_list <- list()
  
  # Iterate through perturbation x time point
  for (pert in unique(rgenes_df$perturbation)) {
    # Filter and pull DEGs from specified column
    genes_raw <- rgenes_df %>% filter(perturbation == pert) %>%
      pull(!!sym(rgenes_df_colname))
        
    # Flatten and split stringified gene lists safely
    genes <- genes_raw %>% na.omit() %>% unlist() %>% strsplit(split = "/") %>% unlist() %>% trimws() %>% unique()

    # Skip if empty or all NAs
    if (length(genes) == 0 || all(is.na(genes)) || all(genes == "")) next

    # Remove empty or NA values before mapping
    genes_clean <- genes[!is.na(genes) & genes != ""]
        
    # Check valid SYMBOL keys
    valid_symbols <- keys(org.Hs.eg.db, keytype = "SYMBOL")
    genes_clean <- intersect(genes_clean, valid_symbols)
        
    # Skip iteration if no valid symbols
    if (length(genes_clean) == 0) next

    # Convert gene symbols to ENTREZ IDs
    genes_entrez <- as.character(na.omit(mapIds(
      org.Hs.eg.db,
      keys = genes_clean,
      column = "ENTREZID",
      keytype = "SYMBOL",
      multiVals = "first"
    )))
        
    if (length(genes_entrez) == 0) next

    # Define universe of genes
    universe_entrez <- as.character(na.omit(mapIds(org.Hs.eg.db, results_df$ID, "ENTREZID", "SYMBOL")))

    # Get gene sets for enrichment analysis
    gene_set <- msigdbr(species = species_name, category = category, subcategory = subcategory)
    gene_set_df <- gene_set %>% dplyr::select(gs_name, entrez_gene)

    # Perform ORA
    ora <- tryCatch({
      enricher(
        gene = genes_entrez,
        universe = universe_entrez,
        TERM2GENE = gene_set_df,
        pAdjustMethod = "fdr",
        qvalueCutoff = 0.10,
        minGSSize = 5,
        maxGSSize = 500
      )
    }, error = function(e) NULL)

      # Check if ORA returned results
      if (!is.null(ora) && nrow(ora@result) > 0) {
      ora_results <- ora@result %>%
        mutate(perturbation = pert,) %>%
      tidyr::separate(GeneRatio, into = c("numerator", "denominator"), sep = "/", remove = FALSE) %>%
        mutate(
          GeneRatioOriginal = GeneRatio,
          GeneRatio = as.numeric(numerator) / as.numeric(denominator)
        ) %>%
        dplyr::select(-numerator, -denominator) %>%
        filter(p.adjust < 0.10 & qvalue < 0.10 & Count > 1)
      ora_results_list[[pert]] <- ora_results
    }
  }
  
  # Combine all ORA results
  ora_results_df <- dplyr::bind_rows(ora_results_list)
  return(ora_results_df)
}
```

Perform GO:BP term ORA of simple rGenes
```{r, message=FALSE, warning=FALSE}
simple_human_up_ora_df <- simple_ora_df(simple_rgenes_df, results_df, 'human_up_rGenes', 'Homo sapiens', 'C5', 'GO:BP')
head(simple_human_up_ora_df)
```

Perform GO:BP term ORA of simple rGenes
```{r, message=FALSE, warning=FALSE}
simple_chimp_up_ora_df <- simple_ora_df(simple_rgenes_df, results_df, 'chimp_up_rGenes', 'Homo sapiens', 'C5', 'GO:BP')
head(simple_chimp_up_ora_df)
```

```{r, message=FALSE, warning=FALSE}
simple_human_down_ora_df <- simple_ora_df(simple_rgenes_df, results_df, 'human_down_rGenes', 'Homo sapiens', 'C5', 'GO:BP')
head(simple_human_down_ora_df)
```

```{r, message=FALSE, warning=FALSE}
simple_chimp_down_ora_df <- simple_ora_df(simple_rgenes_df, results_df, 'chimp_down_rGenes', 'Homo sapiens', 'C5', 'GO:BP')
head(simple_chimp_down_ora_df)
```

```{r}
# Add directions to ORA results
simple_human_up_ora_df$direction <- "Up"
simple_chimp_up_ora_df$direction <- "Up"
simple_human_down_ora_df$direction <- "Down"
simple_chimp_down_ora_df$direction <- "Down"

# Add species to ORA results
simple_human_up_ora_df$species <- "Human"
simple_chimp_up_ora_df$species <- "Chimp"
simple_human_down_ora_df$species <- "Human"
simple_chimp_down_ora_df$species <- "Chimp"

# Combine ORA dfs
combined_simple_ora_df <- bind_rows(simple_human_up_ora_df, simple_human_down_ora_df, simple_chimp_up_ora_df, simple_chimp_down_ora_df)
```

```{r save simple rgenes ora df}
saveRDS(combined_simple_ora_df, paste0(dir_path, "simple_rGenes_ORA_GOBP.rds"))
```

```{r save simple rgenes ora df}
combined_simple_ora_df <- readRDS(paste0(dir_path, "simple_rGenes_ORA_GOBP.rds"))
head(combined_simple_ora_df)
```

```{r set signaling terms}
# Growth factor
growth_factor_terms <- c(
"GOBP_RESPONSE_TO_GROWTH_FACTOR",
"GOBP_REGULATION_OF_RESPONSE_TO_GROWTH_FACTOR",
"GOBP_POSITIVE_REGULATION_OF_RESPONSE_TO_GROWTH_FACTOR",
"GOBP_NEGATIVE_REGULATION_OF_RESPONSE_TO_GROWTH_FACTOR",
"GOBP_CELLULAR_RESPONSE_TO_GROWTH_FACTOR_STIMULUS",
"GOBP_REGULATION_OF_CELLULAR_RESPONSE_TO_GROWTH_FACTOR_STIMULUS",
"GOBP_POSITIVE_REGULATION_OF_CELLULAR_RESPONSE_TO_GROWTH_FACTOR_STIMULUS",
"GOBP_NEGATIVE_REGULATION_OF_CELLULAR_RESPONSE_TO_GROWTH_FACTOR_STIMULUS"
)

# FGF
fgf_terms <- c(
"GOBP_RESPONSE_TO_FIBROBLAST_GROWTH_FACTOR",
"GOBP_CELLULAR_RESPONSE_TO_FIBROBLAST_GROWTH_FACTOR",
"GOBP_FIBROBLAST_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_FIBROBLAST_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_FIBROBLAST_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_FIBROBLAST_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY"
)

# MAPK ERK
mapk_erk_terms <- c(
"GOBP_MAPK_CASCADE",
"GOBP_REGULATION_OF_MAPK_CASCADE",
"GOBP_POSITIVE_REGULATION_OF_MAPK_CASCADE",
"GOBP_NEGATIVE_REGULATION_OF_MAPK_CASCADE",

"GOBP_MAP_KINASE_ACTIVITY",
"GOBP_REGULATION_OF_MAP_KINASE_ACTIVITY",
"GOBP_POSITIVE_REGULATION_OF_MAP_KINASE_ACTIVITY",
"GOBP_NEGATIVE_REGULATION_OF_MAP_KINASE_ACTIVITY",

"GOBP_ERK1_AND_ERK2_CASCADE",
"GOBP_REGULATION_OF_ERK1_AND_ERK2_CASCADE",
"GOBP_POSITIVE_REGULATION_OF_ERK1_AND_ERK2_CASCADE",
"GOBP_NEGATIVE_REGULATION_OF_ERK1_AND_ERK2_CASCADE",

"GOBP_NEGATIVE_REGULATION_OF_RAS_PROTEIN_SIGNAL_TRANSDUCTION"
)

# BMP
bmp_terms <- c(
"GOBP_RESPONSE_TO_BMP",
"GOBP_CELLULAR_RESPONSE_TO_BMP_STIMULUS",

"GOBP_BMP_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_BMP_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_BMP_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_BMP_SIGNALING_PATHWAY",

"GOBP_SMAD_PROTEIN_SIGNAL_TRANSDUCTION",
"GOBP_REGULATION_OF_SMAD_PROTEIN_SIGNAL_TRANSDUCTION",
"GOBP_POSITIVE_REGULATION_OF_SMAD_PROTEIN_SIGNAL_TRANSDUCTION",
"GOBP_NEGATIVE_REGULATION_OF_SMAD_PROTEIN_SIGNAL_TRANSDUCTION"
)

# EGF
egf_terms <- c(
"GOBP_EPIDERMAL_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_EPIDERMAL_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_EPIDERMAL_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_EPIDERMAL_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",

'GOBP_RESPONSE_TO_EPIDERMAL_GROWTH_FACTOR',
'GOBP_CELLULAR_RESPONSE_TO_EPIDERMAL_GROWTH_FACTOR_STIMULUS',

'GOBP_REGULATION_OF_EPIDERMAL_GROWTH_FACTOR_ACTIVATED_RECEPTOR_ACTIVITY',
'GOBP_POSITIVE_REGULATION_OF_EPIDERMAL_GROWTH_FACTOR_ACTIVATED_RECEPTOR_ACTIVITY',
'GOBP_NEGATIVE_REGULATION_OF_EPIDERMAL_GROWTH_FACTOR_ACTIVATED_RECEPTOR_ACTIVITY',

'GOBP_ERBB2_EGFR_SIGNALING_PATHWAY',
'GOBP_ERBB4_EGFR_SIGNALING_PATHWAY'
)

# ERBB
erbb_terms <- c(
'GOBP_ERBB_SIGNALING_PATHWAY',
'GOBP_REGULATION_OF_ERBB_SIGNALING_PATHWAY',
'GOBP_NEGATIVE_REGULATION_OF_ERBB_SIGNALING_PATHWAY',
'GOBP_POSITIVE_REGULATION_OF_ERBB_SIGNALING_PATHWAY',

'GOBP_ERBB3_SIGNALING_PATHWAY',
'GOBP_REGULATION_OF_ERBB3_SIGNALING_PATHWAY',
'GOBP_POSITIVE_REGULATION_OF_ERBB3_SIGNALING_PATHWAY',
'GOBP_NEGATIVE_REGULATION_OF_ERBB3_SIGNALING_PATHWAY',

'GOBP_ERBB2_SIGNALING_PATHWAY',
'GOBP_ERBB4_SIGNALING_PATHWAY',

'GOBP_ERBB2_ERBB3_SIGNALING_PATHWAY',
'GOBP_ERBB2_ERBB4_SIGNALING_PATHWAY',
'GOBP_ERBB3_ERBB4_SIGNALING_PATHWAY',
'GOBP_ERBB4_ERBB4_SIGNALING_PATHWAY'
)

# mTOR
mtor_terms <- c(
"GOBP_TOR_SIGNALING",
"GOBP_REGULATION_OF_TOR_SIGNALING",
"GOBP_POSITIVE_REGULATION_OF_TOR_SIGNALING",
"GOBP_NEGATIVE_REGULATION_OF_TOR_SIGNALING",

"GOBP_TORC1_SIGNALING",
"GOBP_REGULATION_OF_TORC1_SIGNALING",
"GOBP_POSITIVE_REGULATION_OF_TORC1_SIGNALING",
"GOBP_NEGATIVE_REGULATION_OF_TORC1_SIGNALING",

"GOBP_TORC2_SIGNALING",
"GOBP_REGULATION_OF_TORC2_SIGNALING",
"GOBP_POSITIVE_REGULATION_OF_TORC2_SIGNALING",
"GOBP_NEGATIVE_REGULATION_OF_TORC2_SIGNALING"
)

# IGF
igf_terms <- c(
'GOBP_INSULIN_LIKE_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY',
'GOBP_REGULATION_OF_INSULIN_LIKE_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY',
'GOBP_POSITIVE_REGULATION_OF_INSULIN_LIKE_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY',
'GOBP_NEGATIVE_REGULATION_OF_INSULIN_LIKE_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY',

'GOBP_RESPONSE_TO_INSULIN_LIKE_GROWTH_FACTOR_STIMULUS',
'GOBP_CELLULAR_RESPONSE_TO_INSULIN_LIKE_GROWTH_FACTOR_STIMULUS'
)

# Rapamycin
rapamycin_terms <- c(
'GOBP_RESPONSE_TO_RAPAMYCIN',
'GOBP_CELLULAR_RESPONSE_TO_RAPAMYCIN'
)

# HIPPO
hippo_terms <- c(
"GOBP_HIPPO_SIGNALING",
"GOBP_REGULATION_OF_HIPPO_SIGNALING",
"GOBP_POSITIVE_REGULATION_OF_HIPPO_SIGNALING",
"GOBP_NEGATIVE_REGULATION_OF_HIPPO_SIGNALING"
)

# NOTCH
notch_terms <- c(
"GOBP_NOTCH_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_NOTCH_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_NOTCH_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_NOTCH_SIGNALING_PATHWAY",

"GOBP_TRANSCRIPTION_OF_NOTCH_RECEPTOR_TARGET",
"GOBP_REGULATION_OF_TRANSCRIPTION_OF_NOTCH_RECEPTOR_TARGET",
"GOBP_POSITIVE_REGULATION_OF_TRANSCRIPTION_OF_NOTCH_RECEPTOR_TARGET",
"GOBP_NEGATIVE_REGULATION_OF_TRANSCRIPTION_OF_NOTCH_RECEPTOR_TARGET"
)

# RA
ra_terms <- c(
"GOBP_RESPONSE_TO_RETINOIC_ACID",
"GOBP_CELLULAR_RESPONSE_TO_RETINOIC_ACID",

"GOBP_RETINOIC_ACID_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_RETINOIC_ACID_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_RETINOIC_ACID_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_RETINOIC_ACID_RECEPTOR_SIGNALING_PATHWAY"
)

# WNT
wnt_terms <- c(
"GOBP_WNT_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_WNT_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_WNT_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_WNT_SIGNALING_PATHWAY",

"GOBP_CANONICAL_WNT_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_CANONICAL_WNT_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_CANONICAL_WNT_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_CANONICAL_WNT_SIGNALING_PATHWAY",

"GOBP_NON_CANONICAL_WNT_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_NON_CANONICAL_WNT_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_NON_CANONICAL_WNT_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_NON_CANONICAL_WNT_SIGNALING_PATHWAY",

"GOBP_WNT_SIGNALING_PATHWAY_PLANAR_CELL_POLARITY_PATHWAY",
"GOBP_REGULATION_OF_WNT_SIGNALING_PATHWAY_PLANAR_CELL_POLARITY_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_WNT_SIGNALING_PATHWAY_PLANAR_CELL_POLARITY_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_WNT_SIGNALING_PATHWAY_PLANAR_CELL_POLARITY_PATHWAY"
)

# SHH
shh_terms <- c(
"GOBP_SMOOTHENED_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY",

"GOBP_SMOOTHENED_SIGNALING_PATHWAY_INVOLVED_IN_DORSAL_VENTRAL_NEURAL_TUBE_PATTERNING",
"GOBP_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY_INVOLVED_IN_DORSAL_VENTRAL_NEURAL_TUBE_PATTERNING",
"GOBP_POSITIVE_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY_INVOLVED_IN_DORSAL_VENTRAL_NEURAL_TUBE_PATTERNING",
"GOBP_NEGATIVE_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY_INVOLVED_IN_DORSAL_VENTRAL_NEURAL_TUBE_PATTERNING"
)

signaling_terms_list <- list(
  'BMP7' = c(bmp_terms, growth_factor_terms),
  'LDN193189' = bmp_terms,
  'FGF2' = c(fgf_terms, mapk_erk_terms, growth_factor_terms),
  'FGF8b' = c(fgf_terms, mapk_erk_terms,growth_factor_terms),
  'BGJ398' = c(fgf_terms, mapk_erk_terms),
  'NRG1' = c(erbb_terms, mapk_erk_terms,  growth_factor_terms),
  'EGF' = c(egf_terms, erbb_terms, mapk_erk_terms, growth_factor_terms),
  'AZD8931' = c(egf_terms, erbb_terms, mapk_erk_terms),
  'DLL1' = c(notch_terms,  growth_factor_terms),
  'JAG1' = c(notch_terms,  growth_factor_terms),
  'LY411575' = notch_terms,
  'IGF1LR3' = c(igf_terms, mtor_terms, growth_factor_terms),
  'Rapamycin' = c(rapamycin_terms, mtor_terms),
  'TRULI' = c(hippo_terms, growth_factor_terms), 
  'ATRA' = c(ra_terms, growth_factor_terms),
  'HX531' = ra_terms,
  'CHIR99021' = c(wnt_terms, growth_factor_terms),
  'WNTC59' = wnt_terms,
  'SHH' = c(shh_terms,  growth_factor_terms),
  'SANT1' = shh_terms
)
```

```{r}
write.xlsx(combined_simple_ora_df, file = paste0(dir_path, "simple_rGenes_ORA_GOBP_df.xlsx"), rowNames = FALSE)
```


```{r}
# Count matching signaling GO:BP terms per species x perturbation x direction
signaling_term_counts_df <- combined_simple_ora_df %>%
  filter(!(perturbation %in% c('DMSO', 'TreMan'))) %>%
  dplyr::mutate(
    ora_term = ID,
    expected_terms = signaling_terms_list[as.character(.data$perturbation)],
    expected_terms = lapply(expected_terms, function(x) if (is.null(x)) character(0) else unique(x)),
    is_signaling_match = mapply(function(term, expected) term %in% expected,
                               ora_term, expected_terms)
  ) %>%
  dplyr::group_by(species, perturbation, direction) %>%
  dplyr::summarise(
    n_terms_total = dplyr::n(),
    n_signaling_terms = sum(is_signaling_match, na.rm = TRUE),
    .groups = "drop"
  )

signaling_term_counts_df
```




```{r}
signaling_term_counts_df$pert_direct <- paste(signaling_term_counts_df$perturbation, signaling_term_counts_df$direction, sep="_")

plot_df <- signaling_term_counts_df %>% filter(perturbation %in% c('BMP7', 'FGF2', 'FGF8b', 'ATRA', 'WNTC59',  'BGJ398'))

p <- ggplot(plot_df, aes(x=pert_direct, y=species, size = n_signaling_terms, color = 'black', fill=perturbation)) +
  geom_point(shape = 21, color = 'black', stroke = 0.25) + 
  labs(
       x = "",
       y = "",
       size = "Expected Signaling\nGO:BP Term Count",
       fill = "Stimulus") +
  scale_fill_manual(
    name = "Stimulus",
    values = pert_colors,
    guide = "none"
    ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5),
    axis.text.y = element_text(size = 10),
    axis.text.x = element_text(angle = 45, hjust = 1)  # Rotate x-axis labels for readability
  )

# Save the plot.
ggsave(paste0(fig_path, "dotplot_simplerGenes_BySpecies_ByPerturbationDirection_SizeByTermCount.pdf"),plot = p, height = 3, width = 6, bg = 'white')
print(p)
```

## Response GSEA GO:BP

```{r GSEA function}
gsea_df <- function(response_df) {
  
  # Initialize an empty list to store results
  gsea_list <- list()
  
  # Iterate through species x perturbation x time point
  for (pert in unique(response_df$perturbation)) {
    for (time in unique(response_df$time_point)) {
      for (spec in unique(response_df$species)) {
        
        # Pull genes and their logFCs, rank by logFC, and create a named numeric vector for input
        gene_list <- response_df %>% filter(perturbation == pert & time_point == time & species == spec) %>%
          ungroup() %>%
          dplyr::select(ID, logFC) %>%
          dplyr::arrange(desc(logFC)) %>%
          deframe()
        
        # Skip iteration if gene_list is empty
        if (length(gene_list) == 0) next  
    
        # Run GSEA using GO:BP terms
        gsea_go <- gseGO(
          geneList = gene_list,
          OrgDb = org.Hs.eg.db,
          keyType = "SYMBOL",  
          ont = "BP", 
          pvalueCutoff = 1,
          minGSSize = 5,
          eps = 0,
          verbose = FALSE,
          by = "fgsea"
          )
          
        # Check if GSEA returned results
        if (!is.null(gsea_go) && nrow(gsea_go@result) > 0) {
              
        # Format GSEA results data frame.
        gsea_results <- gsea_go@result %>%
          mutate(
            perturbation = pert,
            time_point = time,
            species = spec)
        
        # Add results to list
        gsea_list[[paste(pert, time, spec, sep='_')]] <- gsea_results
        }
      }
    }
  }
  
  # Combine list into a data frame and return
  gsea_df <- bind_rows(gsea_list)
  return(gsea_df)
}
```

Perform GO:BP term GSEA of rGenes
```{r rGenes gsea_df}
response_gsea_df <- gsea_df(response_df)
head(response_gsea_df)
```

Add signaling gene set categories to rGenes GSEA data frame
```{r}
# Create a named vector mapping GO term IDs to their signaling category
id_to_category <- unlist(lapply(names(go_term_sets), function(category) {
  setNames(rep(category, length(go_term_sets[[category]])), go_term_sets[[category]])
}))

# Add the 'signaling_category' column to gsea_df based on the ID mapping
response_gsea_df$signaling_category <- id_to_category[as.character(response_gsea_df$ID)]
head(response_gsea_df)
```

Add signaling category colors to rGenes GSEA data frame
```{r}
# Add signaling category colors.
response_gsea_df <- response_gsea_df %>%
  mutate(signaling_category_color = recode(signaling_category, !!!signaling_category_colors)) %>%
  replace_na(list(signaling_category_color = "darkgrey"))

head(response_gsea_df)
```

Save response GSEA GO:BP results
```{r save rGenes gsea_df}
saveRDS(response_gsea_df, file = paste0(dir_path, "rGenes_GSEA_GOBP_df.rds"))
```

```{r}
write.csv(response_gsea_df, file = paste0(dir_path, "response_GSEA_GOBP_df.csv"), row.names = FALSE)
```

```{r}
write.xlsx(response_gsea_df, file = paste0(dir_path, "response_GSEA_GOBP_df.xlsx"), rowNames = FALSE)
```

Load response GSEA GO:BP results
```{r load rGenes gsea_df}
response_gsea_df <- readRDS(paste0(dir_path, "rGenes_GSEA_GOBP_df.rds"))
head(response_gsea_df)
```

## Volcano plots rGenes GSEA GO:BP

```{r rGenes GSEA volcano plot all time points}
# Set time point factor levels
response_gsea_df$time_point <- factor(response_gsea_df$time_point, levels = c("6hr", "54hr", "7day"))

# Change NES to numeric
response_gsea_df$NES <- as.numeric(response_gsea_df$NES)

for (spec in unique(response_gsea_df$species)) {
  for (pert in unique(response_gsea_df$perturbation)) {

    filt_df <- response_gsea_df %>%
      dplyr::filter(perturbation == pert, species == spec)

    if (nrow(filt_df) == 0) next

    # Split dataset by significance / category (fixed)
    ns_df <- filt_df %>% dplyr::filter(qvalue >= 0.10)

    sig_nocat_df <- filt_df %>% dplyr::filter(qvalue < 0.10, is.na(signaling_category))
    sig_cat_df   <- filt_df %>% dplyr::filter(qvalue < 0.10, !is.na(signaling_category))

    # Symmetric x limits
    max_val <- max(abs(filt_df$NES), na.rm = TRUE)

    # Top labels
    top_pos <- filt_df %>% dplyr::filter(NES > 0, p.adjust < 0.05) %>% dplyr::arrange(p.adjust) %>% dplyr::slice_head(n = 5)
    top_neg <- filt_df %>% dplyr::filter(NES < 0, p.adjust < 0.05) %>% dplyr::arrange(p.adjust) %>% dplyr::slice_head(n = 5)
    top_labeled <- dplyr::bind_rows(top_pos, top_neg)

    p <- ggplot2::ggplot(filt_df, ggplot2::aes(x = NES, y = -log10(p.adjust))) +

      # Not significant (constant fill outside aes)
      ggplot2::geom_point(
        data = ns_df,
        ggplot2::aes(shape = time_point),
        fill = "lightgrey",
        color = "black", stroke = 0, alpha = 0.35, size = 3
      ) +

      # Significant, no signaling category (constant fill outside aes)
      ggplot2::geom_point(
        data = sig_nocat_df,
        ggplot2::aes(shape = time_point),
        fill = "darkgrey",
        color = "black", stroke = 0, alpha = 0.65, size = 3
      ) +

      # Significant + signaling category (mapped fill)
      ggplot2::geom_point(
        data = sig_cat_df,
        ggplot2::aes(fill = signaling_category, shape = time_point),
        color = "black", stroke = 0.25, alpha = 1, size = 3
      ) +

      ggrepel::geom_text_repel(
        data = top_labeled,
        ggplot2::aes(label = Description),
        size = 3, nudge_y = 1, direction = "y", hjust = 0.5, vjust = 0
      ) +

      ggplot2::geom_hline(
        yintercept = -log10(0.10),
        linetype = "dashed",
        color = "darkgrey",
        linewidth = 0.5
      ) +

      ggplot2::labs(
        title = paste0("GO:BP GSEA ", spec, " ", pert),
        x = "NES",
        y = expression("-" ~ Log[10] ~ "(Adjusted P-value)")
      ) +
      ggplot2::scale_y_continuous(limits = c(0, NA)) +
      ggplot2::scale_x_continuous(limits = c(-max_val, max_val)) +

      # Shape by time point
      ggplot2::scale_shape_manual(
        name = "Time Point",
        values = c(21, 22, 23),
        guide = ggplot2::guide_legend(
          override.aes = list(fill = "black", color = "black", size = 4, stroke = 0.25)
        )
      ) +

      # Fill legend ONLY for signaling categories
      ggplot2::scale_fill_manual(
        name = "Signaling Category",
        values = signaling_category_colors,
        guide = ggplot2::guide_legend(
          override.aes = list(shape = 21, stroke = 0.25, color = "black", size = 4)
        )
      ) +

      ggplot2::theme_minimal() +
      ggplot2::theme(
        plot.title = ggplot2::element_text(hjust = 0.5, size = 14),
        legend.position = "right",
        axis.line = ggplot2::element_line(color = "black"),
        panel.grid.major = ggplot2::element_blank(),
        panel.grid.minor = ggplot2::element_blank()
      )

    ggplot2::ggsave(
      filename = file.path(fig_path, "volcanoplot_rGenes_GSEA_GOBP",
                           paste0("volcanoplot_rGenes_GSEA_GOBP_", spec, "_", pert, ".png")),
      plot = p, units = "in", height = 4, width = 6, bg = "white", dpi = 600
    )
  }
}
```

```{r rGenes GSEA volcano plot all time points}
# Set time point factor levels
response_gsea_df$time_point <- factor(response_gsea_df$time_point, levels = c("6hr", "54hr", "7day"))

# Change NES to numeric
response_gsea_df$NES <- as.numeric(response_gsea_df$NES)

for (spec in unique(response_gsea_df$species)) {
  for (pert in unique(response_gsea_df$perturbation)) {
    for (time in unique(response_gsea_df$time_point)) {

      filt_df <- response_gsea_df %>%
        dplyr::filter(perturbation == pert, species == spec, time_point == time)
  
      if (nrow(filt_df) == 0) next
  
      # Split dataset by significance / category (fixed)
      ns_df <- filt_df %>% dplyr::filter(qvalue >= 0.10)
  
      sig_nocat_df <- filt_df %>% dplyr::filter(qvalue < 0.10, is.na(signaling_category))
      sig_cat_df   <- filt_df %>% dplyr::filter(qvalue < 0.10, !is.na(signaling_category))
  
      # Symmetric x limits
      max_val <- max(abs(filt_df$NES), na.rm = TRUE)
  
      # Top labels
      top_pos <- filt_df %>% dplyr::filter(NES > 0, p.adjust < 0.05) %>% dplyr::arrange(p.adjust) %>% dplyr::slice_head(n = 3)
      top_neg <- filt_df %>% dplyr::filter(NES < 0, p.adjust < 0.05) %>% dplyr::arrange(p.adjust) %>% dplyr::slice_head(n = 3)
      top_labeled <- dplyr::bind_rows(top_pos, top_neg)
  
      p <- ggplot2::ggplot(filt_df, ggplot2::aes(x = NES, y = -log10(p.adjust))) +
  
        # Not significant (constant fill outside aes)
        ggplot2::geom_point(
          data = ns_df,
          ggplot2::aes(shape = time_point),
          fill = "lightgrey",
          color = "black", stroke = 0, alpha = 0.35, size = 3
        ) +
  
        # Significant, no signaling category (constant fill outside aes)
        ggplot2::geom_point(
          data = sig_nocat_df,
          ggplot2::aes(shape = time_point),
          fill = "darkgrey",
          color = "black", stroke = 0, alpha = 0.65, size = 3
        ) +
  
        # Significant + signaling category (mapped fill)
        ggplot2::geom_point(
          data = sig_cat_df,
          ggplot2::aes(fill = signaling_category, shape = time_point),
          color = "black", stroke = 0.25, alpha = 1, size = 3
        ) +
  
        ggrepel::geom_text_repel(
          data = top_labeled,
          ggplot2::aes(label = Description),
          size = 3, nudge_y = 1, direction = "y", hjust = 0.5, vjust = 0
        ) +
  
        ggplot2::geom_hline(
          yintercept = -log10(0.10),
          linetype = "dashed",
          color = "darkgrey",
          linewidth = 0.5
        ) +
  
        ggplot2::labs(
          title = paste0("GO:BP GSEA ", spec, " ", pert, " ", time),
          x = "NES",
          y = expression("-" ~ Log[10] ~ "(Adjusted P-value)")
        ) +
        ggplot2::scale_y_continuous(limits = c(0, NA)) +
        ggplot2::scale_x_continuous(limits = c(-max_val, max_val)) +
  
        # Shape by time point
        ggplot2::scale_shape_manual(
          name = "Time Point",
          values = c(21, 22, 23),
          guide = ggplot2::guide_legend(
            override.aes = list(fill = "black", color = "black", size = 4, stroke = 0.25)
          )
        ) +
  
        # Fill legend ONLY for signaling categories
        ggplot2::scale_fill_manual(
          name = "Signaling Category",
          values = signaling_category_colors,
          guide = ggplot2::guide_legend(
            override.aes = list(shape = 21, stroke = 0.25, color = "black", size = 4)
          )
        ) +
  
        ggplot2::theme_minimal() +
        ggplot2::theme(
          plot.title = ggplot2::element_text(hjust = 0.5, size = 14),
          legend.position = "right",
          axis.line = ggplot2::element_line(color = "black"),
          panel.grid.major = ggplot2::element_blank(),
          panel.grid.minor = ggplot2::element_blank()
        )
  
      ggplot2::ggsave(
        filename = file.path(fig_path, "volcanoplot_Response_GSEA_GOBP_ByTimePoint",
                             paste0("volcanoplot_Response_GSEA_GOBP_", spec, "_", pert, "_", time, ".pdf")),
        plot = p, units = "in", height = 4, width = 6, bg = "white", dpi = 600
      )
    }
  }
}
```

```{r response GSEA volcano plot}
response_gsea_df <- response_gsea_df %>%
  mutate(significant = p.adjust < 0.10)

pert <- "BMP7"
time <- "54hr"
spec <- "Human"
n_top_terms <- 5

# Filter dataset
plot_df <- response_gsea_df %>%
  filter(perturbation == pert, time_point == time, species == spec)

# Determine symmetric x-axis limits
max_val <- max(abs(plot_df$NES), na.rm = TRUE)

# Identify top n lowest p.adjust values for both positive and negative NES (significant only)
top_pos <- plot_df %>% filter(NES > 0, significant) %>%
  arrange(p.adjust) %>%
  slice_head(n = n_top_terms)

top_neg <- plot_df %>% filter(NES < 0, significant) %>%
  arrange(p.adjust) %>%
  slice_head(n = n_top_terms)

top_labeled <- bind_rows(top_pos, top_neg)

# Add legend labels
plot_df <- plot_df %>%
  mutate(significant_lab = if_else(significant, "S", "NS"))

p <- ggplot(plot_df, aes(x = NES, y = -log10(p.adjust))) +
  geom_point(aes(fill = significant_lab), shape = 21, color = "black", stroke = 0.25, alpha = 0.8, size = 2) +
  labs(
    title = paste0("GO:BP GSEA\n", spec, " ", pert, " ", time),
    x = "NES",
    y = expression("-" ~ Log[10] ~ "(Adjusted P-value)"),
    fill = "Significance"
  ) +
  scale_y_continuous(limits = c(0, NA)) +
  scale_x_continuous(limits = c(-max_val, max_val)) +
  scale_fill_manual(
    values = c("NS" = "grey80", "S" = "red"),
    guide = guide_legend(override.aes = list(shape = 21, color = "black", stroke = 0.25, size = 4))
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    legend.position = "right",
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  geom_text_repel(
    data = top_labeled,
    aes(label = Description),
    size = 3, nudge_y = 1, direction = "y",
    hjust = 0.5, vjust = 0
  )

ggsave(
  paste0(fig_path, "volcanoplot_rGenes_GSEA_GOBP/volcanoplot_rGenes_GSEA_GOBP_", pert, "_", spec, "_", time, ".pdf"),
  plot = p, units = "in", height = 4, width = 6, bg = "white", dpi = 600
)

print(p)
```

```{r response GSEA volcano plot}
response_gsea_df <- response_gsea_df %>%
  mutate(significant = p.adjust < 0.10)

pert <- "BMP7"
time <- "54hr"
spec <- "Human"
n_top_terms <- 5

# Filter dataset
plot_df <- response_gsea_df %>%
  filter(perturbation == pert, time_point == time, species == spec)

# Determine symmetric x-axis limits
max_val <- max(abs(plot_df$NES), na.rm = TRUE)

# Identify top n lowest p.adjust values for both positive and negative NES (significant only)
top_pos <- plot_df %>% filter(NES > 0, significant) %>%
  arrange(p.adjust) %>%
  slice_head(n = n_top_terms)

top_neg <- plot_df %>% filter(NES < 0, significant) %>%
  arrange(p.adjust) %>%
  slice_head(n = n_top_terms)

top_labeled <- bind_rows(top_pos, top_neg)

# Add legend labels
plot_df <- plot_df %>%
  mutate(significant_lab = if_else(significant, "S", "NS"))

p <- ggplot(plot_df, aes(x = NES, y = -log10(p.adjust))) +
  geom_point(aes(fill = significant_lab), shape = 21, color = "black", stroke = 0.25, alpha = 0.8, size = 2) +
  labs(
    title = paste0("GO:BP GSEA\n", spec, " ", pert, " ", time),
    x = "NES",
    y = expression("-" ~ Log[10] ~ "(Adjusted P-value)"),
    fill = "Significance"
  ) +
  scale_y_continuous(limits = c(0, NA)) +
  scale_x_continuous(limits = c(-max_val, max_val)) +
  scale_fill_manual(
    values = c("NS" = "grey80", "S" = "red"),
    guide = guide_legend(override.aes = list(shape = 21, color = "black", stroke = 0.25, size = 4))
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    legend.position = "right",
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  geom_text_repel(
    data = top_labeled,
    aes(label = Description),
    size = 3, nudge_y = 1, direction = "y",
    hjust = 0.5, vjust = 0
  )

ggsave(
  paste0(fig_path, "volcanoplot_rGenes_GSEA_GOBP/volcanoplot_rGenes_GSEA_GOBP_", pert, "_", spec, "_", time, ".pdf"),
  plot = p, units = "in", height = 4, width = 6, bg = "white", dpi = 600
)

print(p)
```


## rGenes volcano plots

```{r rGene volcano plot all time points}

```

```{r rGene volcano plot split by time points}

```



Define a function for rGene volcano plot
```{r rGene volcano plot function}
rDEGs_volcano_plot <- function(response_df, type, pert, spec, top_n) {

  # Filter data to the current cell type, perturbation, and species.
  filt_df <- response_df %>% filter(assay == type & perturbation == pert & species == spec)

  # Skip iteration if there are no results.
  if (nrow(filt_df) == 0) return(NULL)

  # Determine the global max x and y values for setting the same scale axes.
  max_x_val <- max(abs(filt_df$logFC), na.rm = TRUE)
  max_y_val <- max(-(log10(filt_df$adj.P.Val)), na.rm = TRUE)

  # Identify top n lowest p.adjust values for both positive and negative logFC values per time point.
  top_up_labeled <- filt_df %>% filter(logFC > 0 & significant == TRUE) %>%
    group_by(time_point) %>%
    slice_min(order_by = adj.P.Val, n = top_n, with_ties = FALSE) %>%
    ungroup()

  top_down_labeled <- filt_df %>% filter(logFC < 0 & significant == TRUE) %>%
    group_by(time_point) %>%
    slice_min(order_by = adj.P.Val, n = top_n, with_ties = FALSE) %>%
    ungroup()

  top_labeled <- bind_rows(top_up_labeled, top_down_labeled)

  # Plot a combined volcano plot
  p <- ggplot(filt_df, aes(x = logFC, y = -log10(adj.P.Val))) +

    geom_point(data = filt_df,
               aes(fill = simple_signaling_category),
               color = 'black',
               shape = 21,
               stroke = 0.25,
               alpha = 0.75) +

    # Data labels at moved positions
    geom_text_repel(
      data = top_labeled,
      aes(x = logFC, y = -log10(adj.P.Val), label = ID),
      size = 3,
      segment.color = 'black',
      min.segment.length = 0.5
    ) + 

    # Define scales
    scale_x_continuous(limits = c(-max_x_val * 1.1, max_x_val * 1.1)) +
    scale_y_continuous(limits = c(0, max_y_val * 1.1)) +
    scale_fill_manual(values = signaling_category_colors, name = 'Signaling Category',
                      guide = guide_legend(title = "Signaling Category",
                                           override.aes = list(shape = 21, size = 3, stroke=0.50, color = 'black'))) +

    # Facet by time point
    facet_wrap(~ time_point) +

    # Labels and theme
    labs(
      title = paste0("Dreamlet Results: ", type, " ", pert, " ", spec),
      x = expression(~ Log[2] ~ "(Stimulus/Vehicle)"),
      y = expression("-" ~ Log[10] ~ "(Adjusted p-value)")
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 24),
      strip.text = element_text(hjust = 0.5, size = 18),
      plot.background = element_blank(),
      panel.background = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.line = element_line(color = "black", linewidth = 1),
      legend.position = "right"
    ) +

    # Significance lines
    geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "darkgrey", linewidth = 1) +
    geom_vline(xintercept = 0, linetype = "dashed", color = "darkgrey", linewidth = 1)

  return(p)
}
```

Plot rGene volcano plots
```{r rGenes volcano plots}
# Iterate through cell type x perturbation x species and create one combined plot per combination
for (type in unique(response_df$assay)){
  for (pert in unique(response_df$perturbation)){
    for (spec in unique(response_df$species)){
      p_combined <- rDEGs_volcano_plot(response_df, type, pert, spec, 5)
      if (!is.null(p_combined)) {
        ggsave(filename = paste0(fig_path, paste0("/volcanoplot_rDEGs/volcanoplot_rDEGs_", type, "_", pert, "_", spec, ".png")),
               plot = p_combined, height = 5, width = 15, bg = 'white')
      }
    }
  }
}
```

### Inspect specific response GSEA results

```{r}
filt_df <- response_gsea_df %>% filter(perturbation == 'BMP7', time_point %in% c('6hr', '54hr'), p.adjust < 0.10)
write.csv(filt_df, file = paste0(dir_path, "response_GSEA_GOBP_BMP7_6hr_54hr_df.csv"), row.names = FALSE)
```




## Response scatter plots human vs chimp

```{r scatter plot human vs chimp response colored by perturbation}
# Filter data frame to perturbations of interest
filt_df <- response_logfc_df %>% dplyr::filter(!(perturbation %in% c('DMSO', 'TreMan')))

# Define limit and pad for plotting
lim <- max(abs(c(filt_df$Chimp_logFC, filt_df$Human_logFC)), na.rm = TRUE)
pad <- 0.05 * lim
  
# Calculate Pearson's correlation coefficient
pearson_res <- cor.test(filt_df$Chimp_logFC, filt_df$Human_logFC,, method = "pearson")
pearson_r <- unname(pearson_res$estimate)   # correlation coefficient
pearson_p <- pearson_res$p.value
  
# Plot a scatter plot
p <- ggplot(filt_df, aes(x = Chimp_logFC, y = Human_logFC, fill = perturbation, shape = time_point)) +
  geom_point(size = 2, stroke=0.25, color='black') +
  labs(
    x = expression("Chimp " ~Log[2]~"(Stimulus/Vehicle)"),
    y = expression("Human " ~Log[2]~"(Stimulus/Vehicle)"),
    color = "Stimulus",
    shape = "Time Point"
  ) +
  scale_fill_manual(name = "Stimulus", values = pert_colors,
    guide = guide_legend(ncol=2, override.aes = list(shape=21, size=4, color="black", stroke=0.5))
  ) + 
  scale_shape_manual(name = "Time Point", values = time_point_shapes,
    guide = guide_legend(override.aes = list(fill="black", size=4))
  ) +
  theme_minimal() + 
  theme(
    legend.position = "right",  # Place legend on the right
    axis.line = element_line(color = "black"),  # Add axis lines
    panel.grid.major = element_blank(),  # Remove major gridlines
    panel.grid.minor = element_blank(),   # Remove minor gridlines
    axis.text = element_text(size=14)
    ) +
  
  # lock both axes to [-lim, lim] and remove padding so the line hits the axes exactly at (0,0)
  scale_x_continuous(limits = c(-lim - pad, lim + pad), expand = expansion(mult = 0)) +
  scale_y_continuous(limits = c(-lim - pad, lim + pad), expand = expansion(mult = 0)) +
  coord_equal() +     # square aspect ratio
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") + # Add dashed y=x line
  geom_hline(yintercept = 0, color = "black") +  # Horizontal line at y = 0
  geom_vline(xintercept = 0, color = "black") # Vertical line at x = 0
  annotate(
    "text",
    x = min(filt_df$Human_logFC, na.rm = TRUE),
    y = max(filt_df$Chimp_logFC, na.rm = TRUE),
    hjust = 0, vjust = 1,
    label = paste0("r = ", sprintf("%.2f", pearson_r)),
    size = 4
  )

# Save and plot
ggsave(filename = paste0(fig_path, "scatterplot_HumanVsChimp_Response/scatterplot_HumanVsChimp_Response_ColorByPerturbation.pdf"),
           plot = p, height = 4.5, width = 6, bg = 'white', dpi=600, units='in')
print(p)
```

```{r scatter plot human vs chimp response colored by rGene category}
# Filter data frame to perturbations of interest and rGene category subsets 
filt_df <- response_logfc_df %>% filter(!(perturbation %in% c('DMSO', 'TreMan')))
non_df <- filt_df %>% filter(rGene_category == 'Non-rGene')
other_df <- filt_df %>% filter(rGene_category == 'Other rGene')
cons_df <- filt_df %>% filter(rGene_category == 'Conserved rGene')
div_df <- filt_df %>% filter(rGene_category == 'Divergent rGene')

# Define limit and pad for plotting
lim <- max(abs(c(filt_df$Chimp_logFC, filt_df$Human_logFC)), na.rm = TRUE)
pad <- 0.05 * lim
  
# Plot a scatter plot
p <- ggplot(filt_df, aes(x = Chimp_logFC, y = Human_logFC, fill = rGene_category, shape = time_point)) +
  geom_point(data = non_df, size = 2, alpha = 0.7, na.rm = TRUE, stroke = 0.25, color = 'black') +
  geom_point(data = other_df, size = 2, alpha = 0.7, na.rm = TRUE, stroke = 0.25, color = 'black') +
  geom_point(data = cons_df, size = 2, alpha = 0.8, na.rm = TRUE, stroke = 0.25, color = 'black') +
  geom_point(data = div_df, size = 2, alpha = 0.9, na.rm = TRUE, stroke = 0.25, color = 'black') +
  labs(
    x = expression("Chimp " ~Log[2]~"(Stimulus/Vehicle)"),
    y = expression("Human " ~Log[2]~"(Stimulus/Vehicle)"),
    color = "rGene Category",
    shape = "Time Point"
  ) +
  scale_fill_manual(name = "rGene Category", values = rgene_colors,
    guide = guide_legend(ncol=1, override.aes = list(shape=21, size=4, color="black", stroke=0.5))
  ) + 
  scale_shape_manual(name = "Time Point", values = time_point_shapes,
    guide = guide_legend(override.aes = list(fill="black", size=4))
  ) +
  theme_minimal() + 
  theme(
    legend.position = "right",  # Place legend on the right
    axis.line = element_line(color = "black"),  # Add axis lines
    panel.grid.major = element_blank(),  # Remove major gridlines
    panel.grid.minor = element_blank(),   # Remove minor gridlines
    axis.text = element_text(size=14)
    ) +
  
  # lock both axes to [-lim, lim] and remove padding so the line hits the axes exactly at (0,0)
  scale_x_continuous(limits = c(-lim - pad, lim + pad), expand = expansion(mult = 0)) +
  scale_y_continuous(limits = c(-lim - pad, lim + pad), expand = expansion(mult = 0)) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") + # Add dashed y=x line
  coord_fixed(ratio = 1) +  # Ensure the aspect ratio is 1:1 for square plot
  geom_hline(yintercept = 0, color = "black") +  # Horizontal line at y = 0
  geom_vline(xintercept = 0, color = "black") # Vertical line at x = 0

# Save and plot
ggsave(filename = paste0(fig_path, "scatterplot_HumanVsChimp_Response/scatterplot_HumanVsChimp_Response_ColorByrGeneCategory.pdf"),
           plot = p, height = 4.5, width = 5, bg = 'white', dpi=600, units='in')
print(p)
```

## Pie chart rGene categories

```{r pie chart GxP categories}
pie_df <- response_logfc_df %>%
  filter(!(perturbation %in% c("DMSO", "TreMan"))) %>%
  mutate(rGene_sig = coalesce(Human_rGene_sig, FALSE) | coalesce(Chimp_rGene_sig, FALSE)) %>%
  dplyr::count(rGene_sig, name = "n", sort = TRUE) %>%
  mutate(
    frac      = n / sum(n),
    ymax      = cumsum(frac),
    ymin      = lag(ymax, default = 0),
    ymid      = (ymin + ymax) / 2,
    pct_label = scales::percent(frac, accuracy = 0.1),
    label     = paste0(pct_label, "\n(n = ", scales::comma(n), ")"),
    rGene     = ifelse(rGene_sig, "rGene", "Non-rGene")
  )

# Plot pie chart
p <- ggplot(pie_df, aes(ymax = ymax, ymin = ymin, xmax = 1, xmin = 0, fill = rGene)) +
  geom_rect(color = "white") +
  coord_polar(theta = "y") +
  scale_fill_manual(
    values = c("rGene" = "purple", "Non-rGene" = "black"),
    name = "Gene x Stimulus Category",
    guide = guide_legend(nrow = 2, byrow = TRUE, title.position = 'top')
  ) +
  xlim(c(0, 2.0)) +
  theme_void() +
  theme(
    legend.justification = "center",
    legend.position = "bottom",
    legend.title = element_text(size = 14, hjust = 0.5),
    legend.title.align = 0.5,
    legend.text  = element_text(size = 12),
    legend.text.align = 0.5,
    legend.box.just = "center",
    legend.box = "vertical"
  ) +
  geom_text_repel(
    data = pie_df,
    aes(x = 1.2, y = ymid, label = label),
    inherit.aes = FALSE,
    nudge_x = 0.025,
    direction = "y",
    vjust = 0.5,
    size = 4,
    segment.size = 0.3,
    segment.color = "grey50",
    show.legend = FALSE
  )

  ggsave(filename = paste0(fig_path, "PieChart_GeneXStimulus_ColorByrGeneNonrGene.pdf"),
         plot = p, height = 4, width = 3, bg = 'white')
  
print(p)
```

## Pie charts rGenes

```{r pie chart rGene categories}
pie_df <- response_logfc_df %>% filter(rGene_category != 'Non-rGene') %>%
  mutate(rGene_category = as.character(rGene_category)) %>%
  dplyr::count(rGene_category, name = "n", sort = TRUE) %>%
  arrange(desc(rGene_category)) %>%
  mutate(frac = n / sum(n),
         ymax = cumsum(frac),
         ymin = dplyr::lag(ymax, default = 0),
         ymid = (ymin + ymax) / 2,                                 
         pct_label = scales::percent(frac, accuracy = 0.1),
         label     = paste0(pct_label, "\n(n = ", scales::comma(n), ")"),
         )

# Plot pie chart
p <- ggplot(pie_df, aes(ymax = ymax, ymin = ymin, xmax = 1, xmin = 0, fill = rGene_category)) +
  geom_rect(color = "white") +
  coord_polar(theta = "y") +
  scale_fill_manual(values = rgene_colors,
                    name = "rGene Category",
                    guide = guide_legend(nrow = 3, byrow = TRUE, title.position = 'top')
                    ) +
  xlim(c(0, 2.0)) +
  theme_void() +
  theme(
    legend.justification = "center",
    legend.position = "bottom",
    legend.title = element_text(size = 14, hjust = 0.5),
    legend.title.align = 0.5,
    legend.text  = element_text(size = 12),
    legend.text.align = 0.5,
    legend.box.just = "center",
    legend.box = "vertical"
  ) +
  geom_text_repel(
    data = pie_df,
    aes(x = 1.5, y = ymid, label = label),
    inherit.aes = FALSE,
    nudge_x = 0.025,
    direction = "y",
    vjust = 0.5,
    size = 4,
    segment.size = 0.3,
    segment.color = "grey50",
    show.legend = FALSE
  )

  ggsave(filename = paste0(fig_path, "PieChart_rGenes_ColorByrGeneCategories.pdf"),
         plot = p, height = 4, width = 3, bg = 'white', units = 'in', dpi=600)
  
print(p)
```

## Stacked bar plot rGene categories across time points

```{r}
# Filter data and set factor levels
filt_df <- response_logfc_df %>%
  filter(
    !(perturbation %in% c('DMSO', 'TreMan')),
    !is.na(rGene_category),
    rGene_category != "Non-rGene"
  ) %>%
  mutate(
    rGene_category = factor(rGene_category, levels = c("Conserved rGene", "Divergent rGene", "Other rGene")),
    time_point = factor(time_point, levels = c("6hr", "54hr", "7day"))
  )

# Order perturbation levels by decreasing number of rGenes
order_levels <- filt_df %>%
  dplyr::count(perturbation, time_point, name = "n") %>%
  pivot_wider(names_from = time_point, values_from = n, values_fill = 0) %>%
  arrange(desc(`6hr`), desc(`54hr`), desc(`7day`)) %>%
  pull(perturbation)

# Apply perturbation ordering
filt_df <- filt_df %>%
  mutate(perturbation = factor(perturbation, levels = rev(order_levels)))

# Horizontal stacked bar plot with facets
p <- ggplot(filt_df, aes(x = perturbation, fill = rGene_category)) +
  geom_bar() +
  coord_flip() +                              
  facet_wrap(~ time_point, nrow = 1) +
  scale_fill_manual(
    values   = rgene_colors,
    na.value = 'black',
    drop     = FALSE
  ) +
  labs(
    x    = "",                     
    y    = "N rGenes",
    fill = "rGene Category"
  ) +
  theme_minimal() +
  theme(
    strip.text          = element_text(size = 14),
    axis.text.y         = element_text(size = 12),
    axis.text.x         = element_text(size = 12),
    plot.background     = element_blank(),
    panel.background    = element_blank(),
    panel.grid.major    = element_blank(),
    panel.grid.minor    = element_blank(),
    legend.title        = element_text(size = 14),
    legend.text        = element_text(size = 12)
  )

# Save and print
ggsave(filename = file.path(fig_path, "stacked_barplot_NrGenes_byPerturbation_ColorByrGeneCategory_horizontal_facet.pdf"),
  plot = p, width = 12, height = 6, bg = 'white', dpi=600)

print(p)
```

## Inspect specific rGenes

```{r inspect specific rGenes}
r_df <- response_df %>% dplyr::filter(#significant == TRUE
                               #& abs(logFC) > 1
                               perturbation == 'BMP7'
                               #& time_point %in% c('6hr', '54hr', '7day')
                               & ID %in% c('NRG3', 'ERBB4')
                               )
View(r_df)
```


# Conserved rGene Analysis

## N conserved rGenes across time points box plot

```{r box plot N rGenes across time points}
plot_df <- conserved_rgenes_df %>% filter(!(perturbation %in% c('DMSO', 'TreMan'))) %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))
                                  
p <- ggplot(plot_df, aes(x = time_point, y = N_conserved_rGenes, fill=time_point)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.15, alpha = 0.6, size = 1) +
  scale_fill_manual(values = time_point_colors, name = "Time point") +
  labs(
    x = "",
    y = "N rGenes"
  ) +
  theme_classic() +
  theme(
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
  )
ggsave(filename = paste0(fig_path, "boxplot_NrGenes_ByTimePoint.png"), plot = p, height = 4, width = 3, bg = 'white', dpi=600)
print(p)
```

## N conserved rGenes over time by perturbation connected scatter plot

```{r scatter plot log2 N rGenes over time}
# Set factor levels and filter to data of interest
plot_df <- conserved_rgenes_df %>%
  mutate(time_point = factor(time_point, levels = c("6hr", "54hr", "7day")))

# Plot scatter plot across time points
p <- ggplot(plot_df, aes(x = time_point, y = N_conserved_rGenes, color = perturbation, group = perturbation)) +
  geom_point(size = 3) + # Scatter plot.
  geom_line(size = 1) +  # Connecting lines.
  labs(
    x = "Time Point",
    y = "N Conserved rGenes",
    color = "Stimulus"
    ) +
    scale_color_manual(values = pert_colors) +  # Add custom color map.
    scale_x_discrete(expand = c(0.02, 0.02)) +  # Edit padding around the x-axis
    theme_minimal() +                     # Clean theme
    theme(plot.title = element_text(hjust=0.5, size=16), # Edit plot title.
        plot.background = element_blank(),  # Remove the background around the plot.
        panel.background = element_blank(),  # Remove the background inside the plot.
        panel.grid.major = element_blank(),  # Remove major gridlines
        panel.grid.minor = element_blank(),   # Remove minor gridlines
        axis.text  = element_text(size = 12),
        axis.title = element_text(size = 14),
        legend.position = "right", # Edit legend location.
        axis.line = element_line(color = "black", size = 0.5),  # Add axis lines
        axis.ticks = element_line(color = "black", size = 0.5)  # Add axis ticks
    )
ggsave(filename = paste0(fig_path, "connectedscatterplot_NConservedrGenes_ByPerturbation_ByTimePoint.png"),
         plot = p, height = 5, width = 7, bg = 'white')
print(p)
```

## N unique conserved rGenes across time points box plot

```{r box plot N unique conserved rGenes across time points}
# Find number of conserved rGenes within a perturbation that are unique to a time point
unique_cons_rgenes_df <- conserved_rgenes_df %>%
  # 1) explode "/"-separated gene lists into long format
  tidyr::separate_rows(conserved_rGenes, sep = "/") %>%
  dplyr::mutate(
    conserved_rGenes = stringr::str_trim(conserved_rGenes),
    time_point = as.character(time_point),
    perturbation = as.character(perturbation)
  ) %>%
  dplyr::filter(!is.na(conserved_rGenes), conserved_rGenes != "") %>%
  dplyr::distinct(perturbation, time_point, gene = conserved_rGenes) %>%
  
  # 2) for each perturbation+gene, check how many time points it appears in
  dplyr::group_by(perturbation, gene) %>%
  dplyr::mutate(n_timepoints_in_pert = dplyr::n_distinct(time_point)) %>%
  dplyr::ungroup() %>%
  
  # 3) keep genes that occur in exactly 1 time point within that perturbation
  dplyr::filter(n_timepoints_in_pert == 1) %>%
  
  # 4) count unique genes per perturbation x time_point
  dplyr::count(perturbation, time_point, name = "N_unique_conserved_rGenes") %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))
unique_cons_rgenes_df


# Plot a box plot of unique rGenes within perturbation per time point                                  
p <- ggplot(unique_cons_rgenes_df, aes(x = time_point, y = N_unique_conserved_rGenes, fill=time_point)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.15, alpha = 0.6, size = 1) +
  scale_fill_manual(values = time_point_colors,
                    name = "Time Point",
                    guide = 'none'
                    ) +
  labs(
    x = "",
    y = "N Unique Conserved rGenes\n(Within Stimulus)"
  ) +
  theme_classic() +
  theme(
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    legend.position = "none",
  )
ggsave(filename = paste0(fig_path, "boxplot_NUniqueConservedrGenes_AcrossTimePointsWithinPerturbation_ByTimePoint.pdf"),
       plot = p, height = 4, width = 2.25, bg = 'white', dpi=600)
print(p)
```

## Conserved rGenes upset plot

Plot an upset plot of overlapping conserved rGenes 
```{r upset plot simple rGenes}
binary_df <- simple_conserved_rgenes_df %>%
  dplyr::filter(!perturbation %in% c("TreMan", "DMSO")) %>%
  tidyr::separate_rows(conserved_rGenes, sep = "/") %>%
  dplyr::mutate(conserved_rGenes = trimws(conserved_rGenes)) %>%
  dplyr::filter(!is.na(conserved_rGenes), conserved_rGenes != "") %>%
  dplyr::distinct(gene = conserved_rGenes, perturbation) %>%
  dplyr::mutate(present = 1L) %>%
  tidyr::pivot_wider(
    names_from  = perturbation,
    values_from = present,
    values_fill = 0L
  )

set_cols <- setdiff(names(binary_df), "gene")

# --- NEW: compute intersections with > 25 genes ---
mat <- as.matrix(binary_df[, set_cols, drop = FALSE])

# Build a key per gene that identifies the exact intersection (e.g., "BMP7&FGF2")
intersection_key <- apply(mat, 1, function(v) {
  on <- set_cols[as.logical(v)]
  if (length(on) == 0) return(NA_character_)  # ignore genes in no sets
  paste(on, collapse = "&")
})

intersection_counts <- sort(table(intersection_key, useNA = "no"), decreasing = TRUE)

# Keep only intersection patterns with > 25 genes
keep_keys <- names(intersection_counts)[intersection_counts > 25]

# Convert to UpSetR format: list of character vectors
intersections_keep <- lapply(keep_keys, function(k) strsplit(k, "&", fixed = TRUE)[[1]])

# Optional: if you only want TRUE intersections (>=2 sets), uncomment:
# intersections_keep <- intersections_keep[vapply(intersections_keep, length, integer(1)) >= 2]

png(
  paste0(fig_path, "upsetplot_simple_conserved_rGenes.png"),
  width = 18, height = 8, units = "in", res = 600
)

if (length(intersections_keep) == 0) {
  warning("No intersections with > 25 genes found ensured; plotting default UpSetR output instead.")
  UpSetR::upset(
    as.data.frame(binary_df),
    sets = set_cols,
    keep.order = FALSE,
    sets.bar.color = "gray20",
    order.by = "freq",
    mb.ratio = c(0.35, 0.65),
    sets.x.label = "Set Size",
    mainbar.y.label = "Intersection Size",
    decreasing = c(TRUE, NA)
  )
} else {
  UpSetR::upset(
    as.data.frame(binary_df),
    sets = set_cols,
    intersections = intersections_keep,  # <-- restrict plotted intersections here
    keep.order = FALSE,
    sets.bar.color = "gray20",
    order.by = "freq",
    mb.ratio = c(0.35, 0.65),
    sets.x.label = "Set Size",
    mainbar.y.label = "Intersection Size",
    decreasing = c(TRUE, NA),
    text.scale = c(2.5, #intersection_size_title,   
                   2.0, #intersection_size_ticks,
                   2.5, #set_size_title,
                   2.0, #set_size_ticks,
                   2.0  #set_names
    ),
    show.numbers = FALSE
  )
}
dev.off()
```

## Simple conserved rGenes ORA GO:BP

Define a function for performing ORA with a simple conserved rGenes data frame as input and a results data frame as output.
```{r simple ora_df function}
simple_ora_df <- function(degs_df, results_df, degs_list_colname, species_name, category, subcategory) {
  # Initialize an empty list for storing the ORA results.
  ora_results_list <- list()
  
  # Iterate through perturbations.
  for (pert in unique(degs_df$perturbation)) {
    
    # Filter and pull DEGs from specified column
    degs_raw <- degs_df %>% filter(perturbation == pert) %>%
      pull(!!sym(degs_list_colname))
        
    # Flatten and split stringified gene lists safely
    degs <- degs_raw %>% na.omit() %>% unlist() %>% strsplit(split = "/") %>% unlist() %>% trimws() %>% unique()

    # Skip if empty or all NAs
    if (length(degs) == 0 || all(is.na(degs)) || all(degs == "")) next

    # Remove empty or NA values before mapping
    degs_clean <- degs[!is.na(degs) & degs != ""]
        
    # Check valid SYMBOL keys
    valid_symbols <- keys(org.Hs.eg.db, keytype = "SYMBOL")
    degs_clean <- intersect(degs_clean, valid_symbols)
        
    # Skip iteration if no valid symbols
    if (length(degs_clean) == 0) next

    # Convert gene symbols to ENTREZ IDs
    degs_entrez <- as.character(na.omit(mapIds(
      org.Hs.eg.db,
      keys = degs_clean,
      column = "ENTREZID",
      keytype = "SYMBOL",
      multiVals = "first"
      )))
      
    # Skip iteration if no valid IDs.
    if (length(degs_entrez) == 0) next

    # Define universe of genes
    universe_entrez <- as.character(na.omit(mapIds(org.Hs.eg.db, results_df$ID, "ENTREZID", "SYMBOL")))

    # Get gene sets for enrichment analysis
    gene_set <- msigdbr(species = species_name, category = category, subcategory = subcategory)
    gene_set_df <- gene_set %>% dplyr::select(gs_name, entrez_gene)

    # Perform ORA
    ora <- tryCatch({
      enricher(
        gene = degs_entrez,
        universe = universe_entrez,
        TERM2GENE = gene_set_df,
        pAdjustMethod = "fdr",
        pvalueCutoff = 0.05,
        qvalueCutoff = 0.10,
        minGSSize = 5,
        maxGSSize = 500
        )
     }, error = function(e) NULL)

    # Check if ORA returned results
    if (!is.null(ora) && nrow(ora@result) > 0) {
      ora_results <- ora@result %>%
        mutate(perturbation = pert) %>%
        tidyr::separate(GeneRatio, into = c("numerator", "denominator"), sep = "/", remove = FALSE) %>%
        mutate(
          GeneRatioOriginal = GeneRatio,
          GeneRatio = as.numeric(numerator) / as.numeric(denominator)
        ) %>%
          dplyr::select(-numerator, -denominator) %>%
          filter(p.adjust < 0.05 & qvalue < 0.10 & Count > 1)

        ora_results_list[[pert]] <- ora_results
      }
    }
  
  # Combine all ORA results
  simple_ora_df <- dplyr::bind_rows(ora_results_list)
  return(simple_ora_df)
}
```

Perform GO:BP term ORA of conserved up rGenes
```{r crDEGs simple_ora_df, message=FALSE, warning=FALSE}
simple_up_ora_df <- simple_ora_df(simple_conserved_rgenes_df, results_df, 'up_conserved_rGenes', 'Homo sapiens', 'C5', 'GO:BP')
head(simple_up_ora_df)
```

Perform GO:BP term ORA of conserved down rGenes
```{r crGenes simple_ora_df, message=TRUE, warning=TRUE}
simple_down_ora_df <- simple_ora_df(simple_conserved_rgenes_df, results_df, 'down_conserved_rGenes', 'Homo sapiens', 'C5', 'GO:BP')
head(simple_down_ora_df)
```

Combine conserved up and down rGene GO:BP ORA results
```{r all crGenes simple_ora_df}
simple_up_ora_df <- simple_up_ora_df %>% mutate(direction = "Up")
simple_down_ora_df <- simple_down_ora_df %>% mutate(direction = "Down")
simple_ora_df <- bind_rows(simple_up_ora_df, simple_down_ora_df)
simple_ora_df$perturbation_direction <- paste(simple_ora_df$perturbation, simple_ora_df$direction, sep = "_")
head(simple_ora_df)
```

Save simple conserved rGenes GO:BP ORA data frame
```{r save simple crGenes ora_df}
saveRDS(simple_ora_df, file = paste0(dir_path, "simple_conserved_rGenes_GOBP_ORA_df.rds"))
```

Load simple conserved rGenes GO:BP ORA data frame
```{r load simple crGenes ora_df}
simple_ora_df <- readRDS(file = paste0(dir_path, "simple_conserved_rGenes_GOBP_ORA_df.rds"))
head(simple_ora_df)
```

```{r add GO:BP metaterms}
# Patterns
neural_terms           <- c("NEURON","NERVOUS_SYSTEM","NEUROGENESIS","NEURAL","NEURONAL","RADIAL_GLIA",
                            "ACTION_POTENTIAL","SYNAPTIC","SYNAPSE","NEUROBLAST","AXON",
                            "BRAIN","TELENCEPHALON","DIENCEPHALON","FOREBRAIN","CEREBRAL_CORTEX","CORTEX",
                            "SUBPALLIUM","PALLIUM","SPINAL_CORD","MIDBRAIN","HINDBRAIN")
patterning_terms       <- c("PATTERN","PATTERNING","AXIS_ELONGATION","AXIS_SPECIFICATION",
                            "PATTERN_SPECIFICATION","PATTERN_FORMATION")
developmental_terms    <- c("DEVELOPMENT","DEVELOPMENTAL")
differentiation_terms  <- c("DIFFERENTIATION")
morph_terms            <- c("MORPHOGENESIS")
prolif_growth_cc_terms <- c("GROWTH","PROLIFERATION","CELL_CYCLE","MITOTIC","MEIOTIC")

rx_neural            <- str_c(neural_terms, collapse = "|")
rx_patterning        <- str_c(patterning_terms, collapse = "|")
rx_developmental     <- str_c(developmental_terms, collapse = "|")
rx_differentiation   <- str_c(differentiation_terms, collapse = "|")
rx_morph             <- str_c(morph_terms, collapse = "|")
rx_prolif_growth_cc  <- str_c(prolif_growth_cc_terms, collapse = "|")

has_pat <- function(x, patt) str_detect(x, regex(patt, ignore_case = TRUE))

# Add flags
simple_ora_df <- simple_ora_df %>%
  mutate(
    Neural           = has_pat(Description, rx_neural),
    Patterning       = has_pat(Description, rx_patterning),
    Development      = has_pat(Description, rx_developmental),
    Differentiation  = has_pat(Description, rx_differentiation),
    Morphogenesis    = has_pat(Description, rx_morph),
    Prolif_Growth_CC = has_pat(Description, rx_prolif_growth_cc)
  ) %>%
  rowwise() %>%
  mutate(
    matches = list(names(which(c(
      Neural           = Neural,
      Patterning       = Patterning,
      Development      = Development,
      Differentiation  = Differentiation,
      Morphogenesis    = Morphogenesis,
      Prolif_Growth_CC = Prolif_Growth_CC
    ))))
  ) %>%
  ungroup() %>%
  mutate(
    n_matches = lengths(matches),
    Metaterm = dplyr::case_when(
      n_matches == 0 ~ "Other",
      n_matches == 1 ~ vapply(matches, function(x) dplyr::coalesce(x[1], "Other"), character(1)),
      n_matches  > 1 ~ "Multiple metaterms",
      TRUE          ~ "Other"
    )
  )
```

Save simple conserved rGenes GO:BP ORA data frame
```{r save simple crGenes ora_df}
saveRDS(simple_ora_df, file = paste0(dir_path, "simple_conserved_rGenes_GOBP_ORA_df.rds"))
```

```{r}
write.xlsx(simple_ora_df, file = paste0(dir_path, "simple_conserved_rGenes_ORA_GOBP.xlsx"), rowNames = FALSE)
```



```{r count metaterms}
# Count # of terms for each Metaterm within each pert x time x direction group
metaterm_counts_df <- simple_ora_df %>%
  filter(!is.na(Metaterm), !is.na(ID), ID != "") %>%
  group_by(perturbation, direction, Metaterm) %>%
  summarise(N_terms = n(), .groups = "drop")

head(metaterm_counts_df)
```

```{r diverging stacked bar plot GO:BP term counts colored by metaterm}
library(forcats)
library(scales)

# Make a fill palette where "Other" is gray
metaterm_levels <- sort(unique(metaterm_counts_df$Metaterm))
metaterm_pal <- setNames(hue_pal()(length(metaterm_levels)), metaterm_levels)
metaterm_pal["Other"] <- "black"

# Order perturbation by total terms (Up + Down), descending
plot_df <- metaterm_counts_df %>%
  filter(!(perturbation %in% c('TreMan', 'DMSO'))) %>%
  mutate(N_terms_signed = if_else(direction == "Down", -N_terms, N_terms)) %>%
  group_by(perturbation) %>%
  mutate(N_total = sum(N_terms)) %>%   # Up + Down (since N_terms is positive counts)
  ungroup() %>%
  mutate(
    perturbation = fct_reorder(perturbation, N_total, .desc = FALSE),
    Metaterm = factor(Metaterm,
      levels = c('Development', 'Differentiation', 'Morphogenesis', 'Neural', 'Patterning', 'Prolif_Growth_CC', 'Multiple metaterms', 'Other'))
    )

p <- ggplot(plot_df, aes(y = perturbation, x = N_terms_signed, fill = Metaterm)) +
  geom_col(width = 0.8) +
  geom_vline(xintercept = 0) +
  scale_x_continuous(labels = function(x) abs(x)) +
  scale_fill_manual(values = metaterm_pal, drop = FALSE) +
  labs(x = "N GO:BP Terms", y = NULL, fill = "Metaterm") +
  theme_classic() +
  theme(
    axis.text = element_text(size=12),
    axis.title = element_text(size=14),
    legend.title        = element_text(size = 14),
    legend.text        = element_text(size = 12)
  )

# Save and print
ggsave(filename = file.path(fig_path, "horizontal_stacked_barplot_NGOBPTerms_DivergingDirection_ColorByMetaTerm.png"),
  plot = p, width = 6, height = 4, bg = 'white', dpi=600, units='in')

print(p)
```

## Conserved rGenes GO:BP ORA

Define a function for performing ORA with a conserved rGenes data frame as input and an ORA results data frame as output.
```{r conserved rGenes ORA function}
ora_df <- function(rgenes_df, results_df, rgenes_df_colname, species_name, category, subcategory) {
  # Initialize an empty list for storing the ORA results
  ora_results_list <- list()
  
  # Iterate through perturbation x time point
  for (pert in unique(rgenes_df$perturbation)) {
    for (time in unique(rgenes_df$time_point)) {

      # Filter and pull DEGs from specified column
      genes_raw <- rgenes_df %>% filter(perturbation == pert, time_point == time) %>%
        pull(!!sym(rgenes_df_colname))
        
      # Flatten and split stringified gene lists safely
      genes <- genes_raw %>% na.omit() %>% unlist() %>% strsplit(split = "/") %>% unlist() %>% trimws() %>% unique()

        # Skip if empty or all NAs
        if (length(genes) == 0 || all(is.na(genes)) || all(genes == "")) next

        # Remove empty or NA values before mapping
        genes_clean <- genes[!is.na(genes) & genes != ""]
        
        # Check valid SYMBOL keys
        valid_symbols <- keys(org.Hs.eg.db, keytype = "SYMBOL")
        genes_clean <- intersect(genes_clean, valid_symbols)
        
        # Skip iteration if no valid symbols
        if (length(genes_clean) == 0) next

        # Convert gene symbols to ENTREZ IDs
        genes_entrez <- as.character(na.omit(mapIds(
          org.Hs.eg.db,
          keys = genes_clean,
          column = "ENTREZID",
          keytype = "SYMBOL",
          multiVals = "first"
        )))
        
        if (length(genes_entrez) == 0) next

        # Define universe of genes
        universe_entrez <- as.character(na.omit(mapIds(org.Hs.eg.db, results_df$ID, "ENTREZID", "SYMBOL")))

        # Get gene sets for enrichment analysis
        gene_set <- msigdbr(species = species_name, category = category, subcategory = subcategory)
        gene_set_df <- gene_set %>% dplyr::select(gs_name, entrez_gene)

        # Perform ORA
        ora <- tryCatch({
          enricher(
            gene = genes_entrez,
            universe = universe_entrez,
            TERM2GENE = gene_set_df,
            pAdjustMethod = "fdr",
            pvalueCutoff = 0.05,
            qvalueCutoff = 0.10,
            minGSSize = 5,
            maxGSSize = 500
          )
        }, error = function(e) NULL)

        # Check if ORA returned results
        if (!is.null(ora) && nrow(ora@result) > 0) {
          ora_results <- ora@result %>%
            mutate(
              perturbation = pert,
              time_point = time
            ) %>%
            tidyr::separate(GeneRatio, into = c("numerator", "denominator"), sep = "/", remove = FALSE) %>%
            mutate(
              GeneRatioOriginal = GeneRatio,
              GeneRatio = as.numeric(numerator) / as.numeric(denominator)
            ) %>%
            dplyr::select(-numerator, -denominator) %>%
            filter(p.adjust < 0.05 & qvalue < 0.10 & Count > 1)

          ora_results_list[[paste(pert, time, sep = "_")]] <- ora_results
      }
    }
  }
  
  # Combine all ORA results
  ora_results_df <- dplyr::bind_rows(ora_results_list)
  return(ora_results_df)
}
```

Perform GO:BP term ORA of upregulated conserved rGenes
```{r up_ora_df, message=FALSE, warning=FALSE}
up_ora_df <- ora_df(conserved_rgenes_df, results_df, 'up_conserved_rGenes', 'Homo sapiens', 'C5', 'GO:BP')
head(up_ora_df)
```

Perform GO:BP term ORA of downregulated conserved rGenes
```{r down_ora_df}
down_ora_df <- ora_df(conserved_rgenes_df, results_df, 'down_conserved_rGenes', 'Homo sapiens', 'C5', 'GO:BP')
head(down_ora_df)
```

Combine conserved up and down rGene GO:BP ORA results
```{r all crGenes ora_df}
up_ora_df <- up_ora_df %>% mutate(direction = "up")
down_ora_df <- down_ora_df %>% mutate(direction = "down")
all_ora_df <- bind_rows(up_ora_df, down_ora_df)
all_ora_df$perturbation_direction <- paste(all_ora_df$perturbation, all_ora_df$direction, sep = "_")
head(all_ora_df)
```

```{r}
# Function to map ENTREZ IDs to gene symbols
entrez_to_symbol <- function(entrez_str) {
  ids <- unlist(str_split(entrez_str, "/"))
  symbols <- mapIds(org.Hs.eg.db ,keys = ids, column = "SYMBOL", keytype = "ENTREZID", multiVals = "first")
  paste(symbols, collapse = "/")
}
```
  
```{r}
# Apply function to new column
all_ora_df <- all_ora_df %>%
  rowwise() %>%
  mutate(geneSymbol = entrez_to_symbol(geneID)) %>%
  ungroup()
```

Save conserved rGenes GO:BP ORA data frame
```{r save crGenes ora_df}
saveRDS(all_ora_df, file = paste0(dir_path, "conserved_rGenes_ORA_GOBP_df.rds"))
```

```{r}
write.xlsx(all_ora_df, file = paste0(dir_path, "conserved_rGenes_ORA_GOBP_df.xlsx"), rowNames = FALSE)
```

Load conserved rGenes ORA GO:BP data frame
```{r load all_ora_df}
all_ora_df <- readRDS(file = paste0(dir_path, "conserved_rGenes_ORA_GOBP_df.rds"))
head(all_ora_df)
```

## Annotate conserved rGenes

```{r find_overlap function}
# Function to extract overlapping genes from DEGs
find_overlap <- function(degs_str, gene_list) {
  degs_split <- unlist(strsplit(degs_str, "/"))
  overlap <- degs_split[degs_split %in% gene_list]
  if (length(overlap) > 0) paste(overlap, collapse = "/") else NA
}
```

```{r annotate TFs}
# Load human TFs
data(dorothea_hs, package = "dorothea")
tf_genes <- dorothea_hs %>%
  pull(tf) %>%
  unique()

# Annotate df with TF overlaps
for (tf in tf_genes) {
  conserved_rgenes_df$TF_crGenes <- sapply(conserved_rgenes_df$conserved_rGenes, find_overlap, gene_list = tf_genes)
}

# Annotate df with TF overlaps
for (tf in tf_genes) {
  conserved_rgenes_df$up_TF_crGenes <- sapply(conserved_rgenes_df$up_conserved_rGenes, find_overlap, gene_list = tf_genes)
}

# Annotate df with TF overlaps
for (tf in tf_genes) {
  conserved_rgenes_df$down_TF_crGenes <- sapply(conserved_rgenes_df$down_conserved_rGenes, find_overlap, gene_list = tf_genes)
}


head(conserved_rgenes_df)
```

```{r annotate signaling genes}
# Load gene list
signaling_genes_df <- readRDS(file = paste0(dir_path, "signaling_genes_df.rds"))
signaling_genes <- signaling_genes_df$ID

# Annotate df with NDD overlaps
for (gene in signaling_genes) {
  conserved_rgenes_df$signaling_crGenes <- sapply(conserved_rgenes_df$conserved_rGenes, find_overlap, gene_list = tf_genes)
}

head(conserved_rgenes_df)
```

```{r save convergent_response_degs_df}
saveRDS(conserved_rgenes_df, paste0(dir_path, "conserved_rGenes_df.rds"))
```

## Conserved rGene pert x term network plot

```{r signaling_gobp_terms}
signaling_gobp_terms <- c(

# Growth factor
"GOBP_RESPONSE_TO_GROWTH_FACTOR",
"GOBP_REGULATION_OF_RESPONSE_TO_GROWTH_FACTOR",
"GOBP_POSITIVE_REGULATION_OF_RESPONSE_TO_GROWTH_FACTOR",
"GOBP_NEGATIVE_REGULATION_OF_RESPONSE_TO_GROWTH_FACTOR",

"GOBP_CELLULAR_RESPONSE_TO_GROWTH_FACTOR_STIMULUS",
"GOBP_REGULATION_OF_CELLULAR_RESPONSE_TO_GROWTH_FACTOR_STIMULUS",
"GOBP_POSITIVE_REGULATION_OF_CELLULAR_RESPONSE_TO_GROWTH_FACTOR_STIMULUS",
"GOBP_NEGATIVE_REGULATION_OF_CELLULAR_RESPONSE_TO_GROWTH_FACTOR_STIMULUS",

# FGF
"GOBP_RESPONSE_TO_FIBROBLAST_GROWTH_FACTOR",
"GOBP_CELLULAR_RESPONSE_TO_FIBROBLAST_GROWTH_FACTOR",

"GOBP_FIBROBLAST_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_FIBROBLAST_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_FIBROBLAST_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_FIBROBLAST_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",

# MAPK ERK
"GOBP_MAPK_CASCADE",
"GOBP_REGULATION_OF_MAPK_CASCADE",
"GOBP_POSITIVE_REGULATION_OF_MAPK_CASCADE",
"GOBP_NEGATIVE_REGULATION_OF_MAPK_CASCADE",

"GOBP_MAP_KINASE_ACTIVITY",
"GOBP_REGULATION_OF_MAP_KINASE_ACTIVITY",
"GOBP_POSITIVE_REGULATION_OF_MAP_KINASE_ACTIVITY",
"GOBP_NEGATIVE_REGULATION_OF_MAP_KINASE_ACTIVITY",

"GOBP_ERK1_AND_ERK2_CASCADE",
"GOBP_REGULATION_OF_ERK1_AND_ERK2_CASCADE",
"GOBP_POSITIVE_REGULATION_OF_ERK1_AND_ERK2_CASCADE",
"GOBP_NEGATIVE_REGULATION_OF_ERK1_AND_ERK2_CASCADE",

"GOBP_NEGATIVE_REGULATION_OF_RAS_PROTEIN_SIGNAL_TRANSDUCTION",

# BMP
"GOBP_RESPONSE_TO_BMP",
"GOBP_CELLULAR_RESPONSE_TO_BMP_STIMULUS",

"GOBP_BMP_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_BMP_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_BMP_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_BMP_SIGNALING_PATHWAY",

# TGFB
#"GOBP_RESPONSE_TO_TRANSFORMING_GROWTH_FACTOR_BETA",

#"GOBP_CELLULAR_RESPONSE_TO_TRANSFORMING_GROWTH_FACTOR_BETA_STIMULUS",
#"GOBP_REGULATION_OF_CELLULAR_RESPONSE_TO_TRANSFORMING_GROWTH_FACTOR_BETA_STIMULUS",
#"GOBP_POSITIVE_REGULATION_OF_CELLULAR_RESPONSE_TO_TRANSFORMING_GROWTH_FACTOR_BETA_STIMULUS",
#"GOBP_NEGATIVE_REGULATION_OF_CELLULAR_RESPONSE_TO_TRANSFORMING_GROWTH_FACTOR_BETA_STIMULUS",

#"GOBP_TRANSFORMING_GROWTH_FACTOR_BETA_RECEPTOR_SIGNALING_PATHWAY",
#"GOBP_REGULATION_OF_TRANSFORMING_GROWTH_FACTOR_BETA_RECEPTOR_SIGNALING_PATHWAY",
#"GOBP_POSITIVE_REGULATION_OF_TRANSFORMING_GROWTH_FACTOR_BETA_RECEPTOR_SIGNALING_PATHWAY",
#"GOBP_NEGATIVE_REGULATION_OF_TRANSFORMING_GROWTH_FACTOR_BETA_RECEPTOR_SIGNALING_PATHWAY",

# SMAD
"GOBP_SMAD_PROTEIN_SIGNAL_TRANSDUCTION",
"GOBP_REGULATION_OF_SMAD_PROTEIN_SIGNAL_TRANSDUCTION",
"GOBP_POSITIVE_REGULATION_OF_SMAD_PROTEIN_SIGNAL_TRANSDUCTION",
"GOBP_NEGATIVE_REGULATION_OF_SMAD_PROTEIN_SIGNAL_TRANSDUCTION",

# EGF
"GOBP_EPIDERMAL_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_EPIDERMAL_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_EPIDERMAL_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_EPIDERMAL_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY",

'GOBP_RESPONSE_TO_EPIDERMAL_GROWTH_FACTOR',
'GOBP_CELLULAR_RESPONSE_TO_EPIDERMAL_GROWTH_FACTOR_STIMULUS',

'GOBP_REGULATION_OF_EPIDERMAL_GROWTH_FACTOR_ACTIVATED_RECEPTOR_ACTIVITY',
'GOBP_POSITIVE_REGULATION_OF_EPIDERMAL_GROWTH_FACTOR_ACTIVATED_RECEPTOR_ACTIVITY',
'GOBP_NEGATIVE_REGULATION_OF_EPIDERMAL_GROWTH_FACTOR_ACTIVATED_RECEPTOR_ACTIVITY',

'GOBP_ERBB2_EGFR_SIGNALING_PATHWAY',
'GOBP_ERBB4_EGFR_SIGNALING_PATHWAY',

# ERBB
'GOBP_ERBB_SIGNALING_PATHWAY',
'GOBP_REGULATION_OF_ERBB_SIGNALING_PATHWAY',
'GOBP_NEGATIVE_REGULATION_OF_ERBB_SIGNALING_PATHWAY',
'GOBP_POSITIVE_REGULATION_OF_ERBB_SIGNALING_PATHWAY',

'GOBP_ERBB3_SIGNALING_PATHWAY',
'GOBP_REGULATION_OF_ERBB3_SIGNALING_PATHWAY',
'GOBP_POSITIVE_REGULATION_OF_ERBB3_SIGNALING_PATHWAY',
'GOBP_NEGATIVE_REGULATION_OF_ERBB3_SIGNALING_PATHWAY',

'GOBP_ERBB2_SIGNALING_PATHWAY',
'GOBP_ERBB4_SIGNALING_PATHWAY',

'GOBP_ERBB2_ERBB3_SIGNALING_PATHWAY',
'GOBP_ERBB2_ERBB4_SIGNALING_PATHWAY',
'GOBP_ERBB3_ERBB4_SIGNALING_PATHWAY',
'GOBP_ERBB4_ERBB4_SIGNALING_PATHWAY',

# mTOR
"GOBP_TOR_SIGNALING",
"GOBP_REGULATION_OF_TOR_SIGNALING",
"GOBP_POSITIVE_REGULATION_OF_TOR_SIGNALING",
"GOBP_NEGATIVE_REGULATION_OF_TOR_SIGNALING",

"GOBP_TORC1_SIGNALING",
"GOBP_REGULATION_OF_TORC1_SIGNALING",
"GOBP_POSITIVE_REGULATION_OF_TORC1_SIGNALING",
"GOBP_NEGATIVE_REGULATION_OF_TORC1_SIGNALING",

"GOBP_TORC2_SIGNALING",
"GOBP_REGULATION_OF_TORC2_SIGNALING",
"GOBP_POSITIVE_REGULATION_OF_TORC2_SIGNALING",
"GOBP_NEGATIVE_REGULATION_OF_TORC2_SIGNALING",

# IGF
'GOBP_INSULIN_LIKE_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY',
'GOBP_REGULATION_OF_INSULIN_LIKE_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY',
'GOBP_POSITIVE_REGULATION_OF_INSULIN_LIKE_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY',
'GOBP_NEGATIVE_REGULATION_OF_INSULIN_LIKE_GROWTH_FACTOR_RECEPTOR_SIGNALING_PATHWAY',

'GOBP_RESPONSE_TO_INSULIN_LIKE_GROWTH_FACTOR_STIMULUS',
'GOBP_CELLULAR_RESPONSE_TO_INSULIN_LIKE_GROWTH_FACTOR_STIMULUS',

# Rapamycin
'GOBP_RESPONSE_TO_RAPAMYCIN',
'GOBP_CELLULAR_RESPONSE_TO_RAPAMYCIN',

# HIPPO
"GOBP_HIPPO_SIGNALING",
"GOBP_REGULATION_OF_HIPPO_SIGNALING",
"GOBP_POSITIVE_REGULATION_OF_HIPPO_SIGNALING",
"GOBP_NEGATIVE_REGULATION_OF_HIPPO_SIGNALING",

# NOTCH
"GOBP_NOTCH_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_NOTCH_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_NOTCH_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_NOTCH_SIGNALING_PATHWAY",

"GOBP_TRANSCRIPTION_OF_NOTCH_RECEPTOR_TARGET",
"GOBP_REGULATION_OF_TRANSCRIPTION_OF_NOTCH_RECEPTOR_TARGET",
"GOBP_POSITIVE_REGULATION_OF_TRANSCRIPTION_OF_NOTCH_RECEPTOR_TARGET",
"GOBP_NEGATIVE_REGULATION_OF_TRANSCRIPTION_OF_NOTCH_RECEPTOR_TARGET",

# RA
"GOBP_RESPONSE_TO_RETINOIC_ACID",
"GOBP_CELLULAR_RESPONSE_TO_RETINOIC_ACID",

"GOBP_RETINOIC_ACID_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_RETINOIC_ACID_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_RETINOIC_ACID_RECEPTOR_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_RETINOIC_ACID_RECEPTOR_SIGNALING_PATHWAY",

# WNT
"GOBP_WNT_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_WNT_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_WNT_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_WNT_SIGNALING_PATHWAY",

"GOBP_CANONICAL_WNT_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_CANONICAL_WNT_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_CANONICAL_WNT_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_CANONICAL_WNT_SIGNALING_PATHWAY",

"GOBP_NON_CANONICAL_WNT_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_NON_CANONICAL_WNT_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_NON_CANONICAL_WNT_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_NON_CANONICAL_WNT_SIGNALING_PATHWAY",

"GOBP_WNT_SIGNALING_PATHWAY_PLANAR_CELL_POLARITY_PATHWAY",
"GOBP_REGULATION_OF_WNT_SIGNALING_PATHWAY_PLANAR_CELL_POLARITY_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_WNT_SIGNALING_PATHWAY_PLANAR_CELL_POLARITY_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_WNT_SIGNALING_PATHWAY_PLANAR_CELL_POLARITY_PATHWAY",

# SHH
"GOBP_SMOOTHENED_SIGNALING_PATHWAY",
"GOBP_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY",
"GOBP_POSITIVE_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY",
"GOBP_NEGATIVE_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY",

"GOBP_SMOOTHENED_SIGNALING_PATHWAY_INVOLVED_IN_DORSAL_VENTRAL_NEURAL_TUBE_PATTERNING",
"GOBP_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY_INVOLVED_IN_DORSAL_VENTRAL_NEURAL_TUBE_PATTERNING",
"GOBP_POSITIVE_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY_INVOLVED_IN_DORSAL_VENTRAL_NEURAL_TUBE_PATTERNING",
"GOBP_NEGATIVE_REGULATION_OF_SMOOTHENED_SIGNALING_PATHWAY_INVOLVED_IN_DORSAL_VENTRAL_NEURAL_TUBE_PATTERNING"
)
```

```{r}
library(dplyr)
library(stringr)
library(tidyr)
library(igraph)
library(ggraph)
library(ggplot2)
library(grid)     # for stringWidth/unit conversions
library(ggrepel)  # for geom_text_repel

# Load conserved rGene ORA GO:BP results data frame 
all_ora_df <- readRDS(file = paste0(dir_path, "conserved_rGenes_ORA_GOBP_df.rds"))

# Clean up GOBP term labels
all_ora_df <- all_ora_df %>%
  mutate(
    Description = Description %>%
      str_remove("^GOBP_") %>%
      str_replace_all("NEGATIVE", "(-)") %>%
      str_replace_all("POSITIVE", "(+)") %>%
      str_replace_all("SIGNALING", "SIG.") %>%
      str_replace_all("PATHWAY", "PATH.") %>%
      str_replace_all("TRANSDUCTION", "TRANS.") %>%
      str_replace_all("REGULATION", "REG.") %>%
      str_replace_all("RETINOIC_ACID", "RA") %>%
      str_replace_all("FIBROBLAST_GROWTH_FACTOR", "FGF") %>%
      str_replace_all("PLANAR_CELL_POLARITY", "PCP") %>%
      str_replace_all("ERK1_AND_ERK2", "ERK1/2") %>%
      str_replace_all("CELLULAR", "CELL") %>%
      str_replace_all("GROWTH_FACTOR_STIMULUS", "GROWTH_FACTOR") %>%
      str_replace_all("_", " ")
  )

# Iterate over time points
for (tp in sort(unique(all_ora_df$time_point))) {

  edges <- all_ora_df %>%
    filter(time_point == tp, !is.na(ID), ID != "", ID %in% signaling_gobp_terms) %>%
    transmute(perturbation, Description, direction) %>%
    mutate(direction = tolower(direction)) %>%
    filter(direction %in% c("up", "down")) %>%
    distinct(perturbation, Description, direction) %>%
    tidyr::drop_na() %>%
    mutate(
      edge_color = case_when(
        direction == "up"   ~ "darkgreen",
        direction == "down" ~ "firebrick4",
        TRUE ~ "grey70"
      )
    )

  if (nrow(edges) == 0L) {
    message(sprintf("No edges for time_point = %s  skipping.", tp))
    next
  }

  g <- graph_from_data_frame(d = edges, directed = FALSE)
  E(g)$edge_color <- edges$edge_color
  E(g)$direction  <- edges$direction

  V(g)$type <- ifelse(V(g)$name %in% edges$perturbation, "perturbation", "term")

  # Degree
  deg <- as.numeric(igraph::degree(g))
  names(deg) <- V(g)$name
  V(g)$deg <- deg[V(g)$name]

  # Colors
  pert_cols <- unname(pert_colors[V(g)$name])
  pert_cols[is.na(pert_cols)] <- "gray30"

  term_cols <- dplyr::case_when(
    V(g)$deg == 1 ~ "gray80",
    V(g)$deg == 2 ~ "gray50",
    TRUE          ~ "black"
  )

  V(g)$color <- ifelse(V(g)$type == "perturbation", pert_cols, term_cols)

  # Shapes
  V(g)$shape <- ifelse(V(g)$type == "perturbation", "square", "circle")

  # Constant perturbation node size, large enough to fit text
  pert_labels <- V(g)$name[V(g)$type == "perturbation"]
  pert_fontsize <- 8

  max_label_mm <- if (length(pert_labels) > 0) {
    max(vapply(
      pert_labels,
      function(s) convertWidth(stringWidth(s), "mm", valueOnly = TRUE),
      numeric(1)
    ))
  } else 0

  padding_mm <- 6
  pert_node_size <- max(12, (max_label_mm + padding_mm) / 3.5)

  # Sizes: perturbations constant; terms scale with degree
  V(g)$size <- ifelse(V(g)$type == "perturbation", pert_node_size, 3 + V(g)$deg)

  # Layout
  set.seed(11)
  lay <- ggraph::create_layout(g, layout = "fr")

  # Dummy data to force a legend for term-degree colors
  deg_legend_df <- data.frame(
    x = 0, y = 0,
    term_deg_col = c("gray80", "gray50", "black")
  )

  p <- ggraph(lay) +
    geom_edge_link0(
      aes(edge_colour = edge_color),
      edge_width = 0.8,
      edge_alpha = 0.8
    ) +
    ggraph::scale_edge_colour_identity(
      name   = "Enrichment\nDirection",
      breaks = c("firebrick4", "darkgreen"),
      labels = c("Down", "Up"),
      guide  = "legend"
    ) +
    guides(edge_colour = guide_legend(override.aes = list(edge_width = 2, edge_alpha = 1))) +

    # Nodes (hide legends here)
    geom_node_point(aes(color = color, shape = shape, size = size), show.legend = FALSE) +
    scale_shape_identity() +
    scale_size_identity(guide = "none") +

    # Invisible points to create the "degree" legend
    ggplot2::geom_point(
      data = deg_legend_df,
      aes(x = x, y = y, color = term_deg_col),
      inherit.aes = FALSE,
      alpha = 0,
      size = 4,
      show.legend = TRUE
    ) +

    # Identity color scale with legend for the 3 term-degree colors
    scale_color_identity(
      name   = "Degree of Overlaps",
      breaks = c("gray80", "gray50", "black"),
      labels = c("1", "2", "3+"),
      guide  = guide_legend(override.aes = list(alpha = 1, size = 4, shape = 16))
    ) +

    # Invisible "repel obstacles" at perturbation nodes
    ggplot2::geom_point(
      data = dplyr::filter(lay, type == "perturbation"),
      aes(x = x, y = y),
      inherit.aes = FALSE,
      alpha = 0,
      size = 18
    ) +

    # Term labels: repel (avoid each other + avoid perturbation nodes)
    ggrepel::geom_text_repel(
      data = dplyr::filter(lay, type == "term"),
      aes(x = x, y = y, label = name),
      inherit.aes = FALSE,
      size = 4,
      box.padding = 1.0,
      point.padding = 1.2,
      force = 15,
      max.overlaps = Inf,
      seed = 11,
      segment.size = 0.25,
      segment.alpha = 0.6,
      segment.color = "grey60"
    ) +

    # Perturbation labels  fixed, centered
    geom_node_text(
      data = dplyr::filter(lay, type == "perturbation"),
      aes(x = x, y = y, label = name),
      inherit.aes = FALSE,
      repel = FALSE,
      size = pert_fontsize, fontface = "bold", colour = "black",
      vjust = 0.5, hjust = 0.5
    ) +

    ggtitle(paste0("Perturbation x GO:BP Term Network ", tp)) +
    theme_void() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 16),
      legend.position = c(0, 1),
      legend.justification = c(0, 1),
      legend.box = "vertical",
      legend.title = element_text(size = 14),
      legend.text = element_text(size = 12),
      legend.background = element_rect(fill = "white", colour = NA),
      legend.key = element_rect(fill = "white", colour = NA)
    )

  ggsave(
    paste0(fig_path, "networkgraph_PertXTerm_", tp, ".pdf"),
    plot = p, height = 9, width = 12, units = "in", bg = "white"
  )

  print(p)
}
```


## Conserved rGene pert x gene network plot

```{r conserved signaling rGene PxG network graph}
library(dplyr)
library(tidyr)
library(stringr)
library(igraph)
library(ggraph)
library(ggplot2)
library(ggrepel)
library(grid)

# Expect these to exist:
# conserved_rgenes_df with cols: perturbation, time_point, up_conserved_rGenes, down_conserved_rGenes
# pert_colors named vector: names are perturbations
# fig_path output directory

# Iterate over time points
for (tp in sort(unique(conserved_rgenes_df$time_point))) {

  # Build edge list (Perturbation -> Gene) from up/down conserved rGenes, with direction
  edges <- conserved_rgenes_df %>%
    dplyr::filter(time_point == tp) %>%
    dplyr::select(perturbation, time_point, up_conserved_rGenes, down_conserved_rGenes) %>%
    tidyr::pivot_longer(
      cols = c(up_conserved_rGenes, down_conserved_rGenes),
      names_to = "dir_col",
      values_to = "gene_list"
    ) %>%
    dplyr::mutate(
      direction = dplyr::case_when(
        dir_col == "up_conserved_rGenes"   ~ "up",
        dir_col == "down_conserved_rGenes" ~ "down",
        TRUE ~ NA_character_
      )
    ) %>%
    dplyr::filter(!is.na(gene_list), gene_list != "", !is.na(direction)) %>%
    tidyr::separate_rows(gene_list, sep = "/") %>%
    dplyr::mutate(gene = stringr::str_trim(gene_list)) %>%
    dplyr::filter(!is.na(gene), gene != "", gene %in% signaling_genes_df$ID) %>%
    dplyr::distinct(perturbation, gene, direction) %>%
    dplyr::mutate(
      edge_color = dplyr::case_when(
        direction == "up"   ~ "darkgreen",
        direction == "down" ~ "firebrick4",
        TRUE ~ "grey70"
      )
    ) %>%
    dplyr::select(from = perturbation, to = gene, direction, edge_color)

  if (nrow(edges) == 0L) {
    message(sprintf("No edges for time_point = %s  skipping.", tp))
    next
  }

  # Create graph
  g <- igraph::graph_from_data_frame(edges, directed = FALSE)
  igraph::E(g)$edge_color <- edges$edge_color
  igraph::E(g)$direction  <- edges$direction

  # Node typing
  pert_nodes <- unique(edges$from)
  igraph::V(g)$type <- ifelse(igraph::V(g)$name %in% pert_nodes, "perturbation", "gene")

  # Degree (for gene nodes)
  deg <- as.numeric(igraph::degree(g))
  names(deg) <- igraph::V(g)$name
  igraph::V(g)$deg <- deg[igraph::V(g)$name]

  # Node colors
  pert_cols <- unname(pert_colors[igraph::V(g)$name])
  pert_cols[is.na(pert_cols)] <- "gray30"

  gene_cols <- dplyr::case_when(
    igraph::V(g)$deg == 1 ~ "gray80",
    igraph::V(g)$deg == 2 ~ "gray50",
    TRUE                  ~ "black"
  )

  igraph::V(g)$color <- ifelse(igraph::V(g)$type == "perturbation", pert_cols, gene_cols)

  # Shapes
  igraph::V(g)$shape <- ifelse(igraph::V(g)$type == "perturbation", "square", "circle")

  # Constant perturbation node size, big enough to fit text
  pert_labels <- igraph::V(g)$name[igraph::V(g)$type == "perturbation"]
  pert_fontsize <- 8
  max_label_mm <- if (length(pert_labels) > 0) {
    max(vapply(
      pert_labels,
      function(s) grid::convertWidth(grid::stringWidth(s), "mm", valueOnly = TRUE),
      numeric(1)
    ))
  } else 0
  padding_mm <- 6
  pert_node_size <- max(12, (max_label_mm + padding_mm) / 3.5)

  # Node sizes: perturbations constant; genes scale with degree
  igraph::V(g)$size <- ifelse(igraph::V(g)$type == "perturbation", pert_node_size, 3 + igraph::V(g)$deg)

  # Layout once (lets us use x/y for ggrepel + obstacle points)
  set.seed(11)
  lay <- ggraph::create_layout(g, layout = "fr")

  # Dummy data to force a legend for gene-degree colors
  deg_legend_df <- data.frame(
    x = 0, y = 0,
    gene_deg_col = c("gray80", "gray50", "black")
  )

  # Plot
  p <- ggraph(lay) +
    ggraph::geom_edge_link0(
      aes(edge_colour = edge_color),
      edge_width = 0.8,
      edge_alpha = 0.8
    ) +
    ggraph::scale_edge_colour_identity(
      name   = "Direction",
      breaks = c("firebrick4", "darkgreen"),
      labels = c("Down", "Up"),
      guide  = "legend"
    ) +
    guides(edge_colour = guide_legend(override.aes = list(edge_width = 2, edge_alpha = 1))) +

    # Nodes (hide legend for node aesthetics)
    ggraph::geom_node_point(aes(color = color, shape = shape, size = size), show.legend = FALSE) +
    scale_shape_identity() +
    scale_size_identity(guide = "none") +

    # Invisible points to create the "degree" legend (gene-degree colors)
    ggplot2::geom_point(
      data = deg_legend_df,
      aes(x = x, y = y, color = gene_deg_col),
      inherit.aes = FALSE,
      alpha = 0,
      size = 4,
      show.legend = TRUE
    ) +
    scale_color_identity(
      name   = "Gene degree",
      breaks = c("gray80", "gray50", "black"),
      labels = c("1", "2", "3+"),
      guide  = guide_legend(override.aes = list(alpha = 1, size = 4, shape = 16))
    ) +

    # Invisible "repel obstacles" at perturbation nodes to keep gene labels off squares
    ggplot2::geom_point(
      data = dplyr::filter(lay, type == "perturbation"),
      aes(x = x, y = y),
      inherit.aes = FALSE,
      alpha = 0,
      size = 18
    ) +

    # Gene labels  repel from each other + avoid perturbation nodes
    ggrepel::geom_text_repel(
      data = dplyr::filter(lay, type == "gene"),
      aes(x = x, y = y, label = name),
      inherit.aes = FALSE,
      size = 4,
      box.padding = 0.5,
      point.padding = 1.0,
      force = 5,
      max.overlaps = Inf,
      seed = 11,
      segment.size = 0.25,
      segment.alpha = 0.6,
      segment.color = "grey60"
    ) +

    # Perturbation labels  centered
    ggraph::geom_node_text(
      data = dplyr::filter(lay, type == "perturbation"),
      aes(x = x, y = y, label = name),
      inherit.aes = FALSE,
      repel = FALSE,
      size = pert_fontsize, fontface = "bold", colour = "black",
      vjust = 0.5, hjust = 0.5
    ) +

    ggtitle(paste0("Perturbation x Conserved Signaling rGene Network ", tp)) +
    theme_void() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 16),
      legend.position = c(0, 1),
      legend.justification = c(0, 1),
      legend.box = "vertical",
      legend.title = element_text(size = 14),
      legend.text = element_text(size = 12),
      legend.background = element_rect(fill = "white", colour = NA),
      legend.key = element_rect(fill = "white", colour = NA)
    )

  ggsave(
    paste0(fig_path, "networkgraph_PertXSignalingGene_", tp, ".pdf"),
    plot = p, height = 12, width = 12, units = "in", bg = "white"
  )

  print(p)
}
```

```{r}
library(dplyr)
library(tidyr)
library(stringr)
library(igraph)
library(ggraph)
library(ggplot2)
library(ggrepel)
library(grid)

# Expect these to exist:
# conserved_rgenes_df with cols: perturbation, time_point, up_conserved_rGenes, down_conserved_rGenes
# pert_colors named vector: names are perturbations
# fig_path output directory

# Iterate over time points
for (tp in sort(unique(conserved_rgenes_df$time_point))) {

  # Build edge list (Perturbation -> Gene) from up/down conserved rGenes, with direction
  edges <- conserved_rgenes_df %>%
    dplyr::filter(time_point == tp) %>%
    dplyr::select(perturbation, time_point, up_TF_crGenes, down_TF_crGenes) %>%
    tidyr::pivot_longer(
      cols = c(up_TF_crGenes, down_TF_crGenes),
      names_to = "dir_col",
      values_to = "gene_list"
    ) %>%
    dplyr::mutate(
      direction = dplyr::case_when(
        dir_col == "up_TF_crGenes"   ~ "up",
        dir_col == "down_TF_crGenes" ~ "down",
        TRUE ~ NA_character_
      )
    ) %>%
    dplyr::filter(!is.na(gene_list), gene_list != "", !is.na(direction)) %>%
    tidyr::separate_rows(gene_list, sep = "/") %>%
    dplyr::mutate(gene = stringr::str_trim(gene_list)) %>%
    dplyr::filter(!is.na(gene), gene != "") %>%
    dplyr::distinct(perturbation, gene, direction) %>%
    dplyr::mutate(
      edge_color = dplyr::case_when(
        direction == "up"   ~ "darkgreen",
        direction == "down" ~ "firebrick4",
        TRUE ~ "grey70"
      )
    ) %>%
    dplyr::select(from = perturbation, to = gene, direction, edge_color)

  if (nrow(edges) == 0L) {
    message(sprintf("No edges for time_point = %s  skipping.", tp))
    next
  }

  # Create graph
  g <- igraph::graph_from_data_frame(edges, directed = FALSE)
  igraph::E(g)$edge_color <- edges$edge_color
  igraph::E(g)$direction  <- edges$direction

  # Node typing
  pert_nodes <- unique(edges$from)
  igraph::V(g)$type <- ifelse(igraph::V(g)$name %in% pert_nodes, "perturbation", "gene")

  # Degree (for gene nodes)
  deg <- as.numeric(igraph::degree(g))
  names(deg) <- igraph::V(g)$name
  igraph::V(g)$deg <- deg[igraph::V(g)$name]

  # Node colors
  pert_cols <- unname(pert_colors[igraph::V(g)$name])
  pert_cols[is.na(pert_cols)] <- "gray30"

  gene_cols <- dplyr::case_when(
    igraph::V(g)$deg == 1 ~ "gray80",
    igraph::V(g)$deg == 2 ~ "gray50",
    TRUE                  ~ "black"
  )

  igraph::V(g)$color <- ifelse(igraph::V(g)$type == "perturbation", pert_cols, gene_cols)

  # Shapes
  igraph::V(g)$shape <- ifelse(igraph::V(g)$type == "perturbation", "square", "circle")

  # Constant perturbation node size, big enough to fit text
  pert_labels <- igraph::V(g)$name[igraph::V(g)$type == "perturbation"]
  pert_fontsize <- 8
  max_label_mm <- if (length(pert_labels) > 0) {
    max(vapply(
      pert_labels,
      function(s) grid::convertWidth(grid::stringWidth(s), "mm", valueOnly = TRUE),
      numeric(1)
    ))
  } else 0
  padding_mm <- 6
  pert_node_size <- max(12, (max_label_mm + padding_mm) / 3.5)

  # Node sizes: perturbations constant; genes scale with degree
  igraph::V(g)$size <- ifelse(igraph::V(g)$type == "perturbation", pert_node_size, 3 + igraph::V(g)$deg)

  # Layout once (lets us use x/y for ggrepel + obstacle points)
  set.seed(11)
  lay <- ggraph::create_layout(g, layout = "fr")

  # Dummy data to force a legend for gene-degree colors
  deg_legend_df <- data.frame(
    x = 0, y = 0,
    gene_deg_col = c("gray80", "gray50", "black")
  )

  # Plot
  p <- ggraph(lay) +
    ggraph::geom_edge_link0(
      aes(edge_colour = edge_color),
      edge_width = 0.8,
      edge_alpha = 0.8
    ) +
    ggraph::scale_edge_colour_identity(
      name   = "Direction",
      breaks = c("firebrick4", "darkgreen"),
      labels = c("Down", "Up"),
      guide  = "legend"
    ) +
    guides(edge_colour = guide_legend(override.aes = list(edge_width = 2, edge_alpha = 1))) +

    # Nodes (hide legend for node aesthetics)
    ggraph::geom_node_point(aes(color = color, shape = shape, size = size), show.legend = FALSE) +
    scale_shape_identity() +
    scale_size_identity(guide = "none") +

    # Invisible points to create the "degree" legend (gene-degree colors)
    ggplot2::geom_point(
      data = deg_legend_df,
      aes(x = x, y = y, color = gene_deg_col),
      inherit.aes = FALSE,
      alpha = 0,
      size = 4,
      show.legend = TRUE
    ) +
    scale_color_identity(
      name   = "Gene degree",
      breaks = c("gray80", "gray50", "black"),
      labels = c("1", "2", "3+"),
      guide  = guide_legend(override.aes = list(alpha = 1, size = 4, shape = 16))
    ) +

    # Invisible "repel obstacles" at perturbation nodes to keep gene labels off squares
    ggplot2::geom_point(
      data = dplyr::filter(lay, type == "perturbation"),
      aes(x = x, y = y),
      inherit.aes = FALSE,
      alpha = 0,
      size = 18
    ) +

    # Gene labels  repel from each other + avoid perturbation nodes
    ggrepel::geom_text_repel(
      data = dplyr::filter(lay, type == "gene"),
      aes(x = x, y = y, label = name),
      inherit.aes = FALSE,
      size = 4,
      box.padding = 0.5,
      point.padding = 1.0,
      force = 5,
      max.overlaps = Inf,
      seed = 11,
      segment.size = 0.25,
      segment.alpha = 0.6,
      segment.color = "grey60"
    ) +

    # Perturbation labels  centered
    ggraph::geom_node_text(
      data = dplyr::filter(lay, type == "perturbation"),
      aes(x = x, y = y, label = name),
      inherit.aes = FALSE,
      repel = FALSE,
      size = pert_fontsize, fontface = "bold", colour = "black",
      vjust = 0.5, hjust = 0.5
    ) +

    ggtitle(paste0("Perturbation x Conserved TF rGene Network ", tp)) +
    theme_void() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 16),
      legend.position = c(0, 1),
      legend.justification = c(0, 1),
      legend.box = "vertical",
      legend.title = element_text(size = 14),
      legend.text = element_text(size = 12),
      legend.background = element_rect(fill = "white", colour = NA),
      legend.key = element_rect(fill = "white", colour = NA)
    )

  ggsave(
    paste0(fig_path, "networkgraph_PertXGeneTF_", tp, ".pdf"),
    plot = p, height = 12, width = 12, units = "in", bg = "white"
  )

  print(p)
}
```

```{r}
library(dplyr)
library(tidyr)
library(stringr)
library(igraph)
library(ggraph)
library(ggplot2)
library(ggrepel)
library(grid)

# Expect these to exist:
# conserved_rgenes_df with cols: perturbation, time_point, up_conserved_rGenes, down_conserved_rGenes
# pert_colors named vector: names are perturbations
# fig_path output directory

# Iterate over time points
for (tp in sort(unique(conserved_rgenes_df$time_point))) {

  # Build edge list (Perturbation -> Gene) from up/down conserved rGenes, with direction
  edges <- conserved_rgenes_df %>%
    dplyr::filter(time_point == tp, !(perturbation %in% c('DLL1', 'HX531', 'Rapamycin', 'SANT1'))) %>%
    dplyr::select(perturbation, time_point, up_TF_crGenes, down_TF_crGenes) %>%
    tidyr::pivot_longer(
      cols = c(up_TF_crGenes, down_TF_crGenes),
      names_to = "dir_col",
      values_to = "gene_list"
    ) %>%
    dplyr::mutate(
      direction = dplyr::case_when(
        dir_col == "up_TF_crGenes"   ~ "up",
        dir_col == "down_TF_crGenes" ~ "down",
        TRUE ~ NA_character_
      )
    ) %>%
    dplyr::filter(!is.na(gene), gene != "", gene %in% signaling_genes_df$ID) %>%
    tidyr::separate_rows(gene_list, sep = "/") %>%
    dplyr::mutate(gene = stringr::str_trim(gene_list)) %>%
    dplyr::filter(!is.na(gene), gene != "") %>%
    dplyr::distinct(perturbation, gene, direction) %>%
    dplyr::mutate(
      edge_color = dplyr::case_when(
        direction == "up"   ~ "darkgreen",
        direction == "down" ~ "firebrick4",
        TRUE ~ "grey70"
      )
    ) %>%
    dplyr::select(from = perturbation, to = gene, direction, edge_color)

  if (nrow(edges) == 0L) {
    message(sprintf("No edges for time_point = %s  skipping.", tp))
    next
  }

  # Create graph
  g <- igraph::graph_from_data_frame(edges, directed = FALSE)
  igraph::E(g)$edge_color <- edges$edge_color
  igraph::E(g)$direction  <- edges$direction

  # Node typing
  pert_nodes <- unique(edges$from)
  igraph::V(g)$type <- ifelse(igraph::V(g)$name %in% pert_nodes, "perturbation", "gene")

  # Degree (for gene nodes)
  deg <- as.numeric(igraph::degree(g))
  names(deg) <- igraph::V(g)$name
  igraph::V(g)$deg <- deg[igraph::V(g)$name]

  # Node colors
  pert_cols <- unname(pert_colors[igraph::V(g)$name])
  pert_cols[is.na(pert_cols)] <- "gray30"

  gene_cols <- dplyr::case_when(
    igraph::V(g)$deg == 1 ~ "gray80",
    igraph::V(g)$deg == 2 ~ "gray50",
    TRUE                  ~ "black"
  )

  igraph::V(g)$color <- ifelse(igraph::V(g)$type == "perturbation", pert_cols, gene_cols)

  # Shapes
  igraph::V(g)$shape <- ifelse(igraph::V(g)$type == "perturbation", "square", "circle")

  # Constant perturbation node size, big enough to fit text
  pert_labels <- igraph::V(g)$name[igraph::V(g)$type == "perturbation"]
  pert_fontsize <- 8
  max_label_mm <- if (length(pert_labels) > 0) {
    max(vapply(
      pert_labels,
      function(s) grid::convertWidth(grid::stringWidth(s), "mm", valueOnly = TRUE),
      numeric(1)
    ))
  } else 0
  padding_mm <- 6
  pert_node_size <- max(12, (max_label_mm + padding_mm) / 3.5)

  # Node sizes: perturbations constant; genes scale with degree
  igraph::V(g)$size <- ifelse(igraph::V(g)$type == "perturbation", pert_node_size, 3 + igraph::V(g)$deg)

  # Layout once (lets us use x/y for ggrepel + obstacle points)
  set.seed(11)
  lay <- ggraph::create_layout(g, layout = "fr")

  # Dummy data to force a legend for gene-degree colors
  deg_legend_df <- data.frame(
    x = 0, y = 0,
    gene_deg_col = c("gray80", "gray50", "black")
  )

  # Plot
  p <- ggraph(lay) +
    ggraph::geom_edge_link0(
      aes(edge_colour = edge_color),
      edge_width = 0.8,
      edge_alpha = 0.8
    ) +
    ggraph::scale_edge_colour_identity(
      name   = "Direction",
      breaks = c("firebrick4", "darkgreen"),
      labels = c("Down", "Up"),
      guide  = "legend"
    ) +
    guides(edge_colour = guide_legend(override.aes = list(edge_width = 2, edge_alpha = 1))) +

    # Nodes (hide legend for node aesthetics)
    ggraph::geom_node_point(aes(color = color, shape = shape, size = size), show.legend = FALSE) +
    scale_shape_identity() +
    scale_size_identity(guide = "none") +

    # Invisible points to create the "degree" legend (gene-degree colors)
    ggplot2::geom_point(
      data = deg_legend_df,
      aes(x = x, y = y, color = gene_deg_col),
      inherit.aes = FALSE,
      alpha = 0,
      size = 4,
      show.legend = TRUE
    ) +
    scale_color_identity(
      name   = "Gene degree",
      breaks = c("gray80", "gray50", "black"),
      labels = c("1", "2", "3+"),
      guide  = guide_legend(override.aes = list(alpha = 1, size = 4, shape = 16))
    ) +

    # Invisible "repel obstacles" at perturbation nodes to keep gene labels off squares
    ggplot2::geom_point(
      data = dplyr::filter(lay, type == "perturbation"),
      aes(x = x, y = y),
      inherit.aes = FALSE,
      alpha = 0,
      size = 18
    ) +

    # Gene labels  repel from each other + avoid perturbation nodes
    ggrepel::geom_text_repel(
      data = dplyr::filter(lay, type == "gene"),
      aes(x = x, y = y, label = name),
      inherit.aes = FALSE,
      size = 4,
      box.padding = 0.5,
      point.padding = 1.0,
      force = 5,
      max.overlaps = Inf,
      seed = 11,
      segment.size = 0.25,
      segment.alpha = 0.6,
      segment.color = "grey60"
    ) +

    # Perturbation labels  centered
    ggraph::geom_node_text(
      data = dplyr::filter(lay, type == "perturbation"),
      aes(x = x, y = y, label = name),
      inherit.aes = FALSE,
      repel = FALSE,
      size = pert_fontsize, fontface = "bold", colour = "black",
      vjust = 0.5, hjust = 0.5
    ) +

    ggtitle(paste0("Perturbation x Conserved rGene Network ", tp)) +
    theme_void() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 16),
      legend.position = c(0, 1),
      legend.justification = c(0, 1),
      legend.box = "vertical",
      legend.title = element_text(size = 14),
      legend.text = element_text(size = 12),
      legend.background = element_rect(fill = "white", colour = NA),
      legend.key = element_rect(fill = "white", colour = NA)
    )

  ggsave(
    paste0(fig_path, "networkgraph_PertXSignalingGeneTF_", tp, ".pdf"),
    plot = p, height = 12, width = 12, units = "in", bg = "white"
  )

  print(p)
}
```

## Dot plots conserved rGene ORA GO:BP

```{r function ORA dot plot}
ora_dotplot <- function(df) {
  # Plot dot plot
  p <- ggplot(df, aes(x = Count, y = reorder(Description, Count), size = GeneRatio, fill = -log10(p.adjust))) +
    geom_point(shape = 21, color = 'black', stroke = 0.5) + 
    scale_fill_viridis_c(option = 'inferno',
                         name = expression("-" ~ Log[10] ~"(Adjusted P-value)")
                         ) +
    labs(title = paste0("Conserved rGene ORA", "\n", pert, " ", time),
         x = "Gene Count",
         y = "",
         size = "Gene Ratio",
         fill = expression("-" ~ Log[10] ~"(Adjusted P-value)")) +
         theme_minimal() +
         theme(plot.title = element_text(hjust = 0.5),
               axis.text.y = element_text(size = 10))
  return(p)
}
```

```{r}
pert <- 'BMP7'
time <- '54hr'
n_top_terms <- 10

# Filter dataset
plot_df <- all_ora_df %>% filter(perturbation == pert, time_point == time, direction == 'down') %>%
  arrange(p.adjust) %>%
  slice_head(n = n_top_terms)
  

# Plot ORA dot plot
p <- ora_dotplot(plot_df)
  ggsave(
    paste0(fig_path, "dotplot_conserved_rGene_ORA_", pert, "_", time, ".pdf"),
    plot = p, height = 18, width = 8, units = "in", bg = "white"
  )
print(p)
```

```{r}
pert <- 'FGF8b'
time <- '6hr'
n_top_terms <- 10
direction <- 'down'

# Filter dataset
plot_df <- all_ora_df %>% filter(perturbation == pert, time_point == time, direction == direction) %>%
  arrange(p.adjust) %>%
  slice_head(n = n_top_terms)
  

# Plot ORA dot plot
p <- ora_dotplot(plot_df)
  ggsave(
    paste0(fig_path, "dotplot_conserved_rGene_ORA_", pert, "_", time, "_", direction, ".pdf"),
    plot = p, height = 8, width = 8, units = "in", bg = "white"
  )
print(p)
```


```{r}
pert <- 'BMP7'
time <- '6hr'

# Filter dataset
plot_df <- all_ora_df %>% filter(perturbation == pert, time_point == time, direction == 'down') %>%
  arrange()

# Plot ORA dot plot
p <- ora_dotplot(plot_df)
  ggsave(
    paste0(fig_path, "dotplot_conserved_rGene_ORA_", pert, "_", time, ".pdf"),
    plot = p, height = 18, width = 8, units = "in", bg = "white"
  )
print(p)
```

```{r}
pert <- 'FGF2'
time <- '54hr'

# Filter dataset
plot_df <- all_ora_df %>% filter(perturbation == pert, time_point == time, direction == 'up') %>%
  arrange()

# Plot ORA dot plot
p <- ora_dotplot(plot_df)
  ggsave(
    paste0(fig_path, "dotplot_conserved_rGene_ORA_", pert, "_", time, ".pdf"),
    plot = p, height = 18, width = 8, units = "in", bg = "white"
  )
print(p)
```

















Visualize all crGenes
```{r all crGenes ORA dot plot}
df <- all_ora_df
save_name <- "dotplot_all_crGenes_ORA_GOBP"
# Iterate through cell type x perturbation x time point.
  for (type in cell_types){
    for (pert in unique(df$perturbation)) {
      for (time in unique(df$time_point)) {
        # Filter data to the current cell type x perturbation x time point, to significant results, and to more than one gene count.
        filt_df <- df %>% filter(cell_type == type & perturbation == pert & time_point == time & p.adjust < 0.05 & qvalue < 0.10 & Count > 1) %>%
          arrange(qvalue) %>%
          slice_head(n = 25)
        
        # Check if there are ORA results and skip iteration if not.
        if (nrow(filt_df) == 0) {next}
        
        # Plot dot plot with function
        p <- ora_dotplot(filt_df)
        
        # Save plot
        ggsave(filename = paste0(fig_path, save_name, "/", save_name, "_", type, "_", pert, "_", time, ".png"),
         plot = p, height = 10, width = 20, bg = 'white')
    }
  }
}
```

Visualize up crGenes
```{r up crGenes ORA dot plot}
df <- up_ora_df
save_name <- "dotplot_up_crGenes_ORA_GOBP"
# Iterate through cell type x perturbation x time point.
  for (type in cell_types){
    for (pert in unique(df$perturbation)) {
      for (time in unique(df$time_point)) {
        # Filter data to the current cell type x perturbation x time point, to significant results, and to more than one gene count.
        filt_df <- df %>% filter(cell_type == type & perturbation == pert & time_point == time & p.adjust < 0.05 & qvalue < 0.10 & Count > 1) %>%
          arrange(qvalue) %>%
          slice_head(n = 25)
        
        # Check if there are ORA results and skip iteration if not.
        if (nrow(filt_df) == 0) {next}
        
        # Plot dot plot with function
        p <- ora_dotplot(filt_df)
        
        # Save plot
        ggsave(filename = paste0(fig_path, save_name, "/", save_name, "_", type, "_", pert, "_", time, ".png"),
         plot = p, height = 10, width = 20, bg = 'white')
    }
  }
}
```

Visualize down crGenes
```{r down crGenes ORA dot plot}
df <- down_ora_df
save_name <- "dotplot_down_crGenes_ORA_GOBP"
# Iterate through cell type x perturbation x time point.
  for (type in cell_types){
    for (pert in unique(df$perturbation)) {
      for (time in unique(df$time_point)) {
        # Filter data to the current cell type x perturbation x time point, to significant results, and to more than one gene count.
        filt_df <- df %>% filter(cell_type == type & perturbation == pert & time_point == time & p.adjust < 0.05 & qvalue < 0.10 & Count > 1) %>%
          arrange(qvalue) %>%
          slice_head(n = 25)
        
        # Check if there are ORA results and skip iteration if not.
        if (nrow(filt_df) == 0) {next}
        
        # Plot dot plot with function
        p <- ora_dotplot(filt_df)
        
        # Save plot
        ggsave(filename = paste0(fig_path, save_name, "/", save_name, "_", type, "_", pert, "_", time, ".png"),
         plot = p, height = 10, width = 20, bg = 'white')
    }
  }
}
```

### Inspect specific conserved rGene ORA results

```{r inspect conserved ora_df}
ora_df <- ora_crgenes_df %>% filter(
  #perturbation == 'FGF2' &
  perturbation %in% c('FGF2', 'FGF8b')
  & time_point == '6hr'
  & cell_type == 'Tel_NE_ventral'
  #& ID %in% gobp_terms
)
View(ora_df)
```


## Conserved rGene ORA DisGeNet

```{r}
simple_conserved_rgenes_long_df <- simple_conserved_rgenes_df %>%
  dplyr::select(perturbation, up_conserved_rGenes, down_conserved_rGenes) %>%
  pivot_longer(
    cols = c(up_conserved_rGenes, down_conserved_rGenes),
    names_to  = "direction",
    values_to = "genes"
  ) %>%
  mutate(
    direction = recode(direction,
                       up_conserved_rGenes   = "up",
                       down_conserved_rGenes = "down")
  ) %>%
  filter(!is.na(genes), genes != "") %>%
  separate_rows(genes, sep = "/") %>%
  mutate(
    ID = str_trim(genes)
  ) %>%
  filter(!is.na(ID), ID != "") %>%
  dplyr::select(perturbation, ID, direction) %>%
  distinct()

# optional peek
head(simple_conserved_rgenes_long_df)
```

```{r}
write.csv(simple_conserved_rgenes_long_df, file = paste0(dir_path, "simple_conserved_rGenes_long_df.csv"), row.names = FALSE)

```




https://disgenet2r-medbio-e8c7d57bc0ba433de7ffbbcf2ae597ac42301b1c66845.gitlab.io/#disease-enrichment

```{r disgenet2r api key}
api_key <- "eed75b4e-5d61-45b4-a3b9-eba24323fb8b" 
Sys.setenv(DISGENET_API_KEY= api_key)
```

Define a function for taking a rGenes data frame as input and generating a DisGeNet ORA results data frame as output.
```{r ora_disgenet_df function}
simple_ora_disgenet_df <- function(degs_df, degs_list_colname) {
  ora_results_list <- list()
  # Iterate through perturbations.
  for (pert in unique(degs_df$perturbation)){
    # Filter to current perturbation and tp cell types of interest.
    degs_raw <- degs_df %>% filter(perturbation == pert) %>% 
      pull(!!sym(degs_list_colname))
    
    # Flatten and split stringified gene lists safely
    degs <- degs_raw %>%
      na.omit() %>%
      unlist() %>%
      strsplit(split = "/") %>%
      unlist() %>%
      trimws() %>%
      unique()
    
    # Skip if empty or all NAs
    if (length(degs) == 0 || all(is.na(degs)) || all(degs == "")) next
    
    # Remove empty or NA values before mapping
    degs_clean <- degs[!is.na(degs) & degs != ""]
          
    # Skip iteration if no valid gene symbols
    if (length(degs_clean) == 0) next
  
    # Perform ORA for DEGs with disgenet2r
    ora <- disease_enrichment(entities = degs_clean,
                              common_entities = 3,
                              max_pvalue = 0.10,
                              vocabulary = 'HGNC',
                              database = 'CURATED')
          
    # Check if ORA returned results
    if (!is.null(ora@qresult) && nrow(ora@qresult) > 0) {
      # Add metadata columns to ORA results and store in list
      ora_results <- ora@qresult %>% as.data.frame() %>%
        mutate(perturbation = pert) %>%
        filter(p_adjusted < 0.10 & intersectionSize > 1)
            
      # Add results to results list.
      ora_results_list[[pert]] <- ora_results
    }
  }
  # Combine all results into a single data frame and return the data frame
  ora_results_list <- ora_results_list[lengths(ora_results_list) > 0]
  ora_results_df <- bind_rows(ora_results_list)
  
  # Add Gene ratio
  ora_results_df$GeneRatio <- sapply(ora_results_df$geneRatio, function(x) {
  parts <- strsplit(x, "/")[[1]]
  as.numeric(parts[1]) / as.numeric(parts[2])
  })
  
  return(ora_results_df)
}
```

```{r}
simple_cons_rgenes_disgenet_df <-simple_ora_disgenet_df(simple_conserved_rgenes_df, 'conserved_rGenes')
head(simple_cons_rgenes_disgenet_df)
```

```{r}
for (pert in unique(simple_rgenes_df$perturbation)){
  filt_df <- simple_conserved_rgenes_df %>% filter(perturbation == pert) %>%
    pull('conserved_rGenes') %>%
    na.omit() %>%
    unlist() %>%
    strsplit(split = "/") %>%
    unlist() %>%
    trimws() %>%
    unique() %>%
    tibble::tibble(gene = .) %>%
    dplyr::filter(!stringr::str_detect(gene, "^(AL|AC|AP|AUXG|Z|AF|BX)\\d{4,}"),
                  !stringr::str_detect(gene, "^human_chim_")) %>%
    dplyr::arrange(gene)

  write.csv(filt_df, paste0(dir_path, 'simple_conserved_rGenes_', pert, '.csv'))
}
```

```{r save conserved rGene DisGeNet ORA df}
saveRDS(simple_disgenet_df, file = paste0(dir_path, "conserved_rGenes_DisGeNet_ORA.rds"))
```

```{r}
simple_disgenet_ora_df <- read.csv('/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/dreamlet_v2/conserved_rGene_DisGeNet_ORA_manual_df.csv')
head(simple_disgenet_ora_df)
```

```{r multiple hypothesis testing correction}
simple_disgenet_ora_df <- simple_disgenet_ora_df %>%
  mutate(
    qvalue = p.adjust(pvalue, method = "BH")
  )
head(simple_disgenet_ora_df)
```

```{r add DisGeNet metaterms}
# Patterns
cancer_terms    <- c("Neoplasm", "neoplasm", "Carcinoma", "carcinoma","Hyperplasia", "epithelioma", "Epithelioma", "Cancer", "cancer", "tumor", "Tumor",
                     'sarcoma', 'leukemia', 'Leukemia', 'MENINGIOMA')
ndd_terms       <- c("Autism", "Intellectual Disability", "Neurodevelopmental Disorders", "Rare genetic syndromic intellectual disability",
                     "Autistic Disorder", "Autism Spectrum Disorders", "Neurodevelopmental delay", "Paranoid Schizophrenia", "Epilepsy", "Bipolar Disorder",
                     "Mixed bipolar I disorder", "Bipolar I disorder, most recent episode manic (disorder)", "EPILEPTIC ENCEPHALOPATHY, EARLY INFANTILE, 1",
                     "Developmental and epileptic encephalopathy", "Depression, Bipolar", "Schizoaffective Disorder", "Congenital neurologic anomalies", 
                     "Autosomal dominant non-syndromic intellectual disability", "Rare genetic syndromic intellectual disability", 'Schizophrenia',
                     'Mental Retardation')
brain_prolif    <- c('Brain Neoplasms', "Glioblastoma", "Adult Glioblastoma", "Primary Glioblastoma",
                     'Autosomal Recessive Primary Microcephaly', 'Microcephaly (physical finding)', 'Microcephaly')
brain_structure <- c('Alobar Holoprosencephaly', 'Microform holoprosencephaly', 'Semi-lobar holoprosencephaly', 'Agenesis of corpus callosum',
                     "FOXG1 syndrome", "Enlarged lateral ventricles")

rx_cancer            <- str_c(cancer_terms, collapse = "|")
rx_ndd               <- str_c(ndd_terms, collapse = "|")
rx_brain_prolif      <- str_c(brain_prolif, collapse = "|")
rx_brain_structure   <- str_c(brain_structure, collapse = "|")

has_pat <- function(x, patt) str_detect(x, regex(patt, ignore_case = TRUE))

# Add flags
simple_disgenet_ora_df <- simple_disgenet_ora_df %>%
  mutate(
    Prolif_Cancer     = has_pat(diseaseName, rx_cancer),
    Neurodev_Disorder = has_pat(diseaseName, rx_ndd),
    Brain_Prolif      = has_pat(diseaseName, rx_brain_prolif),
    Brain_Structural  = has_pat(diseaseName, rx_brain_structure)
  ) %>%
  rowwise() %>%
  mutate(
    matches = list(names(which(c(
      Prolif_Cancer     = Prolif_Cancer,
      Neurodev_Disorder = Neurodev_Disorder,
      Brain_Prolif      = Brain_Prolif,
      Brain_Structural  = Brain_Structural
    ))))
  ) %>%
  ungroup() %>%
  mutate(
    n_matches = lengths(matches),
    Metaterm = dplyr::case_when(
      n_matches == 0 ~ "Other",
      n_matches == 1 ~ vapply(matches, function(x) dplyr::coalesce(x[1], "Other"), character(1)),
      n_matches  > 1 ~ "Multiple Metaterms",
      TRUE          ~ "Other"
    )
  )
head(simple_disgenet_ora_df)
```

```{r save conserved rGene DisGeNet ORA df}
saveRDS(simple_disgenet_ora_df, file = paste0(dir_path, "conserved_rGenes_DisGeNet_ORA.rds"))
```

```{r}
write.xlsx(simple_disgenet_ora_df, file = paste0(dir_path, "simple_conserved_rGenes_ORA_DisGeNet_df.xlsx"), rowNames = FALSE)
```

```{r count metaterms}
# Count # of terms for each Metaterm 
metaterm_counts_df <- simple_disgenet_ora_df %>%
  dplyr::filter(!is.na(Metaterm), !is.na(diseaseName), diseaseName != "") %>%
  group_by(perturbation, Metaterm) %>%
  summarise(N_terms = n(), .groups = "drop")

head(metaterm_counts_df)
```

```{r}
# Make a fill palette where "Other" is black
metaterm_levels <- sort(unique(metaterm_counts_df$Metaterm))
metaterm_pal <- setNames(hue_pal()(length(metaterm_levels)), metaterm_levels)
metaterm_pal["Other"] <- "black"

# Order perturbations by descending total terms
pert_levels <- metaterm_counts_df %>%
  dplyr::group_by(perturbation) %>%
  dplyr::summarise(total_terms = sum(N_terms, na.rm = TRUE), .groups = "drop") %>%
  dplyr::arrange(dplyr::desc(total_terms)) %>%
  dplyr::pull(perturbation)

plot_df <- metaterm_counts_df %>%
  dplyr::mutate(
    Metaterm = factor(Metaterm, levels = c('Brain_Prolif', 'Brain_Structural', 'Neurodev_Disorder',
                                          'Prolif_Cancer', 'Multiple Metaterms', 'Other')),
    perturbation = factor(perturbation, levels = rev(pert_levels))
  )

p <- ggplot(plot_df, aes(y = perturbation, x = N_terms, fill = Metaterm)) +
  geom_col(width = 0.8) +
  labs(
    y = "",
    x = "Disease/Disorder Term Count",
    fill = "Metaterm"
  ) +
  scale_fill_manual(values = metaterm_pal, drop = FALSE) +
  theme_classic() +
  theme(
    axis.text = element_text(hjust = 1, size = 12),
    axis.title = element_text(size = 14),
    legend.title = element_text(size = 14),
    legend.text = element_text(size = 12),
    legend.position = "right"
  )

ggsave(paste0(fig_path, "StackedBarplot_Perturbation_Vs_DisGeNetTermCount_ColorByMetaterm.pdf"),
       plot = p, width = 6, height = 4, dpi = 600, bg = "white")

print(p)

```

```{r dot plot disgenet brain prolif}
# Specify terms to plot
terms <- c('Brain Neoplasms', "Glioblastoma", "Adult Glioblastoma", "Primary Glioblastoma",
           'Autosomal Recessive Primary Microcephaly', 'Microcephaly (physical finding)', 'Microcephaly')

# Specify perts to plot
perts <- c('FGF8b', 'FGF2', 'BMP7')
# Filter the data to the current perturbation
filt_df <- simple_disgenet_ora_df %>%
  filter(perturbation %in% perts, diseaseName %in% terms, qvalue <= 0.10) %>%
  mutate(
    diseaseName  = factor(diseaseName, levels = rev(terms)),
    perturbation  = factor(perturbation, levels = perts)
  )
  
  
# Plot a dot plot
p <- ggplot(filt_df, aes(x = perturbation, y = diseaseName, size = intersectionSize, fill = -log10(qvalue))) +
  geom_point(shape = 21, color = 'black', stroke = 0.25) +
  scale_size_continuous(range = c(1, 5)) + 
  scale_fill_viridis(option='inferno') +
  theme_minimal() +
  labs(
    x = "",
    y = "",
    size = 'Gene Count',
    fill = expression ("-" ~ Log[10] ~ "FDR")) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels for readability
    plot.title = element_text(hjust=0.5, size=16), # Edit plot title.
    plot.background = element_blank(),  # Remove the background around the plot.
    panel.background = element_blank(),  # Remove the background inside the plot.
    #panel.grid.major = element_blank(),  # Remove major gridlines
    #panel.grid.minor = element_blank()   # Remove minor gridlines
  )
  
ggsave(filename = paste0(fig_path, "DotPlot_Conserved_rGenes_DisGeNet_ORA_select.pdf"),
         plot = p, units = 'in', height = 4, width = 5, bg = 'white')

print(p)
```



Plot a dot plot of the conserved rGene DisGeNet ORA results
```{r simple_ora_disgenet_df dot plot}
# Filter the data to the current perturbation
filt_df <- simple_disgenet_df %>%
  filter(perturbation %in% c('FGF2', 'BMP7', 'LY411575', 'ATRA') & !(diseaseName %in% c('Fanconi Anemia', 'Anemia, Diamond-Blackfan', 'Aase syndrome')))
  
# Plot a dot plot
p <- ggplot(filt_df, aes(x = perturbation, y = diseaseName, size = intersectionSize, fill = -log10(p_adjusted))) +
  geom_point(shape = 21, color = 'black', stroke = 0.25) +
  scale_size_continuous(range = c(1, 5)) + 
  scale_fill_viridis(option='inferno') +
  theme_minimal() +
  labs(
    x = "",
    y = "",
    size = 'N Genes',
    fill = expression ("-" ~ Log[10] ~ "Adjusted p-value")) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels for readability
    plot.title = element_text(hjust=0.5, size=16), # Edit plot title.
    plot.background = element_blank(),  # Remove the background around the plot.
    panel.background = element_blank(),  # Remove the background inside the plot.
    #panel.grid.major = element_blank(),  # Remove major gridlines
    #panel.grid.minor = element_blank()   # Remove minor gridlines
  )
  
ggsave(filename = paste0(fig_path, "dotplot_conserved_rGenes_DisGeNet_ORA/dotplot_conserved_rGenes_DisGeNet_ORA.png"),
         plot = p, units = 'in', height = 5, width = 6, bg = 'white')

print(p)
```

Plot a dot plot of the conserved rGene DisGeNet ORA results
```{r simple_ora_disgenet_df dot plot}
# Iterate through perturbations
for (pert in unique(simple_disgenet_df$perturbation)){
  
  # Filter the data to the current perturbation
  filt_df <- simple_disgenet_df %>% filter(perturbation == pert)
  
  # Plot a dot plot.
  p <- ggplot(filt_df, aes(x = intersectionSize, y = reorder(diseaseName, intersectionSize), size = GeneRatio, fill = -log10(p_adjusted))) +
    geom_point(shape = 21, color = 'black', stroke = 0.25) +
    scale_size_continuous(range = c(1, 5)) + 
    scale_fill_viridis(option='inferno') +
    theme_minimal() +
    labs(
      title = paste0("Conserved rGene\nDisGeNet Enrichments: ", pert),
      x = "N conserved rGenes",
      y = '',
         size = 'Gene Ratio',
         fill = expression ("-" ~ Log[10] ~ "Adjusted p-value"))
    theme(axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels for readability
            plot.title = element_text(hjust=0.5, size=16), # Edit plot title.
            plot.background = element_blank(),  # Remove the background around the plot.
            panel.background = element_blank(),  # Remove the background inside the plot.
            panel.grid.major = element_blank(),  # Remove major gridlines
            panel.grid.minor = element_blank()   # Remove minor gridlines
          )
  ggsave(filename = paste0(fig_path, "dotplot_conserved_rGenes_DisGeNet_ORA/dotplot_conserved_rGenes_DisGeNet_ORA_", pert, ".png"),
         plot = p, height = 3, width = 8, bg = 'white')
}
```

```{r rGenes volcano plot function}
rGenes_volcano_plot <- function(response_df, type, pert, spec, top_n) {

  # Filter data to the current cell type, perturbation, and species.
  filt_df <- response_df %>% filter(assay == type & perturbation == pert & species == spec)

  # Skip iteration if there are no results.
  if (nrow(filt_df) == 0) return(NULL)

  # Determine the global max x and y values for setting the same scale axes.
  max_x_val <- max(abs(filt_df$logFC), na.rm = TRUE)
  max_y_val <- max(-(log10(filt_df$adj.P.Val)), na.rm = TRUE)

  # Identify top n lowest p.adjust values for both positive and negative logFC values per time point.
  top_up_labeled <- filt_df %>% filter(logFC > 0 & significant == TRUE) %>%
    group_by(time_point) %>%
    slice_min(order_by = adj.P.Val, n = top_n, with_ties = FALSE) %>%
    ungroup()

  top_down_labeled <- filt_df %>% filter(logFC < 0 & significant == TRUE) %>%
    group_by(time_point) %>%
    slice_min(order_by = adj.P.Val, n = top_n, with_ties = FALSE) %>%
    ungroup()

  top_labeled <- bind_rows(top_up_labeled, top_down_labeled)

  # Plot a combined volcano plot
  p <- ggplot(filt_df, aes(x = logFC, y = -log10(adj.P.Val))) +

    geom_point(data = filt_df,
               aes(fill = simple_signaling_category),
               color = 'black',
               shape = 21,
               stroke = 0.25,
               alpha = 0.75) +

    # Data labels at moved positions
    geom_text_repel(
      data = top_labeled,
      aes(x = logFC, y = -log10(adj.P.Val), label = ID),
      size = 3,
      segment.color = 'black',
      min.segment.length = 0.5
    ) + 

    # Define scales
    scale_x_continuous(limits = c(-max_x_val * 1.1, max_x_val * 1.1)) +
    scale_y_continuous(limits = c(0, max_y_val * 1.1)) +
    scale_fill_manual(values = signaling_category_colors, name = 'Signaling Category',
                      guide = guide_legend(title = "Signaling Category",
                                           override.aes = list(shape = 21, size = 3, stroke=0.50, color = 'black'))) +

    # Facet by time point
    facet_wrap(~ time_point) +

    # Labels and theme
    labs(
      title = paste0("Dreamlet Results: ", type, " ", pert, " ", spec),
      x = expression(~ Log[2] ~ "(Stimulus/Vehicle)"),
      y = expression("-" ~ Log[10] ~ "(Adjusted p-value)")
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 24),
      strip.text = element_text(hjust = 0.5, size = 18),
      plot.background = element_blank(),
      panel.background = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.line = element_line(color = "black", linewidth = 1),
      legend.position = "right"
    ) +

    # Significance lines
    geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "darkgrey", linewidth = 1) +
    geom_vline(xintercept = 0, linetype = "dashed", color = "darkgrey", linewidth = 1)

  return(p)
}
```

```{r disease rGenes volcano plots}
pert <- "BMP7"
cell_types <- c('Tel_NE_dorsal', 'Tel_NE_ventral')
microcephaly_genes <- simple_disgenet_df %>% filter(perturbation == pert & diseaseName == "Autosomal Recessive Primary Microcephaly") %>% pull(intersection)
microcephaly_genes <- unlist(strsplit(microcephaly_genes, ","))
microcephaly_genes <- gsub("-", "_", microcephaly_genes)

# Iterate through cell type x species and create one combined plot per combination
for (type in cell_types){
  for (spec in unique(response_df$species)){
    filt_df <- response_df %>% filter(ID %in% microcephaly_genes)
    p_combined <- rGenes_volcano_plot(filt_df, type, pert, spec, 5)
    if (!is.null(p_combined)) {
      ggsave(filename = paste0(fig_path, paste0("/volcanoplot_rDEGs_microcephaly/volcanoplot_rDEGs_microcephaly_", type, "_", pert, "_", spec, ".png")),
        plot = p_combined, height = 5, width = 15, bg = 'white')
    }
  }
}
```


## Conserved response scatter plots

```{r scatter plot response LogFC conserved rGenes colored by signaling category}
# Filter to data of interest
filt_df <- response_logfc_df %>% filter(cell_type == 'Tel_NE_ventral' & time_point == '54hr' & perturbation == 'BMP7' & rGene_category == 'Conserved rGene')
lim <- max(abs(c(filt_df$Chimp_logFC, filt_df$Human_logFC)), na.rm = TRUE)

# Plot scatter plot
p <- ggplot(filt_df, aes(x = Chimp_logFC, y = Human_logFC, shape = time_point)) +
  geom_point(data = filt_df, aes(fill = simple_signaling_category), color = 'black', stroke = 0.2, alpha = 1.0, size = 2) +
  scale_fill_manual(values = signaling_category_colors, name = 'Signaling Category',
                    guide = guide_legend(title = "Signaling Category", override.aes = list(shape = 21, size = 4, stroke=0.4, color = 'black'))) +
  labs(
      x = expression("Chimp " ~Log[2]~"(Stimulus/Vehicle)"),
      y = expression("Human " ~Log[2]~"(Stimulus/Vehicle)"),
      color = "Stimulus",
      shape = "Time Point"
    ) +
  scale_shape_manual(name = "Time Point", values = time_point_shapes,
    guide = guide_legend(override.aes = list(fill = "black"))
  ) +
    theme_minimal() + 
    theme(
      plot.title = element_text(hjust = 0.5, size = 16),  # Center and size the title
      legend.position = "right",  # Place legend on the right
      axis.line = element_line(color = "black"),  # Add axis lines
      panel.grid.major = element_blank(),  # Remove major gridlines
      panel.grid.minor = element_blank()   # Remove minor gridlines
    ) +
    # lock both axes to [-lim, lim] and remove padding so the line hits the axes exactly at (0,0)
    scale_x_continuous(limits = c(-lim, lim), expand = expansion(mult = 0)) +
    scale_y_continuous(limits = c(-lim, lim), expand = expansion(mult = 0)) +
    # square aspect ratio
    coord_equal() +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") + # Add dashed y=x line
    geom_hline(yintercept = 0, color = "black") +  # Horizontal line at y = 0
    geom_vline(xintercept = 0, color = "black") + # Vertical line at x = 0
    coord_fixed(ratio = 1)  # Ensure the aspect ratio is 1:1 for square plot

ggsave(filename = paste0(fig_path, "scatterplot_LogFC_ChimpResponse_HumanResponse/scatterplot_LogFC_ChimpResponse_HumanResponse_ColorBySignalingPathway_BMP7_54hr.png"),
       plot = p, height = 10, width = 10, bg = 'white')
print(p)
```

```{r scatter plot response LogFC conserved rGenes colored by BMP and/or WNT}
# Define colors
signaling_colors <- c(
  "BMP"       = "#40E0D0",
  "WNT"       = "#000080",
  "BMP & WNT" = "seagreen3",
  "Other"     = "grey60"
)

# One row per category for legend
legend_df <- dplyr::tibble(
  label = factor(c("BMP", "WNT", "BMP & WNT", "Other"),
                 levels = c("BMP", "WNT", "BMP & WNT", "Other"))
)

# Filter data to variables of interest
filt_df <- response_logfc_df %>%
  dplyr::filter(time_point == "54hr", perturbation == "BMP7") %>%
  dplyr::mutate(signaling_category = as.character(signaling_category))

# Axis limits (symmetric)
lim <- max(abs(c(filt_df$Human_logFC, filt_df$Chimp_logFC)), na.rm = TRUE)
pad <- 0.08 * lim
x_lo <- -lim - pad; x_hi <- lim + pad
y_lo <- -lim - pad; y_hi <- lim + pad

# Nudges for labels
nudge_y_mag <- 0.30 * lim
nudge_x_mag <- 0.50 * lim

# Categorize rows: BMP, WNT, BMP & WNT, Other
filt_df <- filt_df %>%
  dplyr::mutate(
    has_bmp = !is.na(signaling_category) & grepl("BMP", signaling_category) & !(rGene_category=='Non-rGene'),
    has_wnt = !is.na(signaling_category) & grepl("WNT", signaling_category) & !(rGene_category=='Non-rGene'),
    sig_cat = dplyr::case_when(
      has_bmp & has_wnt ~ "BMP & WNT",
      has_bmp           ~ "BMP",
      has_wnt           ~ "WNT",
      TRUE              ~ "Other"
    ),
    sig_cat = factor(sig_cat, levels = names(signaling_colors))
  )

# Split data set by categories
o_df  <- filt_df %>% dplyr::filter(sig_cat == "Other")
b_df  <- filt_df %>% dplyr::filter(sig_cat == "BMP")
w_df  <- filt_df %>% dplyr::filter(sig_cat == "WNT")
bw_df <- filt_df %>% dplyr::filter(sig_cat == "BMP & WNT")

# Label data (define above/below)
label_df <- filt_df %>%
  dplyr::filter(ID %in% c("TLE3", "ID2", "SMAD9", "LEF1"))

label_df_above <- label_df %>% dplyr::filter(Human_logFC >= 0)
label_df_below <- label_df %>% dplyr::filter(Human_logFC < 0)

p <- ggplot2::ggplot(filt_df, ggplot2::aes(x = Chimp_logFC, y = Human_logFC)) +

  # --- invisible layer ONLY to create the legend ---
  ggplot2::geom_point(
    data = legend_df,
    ggplot2::aes(x = 0, y = 0, fill = label),
    shape = 22, color = "black", stroke = 0.1, size = 4, alpha = 1,
    inherit.aes = FALSE
  ) +

  # Real points (no legend)
  ggplot2::geom_point(data = o_df,  shape = 22, fill = signaling_colors["Other"],     color = "black", stroke = 0.1, alpha = 0.5, size = 2, show.legend = FALSE) +
  ggplot2::geom_point(data = w_df,  shape = 22, fill = signaling_colors["WNT"],       color = "black", stroke = 0.1, alpha = 0.9, size = 2, show.legend = FALSE) +
    ggplot2::geom_point(data = b_df,  shape = 22, fill = signaling_colors["BMP"],       color = "black", stroke = 0.1, alpha = 0.9, size = 2, show.legend = FALSE) +
  ggplot2::geom_point(data = bw_df, shape = 22, fill = signaling_colors["BMP & WNT"], color = "black", stroke = 0.1, alpha = 0.9, size = 2, show.legend = FALSE) +

  # Labels (above)
  ggrepel::geom_text_repel(
    data = label_df_above,
    ggplot2::aes(x = Chimp_logFC, y = Human_logFC, label = ID),
    inherit.aes = FALSE, size = 5,
    box.padding = 0.28, point.padding = 0.22,
    nudge_x = -nudge_x_mag, nudge_y =  nudge_y_mag,
    direction = "both",
    max.overlaps = Inf, force = 8, force_pull = 1.5,
    segment.size = 0.2, segment.color = "black", color = "black",
    xlim = c(x_lo, x_hi), ylim = c(y_lo, y_hi)
  ) +

  # Labels (below)
  ggrepel::geom_text_repel(
    data = label_df_below,
    ggplot2::aes(x = Chimp_logFC, y = Human_logFC, label = ID),
    inherit.aes = FALSE, size = 5,
    box.padding = 0.28, point.padding = 0.22,
    nudge_x =  nudge_x_mag, nudge_y = -nudge_y_mag,
    direction = "both",
    max.overlaps = Inf, force = 8, force_pull = 1.5,
    segment.size = 0.2, segment.color = "black", color = "black",
    xlim = c(x_lo, x_hi), ylim = c(y_lo, y_hi)
  ) +

  ggplot2::geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
  ggplot2::geom_hline(yintercept = 0, color = "black") +
  ggplot2::geom_vline(xintercept = 0, color = "black") +

  ggplot2::labs(
    title = "Species Comparison of\nBMP7 Response at 54 hrs",
    x = expression("Chimp " ~ Log[2] ~ "(Stimulus/Vehicle)"),
    y = expression("Human " ~ Log[2] ~ "(Stimulus/Vehicle)"),
    fill = "Signaling Category"
  ) +

  ggplot2::scale_fill_manual(values = signaling_colors, drop = FALSE) +
  guides(fill = ggplot2::guide_legend(nrow = 2, byrow = TRUE, override.aes = list(shape = 22, size = 4, color = "black", alpha = 1))
  )+
  ggplot2::scale_x_continuous(limits = c(x_lo, x_hi), expand = ggplot2::expansion(mult = 0)) +
  ggplot2::scale_y_continuous(limits = c(y_lo, y_hi), expand = ggplot2::expansion(mult = 0)) +
  ggplot2::coord_fixed(ratio = 1, clip = "on") +

  ggplot2::theme_minimal() +
  ggplot2::theme(
    plot.title   = ggplot2::element_text(hjust = 0.5, size = 16),
    legend.title = ggplot2::element_text(size = 14),
    legend.text  = ggplot2::element_text(size = 12),
    axis.title   = ggplot2::element_text(size = 14),
    legend.position = "bottom",
    axis.line = ggplot2::element_line(color = "black"),
    panel.grid.major = ggplot2::element_blank(),
    panel.grid.minor = ggplot2::element_blank()
  )

ggplot2::ggsave(
  filename = file.path(fig_path, "scatterplot_HumanVsChimpResponse",
                       "scatterplot_HumanVsChimpResponse_ColorBySignalingCategory_BMP7_54hr.pdf"),
  plot = p, height = 5, width = 5, bg = "white", dpi = 600, units = "in"
)

print(p)
```

```{r scatter plot response LogFC conserved rGenes microcephaly}
# Load conserved rGene ORA with DisGeNet results table
simple_disgenet_df <- readRDS(paste0(dir_path, "conserved_rGenes_DisGeNet_ORA.rds"))

# Pull genes
micro_genes <- simple_disgenet_df %>%
  dplyr::filter(perturbation == "BMP7",
                diseaseName == "Autosomal Recessive Primary Microcephaly") %>%
  dplyr::pull(intersection) |>
  strsplit(",") |>
  unlist()

# Filter to data of interest
filt_df <- response_logfc_df %>%
  dplyr::filter(
    cell_type == "Tel_NE_ventral",
    perturbation == "BMP7",
    rGene_category == "Conserved rGene"
  ) %>%
  dplyr::distinct()

lim <- max(abs(c(filt_df$Chimp_logFC, filt_df$Human_logFC)), na.rm = TRUE)

other_df <- dplyr::filter(filt_df, !(ID %in% micro_genes))
micro_df <- dplyr::filter(filt_df,  (ID %in% micro_genes))

# IMPORTANT: map shape in aes(); use filled shapes (2125) in time_point_shapes
p <- ggplot(filt_df, aes(x = Chimp_logFC, y = Human_logFC, shape = time_point)) +
  geom_point(
    data = other_df,
    aes(fill = "Other"),
    color = "black", stroke = 0.2, alpha = 0.9, size = 2
  ) +
  geom_point(
    data = micro_df,
    aes(fill = "Autosomal Recessive \n Primary Microcephaly"),
    color = "black", stroke = 0.2, alpha = 1.0, size = 2
  ) +
  # Fill legend (category)
  scale_fill_manual(
    name   = "Category",
    values = c(
      "Other" = "grey60",
      "Autosomal Recessive \n Primary Microcephaly" = "firebrick"
    ),
    guide = guide_legend(
      override.aes = list(shape = 22, size = 4, stroke = 0.4, color = "black")
    )
  ) +
  # Shape legend (time point)  ensure these are filled shapes (2125)
  scale_shape_manual(
    values = time_point_shapes,
    guide = guide_legend(override.aes = list(fill = "black"), size = 4)
  ) +
  labs(
    x = expression("Chimp " ~ Log[2] ~ "(Stimulus/Vehicle)"),
    y = expression("Human " ~ Log[2] ~ "(Stimulus/Vehicle)")
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    legend.position = "right",
    legend.title = element_blank(),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  # use your precomputed padded limits
  scale_x_continuous(limits = c(x_lo, x_hi), expand = expansion(mult = 0)) +
  scale_y_continuous(limits = c(y_lo, y_hi), expand = expansion(mult = 0)) +
  coord_fixed(ratio = 1, clip = "on") +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
  geom_hline(yintercept = 0, color = "black") +
  geom_vline(xintercept = 0, color = "black")

ggsave(
  filename = paste0(fig_path, "scatterplot_LogFC_ChimpResponse_HumanResponse/scatterplot_LogFC_ChimpResponse_HumanResponse_ColorBySignalingPathway_BMP7_microcephaly.png"),
  plot = p, height = 5, width = 5, bg = "white"
)
print(p)

```

## Conserved rGene logFC heatmap

```{r}
response_heatmap <- function(response_df, time_points, perts, genes,
                             species_order, species_colors, pert_colors) {

  # Helper function for stars from p values
  p_to_stars <- function(p) {
    dplyr::case_when(
      p < 0.001 ~ "***",
      p < 0.01  ~ "**",
      p < 0.10  ~ "*",
      TRUE      ~ ""
    )
  }

  # Filter + set factor levels
  df <- response_df %>%
    dplyr::filter(
      time_point    %in% time_points,
      perturbation  %in% perts,
      ID            %in% genes,
      species       %in% species_order
    ) %>%
    dplyr::mutate(
      species      = factor(species, levels = species_order),
      perturbation = factor(perturbation, levels = perts),
      time_point   = factor(time_point, levels = time_points),
      ID           = factor(ID, levels = genes)
    ) %>%
    dplyr::arrange(species, perturbation, ID, time_point) %>%
    dplyr::select(ID, species, perturbation, time_point, logFC, adj.P.Val)

  # Build logFC matrix with rows = species_perturbation, cols = genes
  heatmap_mat <- df %>%
    tidyr::unite("group", species, perturbation, sep = "_", remove = FALSE) %>%
    dplyr::select(group, ID, logFC) %>%
    tidyr::pivot_wider(names_from = ID, values_from = logFC) %>%
    tibble::column_to_rownames("group") %>%
    as.matrix()

  # Build p-value matrix aligned to heatmap_mat (same rows/cols)
  p_mat <- df %>%
    tidyr::unite("group", species, perturbation, sep = "_", remove = FALSE) %>%
    dplyr::select(group, ID, adj.P.Val) %>%
    tidyr::pivot_wider(names_from = ID, values_from = adj.P.Val) %>%
    tibble::column_to_rownames("group") %>%
    as.matrix()

  # Enforce identical ordering to the heatmap matrix (robust to missing combos)
  p_mat <- p_mat[rownames(heatmap_mat), colnames(heatmap_mat), drop = FALSE]

  # Convert p-values to stars (character matrix)
  star_mat <- matrix(
    p_to_stars(as.vector(p_mat)),
    nrow = nrow(p_mat),
    ncol = ncol(p_mat),
    dimnames = dimnames(p_mat)
  )

  # Row annotation metadata
  metadata_df <- data.frame(group = rownames(heatmap_mat), stringsAsFactors = FALSE)
  parts <- strsplit(metadata_df$group, "_", fixed = TRUE)
  metadata_df$species      <- vapply(parts, `[`, character(1), 1)
  metadata_df$perturbation <- vapply(parts, `[`, character(1), 2)

  # Symmetric diverging colormap about 0
  max_abs <- max(abs(heatmap_mat), na.rm = TRUE)
  max_abs_int <- ceiling(max_abs)

  col_fun <- circlize::colorRamp2(
    c(-max_abs_int, 0, max_abs_int),
    c("darkblue", "white", "darkred")
  )
  legend_at <- c(-max_abs_int, 0, max_abs_int)

  # Row annotations
  ha_row <- ComplexHeatmap::rowAnnotation(
    Species  = metadata_df$species,
    Stimulus = metadata_df$perturbation,
    col = list(
      Species  = species_colors,
      Stimulus = pert_colors
    ),
    annotation_name_side = "bottom"
  )

  # Heatmap + stars centered in each cell
  ht <- ComplexHeatmap::Heatmap(
    heatmap_mat,
    col = col_fun,
    left_annotation = ha_row,
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_row_names = FALSE,
    show_column_names = TRUE,
    column_names_gp = grid::gpar(fontsize = 8),
    column_names_rot = 60,
    heatmap_legend_param = list(
      title = expression(Log[2] ~ "FC"),
      at = legend_at
    ),
    cell_fun = function(j, i, x, y, width, height, fill) {
      lab <- star_mat[i, j]
      if (!is.na(lab) && nzchar(lab)) {
        grid::grid.text(
          lab, x, y,
          gp = grid::gpar(col = "black", fontsize = 10, fontface = "bold")
        )
      }
    }
  )

  ComplexHeatmap::draw(
    ht,
    heatmap_legend_side = "right",
    annotation_legend_side = "right"
  )
  invisible(ht)
}

```

```{r early BMP TFs}
genes <- c('ID1', 'ID2', 'ID3', 'SMAD6', 'SMAD7', 'SMAD9', 'MSX1', 'MSX2', 'SIX3', 'TBX3', 'LEF1', 'TLE3')
time_points <- c('54hr')
perts <- c('BMP7', 'LDN193189')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_LDN193189_54hr_SelectGenes.pdf"), width = 5, height = 3, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r wnt response}
hum_chimp_genes <- all_ora_df %>% filter(perturbation == 'BMP7', time_point=='54hr', ID=='GOBP_WNT_SIGNALING_PATHWAY') %>%
  pull(geneSymbol) %>%
  strsplit("/", fixed = TRUE) %>%
  `[[`(1) %>%
  trimws() %>%
  unique()

orang_genes <- response_df %>% filter(perturbation=='BMP7', time_point=='54hr', species=='Orangutan', significant) %>% pull(ID) %>% unique()

genes <- intersect(hum_chimp_genes, orang_genes)

time_points <- c('54hr')
perts <- c('BMP7', 'LDN193189')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_LDN193189_54hr_WNTSignalingPathway.pdf"), width = 9, height = 3, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r bmp response}
hum_chimp_genes <- all_ora_df %>% filter(perturbation == 'BMP7', time_point=='54hr', ID=='GOBP_RESPONSE_TO_BMP') %>%
  pull(geneSymbol) %>%
  strsplit("/", fixed = TRUE) %>%
  `[[`(1) %>%
  trimws() %>%
  unique()

orang_genes <- response_df %>% filter(perturbation=='BMP7', time_point=='54hr', species=='Orangutan', significant) %>% pull(ID) %>% unique()

genes <- intersect(hum_chimp_genes, orang_genes)

time_points <- c('54hr')
perts <- c('BMP7', 'LDN193189')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_LDN193189_54hr_ResponseToBMP.pdf"), width = 6, height = 3, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r microcephaly genes}
simple_disgenet_ora_df <- read.csv('/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/dreamlet_v2/conserved_rGene_DisGeNet_ORA_manual_df.csv')

micro_genes <- simple_disgenet_ora_df %>% dplyr::filter(perturbation == 'BMP7', diseaseName=='Autosomal Recessive Primary Microcephaly') %>%
  pull(intersection) %>%
  strsplit(";", fixed = TRUE) %>%
  `[[`(1) %>%
  trimws() %>%
  unique()

orang_genes <- response_df %>% dplyr::filter(perturbation=='BMP7', time_point=='54hr', species=='Orangutan', significant==TRUE) %>% pull(ID) %>% unique()
human_genes <- response_df %>% dplyr::filter(perturbation=='BMP7', time_point=='54hr', species=='Human', significant==TRUE) %>% pull(ID) %>% unique()
chimp_genes <- response_df %>% dplyr::filter(perturbation=='BMP7', time_point=='54hr', species=='Chimp', significant==TRUE) %>% pull(ID) %>% unique()

genes <- intersect(human_genes, chimp_genes)
genes <- intersect(genes, orang_genes)
genes <- intersect(genes, micro_genes)

time_points <- c('54hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_Microcephaly.pdf"), width = 6, height = 3, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r microcephaly genes}
simple_disgenet_ora_df <- read.csv('/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/dreamlet_v2/conserved_rGene_DisGeNet_ORA_manual_df.csv')

micro_genes <- simple_disgenet_ora_df %>% filter(perturbation == 'BMP7', diseaseName=='Autosomal Recessive Primary Microcephaly') %>%
  pull(intersection) %>%
  strsplit(";", fixed = TRUE) %>%
  `[[`(1) %>%
  trimws() %>%
  unique()

orang_genes <- response_df %>% filter(perturbation=='BMP7', time_point=='54hr', species=='Orangutan', significant==TRUE) %>% pull(ID) %>% unique()
human_genes <- response_df %>% filter(perturbation=='BMP7', time_point=='54hr', species=='Human', significant==TRUE) %>% pull(ID) %>% unique()
chimp_genes <- response_df %>% filter(perturbation=='BMP7', time_point=='54hr', species=='Chimp', significant==TRUE) %>% pull(ID) %>% unique()

genes <- intersect(human_genes, chimp_genes)
genes <- intersect(genes, orang_genes)
genes <- intersect(genes, micro_genes)

time_points <- c('54hr')
perts <- c('BMP7', 'LDN193189')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_LDN193189_54hr_Microcephaly.pdf"), width = 6, height = 3, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r early wnt response}
genes <- all_ora_df %>% filter(perturbation == 'BMP7', time_point=='6hr', ID=='GOBP_WNT_SIGNALING_PATHWAY') %>%
  pull(geneSymbol) %>%
  strsplit("/", fixed = TRUE) %>%
  `[[`(1) %>%
  trimws() %>%
  unique()

time_points <- c('6hr')
perts <- c('BMP7', 'LDN193189')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_LDN193189_6hr_WNTSignalingPathway.pdf"), width = 9, height = 3, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r early bmp response}
genes <- all_ora_df %>% filter(perturbation == 'BMP7', time_point=='6hr', ID=='GOBP_RESPONSE_TO_BMP') %>%
  pull(geneSymbol) %>%
  strsplit("/", fixed = TRUE) %>%
  `[[`(1) %>%
  trimws() %>%
  unique()

time_points <- c('6hr')
perts <- c('BMP7', 'LDN193189')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_LDN193189_6hr_ResponseToBMP.pdf"), width = 9, height = 3, useDingbats = FALSE)
draw(p)
dev.off()
```

## Conserved rGene totals

```{r}
n_unique_crGenes <- conserved_rgenes_df %>%
  filter(!is.na(crDEGs), crDEGs != "", !perturbation == 'Verteporfin') %>%
  separate_rows(crDEGs, sep = "/") %>%
  mutate(crDEGs = str_trim(crDEGs)) %>%
  filter(crDEGs != "") %>%
  summarise(n_unique = n_distinct(crDEGs)) %>%
  pull(n_unique)

print(n_unique_crGenes)
```

```{r}
n_unique_crGenes <- conserved_rgenes_df %>%
  filter(!is.na(crDEGs), crDEGs != "", !perturbation == 'Verteporfin', cell_type == 'Tel_NE_ventral', time_point == '6hr') %>%
  separate_rows(crDEGs, sep = "/") %>%
  mutate(crDEGs = str_trim(crDEGs)) %>%
  filter(crDEGs != "") %>%
  summarise(n_unique = n_distinct(crDEGs)) %>%
  pull(n_unique)

print(n_unique_crGenes)
```

## Inspect specific conserved rGenes

```{r inspect specific crGenes}
cr_df <- conserved_rgenes_df %>%
  dplyr::filter(perturbation == 'BMP7'
         & time_point %in% c('6hr')
  )
View(cr_df)
```


# Baseline divergence analysis

## N baseline divergent genes
```{r}
baseline_divergence_df <- speciesdiff_df %>% dplyr::filter(perturbation == 'Baseline')
baseline_divergent_genes <- baseline_divergence_df %>% dplyr::filter(significant == TRUE) %>%
  pull(ID) %>%
  unique()

n_baseline_genes <- length(baseline_divergent_genes)
n_universe_genes <- length(unique(baseline_divergence_df$ID))
print(n_baseline_genes)
print(n_baseline_genes / n_universe_genes *100)
```

## N baseline divergent genes over time
```{r}
baseline_divergent_genes_by_time <- baseline_divergence_df %>% filter(significant == TRUE) %>%
  group_by(time_point) %>%
  summarise(n_genes = n_distinct(ID), .groups = "drop")

baseline_divergent_genes_by_time
```

## N unique baseline divergent genes over time

```{r}
unique_baseline_divergent_genes_by_time <- baseline_divergence_df %>% filter(significant == TRUE) %>%
  distinct(ID, time_point) %>%
  group_by(ID) %>% # keep only genes that are significant in exactly one time_point total
  filter(n() == 1) %>%
  ungroup() %>%
  dplyr::count(time_point, name = "n_unique_genes")   # now count how many such genes there are at each time_point

unique_baseline_divergent_genes_by_time
```


## Inspect specific divergent baseline genes

```{r}
basediv_df <- baseline_divergence_df %>% dplyr::filter(ID %in% c('NRG3'))
View(basediv_df)
```


## Baseline divergent gene volcano plots

```{r}
rgene_volcano <- function(df, pert, top_n) {

  # Filter data to the current cell type, perturbation, and species.
  filt_df <- df %>% filter(perturbation == pert)

  # Skip iteration if there are no results.
  if (nrow(filt_df) == 0) return(NULL)

  # Determine the global max x and y values for setting the same scale axes.
  max_x_val <- max(abs(filt_df$logFC), na.rm = TRUE)
  max_y_val <- max(-(log10(filt_df$adj.P.Val)), na.rm = TRUE)
  
  # Split dataset by significance
  sig_df <- filt_df %>% filter(significant == TRUE)
  non_df <- filt_df %>% filter(significant == FALSE)

  # Identify top n lowest p.adjust values for both positive and negative logFC values per time point.
  top_up_labeled <- filt_df %>% filter(logFC > 0 & significant == TRUE) %>%
    group_by(time_point) %>%
    slice_min(order_by = adj.P.Val, n = top_n, with_ties = FALSE) %>%
    ungroup()

  top_down_labeled <- filt_df %>% filter(logFC < 0 & significant == TRUE) %>%
    group_by(time_point) %>%
    slice_min(order_by = adj.P.Val, n = top_n, with_ties = FALSE) %>%
    ungroup()

  top_labeled <- bind_rows(top_up_labeled, top_down_labeled)

  p <- ggplot(filt_df, aes(x = logFC, y = -log10(adj.P.Val), shape = time_point)) +
    # non-significant first (background)
    geom_point(
      data = non_df,
      aes(fill = "Not Significant"),
      color = "black", stroke = 0, alpha = 0.9, size = 3
    ) +
    # significant on top
    geom_point(
      data = sig_df,
      aes(fill = "Significant"),
      color = "black", stroke = 0, alpha = 1.0, size = 3
    ) +
    # labels
    ggrepel::geom_text_repel(
      data = top_labeled,
      aes(label = ID),
      size = 3,
      segment.color = "black",
      min.segment.length = 0.5,
      point.padding = 0.5,
      box.padding = 0.5
    ) +
    # scales
    scale_x_continuous(limits = c(-max_x_val * 1.1, max_x_val * 1.1)) +
    scale_y_continuous(limits = c(0, max_y_val * 1.1)) +
    scale_shape_manual(name = "Time Point",
                       values = time_point_shapes,
                       guide  = guide_legend(override.aes = list(fill = "black", color = "black", alpha = 1, size = 4))
                       ) +
    scale_fill_manual(
      name = "Significance",
      values = c("Not Significant" = "grey50", "Significant" = "firebrick"),
      guide = guide_legend(override.aes = list(shape = 21, color = "black", alpha = 1, size = 4))
    ) +
    # labels/theme
    labs(
      x = expression(~Log[2]~"(Human/Chimp)"),
      y = expression("-"~Log[10]~"(Adjusted p-value)")
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 24),
      strip.text = element_text(hjust = 0.5, size = 18),
      plot.background = element_blank(),
      panel.background = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.line = element_line(color = "black", linewidth = 1),
      legend.position = "right"
    )

  return(p)
}
```

```{r}
p <- rgene_volcano(baseline_divergence_df, "Baseline", 5)
ggsave(paste0(fig_path, "VolcanoPlot_BaselineDivergence_ColorBySignificance.pdf"),
    plot = p, height = 4, width = 5, units = "in", bg = "white", dpi=600)
print(p)
```

```{r drGenes_volcano_plot function}
rgene_volcano_by_time <- function(df, pert, top_n) {

  # Filter data to the current cell type, perturbation, and species.
  filt_df <- df %>% filter(perturbation == pert)

  # Skip iteration if there are no results.
  if (nrow(filt_df) == 0) return(NULL)

  # Determine the global max x and y values for setting the same scale axes.
  max_x_val <- max(abs(filt_df$logFC), na.rm = TRUE)
  max_y_val <- max(-(log10(filt_df$adj.P.Val)), na.rm = TRUE)
  
  # Split dataset by significance
  sig_df <- filt_df %>% filter(significant == TRUE)
  non_df <- filt_df %>% filter(significant == FALSE)

  # Identify top n lowest p.adjust values for both positive and negative logFC values per time point.
  top_up_labeled <- filt_df %>% filter(logFC > 0 & significant == TRUE) %>%
    group_by(time_point) %>%
    slice_min(order_by = adj.P.Val, n = top_n, with_ties = FALSE) %>%
    ungroup()

  top_down_labeled <- filt_df %>% filter(logFC < 0 & significant == TRUE) %>%
    group_by(time_point) %>%
    slice_min(order_by = adj.P.Val, n = top_n, with_ties = FALSE) %>%
    ungroup()

  top_labeled <- bind_rows(top_up_labeled, top_down_labeled)

  # Plot a combined volcano plot
  p <- ggplot(filt_df, aes(x = logFC, y = -log10(adj.P.Val))) +
    geom_point(data = non_df, fill = 'grey50', color = 'black', shape = 21, stroke = 0, alpha = 1.0) +
    geom_point(data = sig_df, fill = 'firebrick', color = 'black', shape = 21, stroke = 0, alpha = 1.0) +

    # Data labels at moved positions
    geom_text_repel(
      data = top_labeled,
      aes(x = logFC, y = -log10(adj.P.Val), label = ID), size = 3, segment.color = 'black', min.segment.length = 0.5
    ) + 

    # Define scales
    scale_x_continuous(limits = c(-max_x_val * 1.1, max_x_val * 1.1)) +
    scale_y_continuous(limits = c(0, max_y_val * 1.1)) +
    # Facet by time point
    facet_wrap(~ time_point) +

    # Labels and theme
    labs(
      x = expression(~ Log[2] ~ "(Human/Chimp)"),
      y = expression("-" ~ Log[10] ~ "(Adjusted p-value)")
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 24),
      strip.text = element_text(hjust = 0.5, size = 18),
      plot.background = element_blank(),
      panel.background = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.line = element_line(color = "black", linewidth = 1),
      legend.position = "right"
    )
  return(p)
}
```

```{r}
p <- rgene_volcano_by_time(baseline_divergence_df, "Baseline", 5)
ggsave(paste0(fig_path, "VolcanoPlot_BaselineDivergence_ByTimePoint_ColorBySignificance.pdf"),
    plot = p, height = 4, width = 12, units = "in", bg = "white", dpi=600)
print(p)
```

## Baseline divergent gene ORA

```{r}
simple_baseline_div_genes_ora <- function(df, species_name, category, subcategory){

  # Safe mapIds wrapper: returns character(0) if no keys (avoids "no ids to receive")
  safe_mapIds <- function(db, keys, column, keytype, multiVals = "first") {
    keys <- unique(stats::na.omit(as.character(keys)))
    if (length(keys) == 0) return(character(0))
    out <- AnnotationDbi::mapIds(db, keys = keys, column = column, keytype = keytype, multiVals = multiVals)
    as.character(stats::na.omit(out))
  }

  # Filter dataset to significant genes
  genes <- df %>%
    dplyr::filter(significant == TRUE) %>%
    dplyr::pull(ID) %>%
    unique()

  # Convert gene symbols to ENTREZ IDs (safe)
  genes_entrez <- safe_mapIds(
    db = org.Hs.eg.db,
    keys = genes,
    column = "ENTREZID",
    keytype = "SYMBOL"
  )

  # If nothing maps, skip
  if (length(genes_entrez) == 0) {next }

  # Define universe genes and convert gene symbols to ENTREZ IDs (safe)
  universe_entrez <- safe_mapIds(
    db = org.Hs.eg.db,
    keys = unique(baseline_divergence_df$ID),
    column = "ENTREZID",
    keytype = "SYMBOL"
    )

  # Get gene sets for enrichment analysis
  gene_set <- msigdbr::msigdbr(species = species_name, category = category, subcategory = subcategory)
  gene_set_df <- gene_set %>% dplyr::select(gs_name, entrez_gene)

  # Perform ORA
  ora <- clusterProfiler::enricher(
    gene = genes_entrez,
    universe = universe_entrez,
    TERM2GENE = gene_set_df,
    pAdjustMethod = "fdr",
    qvalueCutoff = 0.10,
    minGSSize = 5,
    maxGSSize = 10000
  )

  # Format ORA results
  ora_results_df <- ora@result %>%
    dplyr::mutate(time_point = time) %>%
    tidyr::separate(GeneRatio, into = c("numerator", "denominator"), sep = "/", remove = FALSE) %>%
    dplyr::mutate(
      GeneRatioOriginal = GeneRatio,
      GeneRatio = as.numeric(numerator) / as.numeric(denominator)
    ) %>%
    dplyr::select(-numerator, -denominator) %>%
    dplyr::filter(qvalue < 0.10, Count > 1)

  return(ora_results_df)
}
```

```{r}
# Define parameters of interest
species_name = "Homo sapiens"
category = 'C5'
subcategory = 'GO:BP'

# Filter to up in human
filt_df <- baseline_divergence_df %>% dplyr::filter(logFCsign==1)
# Run ORA
human_up_gobp_df <- simple_baseline_div_genes_ora(filt_df, species_name, category, subcategory)
human_up_gobp_df$direction <- 'Up in human'
human_up_gobp_df$gene_set <- 'GOBP'
head(human_up_gobp_df)
```

```{r}
# Define parameters of interest
species_name = "Homo sapiens"
category = 'C5'
subcategory = 'GO:BP'

# Filter to up in human
filt_df <- baseline_divergence_df %>% dplyr::filter(logFCsign==-1)
# Run ORA
chimp_up_gobp_df <- simple_baseline_div_genes_ora(filt_df, species_name, category, subcategory)
chimp_up_gobp_df$direction <- 'Up in chimp'
chimp_up_gobp_df$gene_set <- 'GOBP'
head(chimp_up_gobp_df)
```

```{r}
simple_div_baseline_ora_df <- dplyr::bind_rows(human_up_gobp_df, chimp_up_gobp_df)
head(simple_div_baseline_ora_df)
```

```{r}
# Function to map ENTREZ IDs to gene symbols
entrez_to_symbol <- function(entrez_str) {
  ids <- unlist(str_split(entrez_str, "/"))
  symbols <- mapIds(org.Hs.eg.db ,keys = ids, column = "SYMBOL", keytype = "ENTREZID", multiVals = "first")
  paste(symbols, collapse = "/")
}
```
  
```{r}
# Apply function to new column
simple_div_baseline_ora_df <- simple_div_baseline_ora_df %>%
  rowwise() %>%
  mutate(geneSymbol = entrez_to_symbol(geneID)) %>%
  ungroup()
```

```{r}
saveRDS(simple_div_baseline_ora_df, paste0(dir_path, "simple_divergent_baseline_genes_ORA_df.rds"))
```

```{r}
write.xlsx(simple_div_baseline_ora_df, file = paste0(dir_path, "simple_divergent_baseline_genes_ORA_GOBP.xlsx"), rowNames = FALSE)
```

```{r}
simple_div_baseline_ora_df <- readRDS(paste0(dir_path, "simple_divergent_baseline_genes_ORA_df.rds"))
head(simple_div_baseline_ora_df)
```

```{r function ORA dot plot}
plot_df <- simple_div_baseline_ora_df %>%
  arrange(p.adjust) %>%
  slice_head(n=10) %>%
  mutate(
    Description = Description %>%
      str_replace_all("GOBP", "") %>%   # remove GOBP
      str_replace_all("_", " ") %>%     # replace underscores with spaces
      str_squish() %>%                  # clean up extra spaces
      str_to_lower() %>%                # lower case
      str_to_title() %>%                # capitalize first letter of each word
      str_replace_all("\\bAtp\\b", "ATP")   # fully capitalize ATP
  )

p <- ggplot(plot_df, aes(x = Count, y = reorder(Description, Count), size = GeneRatio, fill = -log10(p.adjust))) +
    geom_point(shape = 21, color = 'black', stroke = 0.5) + 
    scale_fill_viridis_c(option = 'inferno',
                         name = expression("-" ~ Log[10] ~"(Adj. P-val.)")
                         ) +
    labs(title = "Baseline Divergence GO:BP ORA",
         x = "Gene Count",
         y = "",
         size = "Gene Ratio",
         fill = expression("-" ~ Log[10] ~"(Adj. P-val.)")) +
         theme_minimal() +
         theme(plot.title = element_text(hjust = 0.5),
               axis.title = element_text(size = 14),
               axis.text = element_text(size = 12),
               legend.title = element_text(size = 14),
               
               )

ggsave(paste0(fig_path, "DotPlot_BaselineDivergenceORA_GeneCount_Vs_GOBPTerm_SizeByGeneRatio_ColorBySignificance.pdf"),
    plot = p, height = 4, width = 9, units = "in", bg = "white", dpi=600)

print(p)
```

```{r}
baseline_div_genes_ora <- function(df, species_name, category, subcategory){

  ora_results_list <- list()

  # Safe mapIds wrapper: returns character(0) if no keys (avoids "no ids to receive")
  safe_mapIds <- function(db, keys, column, keytype, multiVals = "first") {
    keys <- unique(stats::na.omit(as.character(keys)))
    if (length(keys) == 0) return(character(0))
    out <- AnnotationDbi::mapIds(db, keys = keys, column = column, keytype = keytype, multiVals = multiVals)
    as.character(stats::na.omit(out))
  }

  for (time in unique(df$time_point)) {

    # Filter dataset to current time point and significant genes
    genes <- df %>%
      dplyr::filter(time_point == time, significant == TRUE) %>%
      dplyr::pull(ID) %>%
      unique()

    # Convert gene symbols to ENTREZ IDs (safe)
    genes_entrez <- safe_mapIds(
      db = org.Hs.eg.db,
      keys = genes,
      column = "ENTREZID",
      keytype = "SYMBOL"
    )

    # If nothing maps, skip this time point
    if (length(genes_entrez) == 0) {
      ora_results_list[[as.character(time)]] <- dplyr::tibble()
      next
    }

    # Define universe genes and convert gene symbols to ENTREZ IDs (safe)
    universe_entrez <- safe_mapIds(
      db = org.Hs.eg.db,
      keys = baseline_divergence_df$ID,
      column = "ENTREZID",
      keytype = "SYMBOL"
    )

    # If universe is empty, let enricher run without universe (or skip)
    # Here: if empty, set to NULL so enricher doesn't choke on character(0)
    if (length(universe_entrez) == 0) universe_entrez <- NULL

    # Get gene sets for enrichment analysis
    gene_set <- msigdbr::msigdbr(species = species_name, category = category, subcategory = subcategory)
    gene_set_df <- gene_set %>% dplyr::select(gs_name, entrez_gene)

    # Perform ORA (may return NULL)
    ora <- clusterProfiler::enricher(
      gene = genes_entrez,
      universe = universe_entrez,
      TERM2GENE = gene_set_df,
      pAdjustMethod = "fdr",
      qvalueCutoff = 0.10,
      minGSSize = 5,
      maxGSSize = 10000
    )

    # If no results, store empty tibble and continue
    if (is.null(ora) || nrow(ora@result) == 0) {
      ora_results_list[[as.character(time)]] <- dplyr::tibble()
      next
    }

    # Format ORA results
    ora_results <- ora@result %>%
      dplyr::mutate(time_point = time) %>%
      tidyr::separate(GeneRatio, into = c("numerator", "denominator"), sep = "/", remove = FALSE) %>%
      dplyr::mutate(
        GeneRatioOriginal = GeneRatio,
        GeneRatio = as.numeric(numerator) / as.numeric(denominator)
      ) %>%
      dplyr::select(-numerator, -denominator) %>%
      dplyr::filter(qvalue < 0.10, Count > 1)

    ora_results_list[[as.character(time)]] <- ora_results
  }

  # Combine all ORA results
  ora_results_df <- dplyr::bind_rows(ora_results_list)

  # If nothing came back overall, return early
  if (nrow(ora_results_df) == 0) return(ora_results_df)

  # Function to map ENTREZ IDs to symbols (safe)
  entrez_to_symbol <- function(entrez_str) {
    if (is.null(entrez_str) || is.na(entrez_str) || !nzchar(entrez_str)) return(NA_character_)

    ids <- unlist(stringr::str_split(entrez_str, "/"))
    ids <- unique(stats::na.omit(ids))
    if (length(ids) == 0) return(NA_character_)

    symbols <- AnnotationDbi::mapIds(
      org.Hs.eg.db,
      keys = ids,
      column = "SYMBOL",
      keytype = "ENTREZID",
      multiVals = "first"
    )

    symbols <- as.character(symbols)
    symbols <- symbols[!is.na(symbols) & nzchar(symbols)]
    if (length(symbols) == 0) return(NA_character_)

    paste(symbols, collapse = "/")
  }

  # Apply function to geneID column
  ora_results_df <- ora_results_df %>%
    dplyr::rowwise() %>%
    dplyr::mutate(geneSymbol = entrez_to_symbol(geneID)) %>%
    dplyr::ungroup()

  return(ora_results_df)
}
```

```{r}
# Define parameters of interest
species_name = "Homo sapiens"
category = 'C5'
subcategory = 'GO:BP'

# Filter to up in human
filt_df <- baseline_divergence_df %>% dplyr::filter(logFCsign==1)
# Run ORA
human_up_gobp_df <- baseline_div_genes_ora(filt_df, species_name, category, subcategory)
human_up_gobp_df$direction <- 'Up in human'
human_up_gobp_df$gene_set <- 'GOBP'
head(human_up_gobp_df)
```

```{r}
# Define parameters of interest
species_name = "Homo sapiens"
category = 'C5'
subcategory = 'GO:BP'

# Filter to up in human
filt_df <- baseline_divergence_df %>% dplyr::filter(logFCsign==-1)
# Run ORA
chimp_up_gobp_df <- baseline_div_genes_ora(filt_df, species_name, category, subcategory)
chimp_up_gobp_df$direction <- 'Up in chimp'
chimp_up_gobp_df$gene_set <- 'GOBP'
head(chimp_up_ora_df)
```

```{r}
# Define parameters of interest
species_name = "Homo sapiens"
category = 'C5'
subcategory = 'GO:CC'

# Filter to up in human
filt_df <- baseline_divergence_df %>% dplyr::filter(logFCsign==1)
# Run ORA
human_up_gocc_df <- baseline_div_genes_ora(filt_df, species_name, category, subcategory)
human_up_gocc_df$direction <- 'Up in human'
human_up_gocc_df$gene_set <- 'GOCC'
head(human_up_gocc_df)
```

```{r}
# Define parameters of interest
species_name = "Homo sapiens"
category = 'C5'
subcategory = 'GO:CC'

# Filter to up in human
filt_df <- baseline_divergence_df %>% dplyr::filter(logFCsign==-1)
# Run ORA
chimp_up_gocc_df <- baseline_div_genes_ora(filt_df, species_name, category, subcategory)
chimp_up_gocc_df$direction <- 'Up in chimp'
chimp_up_gocc_df$gene_set <- 'GOCC'
head(chimp_up_gocc_df)
```

```{r}
div_baseline_ora_df <- dplyr::bind_rows(human_up_gobp_df, chimp_up_gobp_df, human_up_gocc_df, chimp_up_gocc_df)
head(div_baseline_ora_df)
```

```{r}
saveRDS(div_baseline_ora_df, paste0(dir_path, "divergent_baseline_genes_ORA_df.rds"))
```

## Baseline divergence GSEA GO:BP

```{r baseline divergent gene GSEA}
# Initiate empty list to store results
gsea_results_list <- list()

# Iterate through tiem points
for (time in baseline_divergence_df$time_point){
# Extract gene and logFC data drame
gene_list <- baseline_divergence_df %>% filter(time_point == time) %>%
  dplyr::select(ID, logFC) %>%
  dplyr::arrange(desc(logFC))
        
  # Skip if empty
  if (nrow(gene_list) == 0) next
        
  gene_list <- deframe(gene_list)
        
  # Skip if gene_list has no names
  if (is.null(names(gene_list)) || length(gene_list) == 0) next
    
  # Run GSEA using GO:BP terms
  gsea_go <- gseGO(
    geneList = gene_list,
    OrgDb = org.Hs.eg.db,
    keyType = "SYMBOL",  
    ont = "BP", 
    pvalueCutoff = 1,
    minGSSize = 10,
    eps = 0,
    verbose = TRUE,
    by = "fgsea"
    )
        
  # Results data frmae
  gsea_results <- gsea_go@result %>%
    filter(qvalue < 0.10) %>%
    mutate(time_point = time)
  
  # Add GSEA results to data frame
  gsea_results_list[[time]] <- gsea_results
}

# Combine all GSEA results
gsea_results_df <- dplyr::bind_rows(gsea_results_list)

head(gsea_results_df)
```

## Top baseline divergent rGene 

```{r}
n_genes <- 100

# Get top  human-upregulated genes
top_human_up_t1 <- baseline_divergence_df %>%
  filter(significant==TRUE, time_point=='6hr') %>%
  arrange(desc(logFC)) %>%
  slice_head(n = n_genes) %>%
  pull(ID) %>%
  unique()

top_human_up_t2 <- baseline_divergence_df %>%
  filter(significant==TRUE, time_point=='54hr') %>%
  arrange(desc(logFC)) %>%
  slice_head(n = n_genes) %>%
  pull(ID) %>%
  unique()

top_human_up_t3 <- baseline_divergence_df %>%
  filter(significant==TRUE, time_point=='7day') %>%
  arrange(desc(logFC)) %>%
  slice_head(n = n_genes) %>%
  pull(ID) %>%
  unique()

top_human_up_t12 <- intersect(top_human_up_t1, top_human_up_t2)
top_human_up_t123 <- intersect(top_human_up_t12, top_human_up_t3)
length(top_human_up_t123)
```

```{r}
n_genes <- 100

# Get top  human-upregulated genes
top_chimp_up_t1 <- baseline_divergence_df %>%
  filter(significant==TRUE, time_point=='6hr') %>%
  arrange(logFC) %>%
  slice_head(n = n_genes) %>%
  pull(ID) %>%
  unique()

top_chimp_up_t2 <- baseline_divergence_df %>%
  filter(significant==TRUE, time_point=='54hr') %>%
  arrange(logFC) %>%
  slice_head(n = n_genes) %>%
  pull(ID) %>%
  unique()

top_chimp_up_t3 <- baseline_divergence_df %>%
  filter(significant==TRUE, time_point=='7day') %>%
  arrange(logFC) %>%
  slice_head(n = n_genes) %>%
  pull(ID) %>%
  unique()

top_chimp_up_t12 <- intersect(top_chimp_up_t1, top_chimp_up_t2)
top_chimp_up_t123 <- intersect(top_chimp_up_t12, top_chimp_up_t3)
length(top_chimp_up_t123)
```

## Top baseline Divergent rGene biotypes

```{r}
biotype_to_broad <- function(bt) {
  bt <- tolower(bt)

  dplyr::case_when(
    is.na(bt) | bt == ""                          ~ "other/unknown",

    bt == "protein_coding"                        ~ "protein-coding",

    # pseudogenes (include translated processed pseudogene explicitly)
    bt == "translated_processed_pseudogene"       ~ "pseudogene",
    str_detect(bt, "pseudogene")                  ~ "pseudogene",

    # antisense
    str_detect(bt, "antisense")                   ~ "antisense",

    # long non-coding / lncRNA-like transcript types (include processed_transcript explicitly)
    bt == "processed_transcript"                  ~ "long non-coding",
    bt %in% c("lincrna", "lncrna", "macro_lncrna", "3prime_overlapping_ncrna") |
      str_detect(bt, "lncrna") |
      str_detect(bt, "sense_intronic|sense_overlapping|bidirectional_promoter") ~
      "long non-coding",

    # small RNAs
    str_detect(bt, "mirna")                       ~ "miRNA",
    str_detect(bt, "snrna|snora|scarna|snorna")   ~ "snRNA/snoRNA",
    str_detect(bt, "rrna")                        ~ "rRNA",
    str_detect(bt, "trna")                        ~ "tRNA",

    # other small ncRNAs often missed
    str_detect(bt,
      "y_rna|vault_rna|srp_rna|7sk|7sl|rnase_p_rna|rnase_mrp_rna|ribozyme|hammerhead|s_rna"
    )                                             ~ "other ncRNA",

    # generic ncRNA buckets
    str_detect(bt, "misc_rna|non_coding")         ~ "other ncRNA",

    TRUE                                          ~ "other/unknown"
  )
}

# --- new function: all genes  (GeneSymbol, GeneSetID) ---
make_biotype_gene_sets <- function(species = "hsapiens") {

  dataset <- paste0(species, "_gene_ensembl")
  mart <- biomaRt::useEnsembl(biomart = "genes", dataset = dataset)

  # pull ALL genes with fine biotype info
  attrs <- c("ensembl_gene_id", "external_gene_name", "gene_biotype")
  res <- biomaRt::getBM(
    attributes = attrs,
    mart       = mart
  )

  gene_set_df <- res %>%
    distinct(ensembl_gene_id, external_gene_name, gene_biotype) %>%
    dplyr::filter(!is.na(external_gene_name), external_gene_name != "") %>%
    mutate(GeneSets = biotype_to_broad(gene_biotype)) %>%
    mutate(
      GeneSets = dplyr::case_when(
        stringr::str_detect(external_gene_name, "(?i)-AS\\d*$") ~ "antisense",
        stringr::str_detect(external_gene_name, "^LINC")        ~ "long non-coding",
        TRUE                                                    ~ GeneSets
      )
    ) %>%
    transmute(GeneSymbol = external_gene_name, GeneSets) %>%
    distinct()

  gene_set_df
}
```

```{r}
n_genes <- 100
# Get top  human-upregulated genes
top_human_up <- baseline_divergence_df %>%
  filter(significant==TRUE) %>%
  arrange(desc(logFC)) %>%
  slice_head(n = n_genes) %>%
  transmute(GeneSymbol = ID,
            logFC = logFC,
            direction = "Human-upregulated") %>%
  distinct(GeneSymbol, .keep_all = TRUE)

# Get top 500 chimp-upregulated genes
top_chimp_up <- baseline_divergence_df %>%
  filter(significant==TRUE) %>%
  arrange(logFC) %>%
  slice_head(n = n_genes) %>%
  transmute(GeneSymbol = ID,
            logFC = logFC,
            direction = "Chimp-upregulated") %>%
  distinct(GeneSymbol, .keep_all = TRUE)

top_genes <- bind_rows(top_human_up, top_chimp_up)

# build biotype lookup table from Ensembl
biotype_gene_sets <- make_biotype_gene_sets(species = "hsapiens")

# Join + summarize biotypes
top_biotypes <- top_genes %>%
  left_join(biotype_gene_sets, by = "GeneSymbol") %>%
  mutate(GeneSets = tidyr::replace_na(GeneSets, "other/unknown"))

# per-gene output (GeneSymbol + direction + broad biotype)
top_biotypes_gene_level <- top_biotypes %>%
  dplyr::select(GeneSymbol, direction, logFC, GeneSets)

# Counts by direction x broad biotype
biotype_counts <- top_biotypes %>%
  dplyr::count(direction, GeneSets, name = "N_genes") %>%
  group_by(direction) %>%
  mutate(frac = N_genes / sum(N_genes)) %>%
  ungroup() %>%
  arrange(direction, desc(N_genes))

# inspect results
biotype_counts
```

```{r}
# Build plotting df with label positions (ymid) for outside labels
plot_df <- biotype_counts %>%
  mutate(
    direction = factor(direction, levels = c("Human-upregulated", "Chimp-upregulated")),
    GeneSets  = factor(GeneSets)
  ) %>%
  group_by(direction) %>%
  arrange(direction, desc(frac), .by_group = TRUE) %>%
  mutate(
    ymax  = cumsum(frac),
    ymin  = dplyr::lag(ymax, default = 0),
    ymid  = (ymin + ymax) / 2,
    label = paste0(round(frac * 100, 1), "%")
  ) %>%
  ungroup()

p <- ggplot(plot_df, aes(ymax = ymax, ymin = ymin, xmax = 1, xmin = 0, fill = GeneSets)) +
  geom_rect(color = "white", linewidth = 0.3) +
  coord_polar(theta = "y", clip = "off") +     # allow text beyond panel (no clipping)
  facet_wrap(~ direction, nrow=1) +
  xlim(0, 1.55) +                              # less extra space => labels closer
  geom_text_repel(
    data = plot_df,
    aes(x = 1.18, y = ymid, label = label),    # bring labels closer to pie
    inherit.aes = FALSE,
    nudge_x = 0.01,
    direction = "y",
    vjust = 0.5,
    size = 3.5,
    segment.size = 0.25,
    segment.color = "grey50",
    box.padding = 0.15,                        # tighter label packing
    point.padding = 0.1,
    min.segment.length = 0,
    show.legend = FALSE
  ) +
  labs(x = NULL, y = NULL) +
  theme_void() +
  theme(
    strip.text = element_text(size = 14),
    legend.position = "bottom",
    plot.margin = margin(t = 12, r = 18, b = 12, l = 12), # prevent facet strip/title cut-off
    legend.title = element_blank()
  ) +
  guides(fill = guide_legend(nrow = 2, byrow = TRUE, title.position = "bottom"))

ggsave(
  filename = paste0(fig_path, "PieChart_Top100Genes_HumanUpOrChimpUp_ColorByGeneBiotype.pdf"),
  plot = p, height = 4, width = 6, units = "in", bg = "white", dpi = 600
)

print(p)
```


```{r}
plot_df <- biotype_counts %>%
  mutate(
    direction = factor(direction, levels = c("Human-upregulated", "Chimp-upregulated")),
    GeneSets  = factor(GeneSets),
    label     = paste0(round(frac*100,1), "%")
  )

p <- ggplot(plot_df, aes(x = 1, y = frac, fill = GeneSets)) +
  geom_col(width = 1, color = "white", linewidth = 0.3) +
  coord_polar(theta = "y") +
  facet_wrap(~ direction) +
  geom_text(
    aes(label = label),
    position = position_stack(vjust = 0.5),
    size = 3
  ) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = NULL, y = NULL, fill = "Gene Biotype") +
  theme_void() +
  theme(
    strip.text = element_text(size = 14),
    legend.position = "right"
  )

ggsave(paste0(fig_path, "PieChart_Top100Genes_HumanUpOrChimpUp_ColorByGeneBiotype.pdf"),
    plot = p, height = 4, width = 8, units = "in", bg = "white", dpi=600)

print(p)
```




# Divergent Response Analysis

## N total telNECs vs. N divergent rGenes scatter plot

```{r scatter plot N rGenes vs N TelNECs by sample group}
# Load in counts data for sample groups
sg_counts_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/scanpy/telencephalon_final/NEMP25_Tel_v5_sample_group_TelNEC_counts.csv")
sg_counts_df <- sg_counts_df %>% filter(!(perturbation %in% c('TreMan', 'Untreated', 'DMSO')) & species != 'Orangutan') %>%
  dplyr::rename(N_cells = n) %>%
  dplyr::group_by(perturbation, time_point) %>%
  dplyr::summarise(
    N_cells = sum(N_cells, na.rm = TRUE),
    .groups = "drop"
  )

# Generate rGene data frame for sample groups
sg_divergent_rgenes_df <- divergent_response_df %>% filter(significant == TRUE & !(perturbation %in% c('TreMan', 'Untreated', 'DMSO'))) %>% 
  dplyr::group_by(perturbation, time_point) %>%
  dplyr::summarise(N_divergent_rGenes = sum(!is.na(ID) & ID != ""), .groups = "drop") %>%
  complete(perturbation, time_point, fill = list(N_divergent_rGenes = 0))
  
# Join rGene and TelNEC counts data frames
sg_gene_cell_counts_df <- dplyr::left_join(sg_counts_df, sg_divergent_rgenes_df, by = c('perturbation', 'time_point')) %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Calculate Spearman's correlation coefficient
spearman_res <- cor.test(
  sg_gene_cell_counts_df$N_cells,
  sg_gene_cell_counts_df$N_divergent_rGenes,
  method = "spearman",
  exact = FALSE
)
spearman_rho <- unname(spearman_res$estimate)
spearman_p <- spearman_res$p.value

# Add padding
x_pad <- diff(range(sg_gene_cell_counts_df$N_cells, na.rm = TRUE)) * 0.005
y_pad <- diff(range(sg_gene_cell_counts_df$N_divergent_rGenes, na.rm = TRUE)) * 0.005

# Plot a scatter plot of N rGenes/sample group vs. N TelNECs/sample group
p <- ggplot(sg_gene_cell_counts_df, aes(x = N_cells, y = N_divergent_rGenes, shape = time_point)) +
  geom_point(size = 4, color = "black", fill='black', stroke = 0.5) +
  labs(x = "Total Cell Number Across Species", y = "N Divergent rGenes") +
  #scale_fill_manual(
    #name = 'Stimulus',
    #values = pert_colors,
    #guide = guide_legend(ncol=2, title.position='top', override.aes = list(shape = 21, color = "black", stroke = 0.5))) +
  scale_shape_manual(
    name = "Time Point",
    values = time_point_shapes,
    guide = guide_legend(override.aes = list(fill = "black"))
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title = element_text(size = 14),
    axis.text  = element_text(size = 12),
    legend.position = "right"
  ) +
  annotate(
    "text",
    x = min(sg_gene_cell_counts_df$N_cells, na.rm = TRUE) + x_pad,
    y = max(sg_gene_cell_counts_df$N_divergent_rGenes, na.rm = TRUE) - y_pad,
    hjust = 0, vjust = 1,
    label = sprintf("rho = %.2f\np = %.2g", spearman_rho, spearman_p),
    size = 5
  )

print(p)
ggsave(filename = paste0(fig_path, "ScatterPlot_NTotalTelNECs_vs_NDivergentrGenes_BySampleGroup.png"), plot = p, units='in', height = 4, width = 5, bg = "white", dpi=600)
```

## Species Difference telNECs vs. N divergent rGenes scatter plot

```{r scatter plot N rGenes vs N TelNECs by sample group}
# Load in counts data for sample groups
sg_counts_df <- read.csv("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/scanpy/telencephalon_final/NEMP25_Tel_v5_sample_group_TelNEC_counts.csv")

# Chimp - Human within each (perturbation, time_point)
sg_counts_diff_df <- sg_counts_df %>%
  dplyr::filter(
    !(perturbation %in% c("TreMan", "Untreated", "DMSO")),
    species %in% c("Human", "Chimp")
  ) %>%
  dplyr::rename(N_cells = n) %>%
  dplyr::group_by(perturbation, time_point, species) %>%
  dplyr::summarise(N_cells = sum(N_cells, na.rm = TRUE), .groups = "drop") %>%
  tidyr::pivot_wider(names_from = species, values_from = N_cells, values_fill = 0) %>%
  dplyr::mutate(N_cells_diff = abs(Chimp - Human))

# Generate rGene data frame for sample groups
sg_divergent_rgenes_df <- divergent_response_df %>% filter(significant == TRUE & !(perturbation %in% c('TreMan', 'Untreated', 'DMSO'))) %>% 
  dplyr::group_by(perturbation, time_point) %>%
  dplyr::summarise(N_divergent_rGenes = sum(!is.na(ID) & ID != ""), .groups = "drop") %>%
  complete(perturbation, time_point, fill = list(N_divergent_rGenes = 0))
  
# Join rGene and TelNEC counts data frames
sg_gene_cell_counts_df <- dplyr::left_join(sg_counts_diff_df, sg_divergent_rgenes_df, by = c('perturbation', 'time_point')) %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Calculate Spearman's correlation coefficient
spearman_res <- cor.test(
  sg_gene_cell_counts_df$N_cells_diff,
  sg_gene_cell_counts_df$N_divergent_rGenes,
  method = "spearman",
  exact = FALSE
)
spearman_rho <- unname(spearman_res$estimate)
spearman_p <- spearman_res$p.value

# Add padding
x_pad <- diff(range(sg_gene_cell_counts_df$N_cells_diff, na.rm = TRUE)) * 0.005
y_pad <- diff(range(sg_gene_cell_counts_df$N_divergent_rGenes, na.rm = TRUE)) * 0.005

# Plot a scatter plot of N rGenes/sample group vs. N TelNECs/sample group
p <- ggplot(sg_gene_cell_counts_df, aes(x = N_cells_diff, y = N_divergent_rGenes, shape = time_point)) +
  geom_point(size = 4, color = "black", fill='black', stroke = 0.5) +
  labs(x = "|Species Difference in Cell Number|", y = "N Divergent rGenes") +
  #scale_fill_manual(
    #name = 'Stimulus',
    #values = pert_colors,
    #guide = guide_legend(ncol=2, title.position='top', override.aes = list(shape = 21, color = "black", stroke = 0.5))) +
  scale_shape_manual(
    name = "Time Point",
    values = time_point_shapes,
    guide = guide_legend(override.aes = list(fill = "black"))
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title = element_text(size = 14),
    axis.text  = element_text(size = 12),
    legend.position = "right"
  ) +
  annotate(
    "text",
    x = min(sg_gene_cell_counts_df$N_cells_diff, na.rm = TRUE) + x_pad,
    y = max(sg_gene_cell_counts_df$N_divergent_rGenes, na.rm = TRUE) - y_pad,
    hjust = 0, vjust = 1,
    label = sprintf("rho = %.2f\np = %.2g", spearman_rho, spearman_p),
    size = 5
  )

print(p)
ggsave(filename = paste0(fig_path, "ScatterPlot_SpeciesDiffNTelNECs_vs_NDivergentrGenes_BySampleGroup.png"), plot = p, units='in', height = 4, width = 5, bg = "white", dpi=600)
```

## N conserved rGenes vs. N divergent rGenes

Plot a scatter plot of N conserved rGenes vs. N divergent rGenes
```{r scatterplot cons vs div rGenes}
c_df <- conserved_rgenes_df %>% filter(!(perturbation %in% c('TreMan', 'DMSO'))) %>%
  dplyr::select(perturbation, time_point, N_conserved_rGenes)

d_df <- divergent_rgenes_df %>% filter(!(perturbation %in% c('TreMan', 'DMSO'))) %>%
  dplyr::select(perturbation, time_point, N_div_rGenes)

rgenes_df <- left_join(c_df, d_df, by = c("perturbation", "time_point"))
rgenes_df <- rgenes_df %>%
  dplyr::select(perturbation, time_point, N_conserved_rGenes, N_div_rGenes) %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Calculate Spearman's correlation coefficient
spearman_res <- cor.test(
  rgenes_df$N_conserved_rGenes,
  rgenes_df$N_div_rGenes,
  method = "spearman",
  exact = FALSE
)

spearman_rho <- unname(spearman_res$estimate)
spearman_p   <- spearman_res$p.value

# Add padding
x_pad <- diff(range(rgenes_df$N_conserved_rGenes, na.rm = TRUE)) * 0.005
y_pad <- diff(range(rgenes_df$N_div_rGenes, na.rm = TRUE)) * 0.005

p <- ggplot(rgenes_df, aes(x = N_conserved_rGenes, y = N_div_rGenes, fill = perturbation, shape = time_point)) +
  geom_point(size = 3, color = "black", stroke = 0.5, show.legend = TRUE) +
  labs(
    x = "N Conserved rGenes",
    y = "N Divergent rGenes",
    fill = "Stimulus"
  ) +
  scale_fill_manual(
    name = "Stimulus",
    values = pert_colors,
    guide = guide_legend(ncol = 2, override.aes = list(shape = 21, color = "black", stroke = 0.5))
  ) +
  scale_shape_manual(
    name = 'Time Point', 
    values = time_point_shapes,
    guide = guide_legend(override.aes = list(fill = "black"))
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title = element_text(size = 14),
    axis.text  = element_text(size = 12),
    legend.position = "right"
  ) +
  annotate(
    "text",
    x = min(rgenes_df$N_conserved_rGenes, na.rm = TRUE) + x_pad,
    y = max(rgenes_df$N_div_rGenes, na.rm = TRUE) - y_pad,
    hjust = 0, vjust = 1,
    label = sprintf("rho = %.2f\np = %.2g", spearman_rho, spearman_p),
    size = 5
  )

print(p)
ggsave(filename = paste0(fig_path, "ScatterPlot_NConservedrGenes_vs_NDivergentrGenes.png"),
       plot = p, height = 5, width = 6, bg = "white", dpi = 600)
```

## N Divergent rGenes across time points box plot

```{r box plot N rGenes across time points}
divergent_rgenes_df <- divergent_rgenes_df %>% mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))
                                  
p <- ggplot(divergent_rgenes_df, aes(x = time_point, y = N_div_rGenes, fill=time_point)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.15, alpha = 0.6, size = 1) +
  scale_fill_manual(values = time_point_colors, name = "Time Point") +
  labs(
    x = "",
    y = "N Divergent rGenes"
  ) +
  theme_classic() +
  theme(
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
  )
ggsave(filename = paste0(fig_path, "boxplot_NrDivergentGenes_ByTimePoint.png"), plot = p, height = 4, width = 3, bg = 'white', dpi=600)
print(p)
```

```{r}
mean_div_rgenes <- divergent_rgenes_df %>%
  mutate(time_point = factor(time_point, levels = c("6hr", "54hr", "7day"))) %>%
  group_by(time_point) %>%
  summarise(
    mean_N_div_rGenes = mean(N_div_rGenes, na.rm = TRUE),
    .groups = "drop"
  )

mean_div_rgenes
```

## N unique rGenes across time points box plot

```{r box plot N unique rGenes across time points}
# Find number of rGenes within a perturbation that are unique to a time point
unique_divergent_rgenes_df <- divergent_response_df %>% filter(significant==TRUE) %>%
  dplyr::select(perturbation, time_point, ID) %>%
  group_by(perturbation, ID) %>%
  mutate(n_timepoints_in_pert = n_distinct(time_point)) %>%
  ungroup() %>%
  filter(n_timepoints_in_pert == 1) %>%                  # unique within perturbation across time points
  dplyr::count(perturbation, time_point, name = "N_unique_divergent_rGenes") %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Plot a box plot of unique rGenes within perturbation per time point                                  
p <- ggplot(unique_divergent_rgenes_df, aes(x = time_point, y = N_unique_divergent_rGenes, fill=time_point)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.15, alpha = 0.6, size = 1) +
  scale_fill_manual(values = time_point_colors, name = "Time point") +
  labs(
    x = "",
    y = "N Unique Divergent rGenes (Within Stimulus)"
  ) +
  theme_classic() +
  theme(
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    legend.position = "none",
  )
ggsave(filename = paste0(fig_path, "boxplot_NUniqueDivergentrGenes_AcrossTimePointsWithinPerturbation_ByTimePoint.png"),
       plot = p, height = 4, width = 2.25, bg = 'white', dpi=600)
print(p)
```

## N divergent rGenes over time

```{r}
sum_div_rgenes_by_time <- divergent_rgenes_df %>%
  group_by(time_point) %>%
  summarise(sum_div_rGenes = sum(N_div_rGenes, na.rm = TRUE)) %>%
  as.data.frame() %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day'))) %>%
  arrange(time_point)

sum_div_rgenes_by_time

# Plot a box plot of unique rGenes within perturbation per time point                                  
p <- ggplot(sum_div_rgenes_by_time, aes(x = time_point, y = sum_div_rGenes, fill=time_point)) +
  geom_col() +
  scale_fill_manual(values = time_point_colors, name = "Time Point") +
  labs(
    x = "",
    y = "Divergent rGene Count"
  ) +
  theme_classic() +
  theme(
    axis.text  = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    legend.position = "none",
  )
ggsave(filename = paste0(fig_path, "boxplot_NTotalDivergentrGenes_ByTimePoint.png"),
       plot = p, height = 4, width = 2.25, bg = 'white', dpi=600)
print(p)
```

## Upset plot divergent rGenes

```{r upset plot simple divergent rGenes}
binary_df <- simple_divergent_rgenes_df %>%
  dplyr::filter(!perturbation %in% c("TreMan", "DMSO")) %>%
  tidyr::separate_rows(div_rGenes, sep = "/") %>%
  dplyr::mutate(div_rGenes = trimws(div_rGenes)) %>%
  dplyr::filter(!is.na(div_rGenes), div_rGenes != "") %>%
  dplyr::distinct(gene = div_rGenes, perturbation) %>%
  dplyr::mutate(present = 1L) %>%
  tidyr::pivot_wider(
    names_from  = perturbation,
    values_from = present,
    values_fill = 0L
  )

set_cols <- setdiff(names(binary_df), "gene")

# --- NEW: compute intersections with > 25 genes ---
mat <- as.matrix(binary_df[, set_cols, drop = FALSE])

# Build a key per gene that identifies the exact intersection (e.g., "BMP7&FGF2")
intersection_key <- apply(mat, 1, function(v) {
  on <- set_cols[as.logical(v)]
  if (length(on) == 0) return(NA_character_)  # ignore genes in no sets
  paste(on, collapse = "&")
})

intersection_counts <- sort(table(intersection_key, useNA = "no"), decreasing = TRUE)

# Keep only intersection patterns with > 25 genes
keep_keys <- names(intersection_counts)[intersection_counts > 3]

# Convert to UpSetR format: list of character vectors
intersections_keep <- lapply(keep_keys, function(k) strsplit(k, "&", fixed = TRUE)[[1]])

# Optional: if you only want TRUE intersections (>=2 sets), uncomment:
# intersections_keep <- intersections_keep[vapply(intersections_keep, length, integer(1)) >= 2]

png(
  paste0(fig_path, "upsetplot_simple_DivergentrGenes.png"),
  width = 12, height = 8, units = "in", res = 600
)

if (length(intersections_keep) == 0) {
  warning("No intersections with > n genes found ensured; plotting default UpSetR output instead.")
  UpSetR::upset(
    as.data.frame(binary_df),
    sets = set_cols,
    keep.order = FALSE,
    sets.bar.color = "gray20",
    order.by = "freq",
    mb.ratio = c(0.35, 0.65),
    sets.x.label = "Set Size",
    mainbar.y.label = "Intersection Size",
    decreasing = c(TRUE, NA)
  )
} else {
  UpSetR::upset(
    as.data.frame(binary_df),
    sets = set_cols,
    intersections = intersections_keep,  # <-- restrict plotted intersections here
    keep.order = FALSE,
    sets.bar.color = "gray20",
    order.by = "freq",
    mb.ratio = c(0.50, 0.50),
    sets.x.label = "Set Size",
    mainbar.y.label = "Intersection Size",
    decreasing = c(TRUE, NA),
    text.scale = c(2.5, #intersection_size_title,   
                   2.0, #intersection_size_ticks,
                   2.5, #set_size_title,
                   2.0, #set_size_ticks,
                   2.0  #set_names
    ),
    show.numbers = FALSE
  )
}
dev.off()
```

```{r}
n_div_rGenes_only_one_pert <- binary_df %>%
  # sum across all perturbation columns (everything except 'gene')
  mutate(n_perts = rowSums(across(-gene))) %>%
  filter(n_perts == 1) %>%
  summarise(n_genes = n()) %>%
  pull(n_genes)

n_div_rGenes_only_one_pert
```

```{r}
pct_only_one_pert <- binary_df %>%
  mutate(n_perts = rowSums(across(-gene))) %>%
  summarise(100 * sum(n_perts == 1) / n()) %>%
  pull()

pct_only_one_pert
```

## Divergent rGene volcano plots

Define a function for divergent rGene volcano plot
```{r drGenes_volcano_plot function}
drGenes_volcano_plot <- function(response_df, pert, top_n) {

  # Filter data to the current perturbation
  filt_df <- response_df %>% filter(perturbation == pert)

  # Skip iteration if there are no results.
  if (nrow(filt_df) == 0) return(NULL)

  # Determine the global max x and y values for setting the same scale axes.
  max_x_val <- max(abs(filt_df$logFC), na.rm = TRUE)
  max_y_val <- max(-(log10(filt_df$adj.P.Val)), na.rm = TRUE)

  # Identify top n lowest p.adjust values for both positive and negative logFC values per time point.
  top_up_labeled <- filt_df %>% filter(logFC > 0 & significant == TRUE) %>%
    group_by(time_point) %>%
    slice_min(order_by = adj.P.Val, n = top_n, with_ties = FALSE) %>%
    ungroup()

  top_down_labeled <- filt_df %>% filter(logFC < 0 & significant == TRUE) %>%
    group_by(time_point) %>%
    slice_min(order_by = adj.P.Val, n = top_n, with_ties = FALSE) %>%
    ungroup()

  top_labeled <- bind_rows(top_up_labeled, top_down_labeled)

  # Plot a volcano plot faceted by time point
  p <- ggplot(filt_df, aes(x = logFC, y = -log10(adj.P.Val))) +

    geom_point(data = filt_df,
               aes(fill = simple_signaling_category),
               color = 'black',
               shape = 21,
               stroke = 0.25,
               alpha = 0.75) +

    # Data labels at moved positions
    geom_text_repel(
      data = top_labeled,
      aes(x = logFC, y = -log10(adj.P.Val), label = ID),
      size = 3,
      segment.color = 'black',
      min.segment.length = 0.5
    ) + 

    # Define scales
    scale_x_continuous(limits = c(-max_x_val * 1.1, max_x_val * 1.1)) +
    scale_y_continuous(limits = c(0, max_y_val * 1.1)) +
    scale_fill_manual(values = signaling_category_colors, name = "Signaling Category",
                      guide = guide_legend(title = "Signaling Category", override.aes = list(shape = 21, size = 4, stroke=0.25, color = 'black'))) +

    # Facet by time point
    facet_wrap(~ time_point) +

    # Labels and theme
    labs(
      x = expression("(Human - Chimp) x (Stimulus - Vehicle) " ~ Log[2] ~ "FC"),
      y = expression("-" ~ Log[10] ~ "(Adjusted p-value)")
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 24),
      strip.text = element_text(hjust = 0.5, size = 18),
      plot.background = element_blank(),
      panel.background = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.line = element_line(color = "black", linewidth = 1),
      legend.position = "right"
    ) +

    # Significance lines
    geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "darkgrey", linewidth = 1) +
    geom_vline(xintercept = 0, linetype = "dashed", color = "darkgrey", linewidth = 1)

  return(p)
}
```

Plot divergent rGenes volcano plots
```{r drGenes volcano plots}
# Iterate through perturbation to plot
for (pert in unique(divergent_response_df$perturbation)){
  p_combined <- drGenes_volcano_plot(divergent_response_df, pert, 5)
  if (!is.null(p_combined)) {
    ggsave(filename = paste0(fig_path, paste0("/volcanoplot_divergent_rGenes/volcanoplot_divergent_rGenes_", pert, ".png")),
           plot = p_combined, height = 5, width = 10, bg = 'white', units='in', dpi=600)
  }
}
```




Plot a volcano plot of divergent rGenes labeleling select genes and coloring by SCZ association
```{r volcano plot divergent rGenes}
library(dplyr)
library(ggplot2)
library(ggrepel)

# Define variables of interest
pert <- "BMP7"
time <- "6hr"
genes <- c('NKD1', 'ROR2', 'SOX4', 'SMAD3', 'GATA3')

# Filter divergent response data frame for select perturbation and time point
filt_df <- divergent_response_df %>%
  dplyr::filter(perturbation == pert, time_point == time) %>%
  dplyr::mutate(neglog10_adjP = -log10(adj.P.Val))

# Split dataset for plotting (handle NA categories safely)
nonsig_df <- filt_df %>% dplyr::filter(significant == FALSE)

label_df <- filt_df %>%
  dplyr::filter(significant == TRUE, ID %in% genes)

sig_df <- filt_df %>%
  dplyr::filter(significant == TRUE, !(ID %in% genes))

# Determine global max x and y values for setting the same scale axes
max_x_val <- max(abs(filt_df$logFC), na.rm = TRUE)
max_y_val <- max(filt_df$neglog10_adjP, na.rm = TRUE)

# Plot a volcano plot
p <- ggplot(filt_df, aes(x = logFC, y = neglog10_adjP)) +
  geom_point(
    data = nonsig_df,
    aes(fill = "Not Significant"),
    shape = 21, color = "black", stroke = 0, alpha = 1, size = 3
  ) +
  geom_point(
    data = sig_df,
    aes(fill = "Significant"),
    shape = 21, color = "black", stroke = 0, alpha = 1.0, size = 3
  ) +
  geom_point(
    data = label_df,
    aes(fill = "Significant, WNT Signaling"),
    shape = 21, color = "black", stroke = 0, alpha = 1.0, size = 3
  ) +
  labs(
    x = expression("Species  Treatment " ~ Log[2] ~ "FC"),
    y = expression("-" ~ Log[10] ~ "(Adjusted p-value)")
  ) +
  scale_x_continuous(limits = c(-max_x_val * 1.1, max_x_val * 1.1)) +
  scale_y_continuous(limits = c(0, max_y_val * 1.1)) +
  scale_fill_manual(
    values = c("Not Significant" = "lightgrey", "Significant, WNT Signaling" = "royalblue", "Significant" = "grey40")
  ) +
  theme_minimal() +
  theme(
    plot.background  = element_blank(),
    panel.background = element_blank(),
    panel.grid       = element_blank(),
    axis.line        = element_line(color = "black", linewidth = 1),
    legend.position  = "bottom",
    legend.title     = element_blank(),
    axis.text        = element_text(size = 12),
    axis.title       = element_text(size = 14),
    legend.text      = element_text(size = 12)
  ) +
  ggrepel::geom_text_repel(
    data = label_df,
    aes(label = ID),
    size = 4,
    segment.color = "black",
    min.segment.length = 0.2,
    box.padding = 0.7,
    point.padding = 0.7,
  )
ggsave(filename = paste0(fig_path, paste0("/VolcanoPlot_DivergentrGene_BMP7_6hrs_ColorBySigWNTSignaling.pdf")),
  plot = p, height = 5, width = 5.5, bg = 'white', units='in', dpi=600)
print(p)
```

```{r volcano plot divergent rGenes}
library(dplyr)
library(ggplot2)
library(ggrepel)

# Define variables of interest
pert <- "BMP7"
time <- "54hr"
genes <- cc_genes

# Filter divergent response data frame for select perturbation and time point
filt_df <- divergent_response_df %>%
  dplyr::filter(perturbation == pert, time_point == time) %>%
  dplyr::mutate(neglog10_adjP = -log10(adj.P.Val))

# Split dataset for plotting (handle NA categories safely)
nonsig_df <- filt_df %>% dplyr::filter(significant == FALSE)
label_df <- filt_df %>% dplyr::filter(significant == TRUE, ID %in% genes)
sig_df <- filt_df %>% dplyr::filter(significant == TRUE, !(ID %in% genes))

# Determine global max x and y values for setting the same scale axes
max_x_val <- max(abs(filt_df$logFC), na.rm = TRUE)
max_y_val <- max(filt_df$neglog10_adjP, na.rm = TRUE)

# Choose nudges (in data units)
nudge_y_mag <- 0.05 * max_y_val   # up
nudge_x_mag <- 0.05 * max_x_val   # left/right

# Plot a volcano plot
p <- ggplot(filt_df, aes(x = logFC, y = neglog10_adjP)) +
  geom_point(
    data = nonsig_df,
    aes(fill = "Not Significant"),
    shape = 21, color = "black", stroke = 0, alpha = 1, size = 3
  ) +
  geom_point(
    data = sig_df,
    aes(fill = "Significant"),
    shape = 21, color = "black", stroke = 0, alpha = 1, size = 3
  ) +
  geom_point(
    data = label_df,
    aes(fill = "Significant, Cell Cycle"),
    shape = 21, color = "black", stroke = 0, alpha = 1.0, size = 3
  ) +
  labs(
    x = expression("Species  Treatment " ~ Log[2] ~ "FC"),
    y = expression("-" ~ Log[10] ~ "(Adjusted p-value)")
  ) +
  scale_x_continuous(limits = c(-max_x_val * 1.1, max_x_val * 1.1)) +
  scale_y_continuous(limits = c(0, max_y_val * 1.1)) +
  scale_fill_manual(
    values = c("Not Significant" = "lightgrey", "Significant, Cell Cycle" = "firebrick", "Significant" = "grey40")
  ) +
  theme_minimal() +
  theme(
    plot.background  = element_blank(),
    panel.background = element_blank(),
    panel.grid       = element_blank(),
    axis.line        = element_line(color = "black", linewidth = 1),
    legend.position  = "bottom",
    legend.title     = element_blank(),
    axis.text        = element_text(size = 12),
    axis.title       = element_text(size = 14),
    legend.text      = element_text(size = 12)
  ) +
  ggrepel::geom_text_repel(
    data = label_df %>% filter(ID %in% c('MKI67', 'RRM2', 'PBK', 'AURKB', 'KNTC1', 'KIF4A', 'BARD1', 'NCAPH', 'SPDL1', 'CLSPN', 'CKS1B')),
    aes(
      label = ID,
      nudge_x = ifelse(logFC >= 0,  nudge_x_mag, -nudge_x_mag),  # right if +, left if -
      nudge_y = nudge_y_mag                                      # always up
    ),
    size = 4,
    segment.color = "black",
    min.segment.length = 0.2,
    box.padding = 0.5,
    point.padding = 0.5,
    max.overlaps = Inf
  )
ggsave(filename = paste0(fig_path, paste0("/VolcanoPlot_DivergentrGene_BMP7_54hrs_ColorBySigCCGenes.pdf")),
  plot = p, height = 5, width = 5.5, bg = 'white', units='in', dpi=600)
print(p)
```

```{r volcano plot divergent rGenes FGF2 FGF8b 6hr}
# Define variables of interest
pert1 <- 'FGF8b'
pert2 <- 'FGF2'
time <- '6hr'

# Filter data frame
filt_df <- divergent_response_df %>% dplyr::filter(time_point == time & perturbation %in% c(pert1, pert2) & significant==TRUE)

# Pull genes for both perturbations
pert1_genes <- filt_df %>% dplyr::filter(perturbation == pert1) %>% pull(ID)
pert2_genes <- filt_df %>% dplyr::filter(perturbation == pert2) %>% pull(ID)

# Find intersection of genes
intersect_genes <- intersect(pert1_genes, pert2_genes)

# Define variables of interest
pert <- 'FGF2'
time <- "6hr"
genes <- intersect_genes

# Filter divergent response data frame for select perturbation and time point
filt_df <- divergent_response_df %>%
  dplyr::filter(perturbation == pert, time_point == time) %>%
  dplyr::mutate(neglog10_adjP = -log10(adj.P.Val))

# Split dataset for plotting (handle NA categories safely)
nonsig_df <- filt_df %>% dplyr::filter(significant == FALSE)
label_df <- filt_df %>% dplyr::filter(significant == TRUE, ID %in% genes)
sig_df <- filt_df %>% dplyr::filter(significant == TRUE, !(ID %in% genes))

# Determine global max x and y values for setting the same scale axes
max_x_val <- max(abs(filt_df$logFC), na.rm = TRUE)
max_y_val <- max(filt_df$neglog10_adjP, na.rm = TRUE)

# Choose nudges (in data units)
nudge_y_mag <- 0.05 * max_y_val   # up
nudge_x_mag <- 0.05 * max_x_val   # left/right

# Plot a volcano plot
p <- ggplot(filt_df, aes(x = logFC, y = neglog10_adjP)) +
  geom_point(
    data = nonsig_df,
    aes(fill = "Not Significant"),
    shape = 21, color = "black", stroke = 0, alpha = 1, size = 3
  ) +
  geom_point(
    data = sig_df,
    aes(fill = "Significant"),
    shape = 21, color = "black", stroke = 0, alpha = 1, size = 3
  ) +
  geom_point(
    data = label_df,
    aes(fill = "Significant, Shared"),
    shape = 21, color = "black", stroke = 0, alpha = 1.0, size = 3
  ) +
  labs(
    x = expression("Species  Treatment " ~ Log[2] ~ "FC"),
    y = expression("-" ~ Log[10] ~ "(Adjusted p-value)")
  ) +
  scale_x_continuous(limits = c(-max_x_val * 1.1, max_x_val * 1.1)) +
  scale_y_continuous(limits = c(0, max_y_val * 1.1)) +
  scale_fill_manual(
    values = c("Not Significant" = "lightgrey", "Significant, Shared" = "#4F7942", "Significant" = "grey40")
  ) +
  theme_minimal() +
  theme(
    plot.background  = element_blank(),
    panel.background = element_blank(),
    panel.grid       = element_blank(),
    axis.line        = element_line(color = "black", linewidth = 1),
    legend.position  = "bottom",
    legend.title     = element_blank(),
    axis.text        = element_text(size = 12),
    axis.title       = element_text(size = 14),
    legend.text      = element_text(size = 12)
  ) +
  ggrepel::geom_text_repel(
    data = filt_df %>% dplyr::filter(ID %in% genes),
    aes(label = ID),
    size = 4,
    segment.color = "black",
    min.segment.length = 0.2,
    box.padding = 0.7,
    point.padding = 0.7,
  )
ggsave(filename = paste0(fig_path, paste0("/VolcanoPlot_DivergentrGene_FGF2_6hrs_ColorBySharedGenes.pdf")),
  plot = p, height = 5, width = 5, bg = 'white', units='in', dpi=600)
print(p)
```

```{r volcano plot divergent rGenes FGF8b 6hr}
# Define variables of interest
pert1 <- 'FGF8b'
pert2 <- 'FGF2'
time <- '6hr'

# Filter data frame
filt_df <- divergent_response_df %>% dplyr::filter(time_point == time & perturbation %in% c(pert1, pert2) & significant==TRUE)

# Pull genes for both perturbations
pert1_genes <- filt_df %>% dplyr::filter(perturbation == pert1) %>% pull(ID)
pert2_genes <- filt_df %>% dplyr::filter(perturbation == pert2) %>% pull(ID)

# Find intersection of genes
intersect_genes <- intersect(pert1_genes, pert2_genes)

# Define variables of interest
pert <- 'FGF8b'
time <- "6hr"
genes <- intersect_genes

# Filter divergent response data frame for select perturbation and time point
filt_df <- divergent_response_df %>%
  dplyr::filter(perturbation == pert, time_point == time) %>%
  dplyr::mutate(neglog10_adjP = -log10(adj.P.Val))

# Split dataset for plotting (handle NA categories safely)
nonsig_df <- filt_df %>% dplyr::filter(significant == FALSE)
label_df <- filt_df %>% dplyr::filter(significant == TRUE, ID %in% genes)
sig_df <- filt_df %>% dplyr::filter(significant == TRUE, !(ID %in% genes))

# Determine global max x and y values for setting the same scale axes
max_x_val <- max(abs(filt_df$logFC), na.rm = TRUE)
max_y_val <- max(filt_df$neglog10_adjP, na.rm = TRUE)

# Choose nudges (in data units)
nudge_y_mag <- 0.05 * max_y_val   # up
nudge_x_mag <- 0.05 * max_x_val   # left/right

# Plot a volcano plot
p <- ggplot(filt_df, aes(x = logFC, y = neglog10_adjP)) +
  geom_point(
    data = nonsig_df,
    aes(fill = "Not Significant"),
    shape = 21, color = "black", stroke = 0, alpha = 1, size = 3
  ) +
  geom_point(
    data = sig_df,
    aes(fill = "Significant"),
    shape = 21, color = "black", stroke = 0, alpha = 1, size = 3
  ) +
  geom_point(
    data = label_df,
    aes(fill = "Significant, Shared"),
    shape = 21, color = "black", stroke = 0, alpha = 1.0, size = 3
  ) +
  labs(
    x = expression("Species  Treatment " ~ Log[2] ~ "FC"),
    y = expression("-" ~ Log[10] ~ "(Adjusted p-value)")
  ) +
  scale_x_continuous(limits = c(-max_x_val * 1.1, max_x_val * 1.1)) +
  scale_y_continuous(limits = c(0, max_y_val * 1.1)) +
  scale_fill_manual(
    values = c("Not Significant" = "lightgrey", "Significant, Shared" = "#4CBB17", "Significant" = "grey40")
  ) +
  theme_minimal() +
  theme(
    plot.background  = element_blank(),
    panel.background = element_blank(),
    panel.grid       = element_blank(),
    axis.line        = element_line(color = "black", linewidth = 1),
    legend.position  = "bottom",
    legend.title     = element_blank(),
    axis.text        = element_text(size = 12),
    axis.title       = element_text(size = 14),
    legend.text      = element_text(size = 12)
  ) +
  ggrepel::geom_text_repel(
    data = filt_df %>% dplyr::filter(ID %in% genes),
    aes(label = ID),
    size = 4,
    segment.color = "black",
    min.segment.length = 0.2,
    box.padding = 0.7,
    point.padding = 0.7,
  )
ggsave(filename = paste0(fig_path, paste0("/VolcanoPlot_DivergentrGene_FGF8b_6hrs_ColorBySharedGenes.pdf")),
  plot = p, height = 5, width = 5, bg = 'white', units='in', dpi=600)
print(p)
```

```{r volcano plot divergent rGenes FGF2 54hr}
# Define variables of interest
pert <- 'FGF2'
time <- "54hr"
genes <- c("PDE4B", "DST", "ST6GALNAC5", 'NPY', 'NRP1', 'IL17RD', 'LPAR1', 'BMPR1A')

# Filter divergent response data frame for select perturbation and time point
filt_df <- divergent_response_df %>%
  dplyr::filter(perturbation == pert, time_point == time) %>%
  dplyr::mutate(neglog10_adjP = -log10(adj.P.Val))

# Split dataset for plotting (handle NA categories safely)
nonsig_df <- filt_df %>% dplyr::filter(significant == FALSE)
label_df <- filt_df %>% dplyr::filter(significant == TRUE, ID %in% genes)
sig_df <- filt_df %>% dplyr::filter(significant == TRUE, !(ID %in% genes))

# Determine global max x and y values for setting the same scale axes
max_x_val <- max(abs(filt_df$logFC), na.rm = TRUE)
max_y_val <- max(filt_df$neglog10_adjP, na.rm = TRUE)

# Choose nudges (in data units)
nudge_y_mag <- 0.05 * max_y_val   # up
nudge_x_mag <- 0.05 * max_x_val   # left/right

# Plot a volcano plot
p <- ggplot(filt_df, aes(x = logFC, y = neglog10_adjP)) +
  geom_point(
    data = nonsig_df,
    aes(fill = "Not Significant"),
    shape = 21, color = "black", stroke = 0, alpha = 1, size = 3
  ) +
  geom_point(
    data = sig_df,
    aes(fill = "Significant"),
    shape = 21, color = "black", stroke = 0, alpha = 1, size = 3
  ) +
  geom_point(
    data = label_df,
    aes(fill = "Significant, MAPK/ERK"),
    shape = 21, color = "black", stroke = 0, alpha = 1.0, size = 3
  ) +
  labs(
    x = expression("Species  Treatment " ~ Log[2] ~ "FC"),
    y = expression("-" ~ Log[10] ~ "(Adjusted p-value)")
  ) +
  scale_x_continuous(limits = c(-max_x_val * 1.1, max_x_val * 1.1)) +
  scale_y_continuous(limits = c(0, max_y_val * 1.1)) +
  scale_fill_manual(
    values = c("Not Significant" = "lightgrey", "Significant, MAPK/ERK" = "#4F7942", "Significant" = "grey40")
  ) +
  theme_minimal() +
  theme(
    plot.background  = element_blank(),
    panel.background = element_blank(),
    panel.grid       = element_blank(),
    axis.line        = element_line(color = "black", linewidth = 1),
    legend.position  = "bottom",
    legend.title     = element_blank(),
    axis.text        = element_text(size = 12),
    axis.title       = element_text(size = 14),
    legend.text      = element_text(size = 12)
  ) +
  ggrepel::geom_text_repel(
    data = filt_df %>% dplyr::filter(ID %in% genes),
    aes(label = ID),
    size = 4,
    segment.color = "black",
    min.segment.length = 0.2,
    box.padding = 0.7,
    point.padding = 0.7,
  )
ggsave(filename = paste0(fig_path, paste0("/VolcanoPlot_DivergentrGene_FGF2_54hrs_ColorByMAPKERK.pdf")),
  plot = p, height = 5, width = 5.25, bg = 'white', units='in', dpi=600)
print(p)
```

```{r volcano plot divergent rGenes FGF2 54hr}
# Define variables of interest
pert <- 'FGF8b'
time <- "54hr"
genes <- c("PDE4B", "DST", "ST6GALNAC5", 'NPY', 'NRP1', 'IL17RD', 'LPAR1', 'BMPR1A')

# Filter divergent response data frame for select perturbation and time point
filt_df <- divergent_response_df %>%
  dplyr::filter(perturbation == pert, time_point == time) %>%
  dplyr::mutate(neglog10_adjP = -log10(adj.P.Val))

# Split dataset for plotting (handle NA categories safely)
nonsig_df <- filt_df %>% dplyr::filter(significant == FALSE)
label_df <- filt_df %>% dplyr::filter(significant == TRUE, ID %in% genes)
sig_df <- filt_df %>% dplyr::filter(significant == TRUE, !(ID %in% genes))

# Determine global max x and y values for setting the same scale axes
max_x_val <- max(abs(filt_df$logFC), na.rm = TRUE)
max_y_val <- max(filt_df$neglog10_adjP, na.rm = TRUE)

# Choose nudges (in data units)
nudge_y_mag <- 0.05 * max_y_val   # up
nudge_x_mag <- 0.05 * max_x_val   # left/right

# Plot a volcano plot
p <- ggplot(filt_df, aes(x = logFC, y = neglog10_adjP)) +
  geom_point(
    data = nonsig_df,
    aes(fill = "Not Significant"),
    shape = 21, color = "black", stroke = 0, alpha = 1, size = 3
  ) +
  geom_point(
    data = sig_df,
    aes(fill = "Significant"),
    shape = 21, color = "black", stroke = 0, alpha = 1, size = 3
  ) +
  geom_point(
    data = label_df,
    aes(fill = "Significant, MAPK/ERK"),
    shape = 21, color = "black", stroke = 0, alpha = 1.0, size = 3
  ) +
  labs(
    x = expression("Species  Treatment " ~ Log[2] ~ "FC"),
    y = expression("-" ~ Log[10] ~ "(Adjusted p-value)")
  ) +
  scale_x_continuous(limits = c(-max_x_val * 1.1, max_x_val * 1.1)) +
  scale_y_continuous(limits = c(0, max_y_val * 1.1)) +
  scale_fill_manual(
    values = c("Not Significant" = "lightgrey", "Significant, MAPK/ERK" = "#4CBB17", "Significant" = "grey40")
  ) +
  theme_minimal() +
  theme(
    plot.background  = element_blank(),
    panel.background = element_blank(),
    panel.grid       = element_blank(),
    axis.line        = element_line(color = "black", linewidth = 1),
    legend.position  = "bottom",
    legend.title     = element_blank(),
    axis.text        = element_text(size = 12),
    axis.title       = element_text(size = 14),
    legend.text      = element_text(size = 12)
  ) +
  ggrepel::geom_text_repel(
    data = filt_df %>% dplyr::filter(ID %in% genes),
    aes(label = ID),
    size = 4,
    segment.color = "black",
    min.segment.length = 0.2,
    box.padding = 0.7,
    point.padding = 0.7,
  )
ggsave(filename = paste0(fig_path, paste0("/VolcanoPlot_DivergentrGene_FGF8b_54hrs_ColorByMAPKERK.pdf")),
  plot = p, height = 5, width = 5.25, bg = 'white', units='in', dpi=600)
print(p)
```

## Divergent rGenes ORA

```{r}
ora_drGenes_df <- function(degs_df,
                           results_df,
                           degs_list_colname,
                           species_name,
                           category,
                           subcategory,
                           universe_symbol_colname = "ID") {
  stopifnot(is.data.frame(degs_df), is.data.frame(results_df))
  if (!degs_list_colname %in% names(degs_df)) {
    stop(sprintf("Column '%s' not found in degs_df.", degs_list_colname))
  }
  if (!universe_symbol_colname %in% names(results_df)) {
    stop(sprintf("Column '%s' not found in results_df (universe).", universe_symbol_colname))
  }

  ora_results_list <- list()

  # Universe of SYMBOLs -> ENTREZ
  universe_symbols <- results_df[[universe_symbol_colname]]
  universe_symbols <- unique(trimws(as.character(universe_symbols)))
  universe_symbols <- universe_symbols[!is.na(universe_symbols) & universe_symbols != ""]

  valid_symbols <- AnnotationDbi::keys(org.Hs.eg.db, keytype = "SYMBOL")
  universe_symbols <- intersect(universe_symbols, valid_symbols)

  universe_entrez <- AnnotationDbi::mapIds(
    org.Hs.eg.db,
    keys      = universe_symbols,
    column    = "ENTREZID",
    keytype   = "SYMBOL",
    multiVals = "first"
  )
  universe_entrez <- unique(as.character(stats::na.omit(universe_entrez)))

  if (length(universe_entrez) == 0) {
    stop("Universe mapping produced 0 ENTREZ IDs. Check your universe_symbol_colname and SYMBOL validity.")
  }

  # Precompute gene sets once
  gene_set <- msigdbr::msigdbr(species = species_name, category = category, subcategory = subcategory)
  term2gene <- gene_set |>
    dplyr::select(gs_name, entrez_gene) |>
    dplyr::rename(term = gs_name, gene = entrez_gene)

  for (pert in unique(degs_df$perturbation)) {
    for (time in unique(degs_df$time_point)) {

      degs_raw <- degs_df |>
        dplyr::filter(perturbation == pert, time_point == time) |>
        dplyr::pull(!!rlang::sym(degs_list_colname))

      # Flatten and split "/"
      degs <- degs_raw |>
        stats::na.omit() |>
        unlist(use.names = FALSE) |>
        as.character() |>
        stringr::str_split(pattern = "/") |>
        unlist(use.names = FALSE) |>
        trimws() |>
        unique()

      degs <- degs[!is.na(degs) & degs != ""]
      if (length(degs) == 0) next

      # Keep only valid SYMBOLs
      degs <- intersect(degs, valid_symbols)
      if (length(degs) == 0) next

      # SYMBOL -> ENTREZ
      degs_entrez <- AnnotationDbi::mapIds(
        org.Hs.eg.db,
        keys      = degs,
        column    = "ENTREZID",
        keytype   = "SYMBOL",
        multiVals = "first"
      )
      degs_entrez <- unique(as.character(stats::na.omit(degs_entrez)))
      if (length(degs_entrez) == 0) next

      ora <- tryCatch(
        clusterProfiler::enricher(
          gene          = degs_entrez,
          universe      = universe_entrez,
          TERM2GENE     = term2gene,
          pAdjustMethod = "fdr",
          qvalueCutoff  = 0.10,
          minGSSize     = 3,
          maxGSSize     = 500
        ),
        error = function(e) NULL
      )

      if (!is.null(ora) && nrow(ora@result) > 0) {
        ora_results <- ora@result |>
          dplyr::mutate(
            perturbation = pert,
            time_point   = time
          ) |>
          tidyr::separate(GeneRatio, into = c("numerator", "denominator"), sep = "/", remove = FALSE) |>
          dplyr::mutate(
            GeneRatioOriginal = GeneRatio,
            GeneRatio = as.numeric(numerator) / as.numeric(denominator)
          ) |>
          dplyr::select(-numerator, -denominator) |>
          dplyr::filter(p.adjust < 0.05, qvalue < 0.10, Count > 1)

        if (nrow(ora_results) > 0) {
          ora_results_list[[paste(degs_list_colname, pert, time, sep = "_")]] <- ora_results
        }
      }
    }
  }

  ora_results_df <- dplyr::bind_rows(ora_results_list, .id = "contrast")

  if (nrow(ora_results_df) == 0) return(ora_results_df)

  # Map ENTREZ IDs in geneID -> SYMBOL string
  entrez_to_symbol <- function(entrez_str) {
    ids <- unlist(stringr::str_split(as.character(entrez_str), "/"))
    ids <- ids[!is.na(ids) & ids != ""]
    syms <- AnnotationDbi::mapIds(
      org.Hs.eg.db,
      keys      = ids,
      column    = "SYMBOL",
      keytype   = "ENTREZID",
      multiVals = "first"
    )
    paste(unique(stats::na.omit(as.character(syms))), collapse = "/")
  }

  ora_results_df <- ora_results_df |>
    dplyr::mutate(geneSymbol = vapply(geneID, entrez_to_symbol, character(1)))

  ora_results_df
}

```

Perform ORA on human-upregulated divergent rGenes

```{r message=FALSE, warning=FALSE}
ora_human_up_gobp_df <- ora_drGenes_df(
  degs_df = divergent_rgenes_df,
  results_df = divergent_response_df,
  degs_list_colname = "up_human_div_rGenes",
  species_name = "Homo sapiens",
  category = "C5",
  subcategory = "GO:BP",
  universe_symbol_colname = "ID"
)
ora_human_up_gobp_df$gene_set <- "GOBP"
```

```{r message=FALSE, warning=FALSE}
ora_human_up_gocc_df <- ora_drGenes_df(
  degs_df = divergent_rgenes_df,
  results_df = divergent_response_df,
  degs_list_colname = "up_human_div_rGenes",
  species_name = "Homo sapiens",
  category = "C5",
  subcategory = "GO:CC",
  universe_symbol_colname = "ID"
)
ora_human_up_gocc_df$gene_set <- "GOCC"
```

```{r message=FALSE, warning=FALSE}
ora_human_up_reactome_df <- ora_drGenes_df(
  degs_df = divergent_rgenes_df,
  results_df = divergent_response_df,
  degs_list_colname = "up_human_div_rGenes",
  species_name = "Homo sapiens",
  category = "C2",
  subcategory = "REACTOME",
  universe_symbol_colname = "ID"
)
ora_human_up_reactome_df$gene_set <- "Reactome"
```

```{r message=FALSE, warning=FALSE}
ora_human_up_wiki_df <- ora_drGenes_df(
  degs_df = divergent_rgenes_df,
  results_df = divergent_response_df,
  degs_list_colname = "up_human_div_rGenes",
  species_name = "Homo sapiens",
  category = "C2",
  subcategory = "WIKIPATHWAYS",
  universe_symbol_colname = "ID"
)
ora_human_up_wiki_df$gene_set <- "Wikipathways"
```

```{r combine drGenes ORA human dfs}
ora_human_up_df <- dplyr::bind_rows(ora_human_up_gobp_df, ora_human_up_gocc_df, ora_human_up_reactome_df, ora_human_up_wiki_df)
ora_human_up_df$direction <- "Up in human"
head(ora_human_up_df)
```

Perform ORA on chimp-upregulated divergent rGenes

```{r message=FALSE, warning=FALSE}
ora_chimp_up_gobp_df <- ora_drGenes_df(
  degs_df = divergent_rgenes_df,
  results_df = divergent_response_df,
  degs_list_colname = "up_chimp_div_rGenes",
  species_name = "Homo sapiens",
  category = "C5",
  subcategory = "GO:BP",
  universe_symbol_colname = "ID"
)
ora_chimp_up_gobp_df$gene_set <- "GOBP"
```

```{r message=FALSE, warning=FALSE}
ora_chimp_up_gocc_df <- ora_drGenes_df(
  degs_df = divergent_rgenes_df,
  results_df = divergent_response_df,
  degs_list_colname = "up_chimp_div_rGenes",
  species_name = "Homo sapiens",
  category = "C5",
  subcategory = "GO:CC",
  universe_symbol_colname = "ID"
)
ora_chimp_up_gocc_df$gene_set <- "GOCC"
```

```{r message=FALSE, warning=FALSE}
ora_chimp_up_reactome_df <- ora_drGenes_df(
  degs_df = divergent_rgenes_df,
  results_df = divergent_response_df,
  degs_list_colname = "up_chimp_div_rGenes",
  species_name = "Homo sapiens",
  category = "C2",
  subcategory = "REACTOME",
  universe_symbol_colname = "ID"
)
ora_chimp_up_reactome_df$gene_set <- "Reactome"
```

```{r message=FALSE, warning=FALSE}
ora_chimp_up_wiki_df <- ora_drGenes_df(
  degs_df = divergent_rgenes_df,
  results_df = divergent_response_df,
  degs_list_colname = "up_chimp_div_rGenes",
  species_name = "Homo sapiens",
  category = "C2",
  subcategory = "WIKIPATHWAYS",
  universe_symbol_colname = "ID"
)
ora_chimp_up_wiki_df$gene_set <- "Wikipathways"
```

```{r combine drGenes ORA chimp dfs}
ora_chimp_up_df <- dplyr::bind_rows(ora_chimp_up_gobp_df, ora_chimp_up_gocc_df, ora_chimp_up_reactome_df, ora_chimp_up_wiki_df)
ora_chimp_up_df$direction <- "Up in chimp"
head(ora_chimp_up_df)
```

```{r}
divergent_ora_df <- dplyr::bind_rows(ora_human_up_df, ora_chimp_up_df)
head(divergent_ora_df)
```

```{r}
saveRDS(divergent_ora_df, paste0(dir_path, "divergent_rGenes_ORA_df.rds"))
```

```{r}
write.xlsx(divergent_ora_df, file = paste0(dir_path, "divergent_rGene_ORA_GOBP.xlsx"), rowNames = FALSE)
```

```{r}
divergent_ora_df <- readRDS(paste0(dir_path, "divergent_rGenes_ORA_df.rds"))
head(divergent_ora_df)
```


## Inspect specific ORA results

```{r}
#direct <- "Up in human"
pert <- 'FGF2'
time <- '6hr'
gene_set_name = 'GOBP'
dr_ora_df <- divergent_ora_df %>% dplyr::filter(perturbation == pert & time_point == time
                                       & gene_set==gene_set_name
                                       #& direction == direct
                                       )
View(dr_ora_df)
```



## Dot plot divergent rGene ORA

```{r}
# Helper function for cleaning enrichment term strings
clean_term <- function(x) {
  x %>%
    stringr::str_remove("^GOBP_") %>%
    stringr::str_replace_all("_", " ") %>%
    stringr::str_to_lower()
}
```

```{r define dotplot_drGenes function}
dotplot_div_rGenes <- function(ora_df){
  # Plot dot plot
  p <- ggplot(filt_df, aes(x = direction, y = Description, size = Count, fill = -log10(p.adjust))) +
    geom_point(shape = 21, color = 'black', stroke = 0.25) +
    scale_fill_viridis_c(option = 'inferno') +
    labs(
      x = "",
      y = "",
      size = "Gene Count",
      fill = expression("-" ~ Log[10] ~"(Adj. P-Val.)")) +
    scale_x_discrete(limits = c("Up in chimp", "Up in human")) +

    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5),
      axis.text.y = element_text(size = 10),
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
    )
}
```

```{r plot drGenes dot plots}
gene_set_name <- c('GOBP')
pert <- 'BMP7'
time <- '6hr'
terms <- c(
  'GOBP_CELL_CELL_JUNCTION_ASSEMBLY',
  'GOBP_AMINO_ACID_TRANSPORT',
  'GOBP_SENSORY_SYSTEM_DEVELOPMENT',
  'GOBP_SKELETAL_SYSTEM_DEVELOPMENT',
  'GOBP_GLIOGENESIS',
  'GOBP_GLIAL_CELL_PROLIFERATION',
  'GOBP_MORPHOGENESIS_OF_AN_EPITHELIUM',
  'GOBP_REGULATION_OF_EPITHELIAL_CELL_PROLIFERATION',
  'GOBP_REGULATION_OF_CELLULAR_RESPONSE_TO_GROWTH_FACTOR_STIMULUS',
  'GOBP_NEGATIVE_REGULATION_OF_CELL_DIFFERENTIATION',
  'GOBP_CANONICAL_WNT_SIGNALING_PATHWAY',
  'GOBP_POSITIVE_REGULATION_OF_WNT_SIGNALING_PATHWAY'
)

term_levels <- clean_term(terms)

filt_df <- divergent_ora_df %>%
  filter(
    perturbation == pert,
    time_point == time,
    gene_set %in% gene_set_name,
    Description %in% terms
  ) %>%
  mutate(
    direction   = factor(direction, levels = c("Up in human", "Up in chimp")),
    Description = factor(clean_term(Description), levels = term_levels)
  )

p <- dotplot_div_rGenes(filt_df) 

ggsave(paste0(fig_path, "DotPlot_DivergentrGene_ORA_", pert, "_", time, ".png"), plot = p, height = 4, width = 6, bg = 'white')
p
```

```{r plot drGenes dot plots}
gene_set_name <- c('GOBP')
pert <- 'BMP7'
time <- '54hr'
terms <- c(
  'GOBP_INTERSTRAND_CROSS_LINK_REPAIR',
  'GOBP_REGULATION_OF_EPITHELIAL_TO_MESENCHYMAL_TRANSITION',
  'GOBP_POSITIVE_REGULATION_OF_EPITHELIAL_TO_MESENCHYMAL_TRANSITION',
  'GOBP_EPITHELIAL_TUBE_MORPHOGENESIS',
  'GOBP_EMBRYONIC_ORGAN_MORPHOGENESIS',
  'GOBP_ACTIN_FILAMENT_BASED_MOVEMENT',
  'GOBP_FOREBRAIN_DEVELOPMENT',
  'GOBP_TRIPEPTIDE_TRANSMEMBRANE_TRANSPORT'
)

term_levels <- clean_term(terms)

filt_df <- divergent_ora_df %>%
  filter(
    perturbation == pert,
    time_point == time,
    gene_set %in% gene_set_name,
    Description %in% terms
  ) %>%
  mutate(
    direction   = factor(direction, levels = c("Up in human", "Up in chimp")),
    Description = factor(clean_term(Description), levels = term_levels)
  )

p <- dotplot_div_rGenes(filt_df) 

ggsave(paste0(fig_path, "DotPlot_DivergentrGene_ORA_", pert, "_", time, ".png"), plot = p, height = 4, width = 6, bg = 'white')
p
```


```{r plot drGenes dot plots}
gene_set_name <- c("Reactome", 'GOBP')
pert <- 'FGF2'
time <- '54hr'


filt_df <- divergent_ora_df %>% filter(perturbation == pert & time_point == time & gene_set %in% gene_set_name) %>%
  arrange(p.adjust) %>%
  slice_head(n = 10) %>%
  arrange(direction)

p <- dotplot_div_rGenes(filt_df)
ggsave(paste0(fig_path, "DotPlot_DivergentrGene_ORA_", pert, "_", time, ".png"), plot = p, height = 4, width = 8, bg = 'white')
p
```

```{r plot drGenes dot plots}
gene_set_name <- c('GOBP')
pert <- 'BMP7'
time <- '54hr'

filt_df <- divergent_ora_df %>% filter(perturbation == pert & time_point == time
                                       & gene_set==gene_set_name
                                       ) %>%
  arrange(p.adjust) %>%
  slice_head(n = 15) %>%
  arrange(desc(direction))

p <- dotplot_div_rGenes(filt_df)
ggsave(paste0(fig_path, "DotPlot_DivergentrGene_ORA_", pert, "_", time, ".png"), plot = p, height = 6, width = 10, bg = 'white')
p
```


## Divergent rGenes GSEA GO:BP

Define a function for performing GSEA with a dreamlet results data frame as input and a GSEA results data frame as output.
```{r single drgene_gsea_df}
# Define sample of interest
pert = 'FGF2'
time = '6hr'

# Extract gene list
gene_list <- divergent_response_df %>% filter(perturbation == pert & time_point == time) %>%
  dplyr::select(ID, logFC) %>%
  dplyr::arrange(desc(logFC))
        
  # Skip if empty
  if (nrow(gene_list) == 0) next
        
  gene_list <- deframe(gene_list)
        
  # Skip if gene_list has no names
  if (is.null(names(gene_list)) || length(gene_list) == 0) next
    
  # Run GSEA using GO:BP terms
  gsea_go <- gseGO(
    geneList = gene_list,
    OrgDb = org.Hs.eg.db,
    keyType = "SYMBOL",  
    ont = "BP", 
    pvalueCutoff = 1,
    minGSSize = 1,
    eps = 0,
    verbose = TRUE,
    by = "fgsea"
    )
        
  # If results were returned, store them
  if (!is.null(gsea_go) && nrow(gsea_go@result) > 0) {
    gsea_results <- gsea_go@result %>%
      filter(qvalue < 0.10) %>%
      mutate(
        perturbation = pert,
        time_point = time)
  }

head(gsea_results)
```

Define a function for performing GSEA with a dreamlet results data frame as input and a GSEA results data frame as output.
```{r gsea_drGenes_df function}
gsea_drGenes_df <- function(dr_df, cell_types) {
  
  # Initialize an empty list
  dr_gsea_list <- list()
  
  # Iterate through cell type x perturbation x time point
  for (type in cell_types) {
    for (pert in unique(dr_df$perturbation)) {
      for (time in unique(dr_df$time_point)) {

        # Extract gene list
        gene_list <- divergent_response_df %>% filter(assay == type & perturbation == pert & time_point == time) %>%
          dplyr::select(ID, logFC) %>%
          dplyr::arrange(desc(logFC))
                
          # Skip if empty
          if (nrow(gene_list) == 0) next
                
          gene_list <- deframe(gene_list)
                
          # Skip if gene_list has no names
          if (is.null(names(gene_list)) || length(gene_list) == 0) next
            
          # Run GSEA using GO:BP terms
          gsea_go <- gseGO(
            geneList = gene_list,
            OrgDb = org.Hs.eg.db,
            keyType = "SYMBOL",  
            ont = "BP", 
            pvalueCutoff = 1,
            minGSSize = 1,
            eps = 0,
            verbose = TRUE,
            by = "fgsea"
            )
                
          # If results were returned, store them
          if (!is.null(gsea_go) && nrow(gsea_go@result) > 0) {
            gsea_results <- gsea_go@result %>%
              filter(qvalue < 0.10) %>%
              mutate(cell_type = type,
              perturbation = pert,
              time_point = time)
          }
          
          dr_gsea_list[[paste(type, pert, time, sep = "_")]] <- gsea_results
        }
      }
    }
  dr_gsea_df <- bind_rows(dr_gsea_list)
  return(dr_gsea_df)
}
```

Perform GO:BP term GSEA of divergent rGenes
```{r drgene_gsea_df}
cell_types <- c('Tel_NE_dorsal', 'Tel_NE_ventral')
drgene_gsea_df <- gsea_drGenes_df(divergent_response_df, cell_types)
head(drgene_gsea_df)
```

Save divergent rDEG GSEA results as a RDS file
```{r save dr_gsea_df rds}
saveRDS(drgene_gsea_df, file = paste0(dir_path, "divergent_rGenes_GSEA_df.rds"))
```

Save significant divergent rGenes GSEA results as a CSV file
```{r save dr_gsea_df csv}
filt_drgene_gsea_df = drgene_gsea_df %>% filter(p.adjust < 0.10 & qvalue < 0.10)
write.csv(filt_drgene_gsea_df, file = paste0(dir_path, "divergent_rGenes_GSEA_df.csv"), row.names = FALSE)
```

Load divergent rGene GSEA results as a RDS file
```{r load dr_gsea_df}
drgenes_gsea_df <- readRDS(paste0(dir_path, "divergent_rDEGs_GSEA_df.rds"))
head(drgenes_gsea_df)
```

Add signaling gene set categories to divergent rGene GSEA data frame
```{r add signaling category}
# Create a named vector mapping GO term IDs to their signaling category
id_to_category <- unlist(lapply(names(go_term_sets), function(category) {
  setNames(rep(category, length(go_term_sets[[category]])), go_term_sets[[category]])
}))

# Add signaling category
drgenes_gsea_df$signaling_category <- id_to_category[as.character(drgenes_gsea_df$ID)]

# Add signaling category colors
drgenes_gsea_df <- drgenes_gsea_df %>%
  mutate(signaling_category_color = recode(signaling_category, !!!signaling_category_colors)) %>%
  replace_na(list(signaling_category_color = "grey"))

head(drgenes_gsea_df)
```

Visualize divergent rGenes GSEA GO:BP results with volcano plots
```{r drGenes GSEA volcano plots}
# Set factor levels.
drgenes_gsea_df$time_point <- factor(drgenes_gsea_df$time_point, levels = c('6hr', '54hr', '7day'))

# Change NES to numeric.
drgenes_gsea_df$NES <- as.numeric(drgenes_gsea_df$NES)

# Iterate through each combination of cell type, perturbation, time point, and species.
for (type in unique(drgenes_gsea_df$cell_type)){
  for (pert in unique(drgenes_gsea_df$perturbation)){

      # Filter the data frame to the current combination.
      filt_df <- drgenes_gsea_df %>% filter(cell_type == type & perturbation == pert)
          
      # Skip iteration if there are no results.
      if (nrow(filt_df) == 0) next 
        
      # Determine symmetric x-axis limits.
      max_val <- max(abs(filt_df$NES), na.rm = TRUE)
          
      # Identify top 3 lowest p.adjust values for both positive and negative NES
      top_pos <- filt_df %>% filter(NES > 0 & p.adjust < 0.05) %>% arrange(p.adjust) %>% slice_head(n = 5)
      top_neg <- filt_df %>% filter(NES < 0 & p.adjust < 0.05) %>% arrange(p.adjust) %>% slice_head(n = 5)
      top_labeled <- bind_rows(top_pos, top_neg)  # Combine both sets
        
      # Plot a volcano plot
      p <- ggplot(filt_df, aes(x = NES, y = -log10(p.adjust))) +
          
        # No signaling category
        geom_point(data = filt_df %>% filter(is.na(signaling_category)),
                  aes(fill = signaling_category, shape = time_point),
                  color = 'black',
                  stroke = 0.25,
                  alpha = 0.25,
                  size = 1) +
          
        # Signaling category
        geom_point(data = filt_df %>% filter(!is.na(signaling_category)),
                  aes(fill = signaling_category, shape = time_point),
                  color = 'black',
                  stroke = 0.25,
                  alpha = 1,
                  size = 1) +
          
        labs(title = paste0("GO:BP GSEA Results \n", type, " ", pert),
          x = "NES",
          y = expression ("-" ~ Log[10] ~"(Adjusted P-value)")) +
        scale_y_continuous(limits = c(0, NA)) +  # Set the bottom of the y-axis to 0
        scale_x_continuous(limits = c(-max_val, max_val)) +  # Center x-axis on 0
        theme_minimal() +
        theme(
          plot.title = element_text(hjust = 0.5, size = 12),  # Center and size the title
          legend.position = "right",  # Place legend on the right
          axis.line = element_line(color = "black"),  # Add axis lines
          panel.grid.major = element_blank(),  # Remove major gridlines
          panel.grid.minor = element_blank()   # Remove minor gridlines
        ) +
    
        # Add text labels for top values for positive and negative NES
        geom_text_repel(data = top_labeled, 
                        aes(label = Description),  # Adjust label column if needed
                        size = 2.5,            # Make text smaller
                        nudge_y = 1,           # Push labels higher
                        direction = "y",       # Force vertical placement
                        hjust = 0.5,           # Center text vertically
                        vjust = 0) +           # Align text towards top
          
        # Shape by time point
        scale_shape_manual(name = "Time Point", values = c(21, 22, 23)) +
          
        # Fill by signaling category color
        scale_fill_manual(name = "GO:BP ID Category", values = signaling_category_colors,
                            guide = guide_legend(override.aes = list(shape = 21, stroke = 0.50, size = 3, color = 'black'))) +
    
        # Add significance threshold lines
        geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "darkgrey", linewidth = 0.5) +
        geom_vline(xintercept = 0, linetype = "dashed", color = "darkgrey", linewidth = 0.5)
        
        # Save plot
        ggsave(filename = paste0(fig_path, "volcanoplot_drGenes_GSEA_GOBP/volcanoplot_drGenes_GSEA_GOBP_", type, "_", pert, ".png"),
                plot = p, height = 5, width = 5, bg = 'white')
  }
}
```


## Divergent response heatmaps

```{r bmp 6 hr genes}
genes <- c('SMAD3', 'SOX4', 'NKD1', 'ROR2', 'GATA3', 'VEGFC', 'PARD6B')
time_points <- c('6hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_6hr_SelectDivergentrGenes.pdf"), width = 3.75, height = 3.25, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r bmp 6 hr genes at 54 hr}
genes <- c('SMAD3', 'SOX4', 'NKD1', 'ROR2', 'GATA3', 'VEGFC', 'PARD6B')
time_points <- c('54hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_SelectDivergentrGenes.pdf"), width = 3.25, height = 3.25, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r bmp 54 hr cc genes}
genes <- c('MKI67', 'AURKB', 'PBK', 'BARD1', 'SPDL1', 'CLSPN', 'SHCBP1', 'KNTC1', 'KIF4A', 'NCAPH', 'CENPI')
  
time_points <- c('54hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_CCGenes.pdf"), width = 6, height = 3.25, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r}
genes <- c('DISC1', 'DGKI')
time_points <- c('6hr')
perts <- c('FGF8b', 'FGF2')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_FGF2_FGF8b_6hr_DivergentrGenes_NearHAQERs_NDDGenes.pdf"),
    width = 3, height = 3, useDingbats = FALSE)
draw(p)
dev.off()
```



```{r}
genes <- c('CUX2', 'TSHZ3', 'MEIS2', 'NRG3', 'CACNA1C', 'CNTN4')
time_points <- c('54hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

#pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_SelectDivergentrGenes.pdf"), width = 5, height = 3, useDingbats = FALSE)
#draw(p)
#dev.off()
```

```{r}
genes <- c('NFIA')
time_points <- c('54hr')
perts <- c('CHIR99021')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

#pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_SelectDivergentrGenes.pdf"), width = 5, height = 3, useDingbats = FALSE)
#draw(p)
#dev.off()
```

```{r}
genes <- divergent_response_df %>% dplyr::filter(significant==TRUE, time_point=='54hr', perturbation=='BMP7', grepl("WNT", signaling_category)) %>% pull(ID)
time_points <- c('54hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_DivergentrGenes_WNTSignaling.pdf"), width = 12, height = 3, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r fgfs shared 6 hr}
# Define variables of interest
pert1 <- 'FGF8b'
pert2 <- 'FGF2'
time <- '6hr'

# Filter data frame
filt_df <- divergent_response_df %>% dplyr::filter(time_point == time & perturbation %in% c(pert1, pert2) & significant==TRUE & logFCsign==1)

# Pull genes for both perturbations
pert1_genes <- filt_df %>% dplyr::filter(perturbation == pert1) %>% pull(ID)
pert2_genes <- filt_df %>% dplyr::filter(perturbation == pert2) %>% pull(ID)

# Find intersection of genes
intersect_genes <- intersect(pert1_genes, pert2_genes)
print(length(intersect_genes))
print(intersect_genes)


genes <- intersect_genes
time_points <- c('6hr')
perts <- c('FGF2', 'FGF8b')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_FGF2_FGF8b_6hr_DivergentrGenes_Shared.pdf"), width = 5, height = 3, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r fgf2 54 hr}
genes <- c("PDE4B", "DST", "ST6GALNAC5", 'NPY', 'NRP1', 'IL17RD', 'LPAR1', 'BMPR1A')
time_points <- c('54hr')
perts <- c('FGF2')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_FGF2_FGF8b_54hr_SelectDivergentrGenes.pdf"), width = 5, height = 3, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r fgf dendrite}
genes <- div_ora_df %>%
  dplyr::filter(perturbation %in% c('FGF2', 'FGF8b') & time_point == '54hr' & Description == 'GOBP_DENDRITE_DEVELOPMENT') %>%
  pull(geneSymbol)%>%
  strsplit("/") %>%
  unlist()
  
time_points <- c('54hr')
perts <- c('FGF2')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_FGF2_54hr_SelectDivergentrGenes_DendriteGenes.pdf"), width = 4.5, height = 3.25, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r fgf gpcr}
genes <- divergent_ora_df %>%
  dplyr::filter(perturbation %in% c('FGF2', 'FGF8b') & time_point == '54hr' & Description == 'REACTOME_SIGNALING_BY_GPCR') %>%
  pull(geneSymbol)%>%
  strsplit("/") %>%
  unlist()
  
time_points <- c('54hr')
perts <- c('FGF2')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_FGF2_54hr_SelectDivergentrGenes_GPCRSignalingGenes.pdf"), width = 4.5, height = 3.25, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r bmp emt}
genes <- divergent_ora_df %>%
  dplyr::filter(perturbation %in% c('BMP7') & time_point == '54hr' & Description == 'GOBP_REGULATION_OF_EPITHELIAL_TO_MESENCHYMAL_TRANSITION') %>%
  pull(geneSymbol)%>%
  strsplit("/") %>%
  unlist()
  
time_points <- c('54hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_SelectDivergentrGenes_EMTGenes.pdf"), width = 4.5, height = 3.25, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r bmp emt select}
genes <- c('GATA3', 'BAMB1', 'ISL1', 'FERMT2', 'MDK')
  
time_points <- c('54hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_DivergentrGenes_EMTGenes_Select.pdf"), width = 3, height = 3.25, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r bmp fb dev}
genes <- divergent_ora_df %>%
  dplyr::filter(perturbation %in% c('BMP7') & time_point == '54hr' & Description == 'GOBP_FOREBRAIN_DEVELOPMENT') %>%
  pull(geneSymbol)%>%
  strsplit("/") %>%
  unlist()
  
time_points <- c('54hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_SelectDivergentrGenes_FBDev.pdf"), width = 4.5, height = 3.25, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r bmp actin}
genes <- divergent_ora_df %>%
  dplyr::filter(perturbation %in% c('BMP7') & time_point == '54hr' & Description == 'GOBP_ACTIN_FILAMENT_BASED_MOVEMENT') %>%
  pull(geneSymbol)%>%
  strsplit("/") %>%
  unlist()
  
time_points <- c('54hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_SelectDivergentrGenes_FBDev.pdf"), width = 4.5, height = 3.25, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r fgf gpcr}
genes <- divergent_ora_df %>%
  dplyr::filter(perturbation %in% c('BMP7') & time_point == '54hr' & Description == 'REACTOME_CARDIAC_CONDUCTION') %>%
  pull(geneSymbol)%>%
  strsplit("/") %>%
  unlist()
  
time_points <- c('54hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_SelectDivergentrGenes_CardiacConduction.pdf"), width = 4.5, height = 3.25, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r bmp morph}
genes <- divergent_ora_df %>%
  dplyr::filter(perturbation %in% c('BMP7') & time_point == '54hr' & Description == 'GOBP_EMBRYONIC_ORGAN_MORPHOGENESIS') %>%
  pull(geneSymbol)%>%
  strsplit("/") %>%
  unlist()
  
time_points <- c('54hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_SelectDivergentrGenes_Morphogenesis.pdf"), width = 4.5, height = 3.25, useDingbats = FALSE)
draw(p)
dev.off()
```

```{r bmp morph}
genes <- divergent_ora_df %>%
  dplyr::filter(perturbation %in% c('BMP7') & time_point == '54hr' & Description == 'GOBP_EPITHELIAL_TUBE_MORPHOGENESIS') %>%
  pull(geneSymbol)%>%
  strsplit("/") %>%
  unlist()
  
time_points <- c('54hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_SelectDivergentrGenes_Morphogenesis.pdf"), width = 4.5, height = 3.25, useDingbats = FALSE)
draw(p)
dev.off()
```
```{r bmp actin morph select}
genes <- c('SRI', 'CACNB2', 'CACNA2D1', 'DCT', 'WNT5A', 'ACVR1', 'MTHFD1L', 'MYCN', 'LRP2')
  
time_points <- c('54hr')
perts <- c('BMP7')
species_order <- c('Human', 'Chimp', 'Orangutan')

p <- response_heatmap(response_df, time_points, perts, genes, species_order, species_colors, pert_colors)

pdf(paste0(fig_path, "complexheatmap_ResponseLogFC_BySampleGroup_BMP7_54hr_DivergentrGenes_ActinMovementMorphogenesis_Select.pdf"), width = 5, height = 3.25, useDingbats = FALSE)
draw(p)
dev.off()
```

## Divergent rGenes STRINGdb enrichment

```{r}
div_string_enrich_df <- read.csv(paste0(dir_path, 'human_upregulated_divergent_BMP7_rGenes_54hrs_STRINDdb_enrichment.csv'))
div_string_enrich_df <- div_string_enrich_df %>% 
  mutate(GeneRatio = observed.gene.count / background.gene.count) %>%
  arrange(desc(signal))

head(div_string_enrich_df)                               
```

```{r stringdb dot plot}
plot_df <- div_string_enrich_df %>% 
  arrange(desc(signal)) %>%
  slice_head(n = 5) %>%
  arrange(desc(GeneRatio)) %>%
  mutate(
    term_label = stringr::str_replace_all(term.description, ",\\s*", ",\n"),
    term_label = stringi::stri_unescape_unicode(term_label),  # ensures real newline
    term_label = factor(term_label, levels = rev(term_label))
  )

p <- ggplot(plot_df, aes(
  x = GeneRatio, y = term_label,
  size = observed.gene.count, fill = -log10(false.discovery.rate)
)) +
  geom_point(shape = 21, color = "black", stroke = 0.25) +
  scale_fill_viridis_c(option = "inferno") +
  labs(x = "Gene Ratio", y = "", size = "Gene Count",
       fill = expression("-" ~ Log[10] ~ "FDR")) +
  theme_minimal() +
  theme(
    axis.text.y = element_text(size = 10)  # <- make sure it's element_text, not element_markdown
  )

ggsave(paste0(fig_path, "DotPlot_STRINGdbEnrichment_ByTerm.pdf"),
             plot = p, height = 4, width = 6.25, dpi = 600, bg = 'white')

print(p)
```




## Divergent rGenes ORA DisGeNet

Perform ORA on divergent rDEGs with DisGeNet gene sets.
https://disgenet2r-medbio-e8c7d57bc0ba433de7ffbbcf2ae597ac42301b1c66845.gitlab.io/#disease-enrichment
```{r disgenet settings}
api_key <- "18e25c4f-58a4-466e-b2b5-3f6d3dcc9546"
Sys.setenv(DISGENET_API_KEY= api_key)
```

```{r ora_disgenet_df function}
ora_disgenet_df <- function(degs_df, degs_list_colname, cell_types) {
  ora_results_list <- list()
  
  # Iterate through perturbations.
  for (pert in unique(degs_df$perturbation)){
    # Filter to current perturbation and cell types of interest.
    degs_raw <- degs_df %>%
      filter(perturbation == pert, cell_type %in% cell_types) %>%
      pull(!!sym(degs_list_colname))
    
    # Flatten and split stringified gene lists safely
    degs <- degs_raw %>%
      na.omit() %>%
      unlist() %>%
      strsplit(split = "/") %>%
      unlist() %>%
      trimws() %>%
      unique()
    
    # Skip if empty or all NAs
    if (length(degs) == 0 || all(is.na(degs)) || all(degs == "")) next
    
    # Remove empty or NA values before mapping
    degs_clean <- degs[!is.na(degs) & degs != ""]
    
    # Skip iteration if no valid gene symbols
    if (length(degs_clean) == 0) next
    
    # Perform ORA for DEGs with disgenet2r
    ora <- disease_enrichment(
      entities = degs_clean,
      common_entities = 1,
      max_pvalue = 0.05,
      vocabulary = 'HGNC',
      database = 'CURATED'
    )
    
    # Check if ORA returned results
    if (!is.null(ora@qresult) && nrow(ora@qresult) > 0) {
      ora_results <- ora@qresult %>%
        as.data.frame() %>%
        mutate(perturbation = pert)
      
      # Correctly append the data frame to the list
      ora_results_list[[length(ora_results_list) + 1]] <- ora_results
    }
  }
  
  # Combine all results into a single data frame and return
  ora_results_df <- bind_rows(ora_results_list)
  return(ora_results_df)
}
```

```{r ora_disgenet_df}
cell_types <- c('Tel_NE_dorsal', 'Tel_NE_ventral')
ora_disgenet_df <- ora_disgenet_df(divergent_rgenes_df, 'all', cell_types)

# Filter to significant results
drgenes_disgenet_df <- ora_disgenet_df %>% filter(p_adjusted < 0.10)

# Create new column 'GeneRatio' by evaluating the string fraction
drgenes_disgenet_df$GeneRatio <- sapply(drgenes_disgenet_df$geneRatio, function(x) {
  parts <- strsplit(x, "/")[[1]]
  as.numeric(parts[1]) / as.numeric(parts[2])
})

head(drgenes_disgenet_df)
```

```{r}
for (pert in unique(simple_divergent_rgenes_df$perturbation)){
  filt_df <- simple_divergent_rgenes_df %>% dplyr::filter(perturbation == pert) %>%
    pull('div_rGenes') %>%
    na.omit() %>%
    unlist() %>%
    strsplit(split = "/") %>%
    unlist() %>%
    trimws() %>%
    unique() %>%
    tibble::tibble(gene = .) %>%
    dplyr::filter(!stringr::str_detect(gene, "^(AL|AC|AP|AUXG|Z|AF|BX)\\d{4,}"),
                  !stringr::str_detect(gene, "^human_chim_")) %>%
    dplyr::arrange(gene)

  write.csv(filt_df, paste0(dir_path, 'simple_divergent_rGenes/simple_divergent_rGenes_', pert, '.csv'))
}
```

Save drGenes DisGeNet ORA results
```{r save ora_disgenet_df}
saveRDS(drgenes_disgenet_df, file = paste0(dir_path, "divergent_rGenes_ORA_DisGeNet.rds"))
write.csv(drgenes_disgenet_df, file = paste0(dir_path, "divergent_rGenes_ORA_DisGeNet.csv"), row.names = FALSE)
```

```{r load ora_disgenet_df}
drgenes_disgenet_df <- readRDS(paste0(dir_path, "divergent_rGenes_ORA_DisGeNet.rds"))
head(drgenes_disgenet_df)
```

Visualize divergent rGenes DisGeNet ORA results with a dot plot.
```{r dot plot divergent rGenes DisGeNet ORA}
# Filter to data of interest
filt_df <- drgenes_disgenet_df %>% filter(perturbation %in% c('FGF2', 'BMP7', 'ATRA'))

# Plot a dot plot.
p <- ggplot(filt_df, aes(x = perturbation, y = diseaseName, size = intersectionSize, fill = -log10(p_adjusted))) +
  geom_point(shape = 21, color = 'black', stroke = 0.25) +
  scale_size_continuous(range = c(1, 5)) + 
  scale_fill_viridis(option='inferno') +
  theme_minimal() +
  labs(
    title = 'Divergent rGene Disorder Enrichments',
    x = "",
    y = '',
    size = 'Gene Count',
    fill = expression ("-" ~ Log[10] ~ "Adjusted p-value")
  ) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels for readability
    plot.title = element_text(hjust=0.5, size=16), # Edit plot title.
    plot.background = element_blank(),  # Remove the background around the plot.
    panel.background = element_blank()
  )

ggsave(filename = paste0(fig_path, "dotplot_divergent_rGenes_ORA_DisGeNet_ByPerturbation_SizeByGeneCount_ColorBySignificance.png"),
       plot = p, units = 'in', height = 4, width = 5, bg = 'white')
print(p)
```

```{r dot plot divergent rGenes DisGeNet ORA}
# Filter to data of interest

# Plot a dot plot.
p <- ggplot(drgenes_disgenet_df,
            aes(x = reorder(perturbation, intersectionSize), y = diseaseName,
            size = intersectionSize,
            fill = -log10(p_adjusted))) +
  geom_point(shape = 21, color = 'black', stroke = 0.25) +
  scale_size_continuous(range = c(1, 5)) + 
  scale_fill_viridis(option='inferno') +
  theme_minimal() +
  labs(title = "Divergent rGene DisGeNet ORA",
       x = 'Perturbation',
       y = 'Disease/Disorder',
       size = 'Gene Count',
       fill = expression ("-" ~ Log[10] ~ "Adjusted p-value")) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels for readability
          plot.title = element_text(hjust=0.5, size=16), # Edit plot title.
          plot.background = element_blank(),  # Remove the background around the plot.
          panel.background = element_blank(),  # Remove the background inside the plot.
          #panel.grid.major = element_blank(),  # Remove major gridlines
          #panel.grid.minor = element_blank()   # Remove minor gridlines
        )
ggsave(filename = paste0(fig_path, "dotplot_divergent_rGenes_ORA_DisGeNet_DiseaseNameByPerturbation_SizeByGeneCount_ColorBySignificance.png"),
       plot = p, height = 5, width = 6, bg = 'white')
print(p)
```

```{r}
simple_div_disgenet_ora_df <- read.csv('/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/dreamlet_v2/simple_divergent_rGenes_DisGeNet_ORA_manual_df.csv')
head(simple_div_disgenet_ora_df)
```

```{r multiple hypothesis testing correction}
simple_div_disgenet_ora_df <- simple_div_disgenet_ora_df %>%
  mutate(
    qvalue = p.adjust(pvalue, method = "BH")
  )
head(simple_div_disgenet_ora_df)
```

```{r}
write.xlsx(simple_div_disgenet_ora_df, file = paste0(dir_path, "simple_divergent_rGenes_ORA_DisGeNet_df.xlsx"), rowNames = FALSE)
```

```{r dot plot disgenet brain prolif}
# Specify terms to plot
terms <- c('Bipolar Disorder', "Colorectal Carcinoma", "Epithelioma", "Stomach Neoplasms", 'Rare genetic syndromic intellectual disability')

# Specify perts to plot
perts <- c('ATRA', 'FGF8b', 'FGF2', 'BMP7', 'TRULI')
# Filter the data to the current perturbation
filt_df <- simple_div_disgenet_ora_df %>%
  dplyr::filter(perturbation %in% perts, diseaseName %in% terms, qvalue <= 0.10) %>%
  mutate(
    diseaseName  = factor(diseaseName, levels = rev(terms)),
    perturbation  = factor(perturbation, levels = perts)
  )
  
  
# Plot a dot plot
p <- ggplot(filt_df, aes(x = perturbation, y = diseaseName, size = intersectionSize, fill = -log10(qvalue))) +
  geom_point(shape = 21, color = 'black', stroke = 0.25) +
  scale_size_continuous(range = c(1, 5)) + 
  scale_fill_viridis(option='inferno') +
  theme_minimal() +
  labs(
    x = "",
    y = "",
    size = 'Gene Count',
    fill = expression ("-" ~ Log[10] ~ "FDR")) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels for readability
    plot.title = element_text(hjust=0.5, size=16), # Edit plot title.
    plot.background = element_blank(),  # Remove the background around the plot.
    panel.background = element_blank(),  # Remove the background inside the plot.
    #panel.grid.major = element_blank(),  # Remove major gridlines
    #panel.grid.minor = element_blank()   # Remove minor gridlines
  )
  
ggsave(filename = paste0(fig_path, "DotPlot_Divergent_rGenes_DisGeNet_ORA_select.pdf"),
         plot = p, units = 'in', height = 4, width = 5, bg = 'white')

print(p)
```

## Annotate divergent rGenes

```{r find_overlap function}
# Function to extract overlapping genes from DEGs
find_overlap <- function(degs_str, gene_list) {
  degs_split <- unlist(strsplit(degs_str, "/"))
  overlap <- degs_split[degs_split %in% gene_list]
  if (length(overlap) > 0) paste(overlap, collapse = "/") else NA
}
```

```{r annotate TF divergent rGenes}
# Load human TFs
data(dorothea_hs, package = "dorothea")
tf_genes <- dorothea_hs %>%
  pull(tf) %>%
  unique()

# Annotate df with TF overlaps
divergent_rgenes_df$div_rGenes <- as.character(divergent_rgenes_df$div_rGenes)

for (tf in tf_genes) {
  divergent_rgenes_df$div_rGene_TFs <- sapply(divergent_rgenes_df$div_rGenes, find_overlap, gene_list = tf_genes)
}

head(divergent_rgenes_df)
```

Annotate NDD genes
```{r annotate NDD genes}
# Define file paths and names
ndd_files <- c(
  SCZ = "disgenet_SCZ_genes.csv",
  ASD = "disgenet_ASD_genes.csv",
  BPD = "disgenet_BPD_genes.csv",
  EPI = "disgenet_EPI_genes.csv",
  ID = "disgenet_ID_genes.csv",
  MACRO = "disgenet_macrocephaly_genes.csv", 
  MICRO = "disgenet_microcephaly_genes.csv"
)

# Load gene lists into a named list
ndd_genes <- lapply(ndd_files, function(file) {
  read.csv(paste0(dir_path, "gene_sets/", file)) %>%
    pull(Gene)
})

# Annotate df with NDD overlaps
for (ndd in names(ndd_genes)) {
  divergent_rgenes_df[[ndd]] <- sapply(divergent_rgenes_df$div_rGenes, find_overlap, gene_list = ndd_genes[[ndd]])
}

head(divergent_rgenes_df)
```

Annotate signaling genes
```{r annotate signaling genes response only rDEGs}
# Load gene list
signaling_genes_df <- readRDS(file = paste0(dir_path, "signaling_genes_df.rds"))
signaling_genes <- signaling_genes_df$ID

# Annotate df with NDD overlaps
for (gene in signaling_genes) {
  divergent_rgenes_df$signaling_gene <- sapply(divergent_rgenes_df$all, find_overlap, gene_list = signaling_genes)
}

head(divergent_rgenes_df)
```





```{r save rds response_degs_df}
saveRDS(divergent_rgenes_df, file = paste0(dir_path, "divergent_rGenes_df.rds"))
```


## Perturbation-gene network plot divergent rGenes

```{r simple divergent rGene PXG network graph colored by species up}
library(dplyr)
library(tidyr)
library(igraph)
library(ggraph)

species_colors <- c(Human = "darkorange2", Chimp = "dodgerblue3")

# --- Build edges from up-in-human/up-in-chimp gene lists ---
edges <- simple_divergent_rgenes_df %>%
  dplyr::select(perturbation, up_human_div_rGenes, up_chimp_div_rGenes) %>%
  pivot_longer(
    c(up_human_div_rGenes, up_chimp_div_rGenes),
    names_to  = "direction",
    values_to = "genes"
  ) %>%
  mutate(
    direction = recode(direction,
                       up_human_div_rGenes = "Up in Human",
                       up_chimp_div_rGenes = "Up in Chimp"),
    genes = as.character(genes)    # ensure not a list-col
  ) %>%
  separate_rows(genes, sep = "/") %>%
  mutate(
    gene       = trimws(genes),
    edge_color = if_else(direction == "Up in human",
                         species_colors["Human"],
                         species_colors["Chimp"])
  ) %>%
  filter(!is.na(gene), gene != "") %>%
  distinct(perturbation, gene, direction, edge_color)

# Skip if nothing to plot
if (nrow(edges) == 0L) next

# --- Create igraph object (carry edge attributes) ---
g <- graph_from_data_frame(
  d = edges %>% dplyr::select(perturbation, gene, direction, edge_color),
  directed = FALSE
)

# Ensure edge attributes are aligned with the constructed graph
E(g)$edge_color <- edges$edge_color
E(g)$direction  <- edges$direction

# Node typing (perturbation vs gene)
V(g)$type <- ifelse(V(g)$name %in% edges$perturbation, "perturbation", "gene")

# Degree for sizing/coloring non-perturbation nodes
deg <- igraph::degree(g)
deg <- as.integer(unlist(deg))  

# Node colors
pert_cols <- unname(pert_colors[V(g)$name])
pert_cols[is.na(pert_cols)] <- "gray30"
gene_cols <- ifelse(deg == 1, "gray80",
                    ifelse(deg == 2, "gray50", "black"))
V(g)$color <- ifelse(V(g)$type == "perturbation", pert_cols, gene_cols)

# Shapes & sizes
V(g)$shape <- ifelse(V(g)$type == "perturbation", "square", "circle")
V(g)$size  <- ifelse(V(g)$type == "perturbation", 25 + deg, 2 + deg)

# --- Plot ---
set.seed(123)
p <- ggraph(g, layout = "fr") +
  # Edges use precomputed colors; identity scale renders them as-is
  geom_edge_link0(
    aes(edge_colour = edge_color),
    edge_width = 0.6,
    edge_alpha = 0.7
  ) +
  ggraph::scale_edge_colour_identity(
    name   = "Direction",
    breaks = c("darkorange2", "dodgerblue3"),
    labels = c("Up in Human", "Up in Chimp"),
    guide  = "legend"
  ) +
  guides(edge_colour = guide_legend(override.aes = list(edge_width = 2, edge_alpha = 1))) +

  # Nodes
  geom_node_point(aes(color = color, shape = shape, size = size)) +

  # Gene labels  small & repelled
  geom_node_text(
    aes(label = name, filter = type == "gene"),
    repel = TRUE, size = 2
  ) +

  # Perturbation labels  big, bold, centered
  geom_node_text(
    aes(label = name, filter = type == "perturbation"),
    repel = FALSE, size = 4, fontface = "bold", colour = "black",
    vjust = 0.5, hjust = 0.5
  ) +

  scale_color_identity() +
  scale_shape_identity() +
  scale_size_continuous(range = c(3, 15), guide = "none") +
  theme_void() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    legend.position = "right",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  ggtitle("Perturbation  Divergent rGene Network")

print(p)
ggsave(
  filename = paste0(fig_path, "networkgraph_divergent_rGenes/networkgraph_divergent_rGenes.png"),
  plot = p, height = 30, width = 30, bg = "white"
)

```

```{r}
divergent_pxg_networkgraph <- function(df, time, height, width){
    # --- Build edges from up-in-human/up-in-chimp gene lists ---
  edges <- df %>%
    filter(time_point == time) %>%
    dplyr::select(perturbation, up_human_div_rGenes, up_chimp_div_rGenes) %>%
    pivot_longer(
      c(up_human_div_rGenes, up_chimp_div_rGenes),
      names_to  = "direction",
      values_to = "genes"
    ) %>%
    mutate(
      direction = recode(direction,
                         up_human_div_rGenes = "Up in Human",
                         up_chimp_div_rGenes = "Up in Chimp"),
      genes = as.character(genes)    # ensure not a list-col
    ) %>%
    separate_rows(genes, sep = "/") %>%
    mutate(
      gene       = trimws(genes),
      edge_color = if_else(direction == "Up in Human",
                           species_colors["Human"],
                           species_colors["Chimp"])
    ) %>%
    filter(!is.na(gene), gene != "") %>%
    distinct(perturbation, gene, direction, edge_color)

  # --- Create igraph object (carry edge attributes) ---
  g <- graph_from_data_frame(
    d = edges %>% dplyr::select(perturbation, gene, direction, edge_color),
    directed = FALSE
  )
  
  # Ensure edge attributes are aligned with the constructed graph
  E(g)$edge_color <- edges$edge_color
  E(g)$direction  <- edges$direction
  
  # Node typing (perturbation vs gene)
  V(g)$type <- ifelse(V(g)$name %in% edges$perturbation, "perturbation", "gene")
  
  # Degree for sizing/coloring non-perturbation nodes
  deg <- igraph::degree(g)
  deg <- as.integer(unlist(deg))  
  
  # Node colors
  pert_cols <- unname(pert_colors[V(g)$name])
  pert_cols[is.na(pert_cols)] <- "gray30"
  gene_cols <- ifelse(deg == 1, "gray80",
                      ifelse(deg == 2, "gray50", "black"))
  V(g)$color <- ifelse(V(g)$type == "perturbation", pert_cols, gene_cols)
  
  # Shapes & sizes
  V(g)$shape <- ifelse(V(g)$type == "perturbation", "square", "circle")
  V(g)$size  <- ifelse(V(g)$type == "perturbation", 50 + deg, 2 + deg)
  
  # --- Plot ---
  set.seed(123)
  p <- ggraph(g, layout = "fr") +
    # Edges use precomputed colors; identity scale renders them as-is
    geom_edge_link0(
      aes(edge_colour = edge_color),
      edge_width = 0.6,
      edge_alpha = 0.7
    ) +
    ggraph::scale_edge_colour_identity(
      name   = "Direction",
      breaks = c("darkorange2", "dodgerblue3"),
      labels = c("Up in Human", "Up in Chimp"),
      guide  = "legend"
    ) +
    guides(edge_colour = guide_legend(override.aes = list(edge_width = 2, edge_alpha = 1))) +
  
    # Nodes
    geom_node_point(aes(color = color, shape = shape, size = size)) +
  
    # Gene labels  small & repelled
    geom_node_text(
      aes(label = name, filter = type == "gene"),
      repel = TRUE, size = 2
    ) +
  
    # Perturbation labels  big, bold, centered
    geom_node_text(
      aes(label = name, filter = type == "perturbation"),
      repel = FALSE, size = 4, fontface = "bold", colour = "black",
      vjust = 0.5, hjust = 0.5
    ) +
  
    scale_color_identity() +
    scale_shape_identity() +
    scale_size_continuous(range = c(3, 15), guide = "none") +
    theme_void() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 16),
      legend.position = "right",
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank()
    )

  print(p)
  ggsave(
    filename = paste0(fig_path, "networkgraph_divergent_rGenes/", time, ".png"),
    plot = p, units = 'in', height = height, width = width, bg = "white"
  )
}
```

```{r}
species_colors <- c(Human = "darkorange2", Chimp = "dodgerblue3")
time <- '6hr'
df <- divergent_rgenes_df %>% filter(!(perturbation %in% c('Verteporfin', 'SANT1')))
divergent_pxg_networkgraph(df, time, 10, 10)
```

```{r}
species_colors <- c(Human = "darkorange2", Chimp = "dodgerblue3")
time <- '54hr'
df <- divergent_rgenes_df %>% filter(!(perturbation %in% c('TreMan', 'DMSO')))
divergent_pxg_networkgraph(df, time, 10, 10)
```

```{r}
species_colors <- c(Human = "darkorange2", Chimp = "dodgerblue3")
time <- '7day'
df <- divergent_rgenes_df %>% filter(!(perturbation %in% c('TreMan', 'DMSO')))
divergent_pxg_networkgraph(df, time, 10, 10)
```

### Perturbation-gene set newtork plot divergent rGenes

```{r divergent rGene PxGS network graph}
# Load conserved rGene ORA GO:BP results data frame 
ora_drgenes_df <- readRDS(paste0(dir_path, "divergent_rGenes_ORA_GOBP_GOCC_Reactome_human_chimp_up.rds"))

# Clean up GOBP term labels
all_ora_df <- ora_drgenes_df %>%
  mutate(
    Description = Description %>%
      str_remove("^GOBP_") %>%
      str_remove("^WP_") %>%
      str_remove("^GOCC_") %>%
      str_remove("^REACTOME_") %>%
      str_replace_all("_", " ")
  ) %>%
  filter(!perturbation %in% c("Verteporfin", "TRULI", "LY411575")
    & cell_type == "Tel_NE_ventral"
    & gene_set == "Wikipathways" #"GOCC", #"Reactome", "Wikipathways"
) 

# Output folder
out_dir <- file.path(fig_path, "networkgraph_perturbation_by_Wikipathways_DivrGenes")
if (!dir.exists(out_dir)) dir.create(out_dir, recursive = TRUE)

# Iterate over time points
for (tp in sort(unique(all_ora_df$time_point))) {

  # --- Build edge list for this time point ---
  edges <- all_ora_df %>%
    filter(
      time_point == tp,
      !is.na(ID), ID != "",
    ) %>%
    transmute(
      perturbation,
      Description = str_trim(Description),
      direction   = tolower(str_trim(direction))
    ) %>%
    filter(direction %in% c("up in human", "up in chimp")) %>%
    distinct(perturbation, Description, direction) %>%
    tidyr::drop_na() %>%
    mutate(
      # Precompute explicit edge color (bullet-proof against scale conflicts)
      edge_color = case_when(
        direction == "up in human"   ~ "darkorange2",
        direction == "up in chimp" ~ "dodgerblue3",
        TRUE ~ "grey70"
      )
    )

  if (nrow(edges) == 0L) {
    message(sprintf("[PxGO] No edges for time_point = %s  skipping.", tp))
    next
  }

  # --- Create igraph object (carry edge attributes) ---
  g <- graph_from_data_frame(d = edges, directed = FALSE)

  # Ensure edge attributes are aligned with the constructed graph
  E(g)$edge_color <- edges$edge_color
  E(g)$direction  <- edges$direction

  # Node typing
  V(g)$type <- ifelse(V(g)$name %in% edges$perturbation, "perturbation", "term")

  # Degree for sizing/coloring non-perturbation nodes
  deg <- degree(g)

  # Node colors
  pert_cols <- unname(pert_colors[V(g)$name])
  pert_cols[is.na(pert_cols)] <- "gray30"
  term_cols <- ifelse(deg == 1, "gray80",
                      ifelse(deg == 2, "gray50", "black"))
  V(g)$color <- ifelse(V(g)$type == "perturbation", pert_cols, term_cols)

  # Shapes & sizes
  V(g)$shape <- ifelse(V(g)$type == "perturbation", "square", "circle")
  V(g)$size  <- ifelse(V(g)$type == "perturbation", 25 + deg, 2 + deg)

  # --- Plot ---
  set.seed(123)
  p <- ggraph(g, layout = "fr") +
    # Edges use precomputed colors; identity scale renders them as-is
    geom_edge_link0(
      aes(edge_colour = edge_color),
      edge_width = 0.6,
      edge_alpha = 0.7
    ) +
    ggraph::scale_edge_colour_identity(
      name   = "Direction",
      breaks = c("darkorange2", "dodgerblue3"),
      labels = c("up in human", "up in chimp"),
      guide  = "legend"
    ) +
    guides(edge_colour = guide_legend(override.aes = list(edge_width = 2, edge_alpha = 1))) +

    # Nodes
    geom_node_point(aes(color = color, shape = shape, size = size)) +

    # GO:BP term labels  small & repelled
    geom_node_text(
      aes(label = name, filter = type == "term"),
      repel = TRUE, size = 2
    ) +

    # Perturbation labels  big, bold, centered
    geom_node_text(
      aes(label = name, filter = type == "perturbation"),
      repel = FALSE, size = 4, fontface = "bold", colour = "black",
      vjust = 0.5, hjust = 0.5
    ) +

    scale_color_identity() +
    scale_shape_identity() +
    scale_size_continuous(range = c(3, 15), guide = "none") +
    theme_void() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 16),
      legend.position = "right",
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank()
    ) +
    ggtitle(sprintf("Perturbation  GO:BP ID Network (%s)", tp))

  print(p)

  ggsave(
    filename = file.path(out_dir, sprintf("networkgraph_perturbation_by_Wikipathways_drGenes_%s.png", tp)),
    plot = p, height = 30, width = 30, units = "cm", bg = "white"
  )
}
```


Plot a network graph plot of overlapping divergent rGenes for each time point for each cell type
```{r network graph divergent rGenes}
  for (time in unique(divergent_rgenes_df$time_point)) {
  
    # Build edge list
    edges <- divergent_rgenes_df %>% 
      filter(time_point == time) %>% 
      tidyr::separate_rows(all, sep = "/") %>%
      dplyr::rename(gene = all) %>%
      distinct(perturbation, gene) %>% 
      drop_na()
    
    # Skip if nothing to plot
    if (nrow(edges) == 0L) next
    
    # Create igraph object
    g <- graph_from_data_frame(edges, directed = FALSE)

    # Flag node type (perturbation vs gene)
    V(g)$type <- ifelse(V(g)$name %in% edges$perturbation, "perturbation", "gene")

    # Calculate degree of overlap
    deg <- degree(g)

    # Color different degrees of overlap
    V(g)$color <- case_when(
      V(g)$type == "perturbation" ~ pert_colors[ V(g)$name ],
      V(g)$type == "gene" & deg == 1 ~ "gray80",
      V(g)$type == "gene" & deg == 2 ~ "gray50",
      TRUE ~ "black"
      )
    
    # Shape perturbation and gene nodes
    V(g)$shape <- ifelse(V(g)$type == "perturbation", "square", "circle")
    
    # Size perturbation and gene nodes
    V(g)$size  <- ifelse(V(g)$type == "perturbation", 25 + deg, 2 + deg)

    # Plot network graph
    set.seed(123)
    p <- ggraph(g, layout = "fr") +
      geom_edge_link0(alpha = 0.5, color = 'gray40') +
      geom_node_point(aes(color = color, shape = shape, size = size)) +
      
      ## Gene labels  small & repelled
      geom_node_text(
        aes(label = name, filter = type == "gene"),
        repel  = TRUE,
        size   = 2
      ) +
      
      ## Perturbation labels  big, bold, centered inside the node
      geom_node_text(
        aes(label = name, filter = type == "perturbation"),
        repel     = FALSE,          # keep exactly on the node
        size      = 4,              # bigger font
        fontface  = "bold",
        colour    = "black",        # good contrast on blue square
        vjust     = 0.5,            # centre vertically
        hjust     = 0.5             # centre horizontally
      ) +
      
      scale_color_identity() +
      scale_shape_identity() +
      scale_size_continuous(range = c(3, 15), guide = "none") +
      theme_void() +
      theme(
        plot.title = element_text(hjust = 0.5, size = 16),  # Center and size the title
        legend.position = "right",  # Place legend on the right
        panel.grid.major = element_blank(),  # Remove major gridlines
        panel.grid.minor = element_blank()   # Remove minor gridlines
      ) +
      ggtitle("Perturbation - Divergent rGene Network")
    print(p)
    ggsave(filename = paste0(fig_path, "networkgraph_divergent_rGenes/networkgraph_divergent_rGenes_", type, "_", time, ".png"), plot = p, height = 15, width = 15, bg = 'white')
  }
```

## Inspect specific divergent rGenes

Isolate specific intersections of genes
```{r intersection between perts}
# Define variables of interest
pert1 <- 'FGF8b'
pert2 <- 'FGF2'
time <- '6hr'

# Filter data frame
filt_df <- divergent_response_df %>% filter(time_point == time & perturbation %in% c(pert1, pert2) & significant==TRUE)

# Pull genes for both perturbations
pert1_genes <- filt_df %>% filter(perturbation == pert1) %>% pull(ID)
pert2_genes <- filt_df %>% filter(perturbation == pert2) %>% pull(ID)

# Find intersection of genes
intersect_genes <- intersect(pert1_genes, pert2_genes)
print(length(intersect_genes))
print(intersect_genes)
```

```{r intersection between perts}
# Define variables of interest
pert1 <- 'FGF8b'
pert2 <- 'FGF2'
time <- '6hr'

# Filter data frame
filt_df <- divergent_response_df %>% filter(time_point == time & perturbation %in% c(pert1, pert2) & significant==TRUE & logFCsign==1)

# Pull genes for both perturbations
pert1_genes <- filt_df %>% filter(perturbation == pert1) %>% pull(ID)
pert2_genes <- filt_df %>% filter(perturbation == pert2) %>% pull(ID)

# Find intersection of genes
intersect_genes <- intersect(pert1_genes, pert2_genes)
print(length(intersect_genes))
print(intersect_genes)
```

```{r intersection between perts}
# Define variables of interest
pert1 <- 'FGF8b'
pert2 <- 'FGF2'
time <- '54hr'

# Filter data frame
filt_df <- divergent_response_df %>% filter(time_point == time & perturbation %in% c(pert1, pert2) & significant==TRUE & logFCsign==1)

# Pull genes for both perturbations
pert1_genes <- filt_df %>% filter(perturbation == pert1) %>% pull(ID)
pert2_genes <- filt_df %>% filter(perturbation == pert2) %>% pull(ID)

# Find intersection of genes
intersect_genes <- intersect(pert1_genes, pert2_genes)
print(length(intersect_genes))
print(intersect_genes)
```

```{r intersection between perts}
# Define variables of interest
pert1 <- 'BMP7'
pert2 <- 'FGF2'
time <- '6hr'

# Filter data frame
filt_df <- divergent_response_df %>% filter(time_point == time & perturbation %in% c(pert1, pert2) & significant==TRUE)

# Pull genes for both perturbations
pert1_genes <- filt_df %>% filter(perturbation == pert1) %>% pull(ID)
pert2_genes <- filt_df %>% filter(perturbation == pert2) %>% pull(ID)

# Find intersection of genes
intersect_genes <- intersect(pert1_genes, pert2_genes)
print(length(intersect_genes))
print(intersect_genes)
```

```{r intersection between perts}
# Define variables of interest
pert1 <- 'BMP7'
pert2 <- 'FGF8b'
time <- '6hr'

# Filter data frame
filt_df <- divergent_response_df %>% filter(time_point == time & perturbation %in% c(pert1, pert2) & significant==TRUE)

# Pull genes for both perturbations
pert1_genes <- filt_df %>% filter(perturbation == pert1) %>% pull(ID)
pert2_genes <- filt_df %>% filter(perturbation == pert2) %>% pull(ID)

# Find intersection of genes
intersect_genes <- intersect(pert1_genes, pert2_genes)
print(length(intersect_genes))
print(intersect_genes)
```

## Inspect specific divergent rGenes

```{r inspect specific drgenes}
dr_df <- divergent_response_df %>% dplyr::filter(
                                          #significant == TRUE
                                          #& abs(logFC) > 0.5
                                          perturbation %in% c('BMP7')
                                          #& time_point == '6hr'
                                          #& time_point %in% c('54hr') # '6hr', , '7day'
                                          & ID %in% c('NRG3', 'ERBB4')
                                          )
View(dr_df)
#write.csv(dr_df, file = paste0(dir_path, "divergent_rGenes_FGF2_df.csv"), row.names = FALSE)
```

## Inspect specific divergent rGene ORA results

```{r inspect specific drgenes}
dr_ora_df <- divergent_ora_df %>% dplyr::filter(perturbation %in% c('BMP7')
                                      #& gene_set == 'GOBP' # 'GOCC', 'Reactome', 'Wikipathways', 'GOBP'
                                       #& perturbation == 'FGF2'
                                      & time_point == '54hr'
                                       #& time_point %in% c('6hr', '54hr')
                                          )
View(dr_ora_df)
```

## Inspect specific divergent rGene ORA results

```{r inspect specific species divergence}
spdiv_df <- speciesdiff_logfc_df %>% dplyr::filter(#div_rGene_sig == TRUE
                                          #& abs(logFC) > 0.5
                                          perturbation %in% c('BMP7')
                                          #& time_point == '6hr'
                                          #& time_point %in% c('54hr') #, '54hr', '7day')
                                          & ID %in% c('CACNA1C', 'GADD45G')
                                          )
View(spdiv_df)
```










# Baseline vs Response Analysis

## Baseline vs response scatter plots

```{r baseline_stimulus_scatter function}
baseline_stimulus_scatter <- function(df,
                                      color_by, color_title, colors_named_vector,
                                      shape_by, shape_title, shapes_named_vector,
                                      save_name) {

  # Ensure output directory exists
  out_dir <- file.path(fig_path, "scatterplot_BaselineSpeciesDiff_ResponseSpeciesDiff")
  if (!dir.exists(out_dir)) dir.create(out_dir, recursive = TRUE)

  p <- ggplot(
    df,
    aes(
      x = Baseline_logFC,
      y = Response_logFC,
      fill = {{ color_by }},
      shape = {{ shape_by }}
    )
  ) +
    geom_point(size = 2, stroke = 0.2, show.legend = TRUE) +  # no outline
    labs(
      x = expression("Baseline" ~ Log[2] ~ "(Human/Chimp)"),
      y = expression("Stimulus" ~ Log[2] ~ "(Human/Chimp)"),
      fill = color_title,
      shape = shape_title
    ) +
    scale_fill_manual(
      values = colors_named_vector,
      guide = guide_legend(override.aes = list(shape = 21), size=6)
    ) +
    scale_shape_manual(
      values = shapes_named_vector,
      guide = guide_legend(override.aes = list(fill = "black"), size=6)
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 16),
      legend.position = "right",
      axis.line = element_line(color = "black"),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      legend.title = element_text(size=14),
      legend.text = element_text(size=12),
      axis.title = element_text(size=14),
      axis.text = element_text(size=12)
    ) +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
    geom_hline(yintercept = 0, color = "black") +
    geom_vline(xintercept = 0, color = "black") +
    coord_fixed(ratio = 1)

  # Save and show
  ggsave(
    filename = file.path(out_dir, paste0(save_name, ".png")),
    plot = p, height = 4, width = 6, bg = "white"
  )
  print(p)
}

```

```{r scatter plot color by div rGene categories}
# Filter dataset to variables of interest
filt_df <- speciesdiff_logfc_df %>% dplyr::filter(rGene_category == "Divergent rGene") %>%
  # Set time point factor levels
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))
# Plot scatter plot using function
baseline_stimulus_scatter(filt_df,
                          div_rGene_category, "Divergent rGene Category", div_rgene_colors,
                          time_point, "Time Point", time_point_shapes,
                          "SpaceOfDivrGenes_ColorByDivrGeneCategory")
```

Plot a baseline vs stimulus logFC scatter plot in the space of divergent rGenes colored by perturbation
```{r scatter plot color by perturbation}
# Filter dataset to variables of interest
filt_df <- speciesdiff_logfc_df %>% dplyr::filter(rGene_category == "Divergent rGene") %>%
  # Set time point factor levels
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))
# Plot scatter plot using function
baseline_stimulus_scatter(filt_df,
                          perturbation, "Stimulus", pert_colors,
                          time_point, "Time Point", time_point_shapes,
                          "SpaceOfDivrGenes_ColorByPerturbation")
```

Plot a baseline vs stimulus logFC scatter plot in the space of divergent rGenes colored by disorder association for a specific perturbation
```{r scatter plot for one perturbation color by cell cycle gene}
# Filter to variables of interest
filt_df <- speciesdiff_logfc_df %>% dplyr::filter(rGene_category == "Divergent rGene", perturbation == 'BMP7', time_point=='54hr') %>%
  # Set time point factor levels
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Split dataset by BPD gene identity
other_df <- dplyr::filter(filt_df, !(ID %in% cc_genes))
bpd_df <- dplyr::filter(filt_df,  (ID %in% cc_genes))

# Padding
lim <- max(abs(c(filt_df$Baseline_logFC, filt_df$Response_logFC)), na.rm = TRUE)
pad_frac <- 0.08
pad_x    <- pad_frac * lim
pad_y    <- pad_frac * lim
x_lo <- -lim - pad_x; x_hi <-  lim + pad_x
y_lo <- -lim - pad_y; y_hi <-  lim + pad_y

# Plot scatter plot
p <- ggplot(filt_df, aes(x = Baseline_logFC, y = Response_logFC, shape = time_point)) +
  geom_point(data = other_df, aes(fill = "Other"), color = "black", stroke = 0.2, alpha = 0.9, size = 2) +
  geom_point(data = bpd_df, aes(fill = "Cell Cycle Associated"), color = "black", stroke = 0.2, alpha = 1.0, size = 2) +
  # Fill legend (category)
  scale_fill_manual(name   = "Category", values = c("Other" = "grey60", "Cell Cycle Associated" = "firebrick"),
    guide = guide_legend(override.aes = list(shape = 22, size = 4, stroke = 0.4, color = "black"))) +
  scale_shape_manual(values = time_point_shapes,
    guide = guide_legend(override.aes = list(fill = "black"), size = 4)) +
  labs(
      x = expression("Baseline" ~ Log[2] ~ "(Human/Chimp)"),
      y = expression("Stimulus" ~ Log[2] ~ "(Human/Chimp)"),
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    legend.position = "right",
    legend.title = element_blank(),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  scale_x_continuous(limits = c(x_lo, x_hi), expand = expansion(mult = 0)) +
  scale_y_continuous(limits = c(y_lo, y_hi), expand = expansion(mult = 0)) +
  coord_fixed(ratio = 1, clip = "on") +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
  geom_hline(yintercept = 0, color = "black") +
  geom_vline(xintercept = 0, color = "black")

ggsave(
  filename = paste0(fig_path, "scatterplot_BaselineVsResponse_BMP7_54hrs_ColorByCCgene.png"),
  plot = p, height = 5, width = 5, bg = "white")
print(p)
```

```{r get cell cycle genes}
# Get all genes that are the descendent of the mitotic cell cycle term
root_id <- "GO:0000278"  # mitotic cell cycle
mitotic_ids <- sort(unique(c(root_id, GO.db::GOBPOFFSPRING[[root_id]])))
mart <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
genes <- getBM(attributes = c("hgnc_symbol", "go_id"), 
                 filters = "go", 
                 values = mitotic_ids, 
                 mart = mart)
cc_genes <- sort(unique(genes$hgnc_symbol))
```

```{r get wnt genes}
# Get all genes that are the descendent of the canonical wnt signaling pathway term
root_id <- "GO:0060070"  # canonical wnt signaling pathway
wnt_ids <- sort(unique(c(root_id, GO.db::GOBPOFFSPRING[[root_id]])))
mart <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
genes <- getBM(attributes = c("hgnc_symbol", "go_id"), 
                 filters = "go", 
                 values = wnt_ids, 
                 mart = mart)
wnt_genes <- sort(unique(genes$hgnc_symbol))
```

```{r scatter plot color by cell cycle and wnt gene}
# Define genes to label
label_genes <- c('MKI67', 'RRM2', 'PBK', 'AURKB', 'KNTC1', 'KIF4A', 'BARD1', 'SPDL1', 'NCAPH', 'CLSPN')

# Filter to variables of interest
filt_df <- speciesdiff_logfc_df %>%
  dplyr::filter(rGene_category == "Divergent rGene", perturbation == 'BMP7', time_point %in% c('54hr')) %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Split dataset by category
other_df <- dplyr::filter(filt_df, !(ID %in% cc_genes))
cc_df <- dplyr::filter(filt_df, (ID %in% cc_genes))

# Find max of x and y axis limits
lim <- max(abs(c(filt_df$Baseline_logFC, filt_df$Response_logFC)), na.rm = TRUE)
nudge_y_mag <- 0.4 * lim
nudge_x_mag <- 0.6 * lim

label_df <- filt_df %>%
  dplyr::filter(ID %in% label_genes)

# Split labels by position relative to y = x
label_df_above <- dplyr::filter(label_df, Response_logFC >= Baseline_logFC)
label_df_below <- dplyr::filter(label_df, Response_logFC <  Baseline_logFC)

# Padding
pad_frac <- 0.05
pad_x <- pad_frac * lim
pad_y <- pad_frac * lim
x_lo <- -lim - pad_x
x_hi <-  lim + pad_x
y_lo <- -lim - pad_y
y_hi <-  lim + pad_y

# Plot scatter plot
p <- ggplot(filt_df, aes(x = Baseline_logFC, y = Response_logFC)) +
  
  # Plot based on category
  geom_point(data = other_df, aes(fill = "Other"), color = "black", stroke = 0.2, alpha = 0.9, size = 3, shape=22) +
  geom_point(data = cc_df, aes(fill = "Mitotic Cell Cycle"), color = "black", stroke = 0.2, alpha = 1.0, size = 3, shape=22) +

  scale_fill_manual(name = "Category",
                    values = c("Other" = "grey60", "Mitotic Cell Cycle" = "firebrick"),
                    guide = guide_legend(nrow=1, title.position = 'top', override.aes = list(shape = 22, size = 4, stroke = 0.4, color = "black"))
  ) +
  labs(
      x = expression("Baseline" ~ Log[2] ~ "(Human/Chimp)"),
      y = expression("Stimulus" ~ Log[2] ~ "(Human/Chimp)"),
  ) +
  theme_minimal() +
  theme(
    legend.position = "bottom",
    legend.title = element_blank(),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  scale_x_continuous(limits = c(x_lo, x_hi), expand = expansion(mult = 0)) +
  scale_y_continuous(limits = c(y_lo, y_hi), expand = expansion(mult = 0)) +
  coord_fixed(ratio = 1, clip = "on") +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
  geom_hline(yintercept = 0, color = "black") +
  geom_vline(xintercept = 0, color = "black") +
  
  ggrepel::geom_text_repel(
    data = label_df,
    aes(x = Baseline_logFC, y = Response_logFC, label = ID),
    size = 3,
    inherit.aes = FALSE,
    box.padding = 0.3,
    point.padding = 0.3,
    nudge_x = -nudge_x_mag,
    nudge_y =  nudge_y_mag,
    direction = "both",
    max.overlaps = Inf,
    force = 7,
    force_pull = 1.5,
    segment.size = 0.3,
    segment.color = "black",
    xlim = c(x_lo, x_hi),
    ylim = c(y_lo, y_hi)
  )

ggsave(paste0(fig_path, "scatterplot_BaselineVsResponse_BMP7_54hrs_ColorByWNTGeneCCgene.pdf"),
  plot = p, height = 4, width = 4, bg = "white", units='in', dpi=600)

print(p)
```

Baseline vs response species divergence scatter plot FGF2 and FGF8b at 6 hrs
```{r scatter plot bmp7 6 hr wnt}
# Define genes to label
genes <- c('NKD1', 'SMAD3', 'ROR2', 'SOX4', 'GATA3')

# Filter to variables of interest
filt_df <- speciesdiff_logfc_df %>% dplyr::filter(rGene_category == "Divergent rGene", perturbation %in% c('BMP7'), time_point %in% c('6hr')) %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Make label df
other_df <- dplyr::filter(filt_df,  !(ID %in% genes))
label_df <- dplyr::filter(filt_df,  (ID %in% genes))

# Find max of x and y axis limits
lim <- max(abs(c(filt_df$Baseline_logFC, filt_df$Response_logFC)), na.rm = TRUE)
nudge_y_mag <- 0.15 * lim
nudge_x_mag <- 0.25 * lim

# Split labels by position relative to y = x
label_df_above <- dplyr::filter(label_df, Response_logFC >= Baseline_logFC)
label_df_below <- dplyr::filter(label_df, Response_logFC <  Baseline_logFC)

# Padding
pad_frac <- 0.05
pad_x <- pad_frac * lim
pad_y <- pad_frac * lim
x_lo <- -lim - pad_x
x_hi <-  lim + pad_x
y_lo <- -lim - pad_y
y_hi <-  lim + pad_y

# Plot scatter plot
p <- ggplot(filt_df, aes(x = Baseline_logFC, y = Response_logFC, fill=perturbation)) +
  geom_point(data=other_df, aes(fill='Other'), color = "black", stroke = 0, alpha = 0.9, size = 2, shape=21) +
  geom_point(data=label_df, aes(fill='WNT Signaling'), color = "black", stroke = 0, alpha = 1.0, size = 2, shape=21) +

  scale_fill_manual(name = NULL,
                    values = c('Other' = 'grey', 'WNT Signaling' = 'royalblue'),
                    guide = guide_legend(override.aes = list(shape = 21, size = 4, stroke = 0.4, color = "black"))) +
  labs(
      x = expression("Baseline" ~ Log[2] ~ "(Human/Chimp)"),
      y = expression("Stimulus" ~ Log[2] ~ "(Human/Chimp)"),
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    legend.position = "bottom",
    legend.title = element_blank(),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  scale_x_continuous(limits = c(x_lo, x_hi), expand = expansion(mult = 0)) +
  scale_y_continuous(limits = c(y_lo, y_hi), expand = expansion(mult = 0)) +
  coord_fixed(ratio = 1, clip = "on") +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
  geom_hline(yintercept = 0, color = "black") +
  geom_vline(xintercept = 0, color = "black") +
  
  ggrepel::geom_text_repel(
    data = label_df,
    aes(x = Baseline_logFC, y = Response_logFC, label = ID),
    size = 3,
    inherit.aes = FALSE,
    box.padding = 0.2,
    point.padding = 0.2,
    nudge_x = -nudge_x_mag,
    nudge_y =  nudge_y_mag,
    direction = "both",
    max.overlaps = Inf,
    force = 2,
    force_pull = 0.75,
    segment.size = 0.3,
    segment.color = "black",
    xlim = c(x_lo, x_hi),
    ylim = c(y_lo, y_hi)
  )

ggsave(
  filename = paste0(fig_path, "scatterplot_BaselineVsResponse_BMP7_6hrs_ColorByPerturbation_LabelWNTGenes.pdf"),
  plot = p, height = 4.25, width = 4, bg = "white", units='in', dpi=600)
print(p)
```

Baseline vs response species divergence scatter plot FGF2 and FGF8b at 6 hrs
```{r scatter plot fgf 6 hr}
# Define genes to label
genes <- c('GAP43', 'PDE4B', 'DNMBP', 'DST', 'ATP2B1', 'ST6GALNAC5', 'DLGAP1')

# Filter to variables of interest
filt_df <- speciesdiff_logfc_df %>% dplyr::filter(rGene_category == "Divergent rGene", perturbation %in% c('FGF2', 'FGF8b'), time_point %in% c('6hr')) %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Make label df
label_df <- dplyr::filter(filt_df,  (ID %in% genes))

# Find max of x and y axis limits
lim <- max(abs(c(filt_df$Baseline_logFC, filt_df$Response_logFC)), na.rm = TRUE)
nudge_y_mag <- 0.3 * lim
nudge_x_mag <- 0.5 * lim

# Split labels by position relative to y = x
label_df_above <- dplyr::filter(label_df, Response_logFC >= Baseline_logFC)
label_df_below <- dplyr::filter(label_df, Response_logFC <  Baseline_logFC)

# Padding
pad_frac <- 0.05
pad_x <- pad_frac * lim
pad_y <- pad_frac * lim
x_lo <- -lim - pad_x
x_hi <-  lim + pad_x
y_lo <- -lim - pad_y
y_hi <-  lim + pad_y

# Plot scatter plot
p <- ggplot(filt_df, aes(x = Baseline_logFC, y = Response_logFC, fill=perturbation)) +
  geom_point(color = "black", stroke = 0, alpha = 1.0, size = 2, shape=21) +
  scale_fill_manual(name = "Stimulus",
                    values = pert_colors,
                    guide = guide_legend(override.aes = list(shape = 21, size = 4, stroke = 0.4, color = "black"))) +
  labs(
      x = expression("Baseline" ~ Log[2] ~ "(Human/Chimp)"),
      y = expression("Stimulus" ~ Log[2] ~ "(Human/Chimp)"),
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    legend.position = "bottom",
    legend.title = element_blank(),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  scale_x_continuous(limits = c(x_lo, x_hi), expand = expansion(mult = 0)) +
  scale_y_continuous(limits = c(y_lo, y_hi), expand = expansion(mult = 0)) +
  coord_fixed(ratio = 1, clip = "on") +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
  geom_hline(yintercept = 0, color = "black") +
  geom_vline(xintercept = 0, color = "black") +
  
  ggrepel::geom_text_repel(
    data = label_df_above,
    aes(x = Baseline_logFC, y = Response_logFC, label = ID),
    size = 3,
    inherit.aes = FALSE,
    box.padding = 0.3,
    point.padding = 0.3,
    nudge_x = -nudge_x_mag,
    nudge_y =  nudge_y_mag,
    direction = "both",
    max.overlaps = Inf,
    force = 4,
    force_pull = 1.5,
    segment.size = 0.3,
    segment.color = "black",
    xlim = c(x_lo, x_hi),
    ylim = c(y_lo, y_hi)
  ) + 
  
    ggrepel::geom_text_repel(
    data = label_df_below,
    aes(x = Baseline_logFC, y = Response_logFC, label = ID),
    size = 3,
    inherit.aes = FALSE,
    box.padding = 0.3,
    point.padding = 0.3,
    nudge_x = +nudge_x_mag,
    nudge_y =  -nudge_y_mag,
    direction = "both",
    max.overlaps = Inf,
    force = 4,
    force_pull = 1.5,
    segment.size = 0.3,
    segment.color = "black",
    xlim = c(x_lo, x_hi),
    ylim = c(y_lo, y_hi)
  )

ggsave(
  filename = paste0(fig_path, "scatterplot_BaselineVsResponse_FGF2_FGF8B_6hrs_ColorByPerturbation_LabelShared.pdf"),
  plot = p, height = 4.25, width = 4, bg = "white", units='in', dpi=600)
print(p)
```

Baseline vs response species divergence scatter plot FGF2 54 hrs
```{r scatter plot fgf2 54hr}
# Define genes to label
genes <- c("PDE4B", "DST", "ST6GALNAC5", 'NPY', 'NRP1', 'IL17RD', 'LPAR1', 'BMPR1A')

# Filter to variables of interest
filt_df <- speciesdiff_logfc_df %>% dplyr::filter(rGene_category == "Divergent rGene", perturbation %in% c('FGF2', 'FGF8b'), time_point %in% c('54hr')) %>%
  mutate(time_point = factor(time_point, levels = c('6hr', '54hr', '7day')))

# Make label df
label_df <- dplyr::filter(filt_df,  (ID %in% genes))

# Find max of x and y axis limits
lim <- max(abs(c(filt_df$Baseline_logFC, filt_df$Response_logFC)), na.rm = TRUE)
nudge_y_mag <- 0.3 * lim
nudge_x_mag <- 0.5 * lim

# Split labels by position relative to y = x
label_df_above <- dplyr::filter(label_df, Response_logFC >= Baseline_logFC)
label_df_below <- dplyr::filter(label_df, Response_logFC <  Baseline_logFC)

# Padding
pad_frac <- 0.05
pad_x <- pad_frac * lim
pad_y <- pad_frac * lim
x_lo <- -lim - pad_x
x_hi <-  lim + pad_x
y_lo <- -lim - pad_y
y_hi <-  lim + pad_y

# Plot scatter plot
p <- ggplot(filt_df, aes(x = Baseline_logFC, y = Response_logFC, fill=perturbation)) +
  geom_point(color = "black", stroke = 0, alpha = 1.0, size = 2, shape=21) +
  scale_fill_manual(name = "Stimulus",
                    values = pert_colors,
                    guide = guide_legend(override.aes = list(shape = 21, size = 4, stroke = 0.4, color = "black"))) +
  labs(
      x = expression("Baseline" ~ Log[2] ~ "(Human/Chimp)"),
      y = expression("Stimulus" ~ Log[2] ~ "(Human/Chimp)"),
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),
    legend.position = "bottom",
    legend.title = element_blank(),
    axis.line = element_line(color = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  scale_x_continuous(limits = c(x_lo, x_hi), expand = expansion(mult = 0)) +
  scale_y_continuous(limits = c(y_lo, y_hi), expand = expansion(mult = 0)) +
  coord_fixed(ratio = 1, clip = "on") +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
  geom_hline(yintercept = 0, color = "black") +
  geom_vline(xintercept = 0, color = "black") +
  
  ggrepel::geom_text_repel(
    data = label_df_above,
    aes(x = Baseline_logFC, y = Response_logFC, label = ID),
    size = 3,
    inherit.aes = FALSE,
    box.padding = 0.3,
    point.padding = 0.3,
    nudge_x = -nudge_x_mag,
    nudge_y =  nudge_y_mag,
    direction = "both",
    max.overlaps = Inf,
    force = 4,
    force_pull = 1.5,
    segment.size = 0.3,
    segment.color = "black",
    xlim = c(x_lo, x_hi),
    ylim = c(y_lo, y_hi)
  ) + 
  
    ggrepel::geom_text_repel(
    data = label_df_below,
    aes(x = Baseline_logFC, y = Response_logFC, label = ID),
    size = 3,
    inherit.aes = FALSE,
    box.padding = 0.3,
    point.padding = 0.3,
    nudge_x = +nudge_x_mag,
    nudge_y =  -nudge_y_mag,
    direction = "both",
    max.overlaps = Inf,
    force = 4,
    force_pull = 1.5,
    segment.size = 0.3,
    segment.color = "black",
    xlim = c(x_lo, x_hi),
    ylim = c(y_lo, y_hi)
  )

ggsave(
  filename = paste0(fig_path, "scatterplot_BaselineVsResponse_FGF2_FGF8B_54hrs_ColorByPerturbation_LabelMAPKERK.pdf"),
  plot = p, height = 4.25, width = 4, bg = "white", units='in', dpi=600)
print(p)
```









## Pie chart divergent rGene categories

```{r pie chart divergent rGenes categories}
pie_df <- speciesdiff_logfc_df %>%
  dplyr::filter(div_rGene_sig, !is.na(div_rGene_category)) %>%
  mutate(div_rGene_category = as.character(div_rGene_category)) %>%
  dplyr::count(div_rGene_category, name = "n", sort = TRUE) %>%
  arrange(desc(div_rGene_category)) %>%
  mutate(frac = n / sum(n),
         ymax = cumsum(frac),
         ymin = dplyr::lag(ymax, default = 0),
         ymid = (ymin + ymax) / 2, 
         pct_label = scales::percent(frac, accuracy = 0.1),
         label = paste0(pct_label, "\n(n = ", scales::comma(n), ")")
         )


# Plot pie chart
p <- ggplot(pie_df, aes(ymax = ymax, ymin = ymin, xmax = 1, xmin = 0, fill = div_rGene_category)) +
  geom_rect(color = "white") +
  coord_polar(theta = "y") +
  scale_fill_manual(values = div_rgene_colors,
                    name = "Divergent rGene Category",
                    guide = guide_legend(nrow = 2, byrow = TRUE, title.position = 'top')
  ) +
  xlim(c(0, 2.0)) +                                                # add room for labels
  theme_void() +
  theme(
    legend.justification = "center",
    #legend.position = "bottom",
    legend.title = element_text(size = 14, hjust = 0.5),
    legend.title.align = 0.5,
    legend.text  = element_text(size = 12),
    legend.text.align = 0.5,
    legend.box.just = "center",
    legend.box = "vertical",
    legend.position = 'none'
  ) +
  geom_text_repel(
    data = pie_df,
    aes(x = 1.5, y = ymid, label = label),
    inherit.aes = FALSE,
    nudge_x = 0.025,
    direction = "y",
    vjust = 0.5,
    size = 4,
    segment.size = 0.3,
    segment.color = "grey50",
    show.legend = FALSE
  )

  ggsave(filename = paste0(fig_path, "PieChart_DivergentrGenes_ColorByDivergentrGeneCategory.png"),
         plot = p, height = 4, width = 4, bg = 'white', units='in', dpi=600)
  
print(p)
```

## Inspect specific baseline vs response divergent gene results

```{r inspect specific drgenes}
bvr_genes_df <- speciesdiff_logfc_df %>% filter(perturbation %in% c('BMP7')
                                       #& perturbation == 'FGF2'
                                       & time_point == '54hr'
                                       #& time_point %in% c('6hr', '54hr')
                                          )
View(bvr_genes_df)
```

## Stacked bar plot divergent rGene categories

```{r stacked bar plot drGene categories}
# Prep & filter once
filt_df <- speciesdiff_logfc_df %>%
  dplyr::filter(div_rGene_sig == TRUE) %>%
  mutate(
    div_rGene_category = as.character(div_rGene_category),
    time_point = as.character(time_point),
    perturbation = as.character(perturbation)
  )

# Build ordering table: counts per perturbation  time
order_tbl <- filt_df %>%
  dplyr::count(perturbation, time_point, name = "n") %>%
  tidyr::pivot_wider(names_from = time_point, values_from = n, values_fill = 0)

# Keep perturbations with any counts; order by 6hr, then 54hr, then 7day (all desc)
keep_levels <- order_tbl %>%
  mutate(total = `6hr` + `54hr` + `7day`) %>%
  dplyr::filter(total > 0) %>%
  arrange(desc(`6hr`), desc(`54hr`), desc(`7day`)) %>%
  pull(perturbation)

# Apply ordering & drop others
filt_df <- filt_df %>%
  dplyr::filter(perturbation %in% keep_levels) %>%
  mutate(perturbation = factor(perturbation, levels = rev(keep_levels)))

# (Optional) Summarize for plotting stacked bars by divergent category
plot_df <- filt_df %>%
  dplyr::count(perturbation, time_point, div_rGene_category, name = "n") %>%
  mutate(
    time_point = factor(time_point, levels = c('6hr', '54hr', '7day'))
  )

# Plot horizontal stacked bar plot with facets by time
p <-ggplot(plot_df, aes(y = perturbation, x = n, fill = div_rGene_category)) +
  geom_col() +
  facet_wrap(~ time_point, nrow = 1) +
  scale_fill_manual(values = div_rgene_colors, na.value = "black", drop = FALSE) +
  labs(
    y = "",
    x = "Divergent rGene Count",
    fill = "Divergent rGene Category"
  ) +
  theme_minimal(base_size = 12) +
  theme(
    strip.text = element_text(size = 14),
    panel.grid = element_blank(),
    panel.grid.major.x = element_line(color = "grey80", linewidth = 0.4),
    axis.text = element_text(size=12),
    legend.title = element_text(size=14),
    legend.text = element_text(size=12),
    legend.position = 'none'
  )

ggsave(
  file.path(fig_path, "stacked_barplot_DivergentrGeneCount_byPerturbation_ColorByDivrGeneCategory_horizontal_facet.pdf"),
  plot = p, width = 7, height = 4, bg = "white", dpi=600, units='in'
)
print(p)
```

## Baseline vs. Response ORA GO:BP

```{r}
# Join baseline and response divergence ORA GO:BP data frames
div_response_ora_df <- readRDS(paste0(dir_path, "divergent_rGenes_ORA_df.rds"))
div_response_ora_df <- div_response_ora_df %>% dplyr::filter(gene_set=='GOBP')
div_response_ora_df$GeneCategory <- 'Divergent rGene'

div_baseline_ora_df <- readRDS(paste0(dir_path, "divergent_baseline_genes_ORA_GOBP_df.rds"))
#div_baseline_ora_df <- div_baseline_ora_df %>% dplyr::filter(gene_set=='GOBP')
div_baseline_ora_df$GeneCategory <- 'Divergent Baseline Gene'

div_ora_df <- dplyr::bind_rows(div_response_ora_df, div_baseline_ora_df)
head(div_ora_df)
```

```{r add GO:BP metaterms}
# Patterns
neural_terms           <- c("NEURON","NERVOUS_SYSTEM","NEUROGENESIS","NEURAL","NEURONAL","RADIAL_GLIA",
                            "ACTION_POTENTIAL","SYNAPTIC","SYNAPSE","NEUROBLAST","AXON",
                            "BRAIN","TELENCEPHALON","DIENCEPHALON","FOREBRAIN","CEREBRAL_CORTEX","CORTEX",
                            "SUBPALLIUM","PALLIUM","SPINAL_CORD","MIDBRAIN","HINDBRAIN")
patterning_terms       <- c("PATTERN","PATTERNING","AXIS_ELONGATION","AXIS_SPECIFICATION",
                            "PATTERN_SPECIFICATION","PATTERN_FORMATION")
developmental_terms    <- c("DEVELOPMENT","DEVELOPMENTAL")
differentiation_terms  <- c("DIFFERENTIATION")
morph_terms            <- c("MORPHOGENESIS")
prolif_growth_cc_terms <- c("GROWTH","PROLIFERATION","CELL_CYCLE","MITOTIC","MEIOTIC")

rx_neural            <- str_c(neural_terms, collapse = "|")
rx_patterning        <- str_c(patterning_terms, collapse = "|")
rx_developmental     <- str_c(developmental_terms, collapse = "|")
rx_differentiation   <- str_c(differentiation_terms, collapse = "|")
rx_morph             <- str_c(morph_terms, collapse = "|")
rx_prolif_growth_cc  <- str_c(prolif_growth_cc_terms, collapse = "|")

has_pat <- function(x, patt) str_detect(x, regex(patt, ignore_case = TRUE))

# Add flags
div_ora_df <- div_ora_df %>%
  mutate(
    Neural           = has_pat(Description, rx_neural),
    Patterning       = has_pat(Description, rx_patterning),
    Development      = has_pat(Description, rx_developmental),
    Differentiation  = has_pat(Description, rx_differentiation),
    Morphogenesis    = has_pat(Description, rx_morph),
    Prolif_Growth_CC = has_pat(Description, rx_prolif_growth_cc)
  ) %>%
  rowwise() %>%
  mutate(
    matches = list(names(which(c(
      Neural           = Neural,
      Patterning       = Patterning,
      Development      = Development,
      Differentiation  = Differentiation,
      Morphogenesis    = Morphogenesis,
      Prolif_Growth_CC = Prolif_Growth_CC
    ))))
  ) %>%
  ungroup() %>%
  mutate(
    n_matches = lengths(matches),
    Metaterm = dplyr::case_when(
      n_matches == 0 ~ "Other",
      n_matches == 1 ~ vapply(matches, function(x) dplyr::coalesce(x[1], "Other"), character(1)),
      n_matches  > 1 ~ "Multiple metaterms",
      TRUE          ~ "Other"
    )
  )
head(div_ora_df)
```

```{r count metaterms}
# Count # of terms for each Metaterm within each pert x time x direction group
metaterm_counts_df <- div_ora_df %>%
  dplyr::filter(!is.na(Metaterm), !is.na(ID), ID != "") %>%
  group_by(GeneCategory, Metaterm) %>%
  summarise(N_terms = n(), .groups = "drop")

head(metaterm_counts_df)
```

```{r}
# Make a fill palette where "Other" is black
metaterm_levels <- sort(unique(metaterm_counts_df$Metaterm))
metaterm_pal <- setNames(hue_pal()(length(metaterm_levels)), metaterm_levels)
metaterm_pal["Other"] <- "black"

plot_df <- metaterm_counts_df %>%
  mutate(
    GeneCategory = factor(GeneCategory),
    Metaterm = factor(Metaterm,
      levels = c('Development', 'Differentiation', 'Morphogenesis', 'Neural', 'Patterning', 'Prolif_Growth_CC', 'Multiple metaterms', 'Other'))
    )

p <- ggplot(plot_df, aes(x = GeneCategory, y = N_terms, fill = Metaterm)) +
  geom_col(width = 0.8) +
  labs(
    x = "",
    y = "GO:BP Term Count",
    fill = "Metaterm"
  ) +
  scale_fill_manual(
    values = metaterm_pal, drop = FALSE,
    guide = guide_legend(nrow = 3, byrow = TRUE, title.position = "top")
  ) +
  theme_classic() +
  theme(
    axis.text.x = element_text(angle = 15, hjust = 1, size=12),
    axis.title = element_text(size=14),
    legend.title = element_text(size=14, hjust=0.5),
    legend.text = element_text(size=12),
    legend.position = "bottom"
  )

ggsave(paste0(fig_path, "StackedBarplot_DivergentGeneCategory_Vs_GOBPTermCount_ColorByMetaterm.pdf"), plot=p, width=5.5, height=5.5, dpi=600, bg='white')
print(p)
```



## Total baseline species DEGs
```{r}
total_genes <- speciesdiff_logfc_df %>% filter(assay == 'Tel_NE_ventral') %>%
  pull(ID) %>%
  unique()

species_degs <- speciesdiff_logfc_df %>% filter(assay == 'Tel_NE_ventral' & Baseline_sig == TRUE & !perturbation=='Verteporfin') %>%
  pull(ID) %>%
  unique()

print(paste0("Total genes: ", length(total_genes)))
print(paste0("Species DEGs: ", length(species_degs)))
print(paste0("Species DEGs / Total genes (%): ", length(species_degs)/length(total_genes)*100))
```

## Total unique divergent rGenes
```{r}
total_genes <- speciesdiff_logfc_df %>% filter(assay == 'Tel_NE_ventral' & !perturbation == 'Verteporfin') %>%
  pull(ID) %>%
  unique()

div_rgenes <- speciesdiff_logfc_df %>% filter(assay == 'Tel_NE_ventral' & div_rGene_sig == TRUE & !perturbation=='Verteporfin') %>%
  pull(ID) %>%
  unique()

print(paste0("Total genes: ", length(total_genes)))
print(paste0("Divergent rGenes: ", length(div_rgenes)))
print(paste0("Divergent rGenes / Total genes (%): ", length(div_rgenes)/length(total_genes)*100))
```

```{r}
div_rgenes <- speciesdiff_logfc_df %>% filter(assay == 'Tel_NE_ventral' & div_rGene_sig == TRUE & !perturbation=='Verteporfin' & time_point == '7day') %>%
  pull(ID) %>%
  unique()

print(paste0("Divergent rGenes: ", length(div_rgenes)))
```

```{r}
response_div_rgenes <- speciesdiff_logfc_df %>%
  filter(assay == 'Tel_NE_ventral' & div_rGene_sig == TRUE & !perturbation=='Verteporfin' & div_rGene_category == 'Response Only') %>%
  pull(ID) %>%
  unique()

print(paste0("Response Only Divergent rGenes: ", length(response_div_rgenes)))
print(paste0("Response Only / Divergent rGenes (%): ", length(response_div_rgenes) / length(div_rgenes)*100))
```

```{r}
baseline_div_rgenes <- speciesdiff_logfc_df %>%
  filter(assay == 'Tel_NE_ventral' & div_rGene_sig == TRUE & !perturbation=='Verteporfin' & div_rGene_category == 'Baseline Only') %>%
  pull(ID) %>%
  unique()

print(paste0("Baseline Only Divergent rGenes: ", length(baseline_div_rgenes)))
print(paste0("Baseline Only / Divergent rGenes (%): ", length(baseline_div_rgenes) / length(div_rgenes)*100))
```

```{r}
enhancing_div_rgenes <- speciesdiff_logfc_df %>%
  filter(assay == 'Tel_NE_ventral' & div_rGene_sig == TRUE & !perturbation=='Verteporfin' & div_rGene_category == 'Enhancing') %>%
  pull(ID) %>%
  unique()

print(paste0("Enhancing Divergent rGenes: ", length(enhancing_div_rgenes)))
print(paste0("Enhancing / Divergent rGenes (%): ", length(enhancing_div_rgenes) / length(div_rgenes)*100))
```

```{r}
opposing_div_rgenes <- speciesdiff_logfc_df %>%
  filter(assay == 'Tel_NE_ventral' & div_rGene_sig == TRUE & !perturbation=='Verteporfin' & div_rGene_category == 'Opposing') %>%
  pull(ID) %>%
  unique()

print(paste0("Opposing Divergent rGenes: ", length(opposing_div_rgenes)))
print(paste0("Opposing / Divergent rGenes (%): ", length(opposing_div_rgenes) / length(div_rgenes)*100))
```

```{r}
reversing_div_rgenes <- speciesdiff_logfc_df %>%
  filter(assay == 'Tel_NE_ventral' & div_rGene_sig == TRUE & !perturbation=='Verteporfin' & div_rGene_category == 'Reversing') %>%
  pull(ID) %>%
  unique()

print(paste0("Reversing Divergent rGenes: ", length(reversing_div_rgenes)))
print(paste0("Reversing / Divergent rGenes (%): ", length(reversing_div_rgenes) / length(div_rgenes)*100))
```

# rGene box plots

```{r load voom_res}
# Read in voom results.
voom_res <- readRDS(paste0(dir_path, "NEMP25_Tel_res_voom_v16.rds"))
```

Define a function for plotting box plots of stimulus vs baseline across species across time points. 
```{r plotStratify function}
# Define plotStratify function.
plotstratify_boxplot <- function(voom_res, cell_type, pert_string, gene_string){
  
  # Assign appropriate vehicle control.
  dmso_perts <- c('TRULI', 'Verteporfin', 'ATRA', 'HX531', 'WNTC59', 'AZD8931', 'LDN193189', 'CHIR99021', 'Rapamycin', 'LY411575', 'BGJ398', 'SANT1')
  treman_perts <- c('BMP7', 'IGF1LR3', 'JAG1', 'DLL1', 'FGF2', 'FGF8b', 'NRG1', 'SHH', 'EGF')
  if (pert_string %in% dmso_perts){
    vehicle <- 'DMSO'
  }
  if (pert_string %in% treman_perts){
    vehicle <- 'TreMan'
  }
  
  # Extract gene data for cell type.
  data = dreamlet::extractData(voom_res, cell_type) 
  
  # Filter the data to the selected perturbation.
  data <- data %>% filter(perturbation == pert_string | perturbation == vehicle) %>%
    # Replace '-' from gene names in columnswith '_'
    rename_with(~ gsub("-", "_", .x))
  
  # Set factor levels.
  data$species <- factor(data$species, levels = c('Human', 'Chimp', 'Orangutan'))
  data$time_point <- factor(data$time_point, levels = c('6hr', '54hr', '7day'))
  data$perturbation <- factor(data$perturbation, levels = c(vehicle, pert_string))
  data$plot_order <- paste(data$species, data$time_point, data$perturbation, sep="_")
  data$plot_order <- factor(data$plot_order, levels = c(
                                                        paste('Human', '6hr', vehicle, sep='_'),
                                                        paste('Human', '54hr', vehicle, sep='_'),
                                                        paste('Human', '7day', vehicle, sep='_'),
                                                        paste('Chimp', '6hr', vehicle, sep='_'),
                                                        paste('Chimp', '54hr', vehicle, sep='_'),
                                                        paste('Chimp', '7day', vehicle, sep='_'),
                                                        paste('Orangutan', '6hr', vehicle, sep='_'),
                                                        paste('Orangutan', '54hr', vehicle, sep='_'),
                                                        paste('Orangutan', '7day', vehicle, sep='_'),
                                                        paste('Human', '6hr', pert_string, sep='_'),
                                                        paste('Human', '54hr', pert_string, sep='_'),
                                                        paste('Human', '7day', pert_string, sep='_'),
                                                        paste('Chimp', '6hr', pert_string, sep='_'),
                                                        paste('Chimp', '54hr', pert_string, sep='_'),
                                                        paste('Chimp', '7day', pert_string, sep='_'),
                                                        paste('Orangutan', '6hr', pert_string, sep='_'),
                                                        paste('Orangutan', '54hr', pert_string, sep='_'),
                                                        paste('Orangutan', '7day', pert_string, sep='_')
                                                        ))
  
  # Convert gene from string.
  gene_raw <- noquote(gene_string)
  
  # Define formula.
  formula <- as.formula(paste(gene_string, "~ plot_order"))

  # Plot plotStratify box plot.
  p <- plotStratify(formula, data, colorBy=data$species, legend=TRUE, x.labels=TRUE, sort = FALSE) +
    scale_fill_manual(values = species_colors) +  # Rename legend categories
    scale_color_manual(values = pert_colors) +
    labs(title = paste0(gene_string, "\n", cell_type, "\n", pert_string, " stimulus")) +
    theme(aspect.ratio=1,
          plot.title = element_text(hjust = 0.5, size=20), #Center plot title.
          axis.text.x=element_text(hjust=1, vjust=1, size=12), #Modify x-axis tick labels.
          legend.position = "right", #Move legend to the top of the plot.
          legend.justification = "center", #Center justify the legend.
          legend.text = element_text(size = 18), # Increase the size of the legend labels
          legend.title = element_blank()) + #Remove the legend title.
    guides(color = "none")  + #Hide the color legend but keeps the fill legend
    ylab(bquote(log[2]~CPM)) + #Add a y-axis label.
    xlab('') + #Remove the x-axis label.
    coord_flip()  # Rotates the plot by 90 degrees
  print(p)
}
```


## Conserved rGene box plots

Plot a box plot for species conserved BMP7 rGenes associated with microcephaly
```{r microcephaly}
# Load conserved rGene ORA with DisGeNet results table
simple_disgenet_df <- readRDS(paste0(dir_path, "conserved_rGenes_DisGeNet_ORA.rds"))

# Set sample group
pert = 'BMP7'
type = 'Tel_NE_ventral'
save_name = paste0("plotstratify_plot_microcephaly_", type, "_", pert)

# Pull genes
degs <- simple_disgenet_df %>% filter(perturbation == pert & diseaseName == "Autosomal Recessive Primary Microcephaly") %>% pull(intersection)
degs <- unlist(strsplit(degs, ","))
degs <- gsub("-", "_", degs)
degs <- intersect(degs, names(voom_res$Tel_NE_ventral$voom.xy$x))

# Iterate through genes and plot
for (deg in degs){
    p <- plotstratify_boxplot(voom_res, type, pert, deg)
    p + ggtitle(paste0(deg, "/n", pert, " stimulus"))
    ggsave(filename = paste0(fig_path, save_name, "/", save_name, "_", deg, ".png"), plot=p, width=12, height=6, dpi=300, bg='white')
}
```

## Divergent rGene box plots

```{r plotStratify divergent rGenes group}
# Plot a box plot for species divergent rGenes for a sample group
# Set sample group
pert = 'FGF2'
type = 'Tel_NE_ventral'
time = '6hr'
save_name = paste0("boxplot_Log2CPM_", type, "_", pert, "_", time)

# Pull divergent rGenes
degs <- divergent_rgenes_df %>% filter(perturbation == pert & cell_type == type & time_point == time) %>% pull(all_up_chimp)
degs <- unlist(strsplit(degs, "/"))
degs <- gsub("-", "_", degs)

# Iterate through genes and plot
for (deg in degs){
    p <- plotstratify_boxplot(voom_res, type, pert, deg)
    p + ggtitle(paste0(deg, "/n", pert, " ", time))
    ggsave(filename = paste0(fig_path, save_name, "/", save_name, "_", deg, ".png"), plot=p, width=8, height=6, dpi=300, bg='white')
}
```

```{r plotStratify divergent rGene single}
# Plot a box plot for species divergent rGene
# Set sample group
pert = 'BMP7'
type = 'Tel_NE_ventral'
save_name = paste0("boxplot_Log2CPM_", type, "_", pert)

# Set divergent rGene
deg = 'NKX2-1'

# Plot gene
p <- plotstratify_boxplot(voom_res, type, pert, deg)
ggsave(filename = paste0(fig_path, save_name, "/", save_name, "_", deg, ".png"), plot=p, width=8, height=6, dpi=300, bg='white')
```

# rGene heatmaps

```{r load voom_df}
voom_df <- extractData(voom_res, assay = "NEC")
```

```{r}
mean_complexheatmap <- function(
  voom_df,
  time_point,
  perts,
  genes,
  save_name,
  title,
  species_order,
  species_colors,
  pert_colors
) {

  # Filter dataset to variables of interest
  filt_voom_df <- voom_df %>%
    dplyr::filter(time_point == time_point, perturbation %in% perts)

  # Mean value per species x perturbation
  mean_v_df <- filt_voom_df %>%
    dplyr::select(species, perturbation, 10:ncol(.)) %>%
    dplyr::group_by(species, perturbation) %>%
    dplyr::summarise(dplyr::across(dplyr::everything(), ~ mean(.x, na.rm = TRUE)), .groups = "drop") %>%
    dplyr::mutate(
      species = factor(species, levels = species_order),
      perturbation = factor(perturbation, levels = perts)
    ) %>%
    dplyr::arrange(species, perturbation)

  # Subset to genes of interest present in matrix
  sel_genes <- intersect(genes, colnames(mean_v_df))
  if (length(sel_genes) == 0) {
    stop("No selected genes found in mean_v_df columns. Check 'genes' and 'mean_v_df' content.")
  }

  # Build heatmap matrix (rows = species_perturbation; cols = selected genes)
  heat_mat <- mean_v_df %>%
    tidyr::unite("group", species, perturbation, sep = "_") %>%
    tibble::column_to_rownames("group") %>%
    dplyr::select(dplyr::all_of(sel_genes)) %>%
    as.matrix()

  # Spectral colormap for values
  rng <- range(heat_mat, na.rm = TRUE)
  col_fun <- colorRamp2(
    seq(rng[1], rng[2], length.out = 11),
    rev(brewer.pal(11, "Spectral"))
  )

  # Build clean row annotation vectors aligned to heat_mat
  meta_rows <- mean_v_df %>%
    tidyr::unite("group", species, perturbation, sep = "_", remove = FALSE) %>%
    tibble::column_to_rownames("group") %>%
    as.data.frame()

  # Align to heat_mat rows
  meta_rows <- meta_rows[rownames(heat_mat), , drop = FALSE]

  # Plain factor vectors
  row_species <- factor(as.character(meta_rows$species), levels = species_order)
  row_perturb <- factor(as.character(meta_rows$perturbation), levels = perts)

  # obust color mapping (names must match levels exactly)
  if (is.null(names(species_colors))) stop("species_colors must be a NAMED vector.")
  if (is.null(names(pert_colors)))    stop("pert_colors must be a NAMED vector.")

  spec_lvls <- levels(row_species)
  pert_lvls <- levels(row_perturb)

  species_colmap <- setNames(unname(species_colors[spec_lvls]), spec_lvls)
  pert_colmap    <- setNames(unname(pert_colors[pert_lvls]),    pert_lvls)

  if (any(is.na(species_colmap))) {
    missing <- spec_lvls[is.na(species_colmap)]
    stop("Missing species_colors for: ", paste(missing, collapse = ", "))
  }
  if (any(is.na(pert_colmap))) {
    missing <- pert_lvls[is.na(pert_colmap)]
    stop("Missing pert_colors for: ", paste(missing, collapse = ", "))
  }

  # Row annotation via explicit vectors
  row_ha <- rowAnnotation(
    Species      = row_species,
    Perturbation = row_perturb,
    col = list(
      Species      = species_colmap,
      Perturbation = pert_colmap
    ),
    annotation_legend_param = list(
      Species      = list(title = "Species"),
      Perturbation = list(title = "Perturbation")
    ),
    show_legend = TRUE
  )

  # Output directory
  out_dir <- file.path(fig_path, "heatmap_MeanLog2CPM")
  if (!dir.exists(out_dir)) dir.create(out_dir, recursive = TRUE)

  perts_str <- paste(perts, collapse = "_")

  # Plot complex heatmap
  pdf(
    file = file.path(out_dir, paste0("heatmap_MeanLog2CPM_", save_name, "_", type, "_", perts_str, "_", time_point, ".pdf")),
    width = 6, height = 4
  )

  ht <- Heatmap(
    heat_mat,
    name = "Log2CPM",
    col = col_fun,
    cluster_rows = FALSE,
    cluster_columns = TRUE,
    show_row_names = FALSE,
    show_column_names = TRUE,
    column_title = title,
    left_annotation = row_ha
  )

  draw(ht)
  dev.off()
}
```

## Conserved rGene heatmap

```{r microcephaly conserved BMP7 rGenes}
# Define input variables
type <- 'Tel_NE_ventral'
time <- '54hr'
pert <- 'BMP7'
cntrl <- 'TreMan'
perts <- c(cntrl, pert)
species_order <- c('Human', 'Chimp', 'Orangutan')


# Load conserved rGene ORA with DisGeNet results table
simple_disgenet_df <- readRDS(paste0(dir_path, "conserved_rGenes_DisGeNet_ORA.rds"))

# Pull genes
genes <- simple_disgenet_df %>% 
  filter(perturbation == pert & diseaseName == "Autosomal Recessive Primary Microcephaly") %>% 
  pull(intersection)
genes <- unlist(strsplit(genes, ","))

# Plot complex heatmap
mean_complexheatmap(voom_df,
                    time,
                    perts,
                    genes,
                    "microcephaly",
                    "Conserved BMP7 rGenes Associated with Autosomal Recessive Primary Microcephaly",
                    species_order,
                    species_colors,
                    pert_colors
                    )
```

```{r WNT-associated conserved BMP7 rGenes}
# Define input variables
type <- 'Tel_NE_ventral'
time <- '54hr'
pert <- 'BMP7'
cntrl <- 'TreMan'
perts <- c(cntrl, pert)
species_order <- c('Human', 'Chimp', 'Orangutan')


# Load conserved rGene ORA with GO:BP results table
all_ora_df <- readRDS(paste0(dir_path, "conserved_rGenes_GOBP_ORA_df.rds"))

# Pull genes
genes <- all_ora_df %>% 
  filter(perturbation == pert & time_point == time & Description == "WNT SIGNALING PATHWAY") %>% 
  pull(geneSymbol)
genes <- unlist(strsplit(genes, "/"))

# Plot complex heatmap
mean_complexheatmap(voom_df,
                    time,
                    perts,
                    genes,
                    "WNT_signaling_pathway",
                    "Conserved BMP7 rGenes Associated with the WNT Signaling Pathway",
                    species_order,
                    species_colors,
                    pert_colors
                    )
```

```{r WNT-associated conserved CHIR99021 rGenes}
# Define input variables
type <- 'Tel_NE_ventral'
time <- '54hr'
pert <- 'CHIR99021'
cntrl <- 'TreMan'
perts <- c(cntrl, pert)
species_order <- c('Human', 'Chimp', 'Orangutan')


# Load conserved rGene ORA with GO:BP results table
all_ora_df <- readRDS(paste0(dir_path, "conserved_rGenes_GOBP_ORA_df.rds"))

# Pull genes
genes <- all_ora_df %>% 
  filter(perturbation == pert & time_point == time & Description == "WNT SIGNALING PATHWAY") %>% 
  pull(geneSymbol)
genes <- unlist(strsplit(genes, "/"))

# Plot complex heatmap
mean_complexheatmap(voom_df,
                    time,
                    perts,
                    genes,
                    "WNT_signaling_pathway",
                    "Conserved CHIR99021 rGenes Associated with the WNT Signaling Pathway",
                    species_order,
                    species_colors,
                    pert_colors
                    )
```

```{r conserved BMP7 rGenes 54 hr panel}
# Define input variables
type <- 'Tel_NE_ventral'
time <- '54hr'
pert <- 'BMP7'
cntrl <- 'TreMan'
perts <- c(cntrl, pert)
species_order <- c('Human', 'Chimp', 'Orangutan')


# Define genes
genes <- c('ID2', 'SMAD9', 'LEF1', 'TLE3')

# Plot complex heatmap
mean_complexheatmap(voom_df,
                    time,
                    perts,
                    genes,
                    "convergent_BMP7_rGenes_54hr_select",
                    "Select Conserved BMP7 rGenes",
                    species_order,
                    species_colors,
                    pert_colors
                    )
```


## Divergent rGene heatmaps

```{r divergent BMP7 rGenes 54 hr panel}
# Define input variables
type <- 'Tel_NE_ventral'
time <- '54hr'
pert <- 'BMP7'
cntrl <- 'TreMan'
perts <- c(cntrl, pert)
species_order <- c('Human', 'Chimp', 'Orangutan')


# Define genes
genes <- c('ID2', # conserved
           'MKI67', 'MCM4', 'RACGAP1', 'SERPINI1', # response only
           'BRCA1', 'LGR5', # enchancing
           'KCNH5', 'CSRNP3' # reversing
           )

# Plot complex heatmap
mean_complexheatmap(voom_df,
                    time,
                    perts,
                    genes,
                    "divergent_BMP7_rGenes_54hr_select",
                    "Select Divergent BMP7 rGenes",
                    species_order,
                    species_colors,
                    pert_colors
                    )
```


```{r}
mean_complexheatmap2 <- function(
  voom_df,
  time_point,
  perts,
  genes,
  save_name,
  title,
  species_order,
  species_colors,
  pert_colors,
  col_df,        # <- NEW: data.frame(gene, category)
  cat_colors     # <- NEW: named vector of colors for categories
) {

  # Filter dataset to variables of interest
  filt_voom_df <- voom_df %>%
    dplyr::filter(time_point == time_point, perturbation %in% perts)

  # Mean value per species x perturbation
  mean_v_df <- filt_voom_df %>%
    dplyr::select(species, perturbation, 10:ncol(.)) %>%
    dplyr::group_by(species, perturbation) %>%
    dplyr::summarise(dplyr::across(dplyr::everything(), ~ mean(.x, na.rm = TRUE)), .groups = "drop") %>%
    dplyr::mutate(
      species = factor(species, levels = species_order),
      perturbation = factor(perturbation, levels = perts)
    ) %>%
    dplyr::arrange(species, perturbation)

  # Subset to genes of interest present in matrix
  sel_genes <- intersect(genes, colnames(mean_v_df))
  if (length(sel_genes) == 0) {
    stop("No selected genes found in mean_v_df columns. Check 'genes' and 'mean_v_df' content.")
  }

  # Build heatmap matrix (rows = species_perturbation; cols = selected genes)
  heat_mat <- mean_v_df %>%
    tidyr::unite("group", species, perturbation, sep = "_") %>%
    tibble::column_to_rownames("group") %>%
    dplyr::select(dplyr::all_of(sel_genes)) %>%
    as.matrix()

  # Color scale for values
  rng <- range(heat_mat, na.rm = TRUE)
  col_fun <- colorRamp2(
    seq(rng[1], rng[2], length.out = 11),
    rev(RColorBrewer::brewer.pal(11, "Spectral"))
  )

  # Build clean row annotation vectors aligned to heat_mat
  meta_rows <- mean_v_df %>%
    tidyr::unite("group", species, perturbation, sep = "_", remove = FALSE) %>%
    tibble::column_to_rownames("group") %>%
    as.data.frame()
  meta_rows <- meta_rows[rownames(heat_mat), , drop = FALSE]

  row_species <- factor(as.character(meta_rows$species), levels = species_order)
  row_perturb <- factor(as.character(meta_rows$perturbation), levels = perts)

  # Robust color mapping (names must match levels exactly)
  if (is.null(names(species_colors))) stop("species_colors must be a NAMED vector.")
  if (is.null(names(pert_colors)))    stop("pert_colors must be a NAMED vector.")
  if (is.null(names(cat_colors)))     stop("cat_colors must be a NAMED vector (category -> color).")

  spec_lvls <- levels(row_species)
  pert_lvls <- levels(row_perturb)

  species_colmap <- setNames(unname(species_colors[spec_lvls]), spec_lvls)
  pert_colmap    <- setNames(unname(pert_colors[pert_lvls]),    pert_lvls)

  if (any(is.na(species_colmap))) {
    missing <- spec_lvls[is.na(species_colmap)]
    stop("Missing species_colors for: ", paste(missing, collapse = ", "))
  }
  if (any(is.na(pert_colmap))) {
    missing <- pert_lvls[is.na(pert_colmap)]
    stop("Missing pert_colors for: ", paste(missing, collapse = ", "))
  }

  # Row annotation
  row_ha <- ComplexHeatmap::rowAnnotation(
    Species      = row_species,
    Perturbation = row_perturb,
    col = list(Species = species_colmap, Perturbation = pert_colmap),
    annotation_legend_param = list(Species = list(title = "Species"), Perturbation = list(title = "Perturbation")),
    show_legend = TRUE,
    show_annotation_name = FALSE
  )

  # -------- NEW: per-gene column annotation (categories) --------
  # Align col_df to the selected genes and ensure all needed categories are mapped
  if (!all(c("gene", "category") %in% colnames(col_df))) {
    stop("col_df must contain columns 'gene' and 'category'.")
  }

  # Keep only the genes in the heatmap, in the same column order
  col_meta <- col_df[match(sel_genes, col_df$gene), , drop = FALSE]

  # Some genes might be missing from col_df; label as NA (they won't be colored) or stop if you prefer
  missing_genes <- sel_genes[is.na(col_meta$gene)]
  if (length(missing_genes) > 0) {
    warning("These genes have no category in col_df and will appear as NA in the column annotation: ",
            paste(missing_genes, collapse = ", "))
    col_meta$gene[is.na(col_meta$gene)] <- sel_genes[is.na(col_meta$gene)]
    col_meta$category[is.na(col_meta$category)] <- NA_character_
  }

  # Ensure category factor levels match cat_colors order (so legend shows in that order)
  known_cats <- names(cat_colors)
  unknown_cats <- setdiff(na.omit(unique(col_meta$category)), known_cats)
  if (length(unknown_cats) > 0) {
    stop("cat_colors missing colors for categories: ", paste(unknown_cats, collapse = ", "))
  }

  col_cat_factor <- factor(col_meta$category, levels = known_cats)

  column_ha <- ComplexHeatmap::HeatmapAnnotation(
    Category = col_cat_factor,
    col = list(Category = cat_colors),
    annotation_legend_param = list(Category = list(title = "Gene Category")),
    show_annotation_name = FALSE
  )
  # --------------------------------------------------------------

  # Output directory
  out_dir <- file.path(fig_path, "heatmap_MeanLog2CPM")
  if (!dir.exists(out_dir)) dir.create(out_dir, recursive = TRUE)
  perts_str <- paste(perts, collapse = "_")

  pdf(
    file = file.path(out_dir, paste0("heatmap_MeanLog2CPM_", save_name, "_", type, "_", perts_str, "_", time_point, ".pdf")),
    width = 6, height = 4
  )

  ht <- ComplexHeatmap::Heatmap(
    heat_mat,
    name = "Log2CPM",
    col = col_fun,
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    show_row_names = FALSE,
    show_column_names = TRUE,
    column_title = title,
    left_annotation = row_ha,
    top_annotation  = column_ha   # <- NEW: add the column annotation
  )

  ComplexHeatmap::draw(ht)
  dev.off()
}

```

```{r}
# Define input variables
type <- 'Tel_NE_ventral'
time <- '54hr'
pert <- 'BMP7'
cntrl <- 'TreMan'
perts <- c(cntrl, pert)
species_order <- c('Human', 'Chimp', 'Orangutan')
genes <- genes <- c('ID2', 'MKI67', 'MCM4', 'RACGAP1', 'SERPINI1', 'BRCA1', 'LGR5', 'KCNH5', 'CSRNP3')
cats <- c('Conserved', 'Response Only', 'Response Only', 'Response Only', 'Response Only', 'Enhancing', 'Enhancing', 'Reversing', 'Reversing')
col_df <- data.frame(gene = genes, category = cats)
cat_colors <- c("Conserved" = 'royalblue2', "Response Only" = 'green', "Reversing" = "navy", "Enhancing" = "cyan", "Opposing"= "red")

# Plot complex heatmap
mean_complexheatmap2(voom_df,
                    time,
                    perts,
                    genes,
                    "divergent_BMP7_rGenes_54hr_select_CatsColored",
                    "Select Divergent BMP7 rGenes",
                    species_order,
                    species_colors,
                    pert_colors,
                    col_df,
                    cat_colors
                    )
```



## Quantitative drGene heatmap

```{r WNT-associated conserved CHIR99021 rGenes}
# Define input variables
type <- 'Tel_NE_ventral'
time <- '54hr'
pert <- 'BMP7'
cntrl <- 'TreMan'
perts <- c(cntrl, pert)
species_order <- c('Human', 'Chimp', 'Orangutan')


# Load conserved rGene ORA with GO:BP results table
all_ora_df <- readRDS(paste0(dir_path, "conserved_rGenes_GOBP_ORA_df.rds"))

# Pull genes
genes <- all_ora_df %>% 
  filter(perturbation == pert & time_point == time & Description == "WNT SIGNALING PATHWAY") %>% 
  pull(geneSymbol)
genes <- unlist(strsplit(genes, "/"))

# Plot complex heatmap
mean_complexheatmap(voom_df,
                    time,
                    perts,
                    genes,
                    "WNT_signaling_pathway",
                    "Conserved CHIR99021 rGenes Associated with the WNT Signaling Pathway",
                    species_order,
                    species_colors,
                    pert_colors
                    )
```


```{r heatmap quantitative drGenes}
# Specify variables
type <- 'Tel_NE_ventral'
time <- '6hr'
pert <- 'BMP7'
cntrl <- 'TreMan'
perts <- c(cntrl, pert)
species_order <- c('Human','Chimp','Orangutan')

# Pull divergent rGenes
genes <- divergent_rgenes_df %>% filter(perturbation == pert & cell_type == type & time_point == time) %>% pull(all)
genes <- unlist(strsplit(genes, "/"))
genes <- gsub("-", "_", genes)

# Filter dataset to variables of interest
v_df <- voom_df %>% filter(time_point == time & perturbation %in% perts)

# Find the mean value for each species
mean_v_df <- v_df %>%
  dplyr::select(species, perturbation, 10:ncol(.)) %>%       # Drop extra metadata columns
  group_by(species, perturbation) %>%
  summarise(across(everything(), ~ mean(.x, na.rm=TRUE)), .groups = 'drop') %>%
  
  # Factor and order
   mutate(species = factor(species, levels = species_order),
          perturbation = factor(perturbation, levels = perts)
   ) %>%
  arrange(species, perturbation)

# Subset to genes of interest
sel_genes <- intersect(genes, colnames(mean_v_df))

# Plot heatmap
heat_mat <- mean_v_df %>%
  unite("group", species, perturbation, sep = "_") %>%
  column_to_rownames("group") %>%
  dplyr::select(all_of(sel_genes)) %>%
  as.matrix()

pdf(paste0(fig_path, "heatmap_MeanLog2CPM/heatmap_MeanLog2CPM_quantitative_", type, "_", pert, "_", time, ".pdf"), width = 60, height = 3)
pheatmap(heat_mat, scale = "none", show_rownames = TRUE, show_colnames = TRUE, fontsize = 8, cluster_rows = FALSE, cluster_cols = TRUE)
dev.off()
```

```{r heatmap quantitative drGenes filter logFC}
# Specify variables
type <- 'Tel_NE_ventral'
time <- '54hr'
pert <- 'BMP7'
cntrl <- 'TreMan'
perts <- c(cntrl, pert)
species_order <- c('Human','Chimp','Orangutan')

# Filter data
filt_df <- divergent_response_df %>% filter(assay == type & time_point == time & perturbation == pert & significant == TRUE & logFC > 1)

# Pull genes
genes <- filt_df %>% pull(ID) 
genes <- gsub("-", "_", genes)

# Filter dataset to variables of interest
v_df <- voom_df %>% filter(time_point == time & perturbation %in% perts)

# Find the mean value for each species
mean_v_df <- v_df %>%
  dplyr::select(species, perturbation, 10:ncol(.)) %>%       # Drop extra metadata columns
  group_by(species, perturbation) %>%
  summarise(across(everything(), ~ mean(.x, na.rm=TRUE)), .groups = 'drop') %>%
  
  # Factor and order
   mutate(species = factor(species, levels = species_order),
          perturbation = factor(perturbation, levels = perts)
   ) %>%
  arrange(species, perturbation)

# Subset to genes of interest
sel_genes <- intersect(genes, colnames(mean_v_df))

# Plot heatmap
heat_mat <- mean_v_df %>%
  unite("group", species, perturbation, sep = "_") %>%
  column_to_rownames("group") %>%
  dplyr::select(all_of(sel_genes)) %>%
  as.matrix()

pdf(paste0(fig_path, "heatmap_MeanLog2CPM/heatmap_MeanLog2CPM_quantitative_", type, "_", pert, "_", time, "_LogFCFilter.pdf"),
    width = 15, height = 3)
pheatmap(heat_mat, scale = "none", show_rownames = TRUE, show_colnames = TRUE, fontsize = 8, cluster_rows = FALSE, cluster_cols = TRUE)
dev.off()
```

```{r load divergent rGene ORA results}
ora_human_up_go_bp_df <- readRDS(paste0(dir_path, "divergent_rGenes_ORA_GOBP_human_up.rds"))

```

```{r pull genes of interest}
cc_gobp_terms <- c('GOBP_CHROMOSOME_ORGANIZATION',
                   'GOBP_CHROMOSOME_SEGREGATION',
                   'GOBP_DNA_REPLICATION',
                   'GOBP_NUCLEAR_CHROMOSOME_SEGREGATION',
                   'GOBP_ORGANELLE_FISSION',
                   'GOBP_PROTEIN_DNA_COMPLEX_ASSEMBLY',
                   'GOBP_NUCLEAR_CHROMOSOME_SEGREGATION',
                   'GOBP_DNA_REPAIR',
                   'GOBP_KINETOCHORE_ORGANIZATION',
                   'GOBP_DNA_TEMPLATED_DNA_REPLICATION',
                   'GOBP_POSITIVE_REGULATION_OF_CELL_CYCLE_PROCESS',
                   'GOBP_CELL_CYCLE_G2_M_PHASE_TRANSITION',
                   'GOBP_MITOTIC_G2_M_TRANSITION_CHECKPOINT',
                   'GOBP_MITOTIC_SISTER_CHROMATID_SEGREGATION',
                   'GOBP_REGULATION_OF_CELL_CYCLE_G2_M_PHASE_TRANSITION',
                   'GOBP_DNA_INTEGRITY_CHECKPOINT_SIGNALING',
                   'GOBP_DNA_TEMPLATED_DNA_REPLICATION_MAINTENANCE_OF_FIDELITY',
                   'GOBP_DNA_REPLICATION_CHECKPOINT_SIGNALING',
                   'GOBP_SISTER_CHROMATID_SEGREGATION',
                   'GOBP_POSITIVE_REGULATION_OF_CHROMOSOME_SEGREGATION',
                   'GOBP_POSITIVE_REGULATION_OF_CHROMOSOME_CONDENSATION',
                   'GOBP_CELL_CYCLE_CHECKPOINT_SIGNALING',
                   'GOBP_POSITIVE_REGULATION_OF_CELL_CYCLE',
                   'GOBP_NEGATIVE_REGULATION_OF_CELL_CYCLE_G2_M_PHASE_TRANSITION',
                   'GOBP_MEIOTIC_CHROMOSOME_SEGREGATION',
                   'GOBP_CENTROMERE_COMPLEX_ASSEMBLY',
                   'GOBP_MITOTIC_NUCLEAR_DIVISION',
                   'GOBP_DNA_SYNTHESIS_INVOLVED_IN_DNA_REPAIR',
                   'GOBP_DOUBLE_STRAND_BREAK_REPAIR',
                   'GOBP_REGULATION_OF_CELL_CYCLE_PHASE_TRANSITION',
                   'GOBP_CELL_CYCLE_PHASE_TRANSITION',
                   'GOBP_REGULATION_OF_DNA_REPLICATION',
                   'GOBP_CHROMOSOME_CONDENSATION',
                   'GOBP_REGULATION_OF_MITOTIC_CELL_CYCLE',
                   'GOBP_REGULATION_OF_CHROMOSOME_CONDENSATION',
                   'GOBP_DNA_STRAND_ELONGATION_INVOLVED_IN_DNA_REPLICATION',
                   'GOBP_HOMOLOGOUS_RECOMBINATION',
                   'GOBP_MICROTUBULE_CYTOSKELETON_ORGANIZATION_INVOLVED_IN_MITOSIS',
                   'GOBP_KINETOCHORE_ASSEMBLY',
                   'GOBP_MITOTIC_CELL_CYCLE_CHECKPOINT_SIGNALING',
                   'GOBP_MITOTIC_CELL_CYCLE_PHASE_TRANSITION',
                   'GOBP_REGULATION_OF_MITOTIC_CELL_CYCLE_PHASE_TRANSITION',
                   'GOBP_POSITIVE_REGULATION_OF_MITOTIC_CELL_CYCLE',
                   'GOBP_NEGATIVE_REGULATION_OF_CELL_CYCLE',
                   'GOBP_MITOTIC_G2_DNA_DAMAGE_CHECKPOINT_SIGNALING',
                   'GOBP_MITOTIC_DNA_REPLICATION_CHECKPOINT_SIGNALING',
                   'GOBP_NEGATIVE_REGULATION_OF_CELL_CYCLE_PROCESS',
                   'GOBP_POSITIVE_REGULATION_OF_CELL_CYCLE_PHASE_TRANSITION',
                   'GOBP_RECOMBINATIONAL_REPAIR',
                   'GOBP_MITOTIC_DNA_INTEGRITY_CHECKPOINT_SIGNALING',
                   'GOBP_MITOTIC_SPINDLE_ORGANIZATION',
                   'GOBP_DNA_REPLICATION_INITIATION'
)
cc_genes <- ora_human_up_go_bp_df %>%
  filter(cell_type == "Tel_NE_ventral",
         perturbation == "BMP7",
         time_point == "54hr",
         Description %in% cc_gobp_terms) %>%
  pull(geneSymbol) %>%                 
  str_split("/", simplify = FALSE) %>%
  unlist(use.names = FALSE) %>%
  str_trim() %>%
  discard(~ .x == "" | is.na(.x)) %>%
  unique() %>%
  sort()
```

```{r heatmap quantitative drGenes CC genes}
# Specify variables
type <- 'Tel_NE_ventral'
time <- '54hr'
pert <- 'BMP7'
cntrl <- 'TreMan'
perts <- c(cntrl, pert)
species_order <- c('Human','Chimp','Orangutan')

genes <- cc_genes

# Filter dataset to variables of interest
v_df <- voom_df %>% filter(time_point == time & perturbation %in% perts)

# Find the mean value for each species
mean_v_df <- v_df %>%
  dplyr::select(species, perturbation, 10:ncol(.)) %>%       # Drop extra metadata columns
  group_by(species, perturbation) %>%
  summarise(across(everything(), ~ mean(.x, na.rm=TRUE)), .groups = 'drop') %>%
  
  # Factor and order
   mutate(species = factor(species, levels = species_order),
          perturbation = factor(perturbation, levels = perts)
   ) %>%
  arrange(species, perturbation)

# Subset to genes of interest
sel_genes <- intersect(genes, colnames(mean_v_df))

# Plot heatmap
heat_mat <- mean_v_df %>%
  unite("group", species, perturbation, sep = "_") %>%
  column_to_rownames("group") %>%
  dplyr::select(all_of(sel_genes)) %>%
  as.matrix()

pdf(paste0(fig_path, "heatmap_MeanLog2CPM/heatmap_MeanLog2CPM_quantitative_", type, "_", pert, "_", time, "_CCgenes.pdf"),
    width = 20, height = 3)
pheatmap(heat_mat, scale = "none", show_rownames = TRUE, show_colnames = TRUE, fontsize = 8, cluster_rows = FALSE, cluster_cols = TRUE)
dev.off()
```

# Qualitative drGenes

```{r}
voom_df <- extractData(voom_res, assay = "Tel_NE_ventral")
```

```{r generate qual_response_df}
qual_response_df <- response_df %>%
  filter(assay %in% c('Tel_NE_ventral', 'Tel_NE_dorsal')) %>%
  group_by(ID, assay, perturbation, time_point) %>%
  filter(n_distinct(species) == 2) %>%
  filter(
    any(significant & abs(logFC) > 2),
    any(!significant)
  ) %>%
  ungroup()
head(qual_response_df)
```

## Heatmap qualitative drGenes

```{r heatmap qualitative drGenes}
# Specify variables
type <- 'Tel_NE_ventral'
time <- '54hr'
pert <- 'BMP7'
cntrl <- 'TreMan'
perts <- c(cntrl, pert)
species_order <- c('Human', 'Chimp', 'Orangutan')

# Pull out genes of interest
genes <- qual_response_df %>% filter(assay == type & perturbation == pert & time_point == time) %>% distinct(ID) %>% pull(ID)

# Filter dataset to variables of interest
v_df <- voom_df %>% filter(time_point == time & perturbation %in% perts)

# Find the mean value for each species
mean_v_df <- v_df %>%
  dplyr::select(species, perturbation, 10:ncol(.)) %>%       # Drop extra metadata columns
  group_by(species, perturbation) %>%
  summarise(across(everything(), ~ mean(.x, na.rm=TRUE)), .groups = 'drop') %>%
  
  # Factor and order
   mutate(species = factor(species, levels = species_order),
          perturbation = factor(perturbation, levels = perts)
   ) %>%
  arrange(species, perturbation)

# Subset to genes of interest
sel_genes <- intersect(genes, colnames(mean_v_df))

# Plot heatmap
heat_mat <- mean_v_df %>%
  unite("group", species, perturbation, sep = "_") %>%
  column_to_rownames("group") %>%
  dplyr::select(all_of(sel_genes)) %>%
  as.matrix()

pdf(paste0(fig_path, "heatmap_MeanLog2CPM/heatmap_MeanLog2CPM_qualitative_", type, "_", pert, "_", time, ".pdf"), width = 24, height = 3)
pheatmap(heat_mat, scale = "none", show_rownames = TRUE, show_colnames = TRUE, fontsize = 8, cluster_rows = FALSE, cluster_cols = TRUE)
dev.off()
```

# Select candidate drGenes

```{r specify select drGenes}
bmp7_6hr_genes <- c(
 "SOX4", "SMAD3", "NKD1", "KREMEN1", "ROR2", "CDH6","GRM8", "CASC1", "ZNF804B")
 #"GNB4", "P3H2",  "CDH18", "PLCB4", "CNTN4", "PTCHD4", "THBS1", 
 #"S1PR1", "GRIK1", "ZNF804B", "ARX",  "KCTD12")

bmp7_54hr_genes <- c(
  'ID2', 'SMAD9',
  'MKI67', 'RACGAP1', 'MCM4', 
  'KCNH5', 'CSRNP3', 'ZNF536',
  'LGR5',  'BRCA1',
  'TYRO3', 'SERPINI1', 'CFAP221', 'KIF26B', 'COL24A1', 'TYMS', 'ETV1', 'SLC16A10'
  #'ZNF521', 'CDH6', 'NAV3', 'MAGI1', 'PAM', 'RORA', 'ST6GALNAC3', 'ZNF536', 'NLGN1', 'PTPRZ1', 'SOX4', 'NXPH1', 'LRRC3B', 'REEP1', 
  #'KIF26B', 'COL24A1', 'VEGFC', 'SERPINI1', 'CDKN1A', 'ETS1', 'ADAMTS18', 'GRIK1', 'PLCB4', 'ETV1', 'SPRED2', 'SMAD3', 'FREM2', 'CUX2',
  #'TMEM132C', 'LRP1B', 'SOX9',  'ATAD2', 'TYMS', 'BRIP1', 'SMC2', 'RRM2', 'KIF11', 'CENPH', 'CENPK', 'CDK1', 'CENPW', 'KCNB2', 'CDH18', 
  #'LAMA1', 'KAT2B', 'NR2F2', 'PCSK5', 'CENPJ', 'S1PR1', 'POLQ',
  #   'PBK', 'SLF1', 'LAMA1', 'PBK', 'SHISA9', 'CHRDL1', 'SOX9-AS1', 'HNMT', 'RGS6', 'POC1A', 'MCM8', 'IQGAP3', 'KCNMA1',
  #'H2AC14', 'H3-4', 'DEPDC1', 'DDIAS', 'KIF20A', 'PDZRN4', 'EMCN', 'H3C7', 'H3C8', 'KIF26B', 'THBS1', 'CLIC6', 'RAB3B', 'KCNH5', 'ENPP2'
)

fgf2_genes <- c(#'PDE4B', "ST6GALNAC5", 'CREB3L4', "DLGAP1", "LGI1", "PDE7B", "SIPA1L2", "WEE1",
                "ST6GALNAC5", "EPHB1", "CNTN4", "DST", "NRP1",  "DGKI", "DNMBP", "STIM2"
                )
                # "PTPRJ", "TMEM158", "ETV3", "NOX4", "RFTN2", "RBFOX1", "PROX1", "AP1S2"
                #"RUNX1T1", "KCND2", "SLC44A5",  "DISC1", "ADCY2", "RUNX1",  "CSMD3", "GALNTL6", "RRS1", "GALNT12", "KCNN3", "WHRN", "PDE1C",
                #"DISC1", "GABRB1", "SHISA9", "NRP1", "GAP43:)

fgf8b_genes <- c(
 "GAP43", "SIPA1L2", "DLGAP1", "ST6GALNAC5", "PTPRJ", "NRP1", "AP1S2", "DST", "WEE1", "DNMBP", "PRICKLE2", "XKR4", "LGR5", "PDE7B", "DISC1",
 "LRRTM3", "MHENCR", "EFHC2", "PPP1R1C", "KCNJ13"
)

int_obvi_cc <- c(
  "ATAD2", "BRIP1", "CDC6", "CDK1", "CENPJ", "CENPK", "CLSPN", "ESCO2", 
  "GMNN", "HJURP", "KCNH5", "KIF11", "KIF15", "KNL1", "MCM4", "MKI67", 
  "NCAPH", "PCLAF", "POLQ", "RACGAP1", "RRM2", "SGO1", "UHRF1"
)
```

```{r select drGenes by category}
# Specify variables
type <- 'Tel_NE_ventral'
time <- '54hr'
pert <- 'BMP7'
category <- 'Enhancing'

filt_df <- speciesdiff_logfc_df %>%
  filter(assay == type & time_point == time & perturbation == pert & div_rGene_sig == TRUE & div_rGene_category == category) %>%
  mutate(abs_residual = abs(residual)) %>%
  arrange(desc(abs_residual))
filt_genes <- filt_df %>% pull(ID)
```

```{r scatterplot select drGenes by category}
# Find top outliers base on absolute value of residuals for labeling
label_df <- filt_df %>% arrange(desc(abs_residual)) %>% slice_head(n = 50)

# Plot a scatter plot
p <- ggplot(filt_df, aes(x = Baseline_logFC, y = Response_logFC, fill = div_rGene_category, shape = time_point)) +
  geom_point(size = 2) +

  # Label top genes
  geom_text_repel(data = label_df, aes(label = ID), size = 2.5, box.padding = 0.1, point.padding = 0.1, max.overlaps = Inf, 
                  force = 5, force_pull = 1, segment.size = 0.2, segment.color = "black", color = "black") +
  
  # Add labels
  labs(
  title = paste0("Species Differences in GEX in Baseline vs. Stimulus \n", pert, ": ", time),
  x = "Baseline Log(Human/Chimp)",
  y = "Stimulus Log(Human/Chimp",
  color = "Perturbation",
  shape = "Time Point"
  ) +
  scale_fill_manual(values = div_rgene_colors) +
  scale_shape_manual(values = time_point_shapes) +
  theme_minimal() + 
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),  # Center and size the title
    legend.position = "right",  # Place legend on the right
    axis.line = element_line(color = "black"),  # Add axis lines
    panel.grid.major = element_blank(),  # Remove major gridlines
    panel.grid.minor = element_blank()   # Remove minor gridlines
  ) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") + # Add dashed y=x line
  geom_hline(yintercept = 0, color = "black") +  # Horizontal line at y = 0
  geom_vline(xintercept = 0, color = "black") #+ # Vertical line at x = 0
  #coord_fixed(ratio = 1)  # Ensure the aspect ratio is 1:1 for square plot
  ggsave(filename = paste0(fig_path, "scatterplot_SpeciesDiffLogFC_BaselineVsStimulus/scatterplot_select_drGenes",
                           time, "_", pert, "_", category, ".png"),
             plot = p, height = 10, width = 10, dpi = 300, bg = 'white')
```

```{r scatterplot select drGenes}
# Specify variables
type <- 'Tel_NE_ventral'
time <- '54hr'
pert <- 'BMP7'
genes <- bmp7_54hr_genes

# Filter data
filt_df <- speciesdiff_logfc_df %>%
  filter(assay == type & time_point == time & perturbation == pert & div_rGene_sig == TRUE) %>%
  mutate(abs_residual = abs(residual)) %>%
  arrange(desc(abs_residual))

# Find top outliers base on absolute value of residuals for labeling
label_df <- filt_df %>% filter(ID %in% genes) %>%
    mutate(
      nudge_x = ifelse(Response_logFC >= Baseline_logFC, -2, 2),
      nudge_y = ifelse(Response_logFC >= Baseline_logFC, 1, -1)
    )


# Plot a scatter plot
p <- ggplot(filt_df, aes(x = Baseline_logFC, y = Response_logFC, fill = div_rGene_category, shape = time_point)) +
  geom_point(size = 2, color = 'black') +

  # Label top genes
  geom_text_repel(
    data = label_df,
    aes(label = ID),
    size = 3,
    box.padding = 0.3,
    point.padding = 0.3,
    nudge_x = label_df$nudge_x,
    nudge_y = label_df$nudge_y,
    segment.size = 0.5,
    segment.color = "black",
    color = "black",
    max.overlaps = Inf,
    force = 5
  ) +
  
  # Add labels
  labs(
  title = paste0("Species Differences in GEX in Baseline vs. Stimulus \n", pert, ": ", time),
  x = "Baseline Log(Human/Chimp)",
  y = "Stimulus Log(Human/Chimp",
  color = "Perturbation",
  shape = "Time Point"
  ) +
  scale_fill_manual(values = div_rgene_colors) +
  scale_shape_manual(values = time_point_shapes) +
  guides(
    fill = guide_legend(override.aes = list(shape = 21, color = "black")), # keep fill legend visible
    shape = guide_legend(override.aes = list(fill = "white"))              # keep shape legend visible
  ) +
  theme_minimal() + 
  theme(
    plot.title = element_text(hjust = 0.5, size = 16),  # Center and size the title
    legend.position = "right",  # Place legend on the right
    axis.line = element_line(color = "black"),  # Add axis lines
    panel.grid.major = element_blank(),  # Remove major gridlines
    panel.grid.minor = element_blank()   # Remove minor gridlines
  ) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") + # Add dashed y=x line
  geom_hline(yintercept = 0, color = "black") +  # Horizontal line at y = 0
  geom_vline(xintercept = 0, color = "black") + # Vertical line at x = 0
  coord_fixed(ratio = 1)  # Ensure the aspect ratio is 1:1 for square plot
  ggsave(filename = paste0(fig_path, "scatterplot_SpeciesDiffLogFC_BaselineVsStimulus/scatterplot_select_drGenes_",
                           time, "_", pert, "_panel.png"),
             plot = p, height = 10, width = 10, dpi = 300, bg = 'white')
```

```{r heatmap select drGenes}
# Specify variables
type <- 'Tel_NE_ventral'
time <- '54hr'
pert <- 'BMP7'
cntrl <- 'TreMan'
perts <- c(cntrl, pert)
species_order <- c('Human', 'Chimp', 'Orangutan')
genes <- bmp7_54hr_genes

# Filter dataset to variables of interest
v_df <- voom_df %>% filter(time_point == time & perturbation %in% perts)

# Find the mean value for each species
mean_v_df <- v_df %>%
  dplyr::select(species, perturbation, 10:ncol(.)) %>%       # Drop extra metadata columns
  group_by(species, perturbation) %>%
  summarise(across(everything(), ~ mean(.x, na.rm=TRUE)), .groups = 'drop') %>%
  
  # Factor and order
   mutate(species = factor(species, levels = species_order),
          perturbation = factor(perturbation, levels = perts)
   ) %>%
  arrange(species, perturbation)

# Subset to genes of interest
sel_genes <- intersect(genes, colnames(mean_v_df))

# Plot heatmap
heat_mat <- mean_v_df %>%
  unite("group", species, perturbation, sep = "_") %>%
  column_to_rownames("group") %>%
  dplyr::select(all_of(sel_genes)) %>%
  as.matrix()

pdf(paste0(fig_path, "heatmap_MeanLog2CPM/heatmap_MeanLog2CPM_", type, "_", pert, "_", time, "_select_panel.pdf"), width = 20, height = 3)
pheatmap(heat_mat, scale = "none", show_rownames = TRUE, show_colnames = TRUE, fontsize = 8, cluster_rows = FALSE, cluster_cols = TRUE)
dev.off()
```



```{r box plot drGene candidates}
# Plot a box plot for species divergent rGenes for a sample group
# Set sample group
pert = 'FGF8b'
type = 'Tel_NE_ventral'
time = '6hr'
save_name = paste0("boxplot_Log2CPM_", type, "_", pert, "_", time, "_candidates")

# Pull divergent rGenes
degs <- fgf8b_genes

# Iterate through genes and plot
for (deg in degs){
    p <- plotstratify_boxplot(voom_res, type, pert, deg)
    p + ggtitle(paste0(deg, "/n", pert, " ", time))
    ggsave(filename = paste0(fig_path, save_name, "/", save_name, "_", deg, ".png"), plot=p, width=8, height=6, dpi=300, bg='white')
}
```

# Response gene category enrichments

```{r define rgene category gene sets}
# Define gene sets of interest

# Universe genes
response_df <- readRDS("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/dreamlet/response_df.rds")
universe_genes <- response_df %>%
  dplyr::filter(assay %in% c("Tel_NE_dorsal", "Tel_NE_ventral") & !(perturbation %in% c('Verteporfin'))) %>%
  pull(ID) %>%
  unique()

# All rGenes
all_rgenes <- response_df %>%
  dplyr::filter(assay %in% c("Tel_NE_dorsal", "Tel_NE_ventral") & !(perturbation %in% c('Verteporfin')) & significant == TRUE) %>%
  pull(ID) %>%
  unique()

# Conserved rGenes
simple_conserved_rgenes_df <- readRDS("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/dreamlet/simple_conserved_rDEGs_df.rds")
all_conserved_rgenes <- simple_conserved_rgenes_df %>%
  dplyr::filter(!(perturbation %in% c('Verteporfin')) & N_crDEGs > 1)%>%
  pull(crDEGs) %>% str_split("/")%>% unlist() %>% na.omit() %>% unique()

# Divergent rGenes
simple_divergent_rgenes_df <- readRDS("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/dreamlet/simple_divergent_rGenes_df.rds")
all_divergent_rgenes <- simple_divergent_rgenes_df %>%
  dplyr::filter(!(perturbation %in% c('Verteporfin')) & N_all > 1)%>%
  pull(all) %>% str_split("/") %>% unlist() %>% na.omit() %>% unique()

# Divergent baseline genes
speciesdiff_df <- readRDS("/Users/reedmcmullen/Desktop/NEMPs/NEMP25/scRNAseq/dreamlet/speciesdiff_df.rds")
all_speciesdiff_genes <- speciesdiff_df %>%
  dplyr::filter(assay %in% c("Tel_NE_dorsal", "Tel_NE_ventral") & !(perturbation %in% c('Verteporfin'))  & significant == TRUE) %>%
  pull(ID) %>%
  unique()

# Define other rGenes
labeled_rgenes <- union(all_conserved_rgenes, all_divergent_rgenes)
other_rgenes <- setdiff(all_rgenes, labeled_rgenes)

# Define non rGenes
non_rgenes <- setdiff(universe_genes, all_rgenes)
non_rgenes <- setdiff(non_rgenes, all_divergent_rgenes)

# Add gene lists to named list
gene_set_list <- list(
  "Conserved rGene"         = all_conserved_rgenes,
  "Non-rGene"               = non_rgenes,
  "Other rGene"             = other_rgenes,
  "Divergent rGene"         = all_divergent_rgenes,
  "Divergent Baseline Gene" = all_speciesdiff_genes
)
```

## DisGeNet ORA

Perform ORA with DisGeNet for all conserved rGenes
```{r all conserved rGene DisGeNet ORA}
# Perform ORA for DEGs with disgenet2r
ora <- disease_enrichment(entities = all_conserved_rgenes,
                          common_entities = 1,
                          max_pvalue = 0.05,
                          vocabulary = 'HGNC',
                          database = 'CURATED')
          
# Check if ORA returned results
if (!is.null(ora@qresult) && nrow(ora@qresult) > 0) {
  # Add metadata columns to ORA results and store in list
      conserved_rgene_disgenet_ora <- ora@qresult %>% as.data.frame() %>%
      filter(p_adjusted < 0.10 & intersectionSize > 1)
}

saveRDS(conserved_rgene_disgenet_ora, paste(dir_path, "conserved_rGene_DisGeNet_ORA_df.rds"))
head(conserved_rgene_disgenet_ora)
```

Perform ORA with DisGeNet for all divergent rGenes
```{r all divergent rGene DisGeNet ORA}
# Perform ORA for DEGs with disgenet2r
ora <- disease_enrichment(entities = all_divergent_rgenes,
                          common_entities = 1,
                          max_pvalue = 0.05,
                          vocabulary = 'HGNC',
                          database = 'CURATED')
          
# Check if ORA returned results
if (!is.null(ora@qresult) && nrow(ora@qresult) > 0) {
  # Add metadata columns to ORA results and store in list
      divergent_rgene_disgenet_ora <- ora@qresult %>% as.data.frame()
      #filter(p_adjusted < 0.10 & intersectionSize > 1)
}
saveRDS(divergent_rgene_disgenet_ora, paste(dir_path, "divergent_rGene_DisGeNet_ORA_df.rds"))
head(divergent_rgene_disgenet_ora)
```

Perform ORA with DisGeNet for all non rGenes
```{r all non rGene DisGeNet ORA}
# Perform ORA for DEGs with disgenet2r
ora <- disease_enrichment(entities = non_rgenes,
                          common_entities = 1,
                          max_pvalue = 0.05,
                          vocabulary = 'HGNC',
                          database = 'CURATED')
          
# Check if ORA returned results
if (!is.null(ora@qresult) && nrow(ora@qresult) > 0) {
  # Add metadata columns to ORA results and store in list
      non_rgene_disgenet_ora <- ora@qresult %>% as.data.frame()
      #filter(p_adjusted < 0.10 & intersectionSize > 1)
}

saveRDS(non_rgene_disgenet_ora, paste(dir_path, "non_rGene_DisGeNet_ORA_df.rds"))
head(non_rgene_disgenet_ora)
```

Perform ORA with DisGeNet for all baseline divergent genes
```{r all species divergent genes DisGeNet ORA}
# Perform ORA for DEGs with disgenet2r
ora <- disease_enrichment(entities = all_speciesdiff_genes,
                          common_entities = 1,
                          max_pvalue = 0.05,
                          vocabulary = 'HGNC',
                          database = 'CURATED')
          
# Check if ORA returned results
if (!is.null(ora@qresult) && nrow(ora@qresult) > 0) {
  # Add metadata columns to ORA results and store in list
      speciesdiff_gene_disgenet_ora <- ora@qresult %>% as.data.frame()
      #filter(p_adjusted < 0.10 & intersectionSize > 1)
}

saveRDS(speciesdiff_gene_disgenet_ora, paste(dir_path, "divergent_baseline_gene_DisGeNet_ORA_df.rds"))
head(speciesdiff_gene_disgenet_ora)
```

```{r}
conserved_rgene_disgenet_ora <- readRDS(paste(dir_path, "conserved_rGene_DisGeNet_ORA_df.rds"))
divergent_rgene_disgenet_ora <- readRDS(paste(dir_path, "divergent_rGene_DisGeNet_ORA_df.rds"))
non_rgene_disgenet_ora <- readRDS(paste(dir_path, "non_rGene_DisGeNet_ORA_df.rds"))
#speciesdiff_gene_disgenet_ora <- readRDS(paste(dir_path, "divergent_baseline_gene_DisGeNet_ORA_df.rds"))
```

Combine results of the DisGeNet ORA for gene categories into one data frame
```{r combine gene category DisGeNet ORA results}
# Add gene category metadata
conserved_rgene_disgenet_ora$GeneCategory <- "Conserved rGene"
divergent_rgene_disgenet_ora$GeneCategory <- "Divergent rGene"
non_rgene_disgenet_ora$GeneCategory <- "Non rGene"
#speciesdiff_gene_disgenet_ora$GeneCategory <- "Baseline Divergent Gene"

# Combine data
gene_category_disgenet_ora <- bind_rows(conserved_rgene_disgenet_ora,
                          divergent_rgene_disgenet_ora,
                          non_rgene_disgenet_ora
                          #speciesdiff_gene_disgenet_ora
                          )
gene_category_disgenet_ora$GeneRatio <- sapply(gene_category_disgenet_ora$geneRatio, function(x) {
  parts <- strsplit(x, "/")[[1]]
  as.numeric(parts[1]) / as.numeric(parts[2])
})

gene_category_disgenet_ora <- gene_category_disgenet_ora %>% filter(p_adjusted < 0.10 & intersectionSize > 1)

head(gene_category_disgenet_ora)
```

```{r save gene_category_disgenet_ora}
saveRDS(gene_category_disgenet_ora, file = paste0(dir_path, "GeneCategory_DisGeNet_ORA_df.rds"))
```

```{r load gene_category_disgenet_ora}
gene_category_disgenet_ora <- readRDS(file = paste0(dir_path, "GeneCategory_DisGeNet_ORA_df.rds"))
head(gene_category_disgenet_ora)
```

Plot a dot plot of gene category DisGeNet ORA results
```{r dot plot conserved rGene genetic classes DisGeNet ORA}
# Filter to dataset of interest
ora_df <- gene_category_disgenet_ora %>% filter(GeneCategory %in% c('Conserved rGene'))

# Plot a dot plot
p <- ggplot(ora_df, aes(x = intersectionSize, y = reorder(diseaseName, intersectionSize), size = GeneRatio, fill = -log10(p_adjusted))) +
  geom_point(shape = 21, color = 'black', stroke = 0.25) +
  scale_size_continuous(range = c(1, 5)) + 
  scale_fill_viridis(option='inferno') +
  theme_minimal() +
  labs(
      #title = paste0("Conserved rGene \nDisease/Disorder Enrichments"),
       x = "N Conserved rGenes",
       y = '',
       size = 'Gene Ratio',
       fill = expression ("-" ~ Log[10] ~ "Adjusted p-value"))
  theme(axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels for readability
        plot.title = element_text(hjust=0.5, size=16), # Edit plot title.
        plot.background = element_blank(),  # Remove the background around the plot.
        panel.background = element_blank(),  # Remove the background inside the plot.
        panel.grid.major = element_blank(),  # Remove major gridlines
        panel.grid.minor = element_blank()   # Remove minor gridlines
      )
ggsave(filename = paste0(fig_path, "dotplot_all_conserved_rGenes_DisGeNet_ORA.png"),
         plot = p, height = 4, width = 8, bg = 'white')
print(p)
```

```{r dot plot gene classes DisGeNet ORA}
# Format data frame
cons_ora$GeneRatio <- sapply(cons_ora$geneRatio, function(x) {
  parts <- strsplit(x, "/")[[1]]
  as.numeric(parts[1]) / as.numeric(parts[2])
})

# Plot a dot plot
p <- ggplot(cons_ora, aes(x = intersectionSize, y = reorder(diseaseName, intersectionSize), size = GeneRatio, fill = -log10(p_adjusted))) +
  geom_point(shape = 21, color = 'black', stroke = 0.25) +
  scale_size_continuous(range = c(1, 5)) + 
  scale_fill_viridis(option='inferno') +
  theme_minimal() +
  labs(title = paste0("Conserved rGene \nDisease/Disorder Enrichments"),
       x = "N Conserved rGenes",
       y = '',
       size = 'Gene Ratio',
       fill = expression ("-" ~ Log[10] ~ "Adjusted p-value"))
  theme(axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis labels for readability
        plot.title = element_text(hjust=0.5, size=16), # Edit plot title.
        plot.background = element_blank(),  # Remove the background around the plot.
        panel.background = element_blank(),  # Remove the background inside the plot.
        panel.grid.major = element_blank(),  # Remove major gridlines
        panel.grid.minor = element_blank()   # Remove minor gridlines
      )
ggsave(filename = paste0(fig_path, "dotplot_all_conserved_rGenes_DisGeNet_ORA.png"),
         plot = p, height = 4, width = 8, bg = 'white')
```